{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "45e9779b-18a7-4024-9d20-8460a554752d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6e640884-8e9e-4095-9d09-edd7c43a7ca2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def clean_text(text):\n",
    "    cleaned_text = text\n",
    "    cleaned_text = re.sub(r'http\\S+', '', cleaned_text) # Remove URLs\n",
    "    cleaned_text = re.sub(r'-\\n', '', cleaned_text) # Remove Hyphenations\n",
    "    cleaned_text = re.sub(\"\\s+\", \" \", cleaned_text) # Remove Duplicate Spaces\n",
    "    cleaned_text = re.sub(r\"\\s+([?.!,])\", r\"\\1\", cleaned_text) # Remove Spaces Before Punctuation\n",
    "    return cleaned_text.strip()\n",
    "\n",
    "def save_df(df, name):\n",
    "    df.to_csv(f'{name}.csv', encoding='utf-8', index=False)\n",
    "    df.to_excel(f'{name}.xlsx', index=False, header=True)\n",
    "    df.to_pickle(f'{name}.pkl')\n",
    "\n",
    "def load_df(name, filetype='csv', verbose=False):\n",
    "    filename = f\"{name}.{filetype}\"\n",
    "    if verbose:\n",
    "        print(f\"File Name: {filename}\")\n",
    "    if filetype == 'csv':\n",
    "        return pd.read_csv(filename)\n",
    "    if filetype == 'xlsx':\n",
    "        return pd.read_excel(filename)\n",
    "    if filetype == 'pkl':\n",
    "        return pd.read_pickle(filename)\n",
    "    raise Exception(\"Invalid File Type\")\n",
    "\n",
    "def remove_unnamed_columns(name, filetype='csv'):\n",
    "    df = load_df(name, filetype)\n",
    "    df.drop(df.columns[df.columns.str.contains('unnamed', case=False)], axis=1, inplace=True)\n",
    "    save_df(df, name)\n",
    "\n",
    "def sample_df(name, filetype='csv', frac=None, n=None, save=False):\n",
    "    df = load_df(name, filetype=filtype)\n",
    "    sub_df = df.copy().sample(frac=frac, n=n)\n",
    "    if save:\n",
    "        save_df(sub_df, f\"Sub{name}\")\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d22ce559-b86d-433a-baad-ded89a2e5165",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\lbeln\\anaconda3\\envs\\3.10\\lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From C:\\Users\\lbeln\\anaconda3\\envs\\3.10\\lib\\site-packages\\tf_keras\\src\\losses.py:2976: The name tf.losses.sparse_softmax_cross_entropy is deprecated. Please use tf.compat.v1.losses.sparse_softmax_cross_entropy instead.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Device set to use cuda:0\n"
     ]
    }
   ],
   "source": [
    "# !python -m pip install tf-keras\n",
    "from transformers import pipeline\n",
    "\n",
    "classifier = pipeline(\"zero-shot-classification\", model=\"MoritzLaurer/DeBERTa-v3-base-mnli-fever-anli\")\n",
    "# classifier = pipeline(\"zero-shot-classification\", model=\"MoritzLaurer/deberta-v3-large-zeroshot-v2.0\")\n",
    "\n",
    "def on_topic(text, verbose=False):\n",
    "    # Topic and Threshold\n",
    "    # I'm using a different model which seems to be more confident,\n",
    "    # so I'm going to increase the threshold.\n",
    "    topics = [(\"ecology\", 0.9), (\"interaction\", 0.9)]\n",
    "            \n",
    "    for topic, threshold in topics:\n",
    "        result = classifier(text, [topic])\n",
    "        if verbose:\n",
    "            print(result[\"scores\"][0])\n",
    "        if result[\"scores\"][0] < threshold:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "# print(on_topic(examples[0][\"Abstract\"], verbose=True))\n",
    "# print(on_topic(examples[1][\"Abstract\"]))\n",
    "# print(on_topic(examples[2][\"Abstract\"]))\n",
    "# print(on_topic(examples[3][\"Abstract\"]))\n",
    "# print(on_topic(\"Fatal familial insomnia (FFI) and a subtype of familial Creutzfeldt-Jakob disease (CJD), two clinically and pathologically distinct diseases, are linked to the same mutation at codon 178 (Asn 178 ) of the prion protein gene. The possibility that a second genetic component modified the phenotypic expression of the Asn 178 mutation was investigated. FFI and the familial CJD subtype segregated with different genotypes determined by the Asn 178 mutation and the methionine-valine polymorphism at codon 129. The Met 129 , Asn 178 allele segregated with FFI in all 15 affected members of five kindreds whereas the Val 129 , Asn 178 allele segregated with the familial CJD subtype in all 15 affected members of six kindreds. Thus, two distinct disease phenotypes linked to a single pathogenic mutation can be determined by a common polymorphism.\", verbose=True))\n",
    "# print(on_topic(\"Metacognition, the ability to reflect upon, and evaluate our own beliefs, can help us avoid making decisions based on unreliable evidence. Here, we provide empirical tests of the importance of human metacognition during the COVID-19 pandemic. ...Metacognition, our ability to reflect on our own beliefs, manifests itself in the confidence we have in these beliefs, and helps us guide our behavior in complex and uncertain environments. Here, we provide empirical tests of the importance of...\", verbose=True))\n",
    "# print(on_topic(\"The serotonin transporter (5-HTT) regulates serotonergic neurotransmission and is thought to influence emotion. A 5-HTT-linked polymorphic region (5-HTTLPR) has two common variants, short (s) and long (l). We previously found population and within-family associations between the lower-expressing s allele and neuroticism, a trait related to anxiety, hostility, and depression, on a standard measure (the NEO Personality Inventory, Revised [NEO-PI-R]) in a primarily male population (n=505), and that the s allele was dominant. We investigated this association in a new sample (n=397, 84% female, primarily sib-pairs). The results robustly replicated the 5-HTTLPR neuroticism association, and the dominance of the s allele. Combined data from the two studies (n=902) showed a highly significant association between the s allele and higher NEO Neuroticism both across individuals and within families. Association between genotype and a related measure, Anxiety on the 16PF inventory, was replicated in the new population and within families in the combined sample. Association to another trait, estimated TPQ Harm Avoidance, was not replicated in the new sample but found only within the combined sibship group. Another association found in our original study, between the s allele and lower scores on NEO-PI-R Agreeableness, was also replicated and was more robust in the current and the combined samples. Associations between the functional 5-HTTLPR polymorphism were similar in women and men. These results help to define specific personality features reproducibly associated with 5-HTTLPR genotype. Such associations were strongest for traits defined by the NEO, enhancing the attractiveness of the five-factor personality model in genetic research on complex behavioral dimensions. Am. J. Med. Genet. (Neuropsychiatr. Genet.) 96:202â€“216, 2000. Published 2000 Wiley-Liss, Inc.\", verbose=True))\n",
    "# print(on_topic(\"This investigation examines the role of trait-mediated indirect interactions in a simple aquatic food web. We conducted the experiments in cattle watering tanks in order to establish whether competitive and predatorâ€“prey interactions between two species are affected by other species in the system; i.e., are pairwise interaction strengths affected by the background species assemblage? We examined the survival and growth response of small bullfrog (Rana catesbeiana) and small green frog (Rana clamitans) tadpoles in the presence and absence of a competitor (large bullfrogs), the lethal presence of the larval odonate predator Tramea lacerata, and the nonlethal (caged) presence of the larval odonate predators Anax junius and Anax longipes. We demonstrate that large bullfrog competitors and caged Anax affect traits (foraging activity level) of small bullfrog and small green frog tadpoles and that these changes in traits, in turn, affect interactions of the small tadpole species with each other and with the other species. In particular, the following four trait-mediated indirect interactions were evident: (1) Presence of large bullfrog competitors increased the predation rate of Tramea on small green frogs and small bullfrogs. (2) Presence of nonlethal Anax reduced the predation rate of Tramea on small green frogs. (3) Presence of nonlethal Anax increased the competitive advantage of bullfrogs over green frogs. (4) Presence of nonlethal Anax facilitated midge invasion of the experimental units. The proposed mechanisms (changes in small tadpole activity) involved in these trait-mediated indirect interactions were supported by observational data on tadpole activity and resource levels in the experimental units, and in laboratory experiments examining tadpole activity responses to predators. The occurrence of strong trait-mediated indirect interactions in this simple food web underscores the potential importance of such interactions in animal communities.\", verbose=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d0022d59-5c8d-4014-9dd8-d7bfbc50d2ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = [\"Title\", \"Abstract\", \"DOI\", \"Score\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5720c824-538a-4ae6-bc16-d3c501286300",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Examples (of TMII Papers)\n",
    "# The abstracts of the examples are only being used.\n",
    "examples = [\n",
    "    {\n",
    "        \"Title\": \"Dynamic population stage structure due to juvenile–adult asymmetry stabilizes complex ecological communities\",\n",
    "        \"Abstract\": clean_text('''\n",
    "            Using food web models that account for juvenile and adult\n",
    "            individuals of species, I show that commonly observed differences between juveniles and adults in foraging capacity\n",
    "            and predation risk result in larger, more complex communities than predicted by models without stage structure.\n",
    "            Based on their species interaction networks these complex\n",
    "            and diverse communities would be expected to be unstable, but these destabilizing effects of species interactions are\n",
    "            overruled by stabilizing changes in juvenile–adult stage structure. Differences between juvenile and adult individuals hence\n",
    "            offer a natural resolution to the diversity–stability enigma of\n",
    "            ecological communities.\n",
    "        '''),\n",
    "        \"DOI\": \"\",\n",
    "        \"Score\": 0\n",
    "    },\n",
    "    {\n",
    "        \"Title\": \"EFFECTS OF TOP PREDATOR SPECIES ON DIRECT AND INDIRECT INTERACTIONS IN A FOOD WEB\",\n",
    "        \"Abstract\": clean_text('''\n",
    "            Current theory on trophic interactions in food webs assumes thatecologically\n",
    "            similar species can be treated collectively as a single functional unit such as a guild or\n",
    "            trophic level. This theory implies that all species within that unit transmit identical direct\n",
    "            and indirect effects throughout the community. We evaluated this assumption by conducting\n",
    "            experiments to compare the direct and indirect effects of three top-predator species, be\n",
    "            longing to the same hunting spider guild, on the same species of grasshopper and on old\n",
    "            field grasses and herbs. Observations under field conditions revealed that each spiderspecies\n",
    "            exhibited different hunting behavior (i.e., sit-and-wait, sit-and-pursue, and active hunting)\n",
    "            and occupied different locations within the vegetation canopy. These differences resulted\n",
    "            in different direct effects on grasshopper prey. Grasshoppers demonstrated significant be\n",
    "            havioral (diet) shifts in the presence of sit-and-wait and sit-and-pursue species but not when\n",
    "            faced with actively hunting species. Grasshopper density was significantly reduced byspider\n",
    "            species that occupied lower parts of the vegetation canopy (sit-and-pursue and actively\n",
    "            hunting species), but it was not significantly reduced by the sit-and-wait spider species that\n",
    "            occupied the upper parts of the canopy. These direct effects manifested themselves differ\n",
    "            ently in the plant trophic level. The sit-and-wait spider caused indirect effects on plants\n",
    "            by changing grasshopper foraging behavior (a trait-mediated effect). The sit-and-pursue\n",
    "            spider caused indirect effects by reducing grasshopper density (density-mediated effects);\n",
    "            the effects of changes in grasshopper behavior were thus not reflected in the plant trophic\n",
    "            level. The actively hunting spiders had strictly density-mediated indirect effects on plants.\n",
    "            The study offers mechanistic insight into how predator species within the same guild can\n",
    "            have very different trophic effects in food webs. Thus classical modeling approaches that\n",
    "            treat all predator species as a single functional unit may not adequately capture biologically\n",
    "            relevant details that influence community dynamics.\n",
    "        '''),\n",
    "        \"DOI\": \"\",\n",
    "        \"Score\": 0\n",
    "    },\n",
    "    {\n",
    "        \"Title\": \"Higher-Order Interaction between Species Inhibits Bacterial Invasion of a Phototroph-Predator Microbial Community\",\n",
    "        \"Abstract\": clean_text('''\n",
    "            The composition of an ecosystem is thought to be\n",
    "            important for determining its resistance to invasion.\n",
    "            Studies of natural ecosystems, from plant to microbial communities, have found that more diverse communities are more resistant to invasion. In some\n",
    "            cases, more diverse communities resist invasion by more completely consuming the resources necessary for the invader. We show that Escherichia\n",
    "            coli can successfully invade cultures of the alga Chlamydomonas reinhardtii (phototroph) or the ciliate\n",
    "            Tetrahymena thermophila (predator) but cannot invade a community where both are present. The invasion resistance of the algae-ciliate community\n",
    "            arises from a higher-order interaction between species (interaction modification) that is unrelated to resource consumption. We show that the mode of\n",
    "            this interaction is the algal inhibition of bacterial aggregation, which leaves bacteria vulnerable to predation. This mode requires both the algae and the ciliate\n",
    "            to be present and provides an example of invasion resistance through an interaction modification.\n",
    "        '''),\n",
    "        \"DOI\": \"\",\n",
    "        \"Score\": 0\n",
    "    },\n",
    "    {\n",
    "        \"Title\": \"Multiple predator effects result in risk reduction for prey across multiple prey densities\",\n",
    "        \"Abstract\": clean_text('''\n",
    "            Investigating how prey density influences a\n",
    "            prey’s combined predation risk from multiple predator\n",
    "            species is critical for understanding the widespread\n",
    "            importance of multiple predator effects. We conducted\n",
    "            experiments that crossed six treatments consisting of\n",
    "            zero, one, or two predator species (hellgrammites,\n",
    "            greenside darters, and creek chubs) with three treat\n",
    "            ments in which we varied the density of mayfly prey.\n",
    "            None of the multiple predator effects in our system were\n",
    "            independent, and instead, the presence of multiple\n",
    "            predator species resulted in risk reduction for the prey\n",
    "            across both multiple predator combinations and all\n",
    "            three levels of prey density. Risk reduction is likely to\n",
    "            have population-level consequences for the prey,\n",
    "            resulting in larger prey populations than would be pre\n",
    "            dicted if the effects of multiple predator species were\n",
    "            independent. For one of the two multiple predator\n",
    "            combinations, the magnitude of risk reduction margin\n",
    "            ally increased with prey density. As a result, models\n",
    "            predicting the combined risk from multiple predator\n",
    "            species in this system will sometimes need to account for\n",
    "            prey density as a factor influencing per-capita prey death\n",
    "            rates.\n",
    "        '''),\n",
    "        \"DOI\": \"\",\n",
    "        \"Score\": 0\n",
    "    },\n",
    "]\n",
    "\n",
    "# # Create Dataset\n",
    "# examples_df = pd.DataFrame([], columns=columns)\n",
    "# for example in examples:\n",
    "#     example = pd.DataFrame([[example[\"Title\"], example[\"Abstract\"], example[\"DOI\"], example[\"Score\"]]],  columns=columns)\n",
    "#     examples_df = pd.concat([examples_df, example])\n",
    "# examples_df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "# # Store Dataset (CSV and Pickle)\n",
    "# save_df(examples_df, \"Examples\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5d238312-9e39-4484-8782-9b1b20ae33bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Broken Keywords: ['mutual', 'trophic', 'intransitive', 'competition', 'keystone', 'predation', 'intraguild', 'resource', 'apparent', 'chain']\n"
     ]
    }
   ],
   "source": [
    "# Level 1 Keywords\n",
    "keywords_1 = [\n",
    "    \"trait\",\n",
    "    \"phenotype\"\n",
    "]\n",
    "\n",
    "# Level 2 Keywords\n",
    "keywords_2 = [\n",
    "    \"trait-mediated\",\n",
    "    \"higher-order interaction\",\n",
    "    \"polymorphism\",\n",
    "    \"interaction modification\",\n",
    "    \"indirect effect\"\n",
    "]\n",
    "\n",
    "# Level 3 Keywords\n",
    "keywords_3 = [\n",
    "    \"apparent competition\",\n",
    "    \"resource competition\",\n",
    "    \"keystone predation\",\n",
    "    \"intraguild predation\",\n",
    "    \"intransitive competition\",\n",
    "    \"trophic chain\",\n",
    "    \"competition chain\",\n",
    "    \"mutual competition\"\n",
    "]\n",
    "\n",
    "# Broken Level 3 Keywords\n",
    "# These are the keywords of before but broken down, so it's\n",
    "# like \"apparent\" and \"competition\" instead of \"apparent competition\".\n",
    "broken_keywords_3 = list(set(\" \".join(keywords_3).split(\" \")))\n",
    "print(f\"Broken Keywords: {broken_keywords_3}\")\n",
    "\n",
    "search_1 = \"|\".join(keywords_1).replace(\" \", \"%20\")\n",
    "search_2 = \"|\".join(keywords_2).replace(\" \", \"%20\")\n",
    "search_3 = \"|\".join(keywords_3).replace(\" \", \"%20\")\n",
    "search_3_broken = \"|\".join(broken_keywords_3).replace(\" \", \"%20\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ce494579-f0bc-4bb4-8368-6ef163a277f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def revert_abstract(inverted_abstract):\n",
    "    if not inverted_abstract:\n",
    "        return \"\"\n",
    "    abstract = \"\"\n",
    "    i = 0\n",
    "    while True:\n",
    "        index_found = False\n",
    "        for k, v in inverted_abstract.items():\n",
    "            # print(k, v)\n",
    "            if i in v:\n",
    "                if abstract and k not in [\".\", \",\", \"?\", \"!\"]:\n",
    "                    abstract += \" \"\n",
    "                abstract += k\n",
    "                i += 1\n",
    "                index_found = True\n",
    "        if not index_found:\n",
    "            break\n",
    "    return abstract\n",
    "\n",
    "def search_papers(searches, verbose=False):\n",
    "    page = 1\n",
    "    search_filter = ','.join([f'title.search:{search}' for search in searches])\n",
    "    url = f\"https://api.openalex.org/works?page={page}&filter={search_filter}\"\n",
    "    if verbose:\n",
    "        print(f\"URL: {url}\")\n",
    "\n",
    "    works = []\n",
    "    data = requests.get(url).json()\n",
    "    data = [] if 'results' not in data.keys() else data['results']\n",
    "    \n",
    "    while len(data) > 0:\n",
    "        works += data\n",
    "        page += 1\n",
    "        url = f\"https://api.openalex.org/works?page={page}&filter={search_filter}\"\n",
    "        if verbose:\n",
    "            print(f\"URL: {url}\")\n",
    "        data = requests.get(url).json()\n",
    "        if 'results' in data.keys():\n",
    "            data = data['results']\n",
    "        else:\n",
    "            break\n",
    "\n",
    "    # Print Size and 1st 10 Titles\n",
    "    if verbose:\n",
    "        num_works = len(works)\n",
    "        print(f\"Number Works: {num_works}\")\n",
    "        for i in range(0, min(10, num_works)):\n",
    "            print(works[i][\"title\"])\n",
    "\n",
    "    # Create Dataset\n",
    "    df = pd.DataFrame([], columns=columns)\n",
    "    for work in works:\n",
    "        work = pd.DataFrame([[work[\"title\"], revert_abstract(work[\"abstract_inverted_index\"]), work[\"doi\"], 0]],  columns=columns)\n",
    "        df = pd.concat([df, work])\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    return df\n",
    "\n",
    "def on_topic_filter(row):\n",
    "    abstract = row['Abstract']\n",
    "    if not abstract:\n",
    "        return False\n",
    "    torch.cuda.empty_cache()\n",
    "    return on_topic(abstract)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2ce26d93-08b8-4721-bcb9-b941463121e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.\n",
      "You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "\n",
    "# Dataset A\n",
    "# df_a = search_papers([search_1], verbose=False)\n",
    "# save_df(df_a, \"A\")\n",
    "\n",
    "# Dataset B\n",
    "# df_b = search_papers([search_1, search_2], verbose=False)\n",
    "# save_df(df_b, \"B\")\n",
    "\n",
    "# Dataset B + Filtered\n",
    "# This portion is unable to run because my memory keeps running out.\n",
    "# I'll try it out on HiPerGator as soon as that works.\n",
    "# df_b_filtered = df_b[df_b.apply(on_topic_filter, axis=1)]\n",
    "# save_df(df_b_filtered, \"BFiltered\")\n",
    "\n",
    "# Dataset C\n",
    "# df_c = search_papers([search_1, search_2, search_3], verbose=False)\n",
    "# save_df(df_c, \"C\")\n",
    "\n",
    "# Dataset C + Filtered\n",
    "# df_c_filtered = df_c[df_c.apply(on_topic_filter, axis=1)]\n",
    "# save_df(df_c_filtered, \"CFiltered\")\n",
    "\n",
    "# Dataset D\n",
    "# df_d = search_papers([search_1, search_2, search_3_broken], verbose=False)\n",
    "# save_df(df_d, \"D\")\n",
    "\n",
    "# Dataset D + Filtered\n",
    "# df_d_filtered = df_d[df_d.apply(on_topic_filter, axis=1)]\n",
    "# save_df(df_d_filtered, \"DFiltered\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "371a0250-168d-4a61-9c44-08cf8c91fbfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A and B Subsets\n",
    "# Datasets A and B are too large (so it takes too long to run).\n",
    "# I'll use a subset of each.\n",
    "# The filtering function should be able to run now.\n",
    "\n",
    "# Subset of Dataset A\n",
    "# sub_df_a = sample_df(\"A\", n=150, save=True)\n",
    "\n",
    "# # Subset of Dataset A + Filtered\n",
    "# sub_df_a_filtered = sub_df_a[sub_df_a.apply(on_topic_filter, axis=1)]\n",
    "# save_df(sub_df_a_filtered, \"SubAFiltered\")\n",
    "\n",
    "# # Subset of Dataset B\n",
    "# sub_df_b = sample_df(\"B\", n=150, save=True)\n",
    "\n",
    "# # Subset of Dataset B + Filtered\n",
    "# sub_df_b_filtered = sub_df_b[sub_df_b.apply(on_topic_filter, axis=1)]\n",
    "# save_df(sub_df_b_filtered, \"SubBFiltered\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
