{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ef38fc58-93aa-4c9d-b492-4a395a687b98",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\lbeln\\anaconda3\\envs\\3.10\\lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import time\n",
    "import pandas as pd\n",
    "from pyalex import Works\n",
    "from itertools import chain\n",
    "from transformers import pipeline\n",
    "from beautifultable import BeautifulTable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e5e7e782-eec8-40dd-8b0e-555507606245",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From C:\\Users\\lbeln\\anaconda3\\envs\\3.10\\lib\\site-packages\\tf_keras\\src\\losses.py:2976: The name tf.losses.sparse_softmax_cross_entropy is deprecated. Please use tf.compat.v1.losses.sparse_softmax_cross_entropy instead.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Device set to use cuda:0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mCreate or Load Labeled Dataset\u001b[0m\n",
      "Start by entering the filepath of the labeled dataset to create or add onto.\n",
      "Ensure that the filepath entered is for a .csv file.\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Filename of Dataset:  Yippee.csv\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading File: 'Yippee.csv'\n",
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mFind Papers to Score\u001b[0m\n",
      "Now, we need to find unscored papers. Enter [1] to view a list of datasets that contain unscored papers or [2] to look for papers on OpenAlex (requires an Internet connection).\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[1] Load Unscored Dataset\n",
      "[2] Search for Unscored Papers\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mSearch Unscored Papers\u001b[0m\n",
      "Enter a search query (i.e. 'intraguild predation') to load papers from OpenAlex or a letter that corresponds to an action (i.e. 'Go Back' and 'Exit Program'). If you want to search 'A', you should instead input 'a', or else you will end up going back a page; the same goes for 'B' and 'b'.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  trait-mediated interaction modification\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number Results: 1000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.\n",
      "You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset\n",
      "C:\\Users\\lbeln\\anaconda3\\envs\\3.10\\lib\\site-packages\\beautifultable\\utils.py:125: FutureWarning: 'BeautifulTable.column_headers' has been deprecated in 'v1.0.0' and will be removed in 'v1.2.0'. Use 'BTColumnCollection.header' instead.\n",
      "  warnings.warn(message, FutureWarning)\n",
      "C:\\Users\\lbeln\\anaconda3\\envs\\3.10\\lib\\site-packages\\beautifultable\\utils.py:125: FutureWarning: 'BeautifulTable.append_row' has been deprecated in 'v1.0.0' and will be removed in 'v1.2.0'. Use 'BTRowCollection.append' instead.\n",
      "  warnings.warn(message, FutureWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Out of Memory. Ending Search\n",
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mPreview Dataset\u001b[0m\n",
      "Before scoring, you can check whether your chosen dataset looks correct. Enter the [1] to continue; [2] to go to the previous page of papers; [3] to go to the next page of papers; or a letter that corresponds to an action (i.e. [A] for 'Go Back' and [B] for 'Exit Program').\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[1] Continue\n",
      "[2] Previous Page\n",
      "[3] Next Page\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "1/1\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "|        Title         |             Abstract              |    DOI    | Score |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| The contribution of  | Many prey modify traits in respon | https://d |   0   |\n",
      "| trait-mediated indir | se to predation risk and this mod | oi.org/10 |       |\n",
      "| ect effects to the n | ification of traits can influence | .1073/pna |       |\n",
      "| et effects of a pred |  the prey's resource acquisition  | s.0710619 |       |\n",
      "|         ator         |       rate. A predator t...       |    98     |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| HABITAT COMPLEXITY D | Despite recognition of the signif | https://d |   0   |\n",
      "| ISRUPTS PREDATOR–PRE | icance of both food web interacti | oi.org/10 |       |\n",
      "| Y INTERACTIONS BUT N | ons and habitat complexity in com | .1890/03- |       |\n",
      "| OT THE TROPHIC CASCA | munity dynamics, current ecologic |   0067    |       |\n",
      "|  DE ON OYSTER REEFS  |       al theory rarely c...       |           |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| Do plant‐ and soil‐m | Summary Biotic interactions of pl | https://d |   0   |\n",
      "| ediated legacy effec | ants can have legacy effects medi | oi.org/10 |       |\n",
      "| ts impact future bio | ated by changes in plant traits a | .1111/136 |       |\n",
      "|  tic interactions?   | nd soil characteristics. Legacy e | 5-2435.12 |       |\n",
      "|                      |       ffects are defined...       |    456    |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| HOW DEPENDENT ARE SP | In ecological theory species inte | https://d |   0   |\n",
      "| ECIES-PAIR INTERACTI | raction strengths are typically d | oi.org/10 |       |\n",
      "| ON STRENGTHS ON OTHE | escribed by constants or function | .1890/03- |       |\n",
      "| R SPECIES IN THE FOO | s that depend on the densities of |   0662    |       |\n",
      "|        D WEB?        |        the two interacti...       |           |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| A twin approach to u | The regulation of gene expression | https://d |   0   |\n",
      "| nraveling epigenetic |  plays a pivotal role in complex  | oi.org/10 |       |\n",
      "|          s           | phenotypes, and epigenetic mechan | .1016/j.t |       |\n",
      "|                      | isms such as DNA methylation are  | ig.2010.1 |       |\n",
      "|                      |       essential to this ...       |   2.005   |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  34\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[91mMust Enter 1, 2, 3, 'A', or 'B'. Try Again.\u001b[0m\n",
      "\n",
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mPreview Dataset\u001b[0m\n",
      "Before scoring, you can check whether your chosen dataset looks correct. Enter the [1] to continue; [2] to go to the previous page of papers; [3] to go to the next page of papers; or a letter that corresponds to an action (i.e. [A] for 'Go Back' and [B] for 'Exit Program').\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[1] Continue\n",
      "[2] Previous Page\n",
      "[3] Next Page\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "1/1\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "|        Title         |             Abstract              |    DOI    | Score |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| The contribution of  | Many prey modify traits in respon | https://d |   0   |\n",
      "| trait-mediated indir | se to predation risk and this mod | oi.org/10 |       |\n",
      "| ect effects to the n | ification of traits can influence | .1073/pna |       |\n",
      "| et effects of a pred |  the prey's resource acquisition  | s.0710619 |       |\n",
      "|         ator         |       rate. A predator t...       |    98     |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| HABITAT COMPLEXITY D | Despite recognition of the signif | https://d |   0   |\n",
      "| ISRUPTS PREDATOR–PRE | icance of both food web interacti | oi.org/10 |       |\n",
      "| Y INTERACTIONS BUT N | ons and habitat complexity in com | .1890/03- |       |\n",
      "| OT THE TROPHIC CASCA | munity dynamics, current ecologic |   0067    |       |\n",
      "|  DE ON OYSTER REEFS  |       al theory rarely c...       |           |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| Do plant‐ and soil‐m | Summary Biotic interactions of pl | https://d |   0   |\n",
      "| ediated legacy effec | ants can have legacy effects medi | oi.org/10 |       |\n",
      "| ts impact future bio | ated by changes in plant traits a | .1111/136 |       |\n",
      "|  tic interactions?   | nd soil characteristics. Legacy e | 5-2435.12 |       |\n",
      "|                      |       ffects are defined...       |    456    |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| HOW DEPENDENT ARE SP | In ecological theory species inte | https://d |   0   |\n",
      "| ECIES-PAIR INTERACTI | raction strengths are typically d | oi.org/10 |       |\n",
      "| ON STRENGTHS ON OTHE | escribed by constants or function | .1890/03- |       |\n",
      "| R SPECIES IN THE FOO | s that depend on the densities of |   0662    |       |\n",
      "|        D WEB?        |        the two interacti...       |           |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| A twin approach to u | The regulation of gene expression | https://d |   0   |\n",
      "| nraveling epigenetic |  plays a pivotal role in complex  | oi.org/10 |       |\n",
      "|          s           | phenotypes, and epigenetic mechan | .1016/j.t |       |\n",
      "|                      | isms such as DNA methylation are  | ig.2010.1 |       |\n",
      "|                      |       essential to this ...       |   2.005   |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mPreview Dataset\u001b[0m\n",
      "Before scoring, you can check whether your chosen dataset looks correct. Enter the [1] to continue; [2] to go to the previous page of papers; [3] to go to the next page of papers; or a letter that corresponds to an action (i.e. [A] for 'Go Back' and [B] for 'Exit Program').\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[1] Continue\n",
      "[2] Previous Page\n",
      "[3] Next Page\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "1/1\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "|        Title         |             Abstract              |    DOI    | Score |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| The contribution of  | Many prey modify traits in respon | https://d |   0   |\n",
      "| trait-mediated indir | se to predation risk and this mod | oi.org/10 |       |\n",
      "| ect effects to the n | ification of traits can influence | .1073/pna |       |\n",
      "| et effects of a pred |  the prey's resource acquisition  | s.0710619 |       |\n",
      "|         ator         |       rate. A predator t...       |    98     |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| HABITAT COMPLEXITY D | Despite recognition of the signif | https://d |   0   |\n",
      "| ISRUPTS PREDATOR–PRE | icance of both food web interacti | oi.org/10 |       |\n",
      "| Y INTERACTIONS BUT N | ons and habitat complexity in com | .1890/03- |       |\n",
      "| OT THE TROPHIC CASCA | munity dynamics, current ecologic |   0067    |       |\n",
      "|  DE ON OYSTER REEFS  |       al theory rarely c...       |           |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| Do plant‐ and soil‐m | Summary Biotic interactions of pl | https://d |   0   |\n",
      "| ediated legacy effec | ants can have legacy effects medi | oi.org/10 |       |\n",
      "| ts impact future bio | ated by changes in plant traits a | .1111/136 |       |\n",
      "|  tic interactions?   | nd soil characteristics. Legacy e | 5-2435.12 |       |\n",
      "|                      |       ffects are defined...       |    456    |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| HOW DEPENDENT ARE SP | In ecological theory species inte | https://d |   0   |\n",
      "| ECIES-PAIR INTERACTI | raction strengths are typically d | oi.org/10 |       |\n",
      "| ON STRENGTHS ON OTHE | escribed by constants or function | .1890/03- |       |\n",
      "| R SPECIES IN THE FOO | s that depend on the densities of |   0662    |       |\n",
      "|        D WEB?        |        the two interacti...       |           |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| A twin approach to u | The regulation of gene expression | https://d |   0   |\n",
      "| nraveling epigenetic |  plays a pivotal role in complex  | oi.org/10 |       |\n",
      "|          s           | phenotypes, and epigenetic mechan | .1016/j.t |       |\n",
      "|                      | isms such as DNA methylation are  | ig.2010.1 |       |\n",
      "|                      |       essential to this ...       |   2.005   |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mPreview Dataset\u001b[0m\n",
      "Before scoring, you can check whether your chosen dataset looks correct. Enter the [1] to continue; [2] to go to the previous page of papers; [3] to go to the next page of papers; or a letter that corresponds to an action (i.e. [A] for 'Go Back' and [B] for 'Exit Program').\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[1] Continue\n",
      "[2] Previous Page\n",
      "[3] Next Page\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "1/1\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "|        Title         |             Abstract              |    DOI    | Score |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| The contribution of  | Many prey modify traits in respon | https://d |   0   |\n",
      "| trait-mediated indir | se to predation risk and this mod | oi.org/10 |       |\n",
      "| ect effects to the n | ification of traits can influence | .1073/pna |       |\n",
      "| et effects of a pred |  the prey's resource acquisition  | s.0710619 |       |\n",
      "|         ator         |       rate. A predator t...       |    98     |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| HABITAT COMPLEXITY D | Despite recognition of the signif | https://d |   0   |\n",
      "| ISRUPTS PREDATOR–PRE | icance of both food web interacti | oi.org/10 |       |\n",
      "| Y INTERACTIONS BUT N | ons and habitat complexity in com | .1890/03- |       |\n",
      "| OT THE TROPHIC CASCA | munity dynamics, current ecologic |   0067    |       |\n",
      "|  DE ON OYSTER REEFS  |       al theory rarely c...       |           |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| Do plant‐ and soil‐m | Summary Biotic interactions of pl | https://d |   0   |\n",
      "| ediated legacy effec | ants can have legacy effects medi | oi.org/10 |       |\n",
      "| ts impact future bio | ated by changes in plant traits a | .1111/136 |       |\n",
      "|  tic interactions?   | nd soil characteristics. Legacy e | 5-2435.12 |       |\n",
      "|                      |       ffects are defined...       |    456    |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| HOW DEPENDENT ARE SP | In ecological theory species inte | https://d |   0   |\n",
      "| ECIES-PAIR INTERACTI | raction strengths are typically d | oi.org/10 |       |\n",
      "| ON STRENGTHS ON OTHE | escribed by constants or function | .1890/03- |       |\n",
      "| R SPECIES IN THE FOO | s that depend on the densities of |   0662    |       |\n",
      "|        D WEB?        |        the two interacti...       |           |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| A twin approach to u | The regulation of gene expression | https://d |   0   |\n",
      "| nraveling epigenetic |  plays a pivotal role in complex  | oi.org/10 |       |\n",
      "|          s           | phenotypes, and epigenetic mechan | .1016/j.t |       |\n",
      "|                      | isms such as DNA methylation are  | ig.2010.1 |       |\n",
      "|                      |       essential to this ...       |   2.005   |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mScore Papers\u001b[0m\n",
      "For each paper, enter an integer between 0 and 3. 0 indicates a lower quality and 3 indicates a higher quality. If the paper has already been scored (you should see a 'Current Score') or if you would just like to skip a paper, press the enter key without any input. Should you want to exit or go back (i.e. [A] for 'Go Back' and [B] for 'Exit Program'), your scores will automatically be saved.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[C] Exit Scoring\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "0/5\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mTitle:\u001b[0m The contribution of trait-mediated indirect effects to the net effects of a predator\n",
      "\u001b[1mAbstract:\u001b[0m Many prey modify traits in response to predation risk and this modification of traits can influence the prey's resource acquisition rate. A predator thus can have a “nonlethal” impact on prey that can lead to indirect effects on other community members. Such indirect interactions are termed trait-mediated indirect interactions because they arise from a predator's influence on prey traits, rather than prey density. Because such nonlethal predator effects are immediate, can influence the entire prey population, and can occur over the entire prey lifetime, we argue that nonlethal predator effects are likely to contribute strongly to the net indirect effects of predators (i.e., nonlethal effects may be comparable in magnitude to those resulting from killing prey). This prediction was supported by an experiment in which the indirect effects of a larval dragonfly ( Anax sp.) predator on large bullfrog tadpoles ( Rana catesbeiana ), through nonlethal effects on competing small bullfrog tadpoles, were large relative to indirect effects caused by density reduction of the small tadpoles (the lethal effect). Treatments in which lethal and nonlethal effects of Anax were manipulated independently indicated that this result was robust for a large range of different combinations of lethal and nonlethal effects. Because many, if not most, prey modify traits in response to predators, our results suggest that the magnitude of interaction coefficients between two species may often be dynamically related to changes in other community members, and that many indirect effects previously attributed to the lethal effects of predators may instead be due to shifts in traits of surviving prey.\n",
      "\u001b[1mDOI:\u001b[0m https://doi.org/10.1073/pnas.071061998\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mScore Papers\u001b[0m\n",
      "For each paper, enter an integer between 0 and 3. 0 indicates a lower quality and 3 indicates a higher quality. If the paper has already been scored (you should see a 'Current Score') or if you would just like to skip a paper, press the enter key without any input. Should you want to exit or go back (i.e. [A] for 'Go Back' and [B] for 'Exit Program'), your scores will automatically be saved.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[C] Exit Scoring\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "1/5\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mTitle:\u001b[0m HABITAT COMPLEXITY DISRUPTS PREDATOR–PREY INTERACTIONS BUT NOT THE TROPHIC CASCADE ON OYSTER REEFS\n",
      "\u001b[1mAbstract:\u001b[0m Despite recognition of the significance of both food web interactions and habitat complexity in community dynamics, current ecological theory rarely couples these two processes. Experimental manipulations of the abundance of the two predators in an oyster-reef trophic cascade, and the structural complexity provided by reefs of living oysters, demonstrated that enhanced habitat complexity weakened the strengths of trophic interactions. The system of tri-trophic interactions included oyster toadfish (Opsanus tau) as the top predator that consumed the mud crab (Panopeus herbstii), which preys upon juvenile oysters (Crassostrea virginica). On reefs of low complexity, toadfish controlled mud crab abundances and indirectly determined the level of mortality of juvenile oysters. The indirect effects of toadfish on oysters emerged through their influence on how intensely mud crabs preyed on oysters. Augmentation of habitat complexity by substituting vertically oriented, living oysters for the flat shells of dead oysters disrupted both of the direct trophic linkages but did not alter the magnitude of the indirect effect of toadfish on juvenile oysters. This paradox can be understood by partitioning the mechanisms by which toadfish influence mud crabs and ultimately juvenile oysters. Trait-mediated indirect interactions (TMIIs; i.e., predator-avoidance behavior in mud crabs) accounted for 95.6–98.2% of toadfish indirect benefits to oyster survival and, consequently, were a much greater contributor than density-mediated indirect interactions (DMIIs; i.e., the reduction in crab abundance by toadfish). Avoidance behavior was unaffected by modification in habitat complexity. Complex reefs increased total oyster survival because added habitat complexity reduced mud crab predation on oysters. Additionally, the magnitude of this effect was much greater than the increase in oyster mortality as a result of complex reefs disrupting toadfish predation on mud crabs. This experimental demonstration of how habitat complexity modifies trophic interactions in a temperate reef community has fundamental implications for our understanding of species interaction webs and community structure. The influence of habitat complexity on the strength of a trophic cascade generally may depend upon whether physical complexity provides actual and perceived refuges for component predator–prey pairs.\n",
      "\u001b[1mDOI:\u001b[0m https://doi.org/10.1890/03-0067\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mScore Papers\u001b[0m\n",
      "For each paper, enter an integer between 0 and 3. 0 indicates a lower quality and 3 indicates a higher quality. If the paper has already been scored (you should see a 'Current Score') or if you would just like to skip a paper, press the enter key without any input. Should you want to exit or go back (i.e. [A] for 'Go Back' and [B] for 'Exit Program'), your scores will automatically be saved.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[C] Exit Scoring\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "2/5\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mTitle:\u001b[0m Do plant‐ and soil‐mediated legacy effects impact future biotic interactions?\n",
      "\u001b[1mAbstract:\u001b[0m Summary Biotic interactions of plants can have legacy effects mediated by changes in plant traits and soil characteristics. Legacy effects are defined as effects that persist after the causal biotic interaction ceases, because plant traits and soil characteristics are plastic and modifications may be maintained for long time periods. To date, studies on biotic interactions of plants with above‐ and below‐ground organisms and their plant‐mediated interactions have mainly focused on short‐term effects in the continuous presence of interacting organisms. Little is known about how long the changes in plant traits and soil characteristics persist after the biotic interaction ceases and indirectly affect future biotic interactions and community assembly. Further, since some organisms switch between above‐ and below‐ground compartments during ontogeny, impacts on soil‐dwelling larval stages can have direct legacy effects on their above‐ground adult stages and vice versa. Elucidating legacy effects of biotic interactions above and below the ground will provide a deeper insight into a better understanding of the importance of plant trait‐ and soil characteristic‐mediated indirect interactions in shaping community assembly, biodiversity and ecosystem function.\n",
      "\u001b[1mDOI:\u001b[0m https://doi.org/10.1111/1365-2435.12456\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  A\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mPreview Dataset\u001b[0m\n",
      "Before scoring, you can check whether your chosen dataset looks correct. Enter the [1] to continue; [2] to go to the previous page of papers; [3] to go to the next page of papers; or a letter that corresponds to an action (i.e. [A] for 'Go Back' and [B] for 'Exit Program').\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[1] Continue\n",
      "[2] Previous Page\n",
      "[3] Next Page\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "1/1\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "|        Title         |             Abstract              |    DOI    | Score |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| The contribution of  | Many prey modify traits in respon | https://d |   0   |\n",
      "| trait-mediated indir | se to predation risk and this mod | oi.org/10 |       |\n",
      "| ect effects to the n | ification of traits can influence | .1073/pna |       |\n",
      "| et effects of a pred |  the prey's resource acquisition  | s.0710619 |       |\n",
      "|         ator         |       rate. A predator t...       |    98     |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| HABITAT COMPLEXITY D | Despite recognition of the signif | https://d |   0   |\n",
      "| ISRUPTS PREDATOR–PRE | icance of both food web interacti | oi.org/10 |       |\n",
      "| Y INTERACTIONS BUT N | ons and habitat complexity in com | .1890/03- |       |\n",
      "| OT THE TROPHIC CASCA | munity dynamics, current ecologic |   0067    |       |\n",
      "|  DE ON OYSTER REEFS  |       al theory rarely c...       |           |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| Do plant‐ and soil‐m | Summary Biotic interactions of pl | https://d |   0   |\n",
      "| ediated legacy effec | ants can have legacy effects medi | oi.org/10 |       |\n",
      "| ts impact future bio | ated by changes in plant traits a | .1111/136 |       |\n",
      "|  tic interactions?   | nd soil characteristics. Legacy e | 5-2435.12 |       |\n",
      "|                      |       ffects are defined...       |    456    |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| HOW DEPENDENT ARE SP | In ecological theory species inte | https://d |   0   |\n",
      "| ECIES-PAIR INTERACTI | raction strengths are typically d | oi.org/10 |       |\n",
      "| ON STRENGTHS ON OTHE | escribed by constants or function | .1890/03- |       |\n",
      "| R SPECIES IN THE FOO | s that depend on the densities of |   0662    |       |\n",
      "|        D WEB?        |        the two interacti...       |           |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| A twin approach to u | The regulation of gene expression | https://d |   0   |\n",
      "| nraveling epigenetic |  plays a pivotal role in complex  | oi.org/10 |       |\n",
      "|          s           | phenotypes, and epigenetic mechan | .1016/j.t |       |\n",
      "|                      | isms such as DNA methylation are  | ig.2010.1 |       |\n",
      "|                      |       essential to this ...       |   2.005   |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mScore Papers\u001b[0m\n",
      "For each paper, enter an integer between 0 and 3. 0 indicates a lower quality and 3 indicates a higher quality. If the paper has already been scored (you should see a 'Current Score') or if you would just like to skip a paper, press the enter key without any input. Should you want to exit or go back (i.e. [A] for 'Go Back' and [B] for 'Exit Program'), your scores will automatically be saved.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[C] Exit Scoring\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "0/5\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mTitle:\u001b[0m The contribution of trait-mediated indirect effects to the net effects of a predator\n",
      "\u001b[1mAbstract:\u001b[0m Many prey modify traits in response to predation risk and this modification of traits can influence the prey's resource acquisition rate. A predator thus can have a “nonlethal” impact on prey that can lead to indirect effects on other community members. Such indirect interactions are termed trait-mediated indirect interactions because they arise from a predator's influence on prey traits, rather than prey density. Because such nonlethal predator effects are immediate, can influence the entire prey population, and can occur over the entire prey lifetime, we argue that nonlethal predator effects are likely to contribute strongly to the net indirect effects of predators (i.e., nonlethal effects may be comparable in magnitude to those resulting from killing prey). This prediction was supported by an experiment in which the indirect effects of a larval dragonfly ( Anax sp.) predator on large bullfrog tadpoles ( Rana catesbeiana ), through nonlethal effects on competing small bullfrog tadpoles, were large relative to indirect effects caused by density reduction of the small tadpoles (the lethal effect). Treatments in which lethal and nonlethal effects of Anax were manipulated independently indicated that this result was robust for a large range of different combinations of lethal and nonlethal effects. Because many, if not most, prey modify traits in response to predators, our results suggest that the magnitude of interaction coefficients between two species may often be dynamically related to changes in other community members, and that many indirect effects previously attributed to the lethal effects of predators may instead be due to shifts in traits of surviving prey.\n",
      "\u001b[1mDOI:\u001b[0m https://doi.org/10.1073/pnas.071061998\n",
      "\u001b[93m\u001b[1m*Current Score: 0\u001b[0m\u001b[0m\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mScore Papers\u001b[0m\n",
      "For each paper, enter an integer between 0 and 3. 0 indicates a lower quality and 3 indicates a higher quality. If the paper has already been scored (you should see a 'Current Score') or if you would just like to skip a paper, press the enter key without any input. Should you want to exit or go back (i.e. [A] for 'Go Back' and [B] for 'Exit Program'), your scores will automatically be saved.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[C] Exit Scoring\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "1/5\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mTitle:\u001b[0m HABITAT COMPLEXITY DISRUPTS PREDATOR–PREY INTERACTIONS BUT NOT THE TROPHIC CASCADE ON OYSTER REEFS\n",
      "\u001b[1mAbstract:\u001b[0m Despite recognition of the significance of both food web interactions and habitat complexity in community dynamics, current ecological theory rarely couples these two processes. Experimental manipulations of the abundance of the two predators in an oyster-reef trophic cascade, and the structural complexity provided by reefs of living oysters, demonstrated that enhanced habitat complexity weakened the strengths of trophic interactions. The system of tri-trophic interactions included oyster toadfish (Opsanus tau) as the top predator that consumed the mud crab (Panopeus herbstii), which preys upon juvenile oysters (Crassostrea virginica). On reefs of low complexity, toadfish controlled mud crab abundances and indirectly determined the level of mortality of juvenile oysters. The indirect effects of toadfish on oysters emerged through their influence on how intensely mud crabs preyed on oysters. Augmentation of habitat complexity by substituting vertically oriented, living oysters for the flat shells of dead oysters disrupted both of the direct trophic linkages but did not alter the magnitude of the indirect effect of toadfish on juvenile oysters. This paradox can be understood by partitioning the mechanisms by which toadfish influence mud crabs and ultimately juvenile oysters. Trait-mediated indirect interactions (TMIIs; i.e., predator-avoidance behavior in mud crabs) accounted for 95.6–98.2% of toadfish indirect benefits to oyster survival and, consequently, were a much greater contributor than density-mediated indirect interactions (DMIIs; i.e., the reduction in crab abundance by toadfish). Avoidance behavior was unaffected by modification in habitat complexity. Complex reefs increased total oyster survival because added habitat complexity reduced mud crab predation on oysters. Additionally, the magnitude of this effect was much greater than the increase in oyster mortality as a result of complex reefs disrupting toadfish predation on mud crabs. This experimental demonstration of how habitat complexity modifies trophic interactions in a temperate reef community has fundamental implications for our understanding of species interaction webs and community structure. The influence of habitat complexity on the strength of a trophic cascade generally may depend upon whether physical complexity provides actual and perceived refuges for component predator–prey pairs.\n",
      "\u001b[1mDOI:\u001b[0m https://doi.org/10.1890/03-0067\n",
      "\u001b[93m\u001b[1m*Current Score: 1\u001b[0m\u001b[0m\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\lbeln\\anaconda3\\envs\\3.10\\lib\\site-packages\\beautifultable\\utils.py:125: FutureWarning: 'BeautifulTable.column_headers' has been deprecated in 'v1.0.0' and will be removed in 'v1.2.0'. Use 'BTColumnCollection.header' instead.\n",
      "  warnings.warn(message, FutureWarning)\n",
      "C:\\Users\\lbeln\\anaconda3\\envs\\3.10\\lib\\site-packages\\beautifultable\\utils.py:125: FutureWarning: 'BeautifulTable.append_row' has been deprecated in 'v1.0.0' and will be removed in 'v1.2.0'. Use 'BTRowCollection.append' instead.\n",
      "  warnings.warn(message, FutureWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mPreview Dataset\u001b[0m\n",
      "Before scoring, you can check whether your chosen dataset looks correct. Enter the [1] to continue; [2] to go to the previous page of papers; [3] to go to the next page of papers; or a letter that corresponds to an action (i.e. [A] for 'Go Back' and [B] for 'Exit Program').\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[1] Continue\n",
      "[2] Previous Page\n",
      "[3] Next Page\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "1/1\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "|        Title         |             Abstract              |    DOI    | Score |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| The contribution of  | Many prey modify traits in respon | https://d |   0   |\n",
      "| trait-mediated indir | se to predation risk and this mod | oi.org/10 |       |\n",
      "| ect effects to the n | ification of traits can influence | .1073/pna |       |\n",
      "| et effects of a pred |  the prey's resource acquisition  | s.0710619 |       |\n",
      "|         ator         |       rate. A predator t...       |    98     |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| HABITAT COMPLEXITY D | Despite recognition of the signif | https://d |   0   |\n",
      "| ISRUPTS PREDATOR–PRE | icance of both food web interacti | oi.org/10 |       |\n",
      "| Y INTERACTIONS BUT N | ons and habitat complexity in com | .1890/03- |       |\n",
      "| OT THE TROPHIC CASCA | munity dynamics, current ecologic |   0067    |       |\n",
      "|  DE ON OYSTER REEFS  |       al theory rarely c...       |           |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| Do plant‐ and soil‐m | Summary Biotic interactions of pl | https://d |   0   |\n",
      "| ediated legacy effec | ants can have legacy effects medi | oi.org/10 |       |\n",
      "| ts impact future bio | ated by changes in plant traits a | .1111/136 |       |\n",
      "|  tic interactions?   | nd soil characteristics. Legacy e | 5-2435.12 |       |\n",
      "|                      |       ffects are defined...       |    456    |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| HOW DEPENDENT ARE SP | In ecological theory species inte | https://d |   0   |\n",
      "| ECIES-PAIR INTERACTI | raction strengths are typically d | oi.org/10 |       |\n",
      "| ON STRENGTHS ON OTHE | escribed by constants or function | .1890/03- |       |\n",
      "| R SPECIES IN THE FOO | s that depend on the densities of |   0662    |       |\n",
      "|        D WEB?        |        the two interacti...       |           |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "| A twin approach to u | The regulation of gene expression | https://d |   0   |\n",
      "| nraveling epigenetic |  plays a pivotal role in complex  | oi.org/10 |       |\n",
      "|          s           | phenotypes, and epigenetic mechan | .1016/j.t |       |\n",
      "|                      | isms such as DNA methylation are  | ig.2010.1 |       |\n",
      "|                      |       essential to this ...       |   2.005   |       |\n",
      "+----------------------+-----------------------------------+-----------+-------+\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  A\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mSearch Unscored Papers\u001b[0m\n",
      "Enter a search query (i.e. 'intraguild predation') to load papers from OpenAlex or a letter that corresponds to an action (i.e. 'Go Back' and 'Exit Program'). If you want to search 'A', you should instead input 'a', or else you will end up going back a page; the same goes for 'B' and 'b'.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  A\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mFind Papers to Score\u001b[0m\n",
      "Now, we need to find unscored papers. Enter [1] to view a list of datasets that contain unscored papers or [2] to look for papers on OpenAlex (requires an Internet connection).\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "[1] Load Unscored Dataset\n",
      "[2] Search for Unscored Papers\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter Input:  1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m\u001b[1mLabel Helper™\u001b[0m\n",
      "Use this program to quickly create labeled datasets.\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "\u001b[1mLoad Unscored Dataset\u001b[0m\n",
      "Enter the number of the dataset to score or a letter that corresponds to an action (i.e. 'Go Back' and 'Exit Program').\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Actions:\n",
      "[A] Go Back\n",
      "[B] Exit Program\n",
      "------------------------------------------------------------------------------------------------------------------------\n",
      "Datasets:\n",
      "[1] DFiltered.csv\n",
      "[2] Search-trait-mediated-interaction-modification.csv\n",
      "[3] SubAFiltered.csv\n",
      "[4] SubBFiltered.csv\n",
      "------------------------------------------------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception ignored in: <bound method IPythonKernel._clean_thread_parent_frames of <ipykernel.ipkernel.IPythonKernel object at 0x000002E53FB131C0>>\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lbeln\\anaconda3\\envs\\3.10\\lib\\site-packages\\ipykernel\\ipkernel.py\", line 788, in _clean_thread_parent_frames\n",
      "    if phase != \"start\":\n",
      "KeyboardInterrupt: \n",
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Constants\n",
    "TERMINAL_SIZE = os.get_terminal_size()\n",
    "HEADER = '\\033[95m'\n",
    "OKBLUE = '\\033[94m'\n",
    "OKCYAN = '\\033[96m'\n",
    "OKGREEN = '\\033[92m'\n",
    "WARNING = '\\033[93m'\n",
    "FAIL = '\\033[91m'\n",
    "ENDC = '\\033[0m'\n",
    "BOLD = '\\033[1m'\n",
    "UNDERLINE = '\\033[4m'\n",
    "\n",
    "# Stack of Pages\n",
    "pages = []\n",
    "\n",
    "# To-Be Labeled DF\n",
    "labeled_df = None\n",
    "labeled_df_filepath = None\n",
    "\n",
    "# Unlabeled DF\n",
    "unlabeled_df = None\n",
    "unlabeled_df_filepath = None\n",
    "\n",
    "# Revert Abstract Function\n",
    "def revert_abstract(inverted_abstract):\n",
    "    if not inverted_abstract:\n",
    "        return \"\"\n",
    "    abstract = \"\"\n",
    "    i = 0\n",
    "    while True:\n",
    "        index_found = False\n",
    "        for k, v in inverted_abstract.items():\n",
    "            # print(k, v)\n",
    "            if i in v:\n",
    "                if abstract and k not in [\".\", \",\", \"?\", \"!\"]:\n",
    "                    abstract += \" \"\n",
    "                abstract += k\n",
    "                i += 1\n",
    "                index_found = True\n",
    "        if not index_found:\n",
    "            break\n",
    "    return abstract\n",
    "\n",
    "# On Topic Function\n",
    "classifier = pipeline(\"zero-shot-classification\", model=\"MoritzLaurer/DeBERTa-v3-base-mnli-fever-anli\")\n",
    "def on_topic(text, verbose=False):\n",
    "    # Topic and Threshold\n",
    "    # I'm using a different model which seems to be more confident,\n",
    "    # so I'm going to increase the threshold.\n",
    "    topics = [(\"ecology\", 0.9), (\"interaction\", 0.9)]\n",
    "            \n",
    "    for topic, threshold in topics:\n",
    "        result = classifier(text, [topic])\n",
    "        if verbose:\n",
    "            print(result[\"scores\"][0])\n",
    "        if result[\"scores\"][0] < threshold:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "# CMD UI Functions\n",
    "def clear_screen():\n",
    "    os.system('cls')\n",
    "\n",
    "def print_border():\n",
    "    print('-' * TERMINAL_SIZE.columns)\n",
    "    \n",
    "def print_header():\n",
    "    print(f\"{OKBLUE}{BOLD}Label Helper™{ENDC}\")\n",
    "    print(\"Use this program to quickly create labeled datasets.\")\n",
    "    print_border()\n",
    "\n",
    "def page_menu():\n",
    "    global pages\n",
    "    global labeled_df\n",
    "    global labeled_df_filepath\n",
    "    \n",
    "    clear_screen()\n",
    "    print_header()\n",
    "    print(f\"{BOLD}Create or Load Labeled Dataset{ENDC}\")\n",
    "    print(\"Start by entering the filepath of the labeled dataset to create or add onto.\")\n",
    "    print(\"Ensure that the filepath entered is for a .csv file.\")\n",
    "    print_border()\n",
    "    while True:\n",
    "        labeled_df_filepath = input(\"Enter Filename of Dataset: \")\n",
    "    \n",
    "        # Existing File\n",
    "        if os.path.isfile(labeled_df_filepath):\n",
    "            print(f\"Loading File: '{labeled_df_filepath}'\")\n",
    "            try:\n",
    "                labeled_df = pd.read_csv(labeled_df_filepath)\n",
    "            except Exception:\n",
    "                print(f\"{FAIL}Error Occurred. Try Again.{ENDC}\\n\")\n",
    "                continue\n",
    "        # Creating New File\n",
    "        else:\n",
    "            if labeled_df_filepath[-4:] != \".csv\":\n",
    "                print(f\"{FAIL}Must Be CSV File (.csv). Try Again.{ENDC}\\n\")\n",
    "                continue\n",
    "            try:\n",
    "                print(f\"Creating File: '{labeled_df_filepath}'\")\n",
    "                labeled_df = pd.DataFrame(columns=['Title','Abstract','DOI','Score'])\n",
    "                labeled_df.to_csv(labeled_df_filepath, index=False)\n",
    "            except Exception:\n",
    "                print(f\"{FAIL}Error Occurred. Try Again.{ENDC}\\n\")\n",
    "                continue\n",
    "\n",
    "        # If this statement is reached,\n",
    "        # it's likely that everything is fine.\n",
    "        break\n",
    "\n",
    "    # Next Page\n",
    "    pages.append(\"Method\")\n",
    "\n",
    "def page_method():\n",
    "    global pages\n",
    "    \n",
    "    clear_screen()\n",
    "    print_header()\n",
    "    print(f\"{BOLD}Find Papers to Score{ENDC}\")\n",
    "    print(\"Now, we need to find unscored papers. Enter [1] to view a list of datasets that contain unscored papers or [2] to look for papers on OpenAlex (requires an Internet connection).\")\n",
    "    print_border()\n",
    "    \n",
    "    # Action\n",
    "    print(\"Actions:\")\n",
    "    print(\"[A] Go Back\")\n",
    "    print(\"[B] Exit Program\")\n",
    "    print(\"[1] Load Unscored Dataset\")\n",
    "    print(\"[2] Search for Unscored Papers\")\n",
    "    print_border()\n",
    "\n",
    "    # Receive Input\n",
    "    while True:\n",
    "        try:\n",
    "            option = input(\"Enter Input: \")\n",
    "            if option in ['1', '2']:\n",
    "                option = int(option)\n",
    "                break\n",
    "            if option in ['A', 'B']:\n",
    "                break\n",
    "            raise ValueError('Invalid Input')\n",
    "        except ValueError:\n",
    "            print(f\"{FAIL}Must Enter 1 or 2. Try Again.{ENDC}\\n\")\n",
    "            continue\n",
    "        break\n",
    "\n",
    "    # Next Page\n",
    "    if option == 1: \n",
    "        pages.append(\"Method Load\")\n",
    "    elif option == 2:\n",
    "        pages.append(\"Method Search\")\n",
    "    elif option == 'A':\n",
    "        pages.pop()\n",
    "    elif option == 'B':\n",
    "        exit()\n",
    "\n",
    "def page_method_load():\n",
    "    global pages\n",
    "    global unlabeled_df\n",
    "    global unlabeled_df_filepath\n",
    "    \n",
    "    clear_screen()\n",
    "    print_header()\n",
    "    print(f\"{BOLD}Load Unscored Dataset{ENDC}\")\n",
    "    print(\"Enter the number of the dataset to score or a letter that corresponds to an action (i.e. 'Go Back' and 'Exit Program').\")\n",
    "    print_border()\n",
    "\n",
    "    # Action\n",
    "    print(\"Actions:\")\n",
    "    print(\"[A] Go Back\")\n",
    "    print(\"[B] Exit Program\")\n",
    "    print_border()\n",
    "\n",
    "    # Datasets\n",
    "    # The unscored datasets should be DataFrames stored as CSV files\n",
    "    # in a directory './UnscoredDatasets'.\n",
    "    print(\"Datasets:\")\n",
    "    \n",
    "    directory = \"./UnscoredDatasets\"\n",
    "    files = [f for f in os.listdir(directory) if os.path.isfile(f\"{directory}/{f}\") and f[-4:] == '.csv']\n",
    "    number_files = len(files)\n",
    "\n",
    "    # Print Files\n",
    "    for i, f in enumerate(files):\n",
    "        print(f\"[{i+1}] {f}\")\n",
    "    \n",
    "    print_border()\n",
    "\n",
    "    # Receive Input\n",
    "    while True:\n",
    "        try:\n",
    "            option = input(\"Enter Input: \")\n",
    "            if option == 'A':\n",
    "                pages.pop()\n",
    "                return\n",
    "            if option == 'B':\n",
    "                exit()\n",
    "            \n",
    "            option = int(option)\n",
    "            if option < 0 or option > number_files:\n",
    "                raise ValueError(f'Number Must Be Between 1 and {number_files}')\n",
    "        except ValueError:\n",
    "            print(f\"{FAIL}Must Enter Number Between 1 and {number_files}. Try Again.{ENDC}\\n\")\n",
    "            continue\n",
    "\n",
    "        unlabeled_df_filepath = f\"{directory}/{files[option - 1]}\"\n",
    "        print(\"Loading File:\",unlabeled_df_filepath)\n",
    "        unlabeled_df = pd.read_csv(unlabeled_df_filepath)\n",
    "        break\n",
    "\n",
    "    # Next Page\n",
    "    pages.append(\"Preview\")\n",
    "\n",
    "def page_method_search():\n",
    "    global pages\n",
    "    global unlabeled_df\n",
    "    global unlabeled_df_filepath\n",
    "    \n",
    "    clear_screen()\n",
    "    print_header()\n",
    "    print(f\"{BOLD}Search Unscored Papers{ENDC}\")\n",
    "    print(\"Enter a search query (i.e. 'intraguild predation') to load papers from OpenAlex or a letter that corresponds to an action (i.e. 'Go Back' and 'Exit Program'). If you want to search 'A', you should instead input 'a', or else you will end up going back a page; the same goes for 'B' and 'b'.\")\n",
    "    print_border()\n",
    "\n",
    "    # Action\n",
    "    print(\"Actions:\")\n",
    "    print(\"[A] Go Back\")\n",
    "    print(\"[B] Exit Program\")\n",
    "    print_border()\n",
    "\n",
    "    while True:\n",
    "        query = input(\"Enter Input: \")\n",
    "\n",
    "        # Go Back\n",
    "        if query == 'A':\n",
    "            pages.pop()\n",
    "            return\n",
    "\n",
    "        # Exit\n",
    "        if query == 'B':\n",
    "            exit()\n",
    "\n",
    "        # Searching OpenAlex\n",
    "        num_pages = 0\n",
    "        max_pages = 1000\n",
    "        works = Works().search(query)\n",
    "        pager = works.paginate(per_page=200)\n",
    "\n",
    "        # Empty DF\n",
    "        unlabeled_df = pd.DataFrame(columns=['Title','Abstract','DOI','Score'])\n",
    "\n",
    "        # Number Results\n",
    "        num_results = min(works.count(), max_pages)\n",
    "        if not num_results:\n",
    "            print(\"No Results. Try Again\\n\")\n",
    "            continue\n",
    "        print(f\"Number Results: {num_results}\")\n",
    "        print(\"Searching...\")\n",
    "\n",
    "        # Populating DF\n",
    "        for page in pager:\n",
    "            for work in page:\n",
    "                if num_pages >= max_pages:\n",
    "                    break\n",
    "                abstract = revert_abstract(work[\"abstract_inverted_index\"])\n",
    "                try:\n",
    "                    if not abstract or not on_topic(abstract):\n",
    "                        continue\n",
    "                except Exception:\n",
    "                    print(\"Out of Memory. Ending Search\")\n",
    "                    max_pages = 0\n",
    "                paper = pd.DataFrame([[work[\"title\"], abstract, work[\"doi\"], 0]],  columns=['Title','Abstract','DOI','Score'])\n",
    "                unlabeled_df = pd.concat([unlabeled_df, paper])\n",
    "                num_pages += 1\n",
    "            if num_pages >= max_pages:\n",
    "                break\n",
    "\n",
    "        # Save DF\n",
    "        unlabeled_df.reset_index(drop=True, inplace=True)\n",
    "        unlabeled_df_filepath = f\"./UnscoredDatasets/Search-{query.replace(' ', '-')}.csv\"\n",
    "        unlabeled_df.to_csv(unlabeled_df_filepath, index=False)\n",
    "        break\n",
    "\n",
    "    # Next Page\n",
    "    pages.append(\"Preview\")\n",
    "\n",
    "def page_preview():\n",
    "    global pages\n",
    "    global unlabeled_df\n",
    "    global unlabeled_df_filepath\n",
    "\n",
    "    NUM_PAPERS_PER_TABLE = 5\n",
    "    ABSTRACT_MAX_LENGTH = 150\n",
    "    \n",
    "    # Tables\n",
    "    table_i = 0\n",
    "    tables = []\n",
    "    for i, row in unlabeled_df.iterrows():\n",
    "        table = None if len(tables) == 0 else tables[-1]\n",
    "        if i % NUM_PAPERS_PER_TABLE == 0:\n",
    "            tables.append(BeautifulTable())\n",
    "            table = tables[-1]\n",
    "            table.column_headers = [\"Title\", \"Abstract\", \"DOI\", \"Score\"]\n",
    "\n",
    "        shortened_abstract = row['Abstract'][:ABSTRACT_MAX_LENGTH] + '...' if len(row['Abstract']) > ABSTRACT_MAX_LENGTH else ''\n",
    "        table.append_row([row['Title'], shortened_abstract,  row['DOI'],  row['Score']])\n",
    "    num_tables = len(tables)\n",
    "    \n",
    "    while True:\n",
    "        clear_screen()\n",
    "        print_header()\n",
    "        print(f\"{BOLD}Preview Dataset{ENDC}\")\n",
    "        print(\"Before scoring, you can check whether your chosen dataset looks correct. Enter the [1] to continue; [2] to go to the previous page of papers; [3] to go to the next page of papers; or a letter that corresponds to an action (i.e. [A] for 'Go Back' and [B] for 'Exit Program').\")\n",
    "        print_border()\n",
    "        \n",
    "        # Actions\n",
    "        print(\"Actions:\")\n",
    "        print(\"[A] Go Back\")\n",
    "        print(\"[B] Exit Program\")\n",
    "        print(\"[1] Continue\")\n",
    "        print(\"[2] Previous Page\")\n",
    "        print(\"[3] Next Page\")\n",
    "        print_border()\n",
    "\n",
    "        # Showing Table Index\n",
    "        print(f\"{table_i+1}/{num_tables}\")\n",
    "        print_border()\n",
    "        \n",
    "        # Showing Table\n",
    "        print(f\"{tables[table_i]}\")\n",
    "        print_border()\n",
    "\n",
    "        # Receive Input\n",
    "        try:\n",
    "            option = input(\"Enter Input: \")\n",
    "\n",
    "            # Go Back\n",
    "            if option == 'A':\n",
    "                pages.pop()\n",
    "                return\n",
    "    \n",
    "            # Exit\n",
    "            if option == 'B':\n",
    "                exit()\n",
    "\n",
    "            # Integer Entered\n",
    "            option = int(option)\n",
    "            \n",
    "            if option == 1:\n",
    "                break\n",
    "            elif option == 2:\n",
    "                table_i = max(0, table_i - 1)\n",
    "                continue\n",
    "            elif option == 3:\n",
    "                table_i = min(len(tables) - 1, table_i + 1)\n",
    "                continue\n",
    "            else:\n",
    "                print(f\"{FAIL}Must Enter 1, 2, 3, 'A', or 'B'. Try Again.{ENDC}\\n\")\n",
    "                continue\n",
    "        except ValueError:\n",
    "            print(f\"{FAIL}Must Enter 1, 2, 3, 'A', or 'B'. Try Again.{ENDC}\\n\")\n",
    "            continue\n",
    "\n",
    "    # Next Page\n",
    "    pages.append(\"Score\")\n",
    "\n",
    "def page_score():\n",
    "    global pages\n",
    "    global unlabeled_df\n",
    "    global unlabeled_df_filepath\n",
    "    global labeled_df\n",
    "    global labeled_df_filepath\n",
    "    \n",
    "    paper_i = 0\n",
    "    num_papers = unlabeled_df.shape[0]\n",
    "\n",
    "    while paper_i < num_papers:\n",
    "        clear_screen()\n",
    "        print_header()\n",
    "        print(f\"{BOLD}Score Papers{ENDC}\")\n",
    "        print(\"For each paper, enter an integer between 0 and 3. 0 indicates a lower quality and 3 indicates a higher quality. If the paper has already been scored (you should see a 'Current Score') or if you would just like to skip a paper, press the enter key without any input. Should you want to exit or go back (i.e. [A] for 'Go Back' and [B] for 'Exit Program'), your scores will automatically be saved.\")\n",
    "        print_border()\n",
    "\n",
    "        # Actions\n",
    "        print(\"Actions:\")\n",
    "        print(\"[A] Go Back\")\n",
    "        print(\"[B] Exit Program\")\n",
    "        print(\"[C] Exit Scoring\")\n",
    "        print_border()\n",
    "\n",
    "        print(f\"{paper_i+1}/{num_papers}\")\n",
    "        print_border()\n",
    "\n",
    "        # Paper Information\n",
    "        title = unlabeled_df.iloc[paper_i]['Title']\n",
    "        abstract = unlabeled_df.iloc[paper_i]['Abstract']\n",
    "        doi = unlabeled_df.iloc[paper_i]['DOI']\n",
    "        print(f\"{BOLD}Title:{ENDC} {title}\")\n",
    "        print(f\"{BOLD}Abstract:{ENDC} {abstract}\")\n",
    "        print(f\"{BOLD}DOI:{ENDC} {doi}\")\n",
    "\n",
    "        # Current Score (if Any)\n",
    "        matching_labeled_row = labeled_df.loc[(labeled_df['Title'] == title) & (labeled_df['DOI'] == doi)]\n",
    "        not_found = matching_labeled_row.empty\n",
    "        if not not_found:\n",
    "            print(f\"{WARNING}{BOLD}*Current Score: {matching_labeled_row.iloc[0]['Score']}{ENDC}{ENDC}\")\n",
    "        print_border()\n",
    "\n",
    "        exit_loop = False\n",
    "        while True:\n",
    "            score = input(\"Enter Input: \")\n",
    "            try:\n",
    "                # Go Back\n",
    "                if score == 'A':\n",
    "                    pages.pop()\n",
    "                    return\n",
    "                # Exit Program\n",
    "                elif score == 'B':\n",
    "                    exit()\n",
    "                # Exit Scoring\n",
    "                elif score == 'C':\n",
    "                    exit_loop = True\n",
    "                    break\n",
    "                elif not_found:\n",
    "                    labeled_paper = pd.DataFrame([[title, abstract, doi, int(score)]],  columns=['Title','Abstract','DOI','Score'])\n",
    "                    labeled_df = pd.concat([labeled_df, labeled_paper])\n",
    "                else:\n",
    "                    labeled_df.loc[(labeled_df['Title'] == title) & (labeled_df['DOI'] == doi), 'Score'] = int(score)\n",
    "                \n",
    "                paper_i += 1\n",
    "                labeled_df.reset_index(drop=True, inplace=True)\n",
    "                labeled_df.to_csv(labeled_df_filepath, index=False)\n",
    "                \n",
    "                break\n",
    "            except Exception:\n",
    "                # No Input Entered (Skip)\n",
    "                if not score:\n",
    "                    paper_i += 1\n",
    "                    break\n",
    "                print(f\"{FAIL}Must Enter 0, 1, 2, 3, or Nothing. Try Again.{ENDC}\\n\")\n",
    "                continue\n",
    "\n",
    "        if exit_loop:\n",
    "            break\n",
    "\n",
    "    print(f\"{OKGREEN}Scored Papers{ENDC}\")\n",
    "    time.sleep(1)\n",
    "    pages = [\"Method\"]\n",
    "\n",
    "while True:\n",
    "    if len(pages) == 0:\n",
    "        page_menu()\n",
    "    else:\n",
    "        last_page = pages[-1]\n",
    "        if last_page == \"Method\":\n",
    "            page_method()\n",
    "        elif last_page == \"Method Load\":\n",
    "            page_method_load()\n",
    "        elif last_page == \"Method Search\":\n",
    "            page_method_search()\n",
    "        elif last_page == \"Preview\":\n",
    "            page_preview()\n",
    "        elif last_page == \"Score\":\n",
    "            page_score()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70f89270-4137-48f5-9521-25878117246c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
