{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "96a71233-b342-4b58-bcd2-ffc199873c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import csv\n",
    "import sys\n",
    "import json\n",
    "import math\n",
    "import spacy\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from taxonerd import TaxoNERD\n",
    "from spacy.matcher import Matcher, DependencyMatcher, PhraseMatcher"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ded48ca0-cc63-4260-a793-b74d4ada04a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "VERBOSE_LEVEL = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c1294dad-2640-43f9-9e94-66840a9fc0c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helper Functions\n",
    "def flatten(arr):\n",
    "    if not isinstance(arr, list):\n",
    "        return [arr]\n",
    "\n",
    "    flat = []\n",
    "    for val in arr:\n",
    "        flat.extend(flatten(val))\n",
    "\n",
    "    return flat\n",
    "\n",
    "def find(arr, foo):\n",
    "    for val in arr:\n",
    "        if foo(val):\n",
    "            return val\n",
    "    return None\n",
    "\n",
    "def find_all(arr, foo):\n",
    "    bar = []\n",
    "    for val in arr:\n",
    "        if foo(val):\n",
    "            bar.append(val)\n",
    "    return bar\n",
    "\n",
    "def find_index(arr, foo):\n",
    "    for i in range(len(arr)):\n",
    "        if foo(arr[i]):\n",
    "            return i\n",
    "    return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "15130c37-d2ad-485f-8258-9bea4fed9731",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Entity:\n",
    "    # Labels\n",
    "    LIST = 1\n",
    "    ITEM = 2\n",
    "    QUOTE = 3\n",
    "    BREAK = 4\n",
    "    END = 5\n",
    "    AND_OR_END = 6\n",
    "    COLON = 7\n",
    "    COLON_BREAK = 8\n",
    "    I_CLAUSE = 9\n",
    "    D_CLAUSE = 10\n",
    "    P_PHRASE = 11\n",
    "    BRACKETS = 12\n",
    "\n",
    "    def __init__(self, doc, label=None, l=None, r=None, children=None):\n",
    "        self.doc = doc\n",
    "        self.label = label\n",
    "        self.l = l\n",
    "        self.r = r\n",
    "        self.children = children or []\n",
    "\n",
    "    def size(self):\n",
    "        return self.r - self.l + 1\n",
    "\n",
    "    def span(self):\n",
    "        return self.doc[self.l:self.r+1]\n",
    "\n",
    "    def lower(self):\n",
    "        return self.doc[self.l:self.r+1].text.lower()\n",
    "\n",
    "    def start(self):\n",
    "        return self.doc[self.l]\n",
    "\n",
    "    def end(self):\n",
    "        return self.doc[self.r]\n",
    "\n",
    "    @staticmethod\n",
    "    def tokens(*, ent=None, ents=None):\n",
    "        if ents:\n",
    "            tokens = flatten([list(ent.span()) for ent in ents])\n",
    "            tokens = sorted(tokens, key=lambda token: token.i)\n",
    "            return tokens\n",
    "        if ent:\n",
    "            tokens = list(ent.span())\n",
    "            return tokens\n",
    "        return None\n",
    "\n",
    "    @staticmethod\n",
    "    def is_conjunction(token):\n",
    "        return token.lower_ in [\"and\", \"or\"]\n",
    "\n",
    "    @staticmethod\n",
    "    def same_speech(speech_1, speech_2):\n",
    "        nouns = [\"NOUN\", \"PRON\", \"PROPN\"]\n",
    "        if speech_1 in nouns and speech_2 in nouns:\n",
    "            return True\n",
    "        return speech_1 == speech_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c43429b9-b46c-4ac9-8927-07923ad86af4",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Quotes:\n",
    "    def __init__(self, main, entities):\n",
    "        self.main = main\n",
    "        self.entities = entities\n",
    "\n",
    "    def is_quote(self, i):\n",
    "        return i < len(self.entities) and self.entities[i].lower() == \"\\\"\"\n",
    "    \n",
    "    def identify(self):\n",
    "        i = 0\n",
    "        \n",
    "        while i < len(self.entities):\n",
    "            if not self.is_quote(i):\n",
    "                i += 1\n",
    "                continue\n",
    "            \n",
    "            self.entities[i].label = Entity.QUOTE\n",
    "            \n",
    "            while not self.is_quote(i+1):\n",
    "                self.entities[i].r = self.entities[i+1].r\n",
    "                self.entities.pop(i+1)\n",
    "\n",
    "            if self.is_quote(i+1):\n",
    "                self.entities[i].r = self.entities[i+1].r\n",
    "                self.entities.pop(i+1)\n",
    "\n",
    "        return self.entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c9b7eca0-6d5c-48ba-b882-d4488f3d466a",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Brackets:\n",
    "    MATCHES = {\n",
    "        \"[\": \"]\", \n",
    "        \"(\": \")\",\n",
    "        \"—\": \"—\",\n",
    "    }\n",
    "\n",
    "    OPENING = MATCHES.keys()\n",
    "    CLOSING = MATCHES.values()\n",
    "\n",
    "    def __init__(self, main, entities):\n",
    "        self.main = main\n",
    "        self.stack = []\n",
    "        self.entities = [*entities]\n",
    "\n",
    "    def is_opening(self, i):\n",
    "        return i < len(self.entities) and self.entities[i].lower()[0] in Brackets.OPENING\n",
    "\n",
    "    def is_closing(self, i):\n",
    "        return i < len(self.entities) and self.entities[i].lower()[0] in Brackets.CLOSING\n",
    "\n",
    "    def closes(self, i):\n",
    "        opener = self.entities[self.stack[-1]].lower()[0]\n",
    "        closer = self.entities[i].lower()[0]\n",
    "        return Brackets.MATCHES[opener] == closer\n",
    "    \n",
    "    def identify(self):\n",
    "        self.stack = []\n",
    "        \n",
    "        i = 0\n",
    "        while i < len(self.entities):\n",
    "            # print(i, self.entities[i].span())\n",
    "            \n",
    "            # Closing\n",
    "            if self.is_closing(i) and self.stack:\n",
    "                j = None if not self.closes(i) else self.stack.pop()\n",
    "                \n",
    "                if not self.stack and j != None:\n",
    "                    self.entities[j].r = self.entities[i].r\n",
    "                    self.entities.pop(i)\n",
    "                    continue\n",
    "                else:\n",
    "                    i += 1\n",
    "\n",
    "            # Opening\n",
    "            elif self.is_opening(i):\n",
    "                if not self.stack:\n",
    "                    self.entities[i].label = Entity.BRACKETS\n",
    "                self.stack.append(i)\n",
    "                i += 1\n",
    "\n",
    "            # Consuming\n",
    "            elif self.stack:\n",
    "                # If you're at the end of the possible entities,\n",
    "                # and the list is unclosed, we must stop.\n",
    "                if i + 1 >= len(self.entities):\n",
    "                    break\n",
    "                self.entities[self.stack[0]].r = self.entities[i+1].r\n",
    "                self.entities.pop(i)\n",
    "\n",
    "            else:\n",
    "                i += 1\n",
    "\n",
    "        # for ent in self.entities:\n",
    "        #     if ent.label == Entity.BRACKETS:\n",
    "        #         print(f\"Bracket: {ent.span()}\")\n",
    "        \n",
    "        return self.entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "41cd2d47-a10e-4a1c-8bca-d9607342aa48",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Separators:\n",
    "    def __init__(self, main, entities):\n",
    "        self.main = main\n",
    "        self.entities = [*entities]\n",
    "\n",
    "    def is_break(self, i):\n",
    "        if i >= len(self.entities):\n",
    "            return False\n",
    "        \n",
    "        if self.entities[i].lower() not in [\";\", \",\"]:\n",
    "            return False\n",
    "\n",
    "        # Breaks cannot have a following conjunction.\n",
    "        # Else, it would be an end and not a break.\n",
    "        return not bool(\n",
    "            i + 1 < len(self.entities) and \n",
    "            self.entities[i+1].size() == 1 and \n",
    "            self.entities[i+1].span()[0].pos_ in [\"CCONJ\"]\n",
    "        )\n",
    "\n",
    "    def is_end(self, i):\n",
    "        if i >= len(self.entities):\n",
    "            return False\n",
    "        \n",
    "        if self.entities[i].lower() not in [\";\", \",\"]:\n",
    "            return False\n",
    "        \n",
    "        return not self.is_break(i)\n",
    "\n",
    "    def identify(self):\n",
    "        i = 0\n",
    "\n",
    "        while i < len(self.entities):\n",
    "            # Break\n",
    "            if self.is_break(i):\n",
    "                self.entities[i].label = Entity.BREAK\n",
    "                i += 1\n",
    "\n",
    "            # End\n",
    "            elif self.is_end(i):\n",
    "                conj = self.entities[i+1].start().lower_\n",
    "                self.entities[i].label = Entity.AND_OR_END if conj in [\"and\", \"or\"] else Entity.END\n",
    "                self.entities[i].r += 1\n",
    "                self.entities.pop(i+1)\n",
    "\n",
    "            else:\n",
    "                i += 1\n",
    "                \n",
    "        return self.entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8e27c7af-f963-4d0f-9293-3521a72ba3f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Colons:\n",
    "    def __init__(self, main, entities):\n",
    "        self.main = main\n",
    "        self.entities = [*entities]\n",
    "\n",
    "    def identify(self):\n",
    "        i = 0\n",
    "\n",
    "        while i < len(self.entities):\n",
    "            if self.entities[i].lower()[-1] != \":\":\n",
    "                i += 1\n",
    "                continue\n",
    "\n",
    "            if not self.entities[i].label:\n",
    "                self.entities[i].label = Entity.COLON_BREAK\n",
    "\n",
    "            if i + 1 < len(self.entities):\n",
    "                self.entities[i+1].label = Entity.COLON\n",
    "                self.entities[i+1].r = self.entities[-1].r\n",
    "                self.entities = self.entities[:i+2]\n",
    "            \n",
    "            break\n",
    "\n",
    "        return self.entities        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "31b2ee9a-2e22-4807-9a03-e954ff0ddf2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Independent_Clauses:\n",
    "    def __init__(self, main, entities):\n",
    "        self.main = main\n",
    "        self.entities = [*entities]\n",
    "        self.allowed = []\n",
    "\n",
    "    def end(self, i):    \n",
    "        if i >= len(self.entities):\n",
    "            return True\n",
    "\n",
    "        if self.entities[i].label in self.allowed:\n",
    "            return True\n",
    "        \n",
    "        # Here, we check if the entity after\n",
    "        # the supposed end is a clause. If it\n",
    "        # is, then we can end at the current entity.\n",
    "        return bool(\n",
    "            i + 1 < len(self.entities) and \n",
    "            self.entities[i+1].label in [\n",
    "                Entity.COLON,\n",
    "                Entity.COLON_BREAK,\n",
    "                Entity.I_CLAUSE,\n",
    "                Entity.D_CLAUSE,\n",
    "                Entity.P_PHRASE\n",
    "            ]\n",
    "        )\n",
    "\n",
    "    def identify(self, allowed):\n",
    "        self.allowed = allowed\n",
    "        \n",
    "        i = 0\n",
    "        \n",
    "        while i < len(self.entities):\n",
    "            if self.entities[i].label not in self.allowed:\n",
    "                i += 1\n",
    "                continue\n",
    "\n",
    "            # Skip Clause\n",
    "            if self.entities[i].label in [\n",
    "                Entity.I_CLAUSE, \n",
    "                Entity.D_CLAUSE, \n",
    "                Entity.P_PHRASE\n",
    "            ]:\n",
    "                i = entities[i].r + 1\n",
    "                continue\n",
    "\n",
    "            # Create Clause\n",
    "            self.entities[i].label = Entity.I_CLAUSE\n",
    "            while not self.end(i+1):\n",
    "                self.entities[i].r = self.entities[i+1].r\n",
    "\n",
    "                # Add Child\n",
    "                if self.entities[i+1].label in [Entity.BRACKETS, Entity.QUOTE]:\n",
    "                    self.entities[i].children.append(self.entities[i+1])\n",
    "                    \n",
    "                self.entities.pop(i+1)\n",
    "\n",
    "            i += 1\n",
    "            \n",
    "        return self.entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "da0ca7d3-879c-41cf-b663-ab9f80a57505",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Dependent_Clauses:\n",
    "    RELATIVE_NOUNS = [\n",
    "        \"who\",\n",
    "        \"whom\",\n",
    "        \"which\",\n",
    "        \"what\",\n",
    "        \"that\",\n",
    "        \"whose\",\n",
    "        \"whomever\",\n",
    "        \"whoever\",\n",
    "        \"whichever\",\n",
    "        \"whatever\"\n",
    "    ]\n",
    "    \n",
    "    def __init__(self, main, entities):\n",
    "        self.main = main\n",
    "        self.entities = entities\n",
    "        self.separator = None\n",
    "\n",
    "    def end(self, i):\n",
    "        if i >= len(self.entities):\n",
    "            return True\n",
    "\n",
    "        # Here, we check if the entity after\n",
    "        # is a clause. As we don't combine two\n",
    "        # clauses, we must end here if that is\n",
    "        # the case.\n",
    "        if bool(\n",
    "            i + 1 < len(self.entities) and \n",
    "            self.entities[i+1].label in [\n",
    "                Entity.COLON, \n",
    "                Entity.COLON_BREAK,\n",
    "                Entity.I_CLAUSE,\n",
    "                Entity.D_CLAUSE,\n",
    "                Entity.P_PHRASE\n",
    "            ]\n",
    "        ):\n",
    "            return True\n",
    "\n",
    "        return bool(\n",
    "            self.entities[i].lower()[0] == self.separator or\n",
    "            self.entities[i].lower() in Dependent_Clauses.RELATIVE_NOUNS or\n",
    "            self.entities[i].start().pos_ in [\"SCONJ\"]\n",
    "        )\n",
    "\n",
    "    def identify(self, separator):\n",
    "        self.separator = separator\n",
    "        \n",
    "        i = 0\n",
    "        \n",
    "        while i < len(self.entities):\n",
    "            # Skip\n",
    "            if self.entities[i].label in [\n",
    "                Entity.COLON,\n",
    "                Entity.COLON_BREAK,\n",
    "                Entity.I_CLAUSE, \n",
    "                Entity.D_CLAUSE, \n",
    "                Entity.P_PHRASE\n",
    "            ]:\n",
    "                i = self.entities[i].r + 1\n",
    "                continue\n",
    "\n",
    "            # Indicators of Dependent Clause\n",
    "            rel = self.entities[i].lower() in Dependent_Clauses.RELATIVE_NOUNS\n",
    "            sub = self.entities[i].start().pos_ == \"SCONJ\"\n",
    "            \n",
    "            if not sub and not rel:\n",
    "                i += 1\n",
    "                continue\n",
    "\n",
    "            # Create Clause\n",
    "            self.entities[i].label = Entity.D_CLAUSE\n",
    "            while not self.end(i+1):\n",
    "                self.entities[i].r = self.entities[i+1].r\n",
    "\n",
    "                # Add Child\n",
    "                if self.entities[i+1].label in [Entity.BRACKETS, Entity.QUOTE]:\n",
    "                    self.entities[i].children.append(self.entities[i+1])\n",
    "                \n",
    "                self.entities.pop(i+1)\n",
    "\n",
    "            i += 1\n",
    "        \n",
    "        return self.entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "37580cac-61ce-4f50-9fde-2e2b11d98e81",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Prepositional_Phrases:\n",
    "    \n",
    "    def __init__(self, main, entities):\n",
    "        self.main = main\n",
    "        self.entities = [*entities]\n",
    "\n",
    "    # A prepositional phrase is typically ended by a noun.\n",
    "    # Therefore, when we run into a noun, we end the phrase.\n",
    "    # We must also check that it is the last of the first noun(s)\n",
    "    # we encounter.\n",
    "    def last_noun(self, i):\n",
    "        if bool(\n",
    "            # 1. End\n",
    "            i >= len(self.entities) or \n",
    "            \n",
    "            # 2. Noun\n",
    "            self.entities[i].start().pos_ not in [\n",
    "                \"NOUN\", \n",
    "                \"PROPN\", \n",
    "                \"PRON\"\n",
    "            ]\n",
    "        ):\n",
    "            return False\n",
    "        \n",
    "        return bool(\n",
    "            i + 1 > len(self.entities) - 1 or \n",
    "            (\n",
    "                self.entities[i+1].size() == 1 and \n",
    "                self.entities[i+1].start().pos_ not in [\n",
    "                    \"NOUN\", \n",
    "                    \"PROPN\", \n",
    "                    \"PRON\", \n",
    "                    \"PART\"\n",
    "                ]\n",
    "            )\n",
    "        )\n",
    "    \n",
    "    def end(self, i):\n",
    "        return bool(\n",
    "            # 1. End of List\n",
    "            i + 1 >= len(self.entities) or\n",
    "            \n",
    "            # 2. Clause\n",
    "            self.entities[i+1].label in [\n",
    "                Entity.COLON,\n",
    "                Entity.COLON_BREAK,\n",
    "                Entity.I_CLAUSE,\n",
    "                Entity.D_CLAUSE,\n",
    "                Entity.P_PHRASE\n",
    "            ] or\n",
    "            \n",
    "            # 3. Noun\n",
    "            self.last_noun(i)\n",
    "        )\n",
    "\n",
    "    def identify(self):    \n",
    "        i = 0\n",
    "        \n",
    "        while i < len(self.entities):\n",
    "            # Skip\n",
    "            if bool(\n",
    "                self.entities[i].size() != 1 or\n",
    "                self.entities[i].start().pos_ != \"ADP\"\n",
    "            ):\n",
    "                i += 1\n",
    "                continue\n",
    "\n",
    "            # Create Clause\n",
    "            self.entities[i].label = Entity.P_PHRASE\n",
    "            while not self.end(i+1):\n",
    "                self.entities[i].r = self.entities[i+1].r\n",
    "\n",
    "                # Add Child\n",
    "                if self.entities[i+1].label in [Entity.BRACKETS, Entity.QUOTE]:\n",
    "                    self.entities[i].children.append(self.entities[i+1])\n",
    "                \n",
    "                self.entities.pop(i+1)\n",
    "\n",
    "            if self.last_noun(i+1):\n",
    "                self.entities[i].r = self.entities[i+1].r\n",
    "                self.entities.pop(i+1)\n",
    "            \n",
    "            i += 1\n",
    "        \n",
    "        return self.entities   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5f32b871-cb89-4dda-91a5-126142a567cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Lists:\n",
    "    NOUNS = [\"NOUN\", \"PRON\", \"PROPN\"]\n",
    "    \n",
    "    def __init__(self, main, entities):\n",
    "        self.main = main\n",
    "        self.entities = [*entities]\n",
    "        self.separator = None\n",
    "\n",
    "    def is_stop(self, entity):\n",
    "        is_break = entity.label == Entity.BREAK and entity.lower()[0] == self.separator\n",
    "        is_clause = entity.label in [\n",
    "            Entity.I_CLAUSE, \n",
    "            Entity.D_CLAUSE, \n",
    "            Entity.P_PHRASE,\n",
    "            Entity.COLON,\n",
    "            Entity.COLON_BREAK\n",
    "        ]\n",
    "        return is_break or is_clause\n",
    "\n",
    "    def find_lists(self, sep):\n",
    "        self.separator = sep\n",
    "        \n",
    "        lists = [\n",
    "            [\n",
    "                [None, None]\n",
    "            ]\n",
    "        ]\n",
    "\n",
    "        i = 0\n",
    "        while i < len(self.entities):\n",
    "            entity = self.entities[i]\n",
    "\n",
    "            opened = lists[-1][0] != [None, None]\n",
    "            remove_list = entity.label in [Entity.COLON, Entity.COLON_BREAK]\n",
    "            close_list = entity.label in [Entity.AND_OR_END] and entity.lower()[0] == sep\n",
    "            close_item = entity.label in [Entity.BREAK] and entity.lower() == sep\n",
    "        \n",
    "            # Close List\n",
    "            if opened and close_list:\n",
    "                # Invalid List, Remove\n",
    "                if len(lists[-1]) < 2:\n",
    "                    lists[-1] = [[None, None]]\n",
    "                    i += 1\n",
    "                    continue\n",
    "                    \n",
    "                # Find the L Index of Last Item\n",
    "                last_item_l = i + 1\n",
    "\n",
    "                # Find the R Index of Last Item\n",
    "                last_item_r = last_item_l\n",
    "                \n",
    "                length = find_index(self.entities[last_item_l:], lambda e: self.is_stop(e))\n",
    "                if length > 0:\n",
    "                    last_item_r += length - 1\n",
    "                elif length == -1:\n",
    "                    last_item_r = len(self.entities) - 1\n",
    "\n",
    "                # Add Last Item\n",
    "                lists[-1].append([last_item_l, last_item_r])\n",
    "                lists.append([[None, None]])\n",
    "                i += 1\n",
    "\n",
    "            # Close Item\n",
    "            elif opened and close_item:\n",
    "                lists[-1].append([i + 1, i])\n",
    "                i += 1\n",
    "                \n",
    "            # Remove List\n",
    "            elif opened and remove_list:\n",
    "                lists[-1] = [[None, None]]\n",
    "                i += 1\n",
    "            \n",
    "            # Continue Item\n",
    "            else:\n",
    "                if not opened:\n",
    "                    lists[-1][0] = [i, i]\n",
    "                else:\n",
    "                    lists[-1][-1][1] += 1\n",
    "                i += 1\n",
    "        \n",
    "        # If we reach the end of the list and the last\n",
    "        # list is invalid (< 3 items), we remove it.\n",
    "        if bool(\n",
    "            lists and len(lists[-1]) < 3 or \n",
    "            (\n",
    "                lists and\n",
    "                not find(self.entities[lists[-1][0][0]:], lambda e: e.label == Entity.AND_OR_END and e.lower()[0] == sep)\n",
    "            )\n",
    "        ):\n",
    "            lists.pop()\n",
    "        \n",
    "        # In each item, we look for pairs (e.g. X and Y).\n",
    "        # We only handle one conjunction.\n",
    "        num_lists = len(lists)\n",
    "        for i, lst in enumerate(lists):\n",
    "            if i >= num_lists:\n",
    "                break\n",
    "            \n",
    "            for l, r in lst:\n",
    "                tokens = Entity.tokens(ents=self.entities[l:r+1])\n",
    "                conj = find_all(tokens, lambda t: Entity.is_conjunction(t))\n",
    "                if len(conj) == 1:\n",
    "                    lists.append([[l, r]])\n",
    "\n",
    "        # If there's no lists at all, we check if there's a pairing.\n",
    "        # We should divvy the entities up by any separators, but\n",
    "        # pairs aren't of too much importance.\n",
    "        # TODO: ADD ABOVE FUNCTIONALITY\n",
    "        tokens = Entity.tokens(ents=self.entities)\n",
    "        num_conj = len(find_all(tokens, lambda t: Entity.is_conjunction(t)))\n",
    "        if not lists and num_conj == 1:\n",
    "            lists.append([[0, len(self.entities) - 1]])\n",
    "\n",
    "        # Here we remove duplicates, I'm not sure if duplicates still\n",
    "        # occur, I observed them once, but this is here in case.\n",
    "        i = 0\n",
    "        while i < len(lists):\n",
    "            if lists[i] in lists[i+1:]:\n",
    "                lists.pop(i)\n",
    "            else:\n",
    "                i += 1\n",
    "\n",
    "        # Remove Invalid Lists\n",
    "        i = 0\n",
    "        while i < len(lists):\n",
    "            # The list contains one item and that item only contains one\n",
    "            # token, or the list has two items.\n",
    "            if bool(\n",
    "                (\n",
    "                    len(lists[i]) == 1 and \n",
    "                    lists[i][0][0] == lists[i][0][1]\n",
    "                ) or\n",
    "                len(lists[i]) == 2\n",
    "            ):\n",
    "                lists.pop(i)\n",
    "            else:\n",
    "                i += 1\n",
    "         \n",
    "        return lists\n",
    "\n",
    "    def clean_lists(self, lists):\n",
    "        overlaps = []\n",
    "\n",
    "        i = 0\n",
    "        while i + 1 < len(lists):\n",
    "            a = lists[i]\n",
    "            b = lists[i+1]\n",
    "                  \n",
    "            if a[-1] != b[0]:\n",
    "                i += 1\n",
    "                continue\n",
    "\n",
    "            if len(a) <= 1 or len(b) <= 1:\n",
    "                i += 1\n",
    "                continue\n",
    "\n",
    "            # No Way to Split\n",
    "            if a[-1][1] - a[-1][0] <= 1:\n",
    "                overlaps.extend([i, i + 1])\n",
    "                i += 2\n",
    "            else:\n",
    "                a[-1][1] = a[-1][0]\n",
    "                b[0][0] = b[0][1]\n",
    "                i += 2\n",
    "        \n",
    "        lists = [l for i, l in enumerate(lists) if i not in overlaps]\n",
    "        return lists\n",
    "\n",
    "    def expand_noun(self, tokens, start, direction):\n",
    "        for group in [*self.main.sp_doc.noun_chunks, *self.main.sp_doc.ents]:\n",
    "            tokens_i = [t.i for t in group]\n",
    "            if tokens[start].i in tokens_i:\n",
    "                while start >= 0 and start < len(tokens) and tokens[start].i in tokens_i:\n",
    "                    start += 1 * direction\n",
    "                start += 1 * direction * -1\n",
    "            break\n",
    "        \n",
    "        return start\n",
    "        \n",
    "    def char_bound_list(self, lst):\n",
    "        # print(\"Character Bound List\")\n",
    "        # We bound each item according to characters or a speech.\n",
    "        # We find these bounds from the \"base item\", the second to last item.\n",
    "        base_tokens = Entity.tokens(ents=self.entities[lst[-2][0]:lst[-2][1]+1])\n",
    "        \n",
    "        # As we're bounding by characters, primarily, the left bound is just\n",
    "        # the characters of the first token\n",
    "        l_bound = base_tokens[0].lower_\n",
    "\n",
    "        # The right bound is the first tag, of the below set of tags, that we\n",
    "        # encounter in the base tokens. If there's not such a token, we cannot\n",
    "        # bound the items.\n",
    "        speech = [\"NOUN\", \"PROPN\", \"PRON\", \"VERB\", \"NUM\"]\n",
    "        r_bound = None\n",
    "        for i in range(len(base_tokens) - 1, -1, -1):\n",
    "            if base_tokens[i].pos_ in speech:\n",
    "                r_bound = base_tokens[i]\n",
    "                break\n",
    "\n",
    "        if not r_bound:\n",
    "            return None\n",
    "\n",
    "        # The inner items are already bounded on the left and right sides.\n",
    "        # All we need to check is whether the start matches with the left bound.\n",
    "        inner_items = lst[1:-2]\n",
    "\n",
    "        for i, item in enumerate(inner_items):\n",
    "            l = item[0]\n",
    "            r = item[1]\n",
    "            \n",
    "            tokens = Entity.tokens(ents=self.entities[l:r+1])\n",
    "\n",
    "            # If it doesn't match, we check if the next set of items can be\n",
    "            # bounded. If not, we cannot bound the list.\n",
    "            if tokens[0].lower_ != l_bound:\n",
    "                if len(inner_items) - i - 1 >= 2:\n",
    "                    return self.bound_list(lst[i+2:])\n",
    "                return None\n",
    "            \n",
    "        # Check for L Bound in Starting Item\n",
    "        start_tokens = Entity.tokens(ents=self.entities[lst[0][0]:lst[0][1]+1])\n",
    "        start_l = len(start_tokens) - 1\n",
    "        while start_l >= 0 and start_tokens[start_l].lower_ != l_bound:\n",
    "            start_l -= 1\n",
    "\n",
    "        # L Bound Not Found\n",
    "        if start_l < 0:\n",
    "            # If the list is greater than 4 items, we can\n",
    "            # cut off the starting item, and try again.\n",
    "            if len(inner_items) >= 2:\n",
    "                return self.bound_list(lst[1:])\n",
    "            return None\n",
    "\n",
    "        # If the first of the start tokens is a noun, there may be more\n",
    "        # to include.\n",
    "        if start_tokens[start_l].pos_ in Lists.NOUNS:\n",
    "            start_l = self.expand_noun(start_tokens, start_l, -1)\n",
    "                    \n",
    "        # Check for R Bound in Ending Item\n",
    "        end_tokens = Entity.tokens(ents=self.entities[lst[-1][0]:lst[-1][1]+1])\n",
    "        end_r = 0\n",
    "        num_end_tokens = len(end_tokens)\n",
    "        while end_r < num_end_tokens and end_tokens[end_r].pos_ not in speech:\n",
    "            end_r += 1\n",
    "\n",
    "        if end_r >= num_end_tokens:\n",
    "            return None\n",
    "\n",
    "        # If the last of the end tokens is a noun, there may be more\n",
    "        # to include.\n",
    "        if end_tokens[end_r].pos_ in Lists.NOUNS:\n",
    "            end_r = self.expand_noun(end_tokens, end_r, 1)\n",
    "        \n",
    "        # Create List\n",
    "        entity_start_item = Entity(self.main.sp_doc, label=Entity.ITEM, l=start_tokens[start_l].i, r=start_tokens[-1].i)\n",
    "        entity_end_item = Entity(self.main.sp_doc, label=Entity.ITEM, l=end_tokens[0].i, r=end_tokens[end_r].i)\n",
    "        \n",
    "        entity_list = Entity(self.main.sp_doc, label=Entity.LIST, l=start_tokens[start_l].i, r=end_tokens[end_r].i)\n",
    "        entity_list.children.extend([entity_start_item, entity_end_item])\n",
    "        \n",
    "        for item in lst[1:-1]:\n",
    "            tokens = Entity.tokens(ents=self.entities[item[0]:item[1]+1])\n",
    "            entity_item = Entity(self.main.sp_doc, label=Entity.ITEM, l=tokens[0].i, r=tokens[-1].i)\n",
    "            entity_list.children.append(entity_item)\n",
    "\n",
    "        return entity_list\n",
    "            \n",
    "    def bound_list(self, lst):\n",
    "        # print(\"Bound List\")\n",
    "        # Base Item (2nd to Last Item) Tokens\n",
    "        base_tokens = Entity.tokens(ents=self.entities[lst[-2][0]:lst[-2][1]+1])\n",
    "        num_base_tokens = len(base_tokens)\n",
    "\n",
    "        # print(f\"Base Tokens: {base_tokens}\")\n",
    "        # print(f\"Number Base Tokens: {num_base_tokens}\")\n",
    "        \n",
    "        # Bound\n",
    "        speech = [\"NOUN\", \"PROPN\", \"PRON\", \"VERB\", \"NUM\"]\n",
    "\n",
    "        # Find L Bound\n",
    "        l_bound = None\n",
    "        for i in range(0, num_base_tokens):\n",
    "            if base_tokens[i].pos_ in speech:\n",
    "                l_bound = base_tokens[i]\n",
    "                break\n",
    "\n",
    "        if not l_bound:\n",
    "            return None\n",
    "        \n",
    "        # Find R Bound\n",
    "        r_bound = None\n",
    "        for i in range(num_base_tokens - 1, -1, -1):\n",
    "            if base_tokens[i].pos_ in speech:\n",
    "                r_bound = base_tokens[i]\n",
    "                break\n",
    "\n",
    "        if not r_bound:\n",
    "            return None\n",
    "\n",
    "        # print(f\"L Bound Speech: {l_bound.pos_}\")\n",
    "        # print(f\"R Bound Speech: {r_bound.pos_}\")\n",
    "        \n",
    "        # Check Inner Items\n",
    "        # The inner items must have the left bound,\n",
    "        # the right bound isn't as important.\n",
    "        inner_items = lst[1:-1]\n",
    "\n",
    "        verb_seen = False\n",
    "        for i, item in enumerate(inner_items):\n",
    "            l = item[0]\n",
    "            r = item[1]\n",
    "            \n",
    "            item_tokens = Entity.tokens(ents=self.entities[l:r+1])\n",
    "            item_speech = [token.pos_ for token in item_tokens]\n",
    "\n",
    "            # Must be Homogeneous\n",
    "            if \"VERB\" not in item_speech and verb_seen:\n",
    "                if len(inner_items) >= 2:\n",
    "                    return self.bound_list(lst[1:])  \n",
    "                else:\n",
    "                    return None\n",
    "            elif \"VERB\" in item_speech:\n",
    "                verb_seen = True\n",
    "\n",
    "            # Not Found\n",
    "            if l_bound.pos_ not in item_speech:\n",
    "                # We check if the list starting at the next\n",
    "                # item has a chance. If it does, that becomes\n",
    "                # the list.\n",
    "                if len(inner_items) - i + 1 >= 2:\n",
    "                    return self.bound_list(lst[i+2:])\n",
    "                return None\n",
    "        \n",
    "        # Check Starting Item\n",
    "        start_tokens = Entity.tokens(ents=self.entities[lst[0][0]:lst[0][1]+1])\n",
    "        start_l = len(start_tokens) - 1\n",
    "        \n",
    "        while start_l >= 0 and not Entity.same_speech(start_tokens[start_l].pos_, l_bound.pos_):\n",
    "            start_l -= 1\n",
    "\n",
    "        if start_l < 0:\n",
    "            if len(inner_items) >= 2:\n",
    "                return self.bound_list(lst[1:])\n",
    "            return None\n",
    "\n",
    "        # Adjust Starting Item\n",
    "        if l_bound.pos_ in Lists.NOUNS:\n",
    "            start_l = self.expand_noun(start_tokens, start_l, -1)\n",
    "        \n",
    "        # Check Ending Item\n",
    "        end_tokens = Entity.tokens(ents=self.entities[lst[-1][0]:lst[-1][1]+1])\n",
    "        end_r = 0\n",
    "        num_end_tokens = len(end_tokens)\n",
    "\n",
    "        while end_r < num_end_tokens and not Entity.same_speech(end_tokens[end_r].pos_, r_bound.pos_):\n",
    "            end_r += 1\n",
    "\n",
    "        if end_r >= num_end_tokens:\n",
    "            return None\n",
    "\n",
    "        # Adjust Ending Item\n",
    "        if r_bound.pos_ in Lists.NOUNS:\n",
    "            end_r = self.expand_noun(end_tokens, end_r, 1)\n",
    "\n",
    "        # Create List\n",
    "        entity_start_item = Entity(self.main.sp_doc, label=Entity.ITEM, l=start_tokens[start_l].i, r=start_tokens[-1].i)\n",
    "        entity_end_item = Entity(self.main.sp_doc, label=Entity.ITEM, l=end_tokens[0].i, r=end_tokens[end_r].i)\n",
    "        \n",
    "        entity_list = Entity(self.main.sp_doc, label=Entity.LIST, l=start_tokens[start_l].i, r=end_tokens[end_r].i)\n",
    "        entity_list.children.extend([entity_start_item, entity_end_item])\n",
    "\n",
    "        for item in lst[1:-1]:\n",
    "            tokens = Entity.tokens(ents=self.entities[item[0]:item[1]+1])\n",
    "            entity_item = Entity(self.main.sp_doc, label=Entity.ITEM, l=tokens[0].i, r=tokens[-1].i)\n",
    "            entity_list.children.append(entity_item)\n",
    "\n",
    "        return entity_list\n",
    "\n",
    "    def char_bound_pair(self, pair):\n",
    "        # print(\"Character Bound Pair\")\n",
    "        tokens = Entity.tokens(ents=self.entities[pair[0][0]:pair[0][1]+1])\n",
    "        num_tokens = len(tokens)\n",
    "        \n",
    "        m = find_index(tokens, lambda t: Entity.is_conjunction(t))\n",
    "        l = m - 1\n",
    "        r = m + 1\n",
    "\n",
    "        # Bound L by R Token Characters\n",
    "        i = m - 1\n",
    "        while i >= 0 and tokens[i].lower_ != tokens[m + 1].lower_:\n",
    "            i -= 1\n",
    "\n",
    "        if i < 0:\n",
    "            return None\n",
    "\n",
    "        # Bound R by L Token Speech\n",
    "        j =  m + 1\n",
    "        while j < num_tokens and not Entity.same_speech(tokens[m-1].pos_, tokens[j].pos_):\n",
    "            j += 1\n",
    "\n",
    "        if j >= num_tokens:\n",
    "            return None\n",
    "        \n",
    "        e_item_l = Entity(self.main.sp_doc, label=Entity.ITEM, l=tokens[i].i, r=tokens[m-1].i)\n",
    "        e_item_r = Entity(self.main.sp_doc, label=Entity.ITEM, l=tokens[m+1].i, r=tokens[j].i)\n",
    "        e_list = Entity(self.main.sp_doc, label=Entity.LIST, l=tokens[i].i, r=tokens[j].i, children=[e_item_l, e_item_r])\n",
    "        return e_list\n",
    "    \n",
    "    def bound_pair(self, pair):\n",
    "        # print(\"Bound Pair\")\n",
    "        tokens = Entity.tokens(ents=self.entities[pair[0][0]:pair[0][1]+1])\n",
    "        tokens = sorted(tokens, key=lambda t: t.i)\n",
    "        num_tokens = len(tokens)\n",
    "        \n",
    "        m = find_index(tokens, lambda t: Entity.is_conjunction(t))\n",
    "        l = m - 1\n",
    "        r = m + 1\n",
    "\n",
    "        speech = [\"NOUN\", \"PROPN\", \"PRON\", \"VERB\", \"NUM\"]\n",
    "\n",
    "        # Find L Bound\n",
    "        l_bound = None\n",
    "        l_bound_i = None\n",
    "        for i in range(m + 1, num_tokens):\n",
    "            if tokens[i].pos_ in speech:\n",
    "                l_bound = tokens[i].pos_\n",
    "                l_bound_i = tokens[i].i\n",
    "                break\n",
    "\n",
    "        if not l_bound:\n",
    "            return None\n",
    "\n",
    "        # Find R Bound\n",
    "        r_bound = None\n",
    "        r_bound_i = None\n",
    "        for i in range(m - 1, -1, -1):\n",
    "            if tokens[i].pos_ in speech:\n",
    "                r_bound = tokens[i].pos_\n",
    "                r_bound_i = tokens[i].i\n",
    "                break\n",
    "\n",
    "        if not r_bound:\n",
    "            return None\n",
    "\n",
    "        # Bound L\n",
    "        while l >= 0 and not Entity.same_speech(tokens[l].pos_, l_bound):\n",
    "            l -= 1\n",
    "\n",
    "        if l < 0:\n",
    "            return None\n",
    "\n",
    "        # Adjust L if Noun\n",
    "        if l_bound in Lists.NOUNS:\n",
    "            l = self.expand_noun(tokens, l, -1)\n",
    "            \n",
    "        # Bound R\n",
    "        while r < num_tokens and not Entity.same_speech(tokens[r].pos_, r_bound):\n",
    "            r += 1\n",
    "        \n",
    "        if r >= num_tokens:\n",
    "            return None\n",
    "\n",
    "        # Adjust R if Noun\n",
    "        if r_bound in Lists.NOUNS:\n",
    "            r = self.expand_noun(tokens, r, 1)\n",
    "        \n",
    "        e_item_l = Entity(self.main.sp_doc, label=Entity.ITEM, l=tokens[l].i, r=r_bound_i)\n",
    "        e_item_r = Entity(self.main.sp_doc, label=Entity.ITEM, l=l_bound_i, r=tokens[r].i)\n",
    "\n",
    "        e_list = Entity(self.main.sp_doc, label=Entity.LIST, l=tokens[l].i, r=tokens[r].i)\n",
    "        e_list.children.extend([e_item_l, e_item_r])\n",
    "        \n",
    "        return e_list\n",
    "\n",
    "    def bound_lists(self, lists):\n",
    "        # print(f\"Lists: {lists}\")\n",
    "        # for items in lists:\n",
    "        #     print(f\"\\tList: {items}\")\n",
    "        #     for item in items:\n",
    "        #         print(f\"\\t\\tItem: {Entity.tokens(ents=self.entities[item[0]:item[1]+1])}\")\n",
    "        \n",
    "        bound_lists = []\n",
    "        \n",
    "        for lst in lists:\n",
    "            bound = None\n",
    "        \n",
    "            if len(lst) == 1:\n",
    "                bound = self.char_bound_pair(lst)\n",
    "                if not bound:\n",
    "                    bound = self.bound_pair(lst)\n",
    "            else:\n",
    "                bound = self.char_bound_list(lst)\n",
    "                if not bound:\n",
    "                    bound = self.bound_list(lst)\n",
    "            \n",
    "            if bound:\n",
    "                bound_lists.append(bound)\n",
    "\n",
    "        # print(f\"Bounded Lists: {bound_lists}\")\n",
    "        # for bound_list in bound_lists:\n",
    "        #     print(f\"\\tBound List ({bound_list.l}, {bound_list.r}): {bound_list.span()}\")\n",
    "        \n",
    "        # Map (L, R) to Entity List\n",
    "        mapped_bounds = {}\n",
    "        for lst in bound_lists:\n",
    "            mapped_bounds[(lst.l, lst.r)] = lst\n",
    "        bounds = list(mapped_bounds.keys())\n",
    "\n",
    "        # Find Largest Coverage of Bounds\n",
    "        max_coverage = []\n",
    "        \n",
    "        for bound in bounds:\n",
    "            overlap = False\n",
    "            for i, max_bound in enumerate(max_coverage):\n",
    "                contains = max_bound[0] <= bound[0] <= max_bound[1] or max_bound[0] <= bound[1] <= max_bound[1]\n",
    "                surround = bound[0] <= max_bound[0] <= bound[1] or bound[0] <= max_bound[1] <= bound[1]\n",
    "                \n",
    "                if contains or surround:\n",
    "                    overlap = True\n",
    "                \n",
    "                    if bound[1] - bound[0] > max_bound[1] - max_bound[0]:\n",
    "                        max_coverage[i] = bound\n",
    "            \n",
    "            if not overlap:\n",
    "                max_coverage.append(bound)\n",
    "\n",
    "        # print(f\"(Max Coverage) Bounds: {max_coverage}\")\n",
    "        # print(f\"Entities: {[(e.l, e.r) for e in self.entities]}\")\n",
    "        \n",
    "        # Integrate Lists\n",
    "        for bound in max_coverage:\n",
    "            l_overlap = None\n",
    "            l_overlap_i = None\n",
    "            \n",
    "            r_overlap = None\n",
    "            r_overlap_i = None\n",
    "            \n",
    "            i = 0\n",
    "            while i < len(self.entities):\n",
    "                entity = self.entities[i]\n",
    "                \n",
    "                # Overlap w/ Left\n",
    "                if not l_overlap and entity.l <= bound[0] <= entity.r:\n",
    "                    l_overlap = entity\n",
    "                    l_overlap_i = i\n",
    "    \n",
    "                # Overlap w/ Right\n",
    "                if entity.l <= bound[1] <= entity.r:\n",
    "                    r_overlap = entity\n",
    "                    r_overlap_i = i\n",
    "\n",
    "                if l_overlap and r_overlap:\n",
    "                    break\n",
    "\n",
    "                i += 1\n",
    "\n",
    "            # if l_overlap:\n",
    "            #     print(f\"L Overlap: {l_overlap.span()}, {l_overlap.label}\")\n",
    "\n",
    "            # if r_overlap:\n",
    "            #     print(f\"R Overlap: {r_overlap.span()}, {r_overlap.label}\")\n",
    "            \n",
    "            if l_overlap.label in [Entity.I_CLAUSE, Entity.D_CLAUSE, Entity.P_PHRASE]:\n",
    "                if l_overlap.l == mapped_bounds[bound].l:\n",
    "                    # Replace (Not in Use)\n",
    "                    # self.entities = self.entities[:l_overlap_i] + self.entities[r_overlap_i+1:]\n",
    "                    # self.entities.insert(l_overlap_i, mapped_bounds[bound])\n",
    "\n",
    "                    # Add Children\n",
    "                    l_overlap.r = max(l_overlap.r, mapped_bounds[bound].r)\n",
    "                    l_overlap.children.append(mapped_bounds[bound])\n",
    "                    self.entities = self.entities[:l_overlap_i+1] + self.entities[r_overlap_i+1:]\n",
    "                else:\n",
    "                    # Split (Not in Use)\n",
    "                    # l_overlap.r = mapped_bounds[bound].l - 1\n",
    "                    # self.entities = self.entities[:l_overlap_i+1] + self.entities[r_overlap_i+1:]\n",
    "                    # self.entities.insert(l_overlap_i + 1, mapped_bounds[bound])\n",
    "                    \n",
    "                    # Add Children\n",
    "                    l_overlap.r = max(l_overlap.r, mapped_bounds[bound].r)\n",
    "                    l_overlap.children.append(mapped_bounds[bound])\n",
    "                    self.entities = self.entities[:l_overlap_i+1] + self.entities[r_overlap_i+1:]\n",
    "                    \n",
    "            else:\n",
    "                self.entities = self.entities[:l_overlap_i] + self.entities[r_overlap_i+1:]\n",
    "                self.entities.insert(l_overlap_i, mapped_bounds[bound])\n",
    "        \n",
    "        return self.entities\n",
    "        \n",
    "    def identify(self, sep):\n",
    "        lists = self.find_lists(sep)\n",
    "        lists = self.clean_lists(lists)\n",
    "        lists = self.bound_lists(lists)   \n",
    "        return lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "92fe9a25-ea3b-4a12-a45e-6d75f5582c8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Parts:\n",
    "    def __init__(self, main):\n",
    "        self.main = main\n",
    "        self.root = Entity(self.main.sp_doc)\n",
    "        self.reg = []\n",
    "\n",
    "    def matches(self, token_i, sent_i, label):\n",
    "        sent_reg = self.reg[sent_i]\n",
    "        for k, v in sent_reg:\n",
    "            if k[0] <= token_i <= k[1]:\n",
    "                return v.label == label\n",
    "        return False\n",
    "\n",
    "    def load_registry(self, ent):\n",
    "        reg = {(ent.l, ent.r): ent}\n",
    "        for child in ent.children:\n",
    "            if not child.label:\n",
    "                continue\n",
    "            reg.update(self.load_registry(child))\n",
    "        return reg\n",
    "    \n",
    "    def update(self):\n",
    "        reg = []\n",
    "        for sent in self.main.sp_doc.sents:\n",
    "            tokens = list(sent)\n",
    "            # print(f\"Parsing Sentence: {tokens}\")\n",
    "            # print(f\"\\tSize: {len(tokens)}\")\n",
    "            reg.append(self.load_entities(tokens))\n",
    "        self.reg = reg\n",
    "    \n",
    "    def load_entities(self, tokens):\n",
    "        entities = []\n",
    "        for token in tokens:\n",
    "            entity = Entity(\n",
    "                self.main.sp_doc, \n",
    "                l=token.i, \n",
    "                r=token.i\n",
    "            )\n",
    "            entities.append(entity)\n",
    "\n",
    "        entities = Quotes(self.main, entities).identify()\n",
    "        entities = Brackets(self.main, entities).identify()\n",
    "        entities = Separators(self.main, entities).identify()\n",
    "        \n",
    "        sep = \",\"\n",
    "        for entity in entities:\n",
    "            if \";\" == entity.lower()[0]:\n",
    "                sep = \";\"\n",
    "                break\n",
    "        \n",
    "        entities = Colons(self.main, entities).identify()\n",
    "        entities = Dependent_Clauses(self.main, entities).identify(sep)\n",
    "        entities = Independent_Clauses(self.main, entities).identify([Entity.END])\n",
    "        entities = Prepositional_Phrases(self.main, entities).identify()\n",
    "        entities = Lists(self.main, entities).identify(sep)\n",
    "        entities = Independent_Clauses(self.main, entities).identify([Entity.AND_OR_END])\n",
    "        \n",
    "        # Creating Registry\n",
    "        # The top entity contains the entities above for\n",
    "        # convenience.\n",
    "        top = Entity(self.main.sp_doc, l=-1, r=-1, children=entities)\n",
    "        \n",
    "        reg = self.load_registry(top)\n",
    "        del reg[(-1, -1)]\n",
    "        \n",
    "        return reg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5bc336a4-657a-4704-be40-342b996dc423",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Base:\n",
    "    # There is not a defined conversion method for these words.\n",
    "    # This is the default list of irregular nouns. It maps the\n",
    "    # the singular version to the plural version (SP).\n",
    "    IRREGULAR_NOUNS_SP = {\n",
    "        \"ox\": \"oxen\",\n",
    "        \"goose\": \"geese\",\n",
    "        \"mouse\": \"mice\",\n",
    "        \"bacterium\": \"bacteria\"\n",
    "    }\n",
    "\n",
    "    # This is the reversed version of the dictionary above, meaning \n",
    "    # that the plural version is mapped to the singular version \n",
    "    # (PS).\n",
    "    IRREGULAR_NOUNS_PS = {v: k for k, v in IRREGULAR_NOUNS_SP.items()}\n",
    "    \n",
    "    # The singular and plural versions of these words are the same. \n",
    "    # This is the default list of zero plural nouns.\n",
    "    ZERO_PLURAL_NOUNS = [\n",
    "        \"species\", \n",
    "        \"deer\", \n",
    "        \"fish\", \n",
    "        \"moose\", \n",
    "        \"sheep\", \n",
    "        \"swine\", \n",
    "        \"buffalo\", \n",
    "        \"trout\", \n",
    "        \"cattle\"\n",
    "    ]\n",
    "\n",
    "    # These pairs of characters define symbols that enclose other\n",
    "    # information in a text.\n",
    "    ENCLOSURES = {\n",
    "        \"(\": \")\",\n",
    "        \"[\": \"]\",\n",
    "        \"{\": \"}\"\n",
    "    }\n",
    "\n",
    "    LAX_ENCLOSURES = {\n",
    "        \"(\": \")\",\n",
    "        \"[\": \"]\",\n",
    "        \"{\": \"}\",\n",
    "        \"—\": \"—\"\n",
    "    }\n",
    "\n",
    "\n",
    "    \n",
    "    def __init__(self, main, irregular_nouns_sp=IRREGULAR_NOUNS_SP, irregular_nouns_ps=IRREGULAR_NOUNS_PS, zero_plural_nouns=ZERO_PLURAL_NOUNS):\n",
    "        self.main = main\n",
    "        self.zero_plural_nouns = zero_plural_nouns\n",
    "        self.irregular_nouns_sp = irregular_nouns_sp\n",
    "        self.irregular_nouns_ps = irregular_nouns_ps\n",
    "        self.irregular_plural_nouns = list(self.irregular_nouns_sp.values())\n",
    "        self.irregular_singular_nouns = list(self.irregular_nouns_sp.keys())\n",
    "\n",
    "\n",
    "\n",
    "    def delete_extra_whitespace(self, string):\n",
    "        # Duplicate spaces, spaces before punctuation marks,\n",
    "        # and outside spaces are removed.\n",
    "        string = re.sub(r\"\\s+\", \" \", string)\n",
    "        string = re.sub(r\"\\s+([?.!,])\", r\"\\1\", string)\n",
    "        string = string.strip()\n",
    "        return string\n",
    "\n",
    "\n",
    "\n",
    "    def delete_outer_non_alnum(self, string):\n",
    "        while string:\n",
    "            start_len = len(string)\n",
    "            # Remove Leading Non-Alphanumeric Character\n",
    "            if string and not string[0].isalnum():\n",
    "                string = string[1:]\n",
    "            # Remove Trailing Non-Alphanumeric Character\n",
    "            if string and not string[-1].isalnum():\n",
    "                string = string[:-1]\n",
    "            # No Changes Made\n",
    "            if start_len == len(string):\n",
    "                break\n",
    "        return string\n",
    "\n",
    "\n",
    "\n",
    "    def get_parentheticals(self, text, enclosures=ENCLOSURES, flatten=False):\n",
    "        # The parenthetical would be the content inside of a pair\n",
    "        # of matching parentheses, brackets, or braces.\n",
    "        parentheticals = []\n",
    "        \n",
    "        # This contains the text that's not inside of any\n",
    "        # enclosure.\n",
    "        base_text = []\n",
    "        \n",
    "        # This is used for building groups, which often has a \n",
    "        # nested structure.\n",
    "        stacks = []\n",
    "        \n",
    "        # These are the pairs of characters that we recognize\n",
    "        # as defining the parenthetical.\n",
    "        openers = list(enclosures.keys())\n",
    "        closers = list(enclosures.values())\n",
    "        \n",
    "        # This contains the opening characters of the groups \n",
    "        # that are currently open (e.g. '(', '['). We use it \n",
    "        # so that we know whether to open or close a group.\n",
    "        opened = []\n",
    "        \n",
    "        for i, char in enumerate(text):\n",
    "            # Open Group\n",
    "            if char in openers:\n",
    "                stacks.append([])\n",
    "                opened.append(char)\n",
    "            # Close Group\n",
    "            elif opened and char == enclosures.get(opened[-1], \"\"):\n",
    "                parentheticals.append(stacks.pop())\n",
    "                opened.pop()\n",
    "            # Add to Group\n",
    "            elif opened:\n",
    "                stacks[-1].append(i)\n",
    "            # Add to Base Text\n",
    "            else:\n",
    "                base_text.append(i)\n",
    "        \n",
    "        # We close the remaining groups that have not\n",
    "        # been closed.\n",
    "        while stacks:\n",
    "            parentheticals.append(stacks.pop())\n",
    "            \n",
    "        # Cluster Groups' Indices\n",
    "        # A list in the lists of indices (where each list represents a group of text) could have \n",
    "        # an interruption (e.g. [0, 1, 2, 10 15]) because of a parenthetical. So, we cluster the\n",
    "        # indices in each list to make the output more useful (e.g. [(0, 3), (10, 16)]).\n",
    "        lists_of_indices = [*parentheticals, base_text]        \n",
    "        lists_of_clustered_indices = []\n",
    "\n",
    "        for list_of_indices in lists_of_indices:\n",
    "            if not list_of_indices:\n",
    "                continue\n",
    "\n",
    "            # We start off with a single cluster that is made up of the\n",
    "            # first index. If the next index follows the first index, \n",
    "            # we continue the cluster. If it doesn't, we create a new cluster.\n",
    "            clustered_indices = [[list_of_indices[0], list_of_indices[0] + 1]]\n",
    "            \n",
    "            for index in list_of_indices[1:]:\n",
    "                if clustered_indices[-1][1] == index:\n",
    "                    clustered_indices[-1][1] = index + 1\n",
    "                else:\n",
    "                    clustered_indices.append([index, index + 1])\n",
    "\n",
    "            # Add Clustered Indices\n",
    "            lists_of_clustered_indices.append(clustered_indices)\n",
    "            \n",
    "        if flatten:\n",
    "            flattened_clusters = []\n",
    "            # We are placing each cluster of indices into one list.\n",
    "            # This removes the context of the larger parenthetical,\n",
    "            # but the context may be cumbersome instead of useful.\n",
    "            for list_of_clustered_indices in lists_of_clustered_indices:\n",
    "                for clustered_indices in list_of_clustered_indices:\n",
    "                    flattened_clusters.append(clustered_indices)\n",
    "            lists_of_clustered_indices = flattened_clusters\n",
    "        \n",
    "        return lists_of_clustered_indices\n",
    "\n",
    "\n",
    "\n",
    "    def separate_span_by_parenthetical(self, span):\n",
    "        span_parentheticals = []\n",
    "        \n",
    "        # The clusters of the span represented with tuples of char indices\n",
    "        # (e.g. [(0, 1), (1, 5), (5, 10)]. This is a list of clustered\n",
    "        # indices (like above).\n",
    "        text_clusters = self.get_parentheticals(span.text, flatten=True)\n",
    "        \n",
    "        for cluster in text_clusters:\n",
    "            if span.text[cluster[0]:cluster[1]].isspace():\n",
    "                continue\n",
    "\n",
    "            l_char_index = span[0].idx + cluster[0]\n",
    "            r_char_index = span[0].idx + cluster[1] - 1\n",
    "\n",
    "            # Instead of having a tuple dictating the start and end of a cluster,\n",
    "            # we can use a span -- it's much simpler.\n",
    "            cluster_as_span = self.get_span_at_indices(l_char_index, r_char_index)\n",
    "            if not cluster_as_span:\n",
    "                continue\n",
    "            \n",
    "            span_parentheticals.append(cluster_as_span)\n",
    "\n",
    "        return span_parentheticals\n",
    "\n",
    "\n",
    "\n",
    "    def separate_spans_by_parenthetical(self, spans):\n",
    "        all_span_parentheticals = []\n",
    "        for span in spans:\n",
    "            all_span_parentheticals.extend(self.separate_span_by_parenthetical(span))\n",
    "        return all_span_parentheticals\n",
    "\n",
    "    \n",
    " \n",
    "    def singularize(self, string):\n",
    "        string = string.lower()\n",
    "        \n",
    "        # The string to singularize should not have any\n",
    "        # non-alphanumeric characters at the end, or else\n",
    "        # the algorithm will not work.\n",
    "        words = re.split(r\" \", string)\n",
    "\n",
    "        if not words:\n",
    "            return [string]\n",
    "\n",
    "        # If the last word in the string is a zero plural\n",
    "        # or a singular irregular noun, there's no changes\n",
    "        # to make. For example, \"red sheep\" and \"ox\" are \n",
    "        # already singular.\n",
    "        if (\n",
    "            words[-1] in self.zero_plural_nouns or \n",
    "            words[-1] in self.irregular_singular_nouns\n",
    "        ):\n",
    "            return [string]\n",
    "\n",
    "        # If the last word in the string is an irregular\n",
    "        # plural noun, we rely on a dictionary with the\n",
    "        # corresponding mapping.\n",
    "        if words[-1] in self.irregular_plural_nouns:\n",
    "            words[-1] = self.irregular_nouns_ps[words[-1]]\n",
    "            singulars = [self.delete_extra_whitespace(\" \".join(words))]\n",
    "            return singulars\n",
    "        \n",
    "        # We take the singular form of the last word and\n",
    "        # add it back in to the other words. As there could\n",
    "        # be multiple forms (due to uncertainty), we need to\n",
    "        # include all possible versions.\n",
    "        singulars = []\n",
    "        singular_endings = self.get_singular(words[-1])\n",
    "\n",
    "        if not singular_endings:\n",
    "            return [string]\n",
    "        \n",
    "        for singular_ending in singular_endings:\n",
    "            singular = self.delete_extra_whitespace(\" \".join([*words[:-1], singular_ending]))\n",
    "            singulars.append(singular)\n",
    "            \n",
    "        return singulars\n",
    "\n",
    "\n",
    "\n",
    "    def get_singular(self, string):\n",
    "        versions = []\n",
    "\n",
    "        # Replace -ies with -y\n",
    "        if re.fullmatch(r\".*ies$\", string):\n",
    "            versions.append(f'{string[:-3]}y')\n",
    "            return versions\n",
    "\n",
    "        # Replace -ves with -f and -fe\n",
    "        if re.fullmatch(r\".*ves$\", string):\n",
    "            versions.append(f'{string[:-3]}f')\n",
    "            versions.append(f'{string[:-3]}fe')\n",
    "            return versions\n",
    "\n",
    "        # Delete -es \n",
    "        if re.fullmatch(r\".*es$\", string):\n",
    "            versions.append(f'{string[:-2]}')\n",
    "            return versions\n",
    "\n",
    "        # Replace -i with -us\n",
    "        if re.fullmatch(r\".*i$\", string):\n",
    "            versions.append(f'{string[:-1]}us')\n",
    "            return versions\n",
    "\n",
    "        # Delete -s\n",
    "        if re.fullmatch(r\".*s$\", string):\n",
    "            versions.append(f'{string[:-1]}')\n",
    "            return versions\n",
    "\n",
    "        return versions\n",
    "\n",
    "\n",
    "    \n",
    "    def pluralize(self, string):\n",
    "        string = string.lower()\n",
    "        \n",
    "        # The string to pluralize should not have any\n",
    "        # non-alphanumeric characters at the end, or else\n",
    "        # the algorithm will not work.\n",
    "        words = re.split(r\" \", string)\n",
    "\n",
    "        if not words:\n",
    "            return [string]\n",
    "\n",
    "        # If the last word in the string is a zero plural\n",
    "        # or a plural irregular noun, there's no changes\n",
    "        # to make. For example, \"red sheep\" and \"oxen\" are \n",
    "        # already singular.\n",
    "        if (\n",
    "            words[-1] in self.zero_plural_nouns or \n",
    "            words[-1] in self.irregular_plural_nouns\n",
    "        ):\n",
    "            return [string]\n",
    "\n",
    "        # If the last word in the string is an irregular\n",
    "        # singular noun, we rely on a dictionary with the\n",
    "        # corresponding mapping.\n",
    "        if words[-1] in self.irregular_singular_nouns:\n",
    "            words[-1] = self.irregular_nouns_sp[words[-1]]\n",
    "            return [self.delete_extra_whitespace(\" \".join(words))]\n",
    "        \n",
    "        # We take the singular form of the last word and\n",
    "        # add it back in to the other words. As there could\n",
    "        # be multiple forms (due to error), we need to\n",
    "        # handle them all.\n",
    "        plurals = []\n",
    "        plural_endings = self.get_plural(words[-1])\n",
    "\n",
    "        if not plural_endings:\n",
    "            return [string]\n",
    "            \n",
    "        for plural_ending in plural_endings:\n",
    "            plural = self.delete_extra_whitespace(\" \".join([*words[:-1], plural_ending]))\n",
    "            plurals.append(plural)\n",
    "            \n",
    "        return plurals\n",
    "\n",
    "    \n",
    "  \n",
    "    def get_plural(self, string):\n",
    "        versions = []\n",
    "\n",
    "        # Words that end with -us often have\n",
    "        # two different plural versions: -es and -i.\n",
    "        # For example, the plural version of cactus \n",
    "        # can be cactuses or cacti.\n",
    "        if re.fullmatch(r\".*us$\", string):\n",
    "            versions.append(f'{string}es')\n",
    "            versions.append(f'{string[:-2]}i')\n",
    "            return versions\n",
    "\n",
    "        # The -es ending is added to the words below.\n",
    "        if re.fullmatch(r\".*([^l]s|sh|ch|x|z)$\", string):\n",
    "            versions.append(f'{string}es')\n",
    "            return versions\n",
    "\n",
    "        # Words that end with a consonant followed by 'y'\n",
    "        # are made plural by replacing the 'y' with -ies.\n",
    "        # For example, the plural version of canary is\n",
    "        # canaries.\n",
    "        if re.fullmatch(r\".*([^aeiou])(y)$\", string):\n",
    "            versions.append(f'{string[:-1]}ies')\n",
    "            return versions\n",
    "            \n",
    "        # The plural version of words ending with -f\n",
    "        # and -fe aren't clear. To be safe, I will add\n",
    "        # both versions.\n",
    "        if (re.fullmatch(r\".*(f)(e?)$\", string) and not re.fullmatch(r\".*ff$\", string)):\n",
    "            last_clean = re.sub(r\"(f)(e?)$\", \"\", string)\n",
    "            versions.append(f'{last_clean}fs')\n",
    "            versions.append(f'{last_clean}ves')\n",
    "            return versions\n",
    "\n",
    "        # People add -s or -es to words that end with 'o'.\n",
    "        # To be safe, both versions are added.\n",
    "        if re.fullmatch(r\".*([^aeiou])o$\", string):\n",
    "            versions.append(f'{string}s')\n",
    "            versions.append(f'{string}es')\n",
    "            return versions\n",
    "\n",
    "        # If there's no -s at the end of the string and\n",
    "        # the other cases didn't run, we add an -s.\n",
    "        if re.fullmatch(r\".*[^s]$\", string):\n",
    "            versions.append(f'{string}s')\n",
    "        \n",
    "        return versions\n",
    "\n",
    "\n",
    " \n",
    "    def expand_unit(self, *, il_unit, ir_unit, il_boundary, ir_boundary, speech=[], literals=[], include=True, direction='BOTH', verbose=False):\n",
    "        UNIT = self.main.sp_doc[il_unit:ir_unit+1]\n",
    "        \n",
    "        if il_unit > ir_unit:\n",
    "            print(f\"Error: il_unit of {il_unit} greater than ir_unit of {ir_unit}\")\n",
    "            return None\n",
    "        \n",
    "        if direction in ['BOTH', 'LEFT'] and il_boundary > il_unit:\n",
    "            print(f\"Error: il_unit of {il_unit} less than il_boundary of {il_boundary}\")\n",
    "            return None\n",
    "        \n",
    "        if direction in ['BOTH', 'RIGHT'] and ir_boundary < ir_unit:\n",
    "            print(f\"Error: ir_unit of {ir_unit} greater than ir_boundary of {ir_boundary}\")\n",
    "            return None\n",
    "        \n",
    "        # Move Left\n",
    "        if direction in ['BOTH', 'LEFT']:\n",
    "            # The indices are inclusive, therefore, when \n",
    "            # the condition fails, il_unit will be equal\n",
    "            # to il_boundary.\n",
    "            while il_unit > il_boundary:\n",
    "                # We assume that the current token is allowed,\n",
    "                # and look to the token to the left.\n",
    "                l_token = self.main.sp_doc[il_unit-1]\n",
    "\n",
    "                # If the token is invalid, we stop expanding.\n",
    "                in_set = l_token.pos_ in speech or l_token.lower_ in literals\n",
    "\n",
    "                # Case 1: include=False, in_set=True\n",
    "                # If we're not meant to include the defined tokens, and the\n",
    "                # current token is in that set, we stop expanding.\n",
    "                # Case 2: include=True, in_set=False\n",
    "                # If we're meant to include the defined tokens, and the current\n",
    "                # token is not in that set, we stop expanding.\n",
    "                # Case 3: include=in_set\n",
    "                # If we're meant to include the defined tokens, and the current\n",
    "                # token is in that set, we continue expanding. If we're not meant\n",
    "                # to include the defined tokens, and the current token is not\n",
    "                # in that set, we continue expanding.\n",
    "                if include ^ in_set:\n",
    "                    break\n",
    "                \n",
    "                # Else, the left token is valid, and\n",
    "                # we continue to expand.\n",
    "                il_unit -= 1\n",
    "\n",
    "        # Move Right\n",
    "        if direction in ['BOTH', 'RIGHT']:\n",
    "            # Likewise, when the condition fails,\n",
    "            # ir_unit will be equal to the ir_boundary.\n",
    "            # The ir_boundary is also inclusive.\n",
    "            while ir_unit < ir_boundary:\n",
    "                # Assuming that the current token is valid,\n",
    "                # we look to the right to see if we can\n",
    "                # expand.\n",
    "                r_token = self.main.sp_doc[ir_unit+1]\n",
    "\n",
    "                # If the token is invalid, we stop expanding.\n",
    "                in_set = r_token.pos_ in speech or r_token.lower_ in literals\n",
    "                if include ^ in_set:\n",
    "                    break\n",
    "\n",
    "                # Else, the token is valid and\n",
    "                # we continue.\n",
    "                ir_unit += 1\n",
    "\n",
    "        assert il_unit >= il_boundary and ir_unit <= ir_boundary\n",
    "        \n",
    "        expanded_unit = self.main.sp_doc[il_unit:ir_unit+1]\n",
    "\n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(f\"Expanded Unit of '{UNIT}': {expanded_unit}\")\n",
    "        \n",
    "        return expanded_unit\n",
    "\n",
    "\n",
    "    \n",
    "    def contract_unit(self, *, il_unit, ir_unit, speech=[], literals=[], include=True, direction='BOTH', verbose=False):\n",
    "        UNIT = self.main.sp_doc[il_unit:ir_unit+1]\n",
    "        \n",
    "        if il_unit > ir_unit:\n",
    "            print(f\"Error: il_unit of {il_unit} greater than ir_unit of {ir_unit}\")\n",
    "            return None\n",
    "        \n",
    "        # Move Right\n",
    "        if direction in ['BOTH', 'LEFT']:\n",
    "            while il_unit < ir_unit:\n",
    "                # We must check if the current token is not allowed. If it's\n",
    "                # not allowed, we contract (remove).\n",
    "                token = self.main.sp_doc[il_unit]\n",
    "\n",
    "                # include = True means that we want the tokens that match\n",
    "                # the speech and/or literals in the contracted unit.\n",
    "                \n",
    "                # include = False means that we don't want the tokens that\n",
    "                # match the speech and/or literals in the contracted unit.\n",
    "                \n",
    "                # Case 1: include = True, in_set = True\n",
    "                # We have a token that's meant to be included in the set.\n",
    "                # However, we're contracting, which means we would end up\n",
    "                # removing the token if we continue. Therefore, we break.\n",
    "                \n",
    "                # Case 2: include = False, in_set = False\n",
    "                # We have a token that's not in the set which defines the\n",
    "                # tokens that aren't meant to be included. Therefore, we \n",
    "                # have a token that is meant to be included. If we continue,\n",
    "                # we would end up removing this token. Therefore, we break.\n",
    "                \n",
    "                # Default:\n",
    "                # If we have a token that's in the set (in_set=True) of\n",
    "                # tokens we're not supposed to include in the contracted \n",
    "                # unit (include=False), we need to remove it. Likewise, if\n",
    "                # we have a token that's not in the set (in_set=False) of\n",
    "                # tokens to include in the contracted unit (include=True),\n",
    "                # we need to remove it.\n",
    "                \n",
    "                in_set = token.pos_ in speech or token.lower_ in literals\n",
    "                if include == in_set:\n",
    "                    break\n",
    "\n",
    "                # The token is valid, thus we continue.\n",
    "                il_unit += 1\n",
    "\n",
    "        # Move Left      \n",
    "        if direction in ['BOTH', 'RIGHT']:\n",
    "            while ir_unit > il_unit:\n",
    "                token = self.main.sp_doc[ir_unit]\n",
    "\n",
    "                # The token is invalid and we\n",
    "                # stop contracting.\n",
    "                in_set = token.pos_ in speech or token.lower_ in literals\n",
    "                if include == in_set:\n",
    "                    break\n",
    "\n",
    "                # The token is valid and we continue.\n",
    "                ir_unit -= 1\n",
    "\n",
    "        assert il_unit <= ir_unit\n",
    "        \n",
    "        contracted_unit = self.main.sp_doc[il_unit:ir_unit+1]\n",
    "\n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(f\"Contracted Unit of '{UNIT}': {contracted_unit}\")\n",
    "        \n",
    "        return contracted_unit\n",
    "\n",
    "\n",
    "    \n",
    "    def find_unit_context(self, *, il_unit, ir_unit, il_boundary, ir_boundary, speech=[\"ADJ\", \"NOUN\", \"ADP\", \"ADV\", \"PART\", \"PROPN\", \"VERB\", \"PRON\", \"DET\", \"AUX\", \"PART\", \"SCONJ\"], literals=[], include=True, enclosures=LAX_ENCLOSURES, comma_encloses=False, verbose=False):\n",
    "        UNIT = self.main.sp_doc[il_unit:ir_unit+1]\n",
    "        \n",
    "        if il_unit > ir_unit:\n",
    "            print(f\"Error: il_unit of {il_unit} greater than ir_unit of {ir_unit}\")\n",
    "            return None\n",
    "        \n",
    "        if il_boundary > il_unit:\n",
    "            print(f\"Error: il_unit of {il_unit} less than il_boundary of {il_boundary}\")\n",
    "            return None\n",
    "        \n",
    "        if ir_boundary < ir_unit:\n",
    "            print(f\"Error: ir_unit of {ir_unit} greater than ir_boundary of {ir_boundary}\")\n",
    "            return None\n",
    "        \n",
    "        # Caveat: Parentheticals\n",
    "        # The context of a unit inside a set of enclosures should\n",
    "        # not go farther than the boundaries of those enclosures.\n",
    "        # However, we need to manually determine whether the unit\n",
    "        # is in parentheses (or any set of the matching symbols\n",
    "        # below).\n",
    "        openers = list(enclosures.keys())\n",
    "        closers = list(enclosures.values())\n",
    "        enclosing_chars = [*closers, *openers]\n",
    "\n",
    "        # Look for Group Punctuation on the Left\n",
    "        i = il_unit\n",
    "        opener = None\n",
    "        while i > il_boundary:\n",
    "            token = self.main.sp_doc[i]\n",
    "            if token.lower_ in enclosing_chars and token.lower_ != \",\":\n",
    "                opener = token\n",
    "                break\n",
    "            i -= 1\n",
    "\n",
    "        # Look for Group Punctuation on the Right\n",
    "        i = ir_unit\n",
    "        closer = None\n",
    "        while i < ir_boundary:\n",
    "            token = self.main.sp_doc[i]\n",
    "            if token.lower_ in enclosing_chars and token.lower_ != \",\":\n",
    "                closer = token\n",
    "                break\n",
    "            i += 1\n",
    "\n",
    "        # If there's a group punctuation on the left\n",
    "        # and right, and they match each other (e.g. '(' and ')'),\n",
    "        # we return the text between the punctuations.\n",
    "        parenthetical = opener and closer and enclosures.get(opener.lower_) == closer.text\n",
    "        if parenthetical:\n",
    "            context = [t for t in self.main.sp_doc[opener.i:closer.i+1]]\n",
    "            \n",
    "            if verbose and VERBOSE_LEVEL >= 1:\n",
    "                print(f\"Parenthetical - Unit Context of '{UNIT}': {context}\")\n",
    "            \n",
    "            return context\n",
    "\n",
    "        # We can also check whether the unit it enclosed\n",
    "        # in a comma or two, only if a comma can enclose.\n",
    "        if comma_encloses:\n",
    "            i = il_unit\n",
    "            i_token = self.main.sp_doc[i]\n",
    "            while i > il_boundary:\n",
    "                i_token = self.main.sp_doc[i]\n",
    "                if i_token.lower_ in [\",\", \";\", \"—\"]:\n",
    "                    break\n",
    "                i -= 1\n",
    "\n",
    "            j = ir_unit\n",
    "            j_token = self.main.sp_doc[j]\n",
    "            while j < ir_boundary:\n",
    "                j_token = self.main.sp_doc[j]\n",
    "                if j_token.lower_ in [\",\", \";\", \"—\"]:\n",
    "                    break\n",
    "                j += 1\n",
    "\n",
    "            if (i_token and i_token.lower_ == \",\") or (j_token and j_token.lower_ == \",\"):\n",
    "                context = [t for t in self.main.sp_doc[i:j+1]]\n",
    "            \n",
    "                if verbose and VERBOSE_LEVEL >= 1:\n",
    "                    print(f\"Comma - Unit Context of '{UNIT}': {context}\")\n",
    "                    \n",
    "                return context\n",
    "            \n",
    "        # As the unit is not a parenthetical, we will expand\n",
    "        # outwards until we run into a stopping token. The exclude\n",
    "        # list contains tokens that should be excluded from the\n",
    "        # context. Currently, it will contain any parentheticals\n",
    "        # that we run into.\n",
    "        exclude = []\n",
    "\n",
    "        # We can modify the enclosures after handling the parenthetical\n",
    "        # situation to make the code easier.\n",
    "        if comma_encloses:\n",
    "            enclosures[\",\"] : \",\"\n",
    "        \n",
    "        # Expand Left\n",
    "        while il_unit > il_boundary:\n",
    "            # Assuming that the current token is fine,\n",
    "            # we look to the left.\n",
    "            l_token = self.main.sp_doc[il_unit-1]\n",
    "\n",
    "            if l_token.lower_ not in closers:\n",
    "                in_set = l_token.pos_ in speech or l_token.lower_ in literals\n",
    "                if in_set ^ include:\n",
    "                    break\n",
    "                il_unit -= 1\n",
    "            # If it's a closing enclosure (e.g. ')', ']'),\n",
    "            # we need to skip over whatever is contained in\n",
    "            # that punctuation.\n",
    "            else:\n",
    "                i = il_unit - 1\n",
    "                \n",
    "                token = self.main.sp_doc[i]\n",
    "                exclude.append(token)\n",
    "\n",
    "                # We continue until we reach the boundary or\n",
    "                # we find the matching opening character.\n",
    "                closed = []\n",
    "                \n",
    "                while i > il_boundary:\n",
    "                    token = self.main.sp_doc[i]\n",
    "                    # Found Closer\n",
    "                    if token.lower_ in closers:\n",
    "                        exclude.append(token)\n",
    "                        closed.append(token.lower_)\n",
    "                    # Currently Closed\n",
    "                    elif closed:\n",
    "                        exclude.append(token)\n",
    "                        # Found Opener\n",
    "                        if enclosures.get(token.lower_) == closed[-1]:\n",
    "                            closed.pop()\n",
    "                    else:\n",
    "                        break\n",
    "                    i -= 1\n",
    "                \n",
    "                il_unit = i\n",
    "\n",
    "        # Expand Right\n",
    "        while ir_unit < ir_boundary:\n",
    "            # We're checking the token to the right\n",
    "            # to see if we can expand or not.\n",
    "            r_token = self.main.sp_doc[ir_unit+1]\n",
    "\n",
    "            if r_token.lower_ not in openers:\n",
    "                in_set = r_token.pos_ in speech or r_token.lower_ in literals\n",
    "                if in_set ^ include:\n",
    "                    break\n",
    "                ir_unit += 1\n",
    "            # If the token to the right is an opener (e.g. '(', '['), we must skip\n",
    "            # it, the parenthetical inside, and the closer.\n",
    "            else:\n",
    "                i = ir_unit + 1\n",
    "                \n",
    "                token = self.main.sp_doc[i]\n",
    "                exclude.append(token)\n",
    "\n",
    "                # We continue until we reach the boundary or\n",
    "                # we find all the closers for the openers.\n",
    "                opened = []\n",
    "                \n",
    "                while i < ir_boundary:\n",
    "                    token = self.main.sp_doc[i]\n",
    "                    # Found Opener\n",
    "                    if token.lower_ in openers:\n",
    "                        exclude.append(token)\n",
    "                        opened.append(token.lower_)\n",
    "                    # Currently Opened\n",
    "                    elif opened:\n",
    "                        exclude.append(token)\n",
    "                        # Found Closer\n",
    "                        if token.lower_ == enclosures.get(opened[-1]):\n",
    "                            opened.pop()\n",
    "                    else:\n",
    "                        break\n",
    "                    i += 1\n",
    "                \n",
    "                ir_unit = i\n",
    "        \n",
    "        # We remove the excluded tokens and return the context.\n",
    "        context = [t for t in self.main.sp_doc[il_unit:ir_unit+1] if t not in exclude]\n",
    "\n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(f\"Unit Context of '{UNIT}': {context}\")\n",
    "        \n",
    "        return context\n",
    "\n",
    "\n",
    "    \n",
    "    def get_span_at_indices(self, l_index, r_index):\n",
    "        text = self.main.sp_doc.text.lower()\n",
    "\n",
    "        while text[l_index].isspace():\n",
    "            l_index += 1\n",
    "\n",
    "        while text[r_index].isspace():\n",
    "            r_index -= 1\n",
    "\n",
    "        if l_index > r_index:\n",
    "            print(f\"Error: l_index of {l_index} greater than r_index of {r_index}\")\n",
    "            return None\n",
    "            \n",
    "        l_token_i = self.main.token_at_char(l_index).i\n",
    "        r_token_i = self.main.token_at_char(r_index).i\n",
    "        \n",
    "        return self.main.sp_doc[l_token_i:r_token_i+1]\n",
    "\n",
    "\n",
    "    \n",
    "    def get_base_nouns(self, span, return_tokens=False, immediate_stop=False):\n",
    "        ending_nouns = []\n",
    "        \n",
    "        reversed_span = [t for t in span]\n",
    "        reversed_span.reverse()\n",
    "        \n",
    "        for token in reversed_span:\n",
    "            if token.pos_ in [\"NOUN\", \"PROPN\"]:\n",
    "                ending_nouns.append(token if return_tokens else self.main.sp_doc[token.i:token.i+1])\n",
    "                if immediate_stop:\n",
    "                    break\n",
    "            else:\n",
    "                break\n",
    "\n",
    "        return ending_nouns\n",
    "\n",
    "\n",
    "\n",
    "    def flatten(self, arr):\n",
    "        flat_arr = []\n",
    "\n",
    "        if not isinstance(arr, list):\n",
    "            return [arr]\n",
    "\n",
    "        for element in arr:\n",
    "            flat_arr.extend(self.flatten(element))\n",
    "\n",
    "        return flat_arr\n",
    "\n",
    "\n",
    "    def is_same_text(self, sp_a, sp_b):\n",
    "        sp_b_text = sp_b.text.lower()\n",
    "        sp_a_text = sp_a.text.lower()\n",
    "\n",
    "        if sp_a_text == sp_b_text:\n",
    "            return True\n",
    "            \n",
    "        sp_a_singular_texts = [sp_a_text] if sp_a[-1].tag_ in [\"NN\", \"NNP\"] else self.main.singularize(sp_a_text)\n",
    "        sp_b_singular_texts = [sp_b_text] if sp_b[-1].tag_ in [\"NN\", \"NNP\"] else self.main.singularize(sp_b_text)\n",
    "\n",
    "        if set(sp_a_singular_texts).intersection(sp_b_singular_texts):\n",
    "            return True\n",
    "        return False\n",
    "\n",
    "\n",
    "\n",
    "    def has_same_base_nouns(self, sp_a, sp_b):\n",
    "        sp_b_text = sp_b.text.lower()\n",
    "        sp_b_0_text = sp_b[0].lower_\n",
    "        sp_b_0_is_noun = sp_b[0].pos_ in [\"NOUN\", \"PROPN\"]\n",
    "        \n",
    "        sp_b_nouns = []\n",
    "        sp_b_num_adjectives = 0\n",
    "        \n",
    "        for token in sp_b:\n",
    "            if not sp_b_nouns and token.pos_ == \"ADJ\":\n",
    "                sp_b_num_adjectives += 1\n",
    "            elif token.pos_ in [\"PROPN\", \"NOUN\"]:\n",
    "                sp_b_nouns.append(token)\n",
    "\n",
    "        if not sp_b_nouns:\n",
    "            return False\n",
    "\n",
    "        sp_b_nouns_text = [noun.lower_ for noun in sp_b_nouns]\n",
    "        sp_b_singular_texts = [\" \".join(sp_b_nouns_text)] if sp_b_nouns[-1].tag_ in [\"NN\", \"NNP\"] else self.main.singularize(\" \".join(sp_b_nouns_text))\n",
    "\n",
    "        sp_a_text = sp_a.text.lower()\n",
    "        sp_a_0_text = sp_a[0].lower_\n",
    "        sp_a_0_is_noun = sp_a[0].pos_ in [\"NOUN\", \"PROPN\"]\n",
    "\n",
    "        # Case Example: 'Hyla' v. 'Hyla tadpoles'\n",
    "        if sp_a_0_text == sp_b_0_text and (sp_a_0_is_noun or sp_b_0_is_noun):\n",
    "            if sp_a_text in sp_b_text or sp_b_text in sp_a_text:\n",
    "                return True\n",
    "        \n",
    "        # Case Example: 'dogs' v. 'red dogs'\n",
    "        sp_a_nouns = []\n",
    "        sp_a_num_adjectives = 0\n",
    "        for token in sp_a:\n",
    "            if not sp_a_nouns and token.pos_ == \"ADJ\":\n",
    "                sp_a_num_adjectives += 1\n",
    "            elif token.pos_ in [\"PROPN\", \"NOUN\"]:\n",
    "                sp_a_nouns.append(token)\n",
    "        \n",
    "        if not sp_a_nouns:\n",
    "            return False\n",
    "        \n",
    "        sp_a_nouns_text = [noun.lower_ for noun in sp_a_nouns]\n",
    "        \n",
    "        if sp_a_nouns and sp_b_nouns and (\n",
    "            (sp_a_num_adjectives == 1 and sp_b_num_adjectives == 0) or \n",
    "            (sp_b_num_adjectives == 1 and sp_a_num_adjectives == 0)\n",
    "        ):\n",
    "            sp_a_singular_texts = [\" \".join(sp_a_nouns_text)] if sp_a_nouns[-1].tag_ in [\"NN\", \"NNP\"] else self.main.singularize(\" \".join(sp_a_nouns_text))\n",
    "            if set(sp_a_singular_texts).intersection(sp_b_singular_texts):\n",
    "                return True\n",
    "\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7290c329-6929-4dce-9fc6-a026c8b449be",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Species:\n",
    "    def __init__(self, main):\n",
    "        # Tools\n",
    "        self.main = main\n",
    "        self.tn_nlp = TaxoNERD(prefer_gpu=False).load(model=\"en_ner_eco_biobert\", exclude=[\"tagger\", \"parser\", \"attribute_ruler\"])\n",
    "        self.tn_doc = None\n",
    "        \n",
    "        # Contains any spans that have been identified\n",
    "        # as a species.\n",
    "        self.spans = None\n",
    "        self.span_starts = None\n",
    "        \n",
    "        # Contains any tokens that have been identified\n",
    "        # as a species or being a part of a species.\n",
    "        self.tokens = None\n",
    "        \n",
    "        # Used to quickly access the span that a token\n",
    "        # belongs to.\n",
    "        self.token_to_span = None\n",
    "        \n",
    "        # Maps a string to an array of strings wherein\n",
    "        # the strings involved in the key-value pair \n",
    "        # have been identified as an alternate name of each other.\n",
    "        self.alternate_names = None\n",
    "        \n",
    "        # Includes words that (1) are to be identified as species; and\n",
    "        # (2) are sometimes not identified as species, more or less.\n",
    "        self.dictionary = [\"juvenile\", \"juveniles\", \"adult\", \"adults\", \"prey\", \"predator\", \"predators\", \"species\", \"tree\", \"cat\", \"dog\", \"fly\", \"flies\", \"plant\", \"plants\"]\n",
    "\n",
    "\n",
    "\n",
    "    def update(self, text, verbose=False):\n",
    "        if not self.main.sp_doc or not self.main.index_map:\n",
    "            raise Exception(\"DNE\")\n",
    "        self.tn_doc = self.tn_nlp(text)\n",
    "        self.spans, self.tokens, self.token_to_span, self.span_starts = self.load_species(verbose=verbose)\n",
    "        self.alternate_names = self.load_alternate_names(self.spans)\n",
    "\n",
    "\n",
    "\n",
    "    def convert_tn_spans_to_sp_spans(self, tn_spans):\n",
    "        sp_spans = []\n",
    "\n",
    "        for tn_span in tn_spans:\n",
    "            l_char_index = self.tn_doc[tn_span.start].idx\n",
    "            r_char_index = l_char_index + len(tn_span.text) - 1\n",
    "\n",
    "            try:\n",
    "                l_sp_token_i = self.main.token_at_char(l_char_index).i\n",
    "                r_sp_token_i = self.main.token_at_char(r_char_index).i\n",
    "            except Exception as e:\n",
    "                print(f\"Error: Couldn't find token at character index of {l_char_index} and token index of {l_sp_token_i}.\")\n",
    "                print(f\"Error: Couldn't find token at character index of {r_char_index} and token index of {r_sp_token_i}.\")\n",
    "                print(e)\n",
    "                continue\n",
    "\n",
    "            sp_span = self.main.sp_doc[l_sp_token_i:r_sp_token_i+1]\n",
    "            if sp_span.text != tn_span:\n",
    "                print(f\"Error: SpaCy span does not match TaxoNerd span.\")\n",
    "                continue\n",
    "            \n",
    "            sp_spans.append(sp_span)\n",
    "\n",
    "        return sp_spans\n",
    "\n",
    "\n",
    "\n",
    "    def load_search_strings(self, verbose=False):\n",
    "        search_strings = [*self.dictionary]\n",
    "        \n",
    "        # Creating a Broad Set of Species\n",
    "        spans = self.convert_tn_spans_to_sp_spans(self.tn_doc.ents)\n",
    "        spans = self.main.separate_spans_by_parenthetical(spans)\n",
    "\n",
    "        # Add Ending Nouns to Set\n",
    "        all_nouns = []\n",
    "        for span in spans:\n",
    "            nouns = self.main.get_base_nouns(span, immediate_stop=True)\n",
    "            # print(f\"Base Nouns for '{span}': {nouns}\")\n",
    "            if nouns:\n",
    "                all_nouns.extend(nouns)\n",
    "        # print(f\"'All Nouns': {all_nouns}\")\n",
    "        spans.extend(all_nouns)\n",
    "\n",
    "        # Adding Plural and Singular Versions of Spans\n",
    "        for span in spans:\n",
    "            text = span.text.lower()\n",
    "            text = self.main.delete_extra_whitespace(self.main.delete_outer_non_alnum(text))\n",
    "\n",
    "            # Blank Text or No Letters\n",
    "            if not text or not [c for c in text if c.isalpha()]:\n",
    "                continue\n",
    "\n",
    "            search_strings.append(text)\n",
    "\n",
    "            # Add Plural Version\n",
    "            singular = span[-1].pos_ == \"NOUN\" and span[-1].tag_ == \"NN\"\n",
    "            if singular:\n",
    "                plural_version = self.main.pluralize(text)\n",
    "                search_strings.extend(plural_version)\n",
    "\n",
    "            # Add Singular Version\n",
    "            plural = span[-1].pos_ == \"NOUN\" and span[-1].tag_ == \"NNS\"\n",
    "            if plural:\n",
    "                singular_version = self.main.singularize(text)\n",
    "                search_strings.extend(singular_version)\n",
    "\n",
    "        # Remove Duplicates\n",
    "        search_strings = list(set(search_strings))\n",
    "\n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(f\"Search Strings: {search_strings}\")\n",
    "        \n",
    "        return search_strings\n",
    "\n",
    "\n",
    "\n",
    "    def load_alternate_names(self, spans, verbose=False):\n",
    "        spans.sort(key=lambda span: span.start)\n",
    "\n",
    "        # It's useful to know if a different name refers to a\n",
    "        # species we have already seen. For example, in\n",
    "        # \"predatory crab (Carcinus maenas)\", \"predatory crab\"\n",
    "        # is an alternative name for \"Carcinus maenas\" and\n",
    "        # vice versa. This is used so that the species can be\n",
    "        # properly tracked and redundant points are less\n",
    "        # likely to be given.\n",
    "        alternate_names = {}\n",
    "        \n",
    "        # Finding and Storing Alternative Names\n",
    "        for i, species_span in enumerate(spans):\n",
    "            # There's not a next species to\n",
    "            # evaluate.\n",
    "            if i + 1 >= len(spans):\n",
    "                break\n",
    "            \n",
    "            next_species_span = spans[i+1]\n",
    "            \n",
    "            # If there's one token between the species and the next species,\n",
    "            # we check if the next species is surrounded by punctuation.\n",
    "            if next_species_span.start - species_span.end == 1:\n",
    "                # Token Before and After the Next Species\n",
    "                before_next = self.main.sp_doc[next_species_span.start-1]\n",
    "\n",
    "                # Out of Bounds Error\n",
    "                if next_species_span.end >= len(self.main.sp_doc):\n",
    "                    continue\n",
    "                    \n",
    "                after_next = self.main.sp_doc[next_species_span.end]\n",
    "\n",
    "                if before_next.lower_ in [\"(\"] and after_next.lower_ in [\")\"]:\n",
    "                    sp_1_text = species_span.text.lower()\n",
    "                    sp_2_text = next_species_span.text.lower()\n",
    "                    \n",
    "                    if sp_1_text not in alternate_names:\n",
    "                        alternate_names[sp_1_text] = []\n",
    "                    \n",
    "                    if sp_2_text not in alternate_names:\n",
    "                        alternate_names[sp_2_text] = []\n",
    "                    \n",
    "                    alternate_names[sp_1_text].append(sp_2_text)\n",
    "                    alternate_names[sp_2_text].append(sp_1_text)\n",
    "            \n",
    "            # If there's no token between the species and the next,\n",
    "            # species we assume that they refer to the same species.\n",
    "            elif next_species_span.start - species_span.end == 0:\n",
    "                sp_1_text = species_span.text.lower()\n",
    "                sp_2_text = next_species_span.text.lower()\n",
    "                \n",
    "                if sp_1_text not in alternate_names:\n",
    "                    alternate_names[sp_1_text] = []\n",
    "                \n",
    "                if sp_2_text not in alternate_names:\n",
    "                    alternate_names[sp_2_text] = []\n",
    "\n",
    "                alternate_names[sp_1_text].append(sp_2_text)\n",
    "                alternate_names[sp_2_text].append(sp_1_text)\n",
    "\n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(f\"Alternate Names: {alternate_names}\")\n",
    "\n",
    "        return alternate_names\n",
    "\n",
    "\n",
    "\n",
    "    def bar(self, foo):\n",
    "        if not foo or len(foo) == 1:\n",
    "            return foo\n",
    "\n",
    "        foo.sort()\n",
    "        b = [foo[0]]\n",
    "        \n",
    "        l = 0\n",
    "        \n",
    "        for i in range(1, len(foo)):\n",
    "            if foo[i] - foo[l] <= 1:\n",
    "                l = i\n",
    "            else:\n",
    "                b.append(foo[i])\n",
    "                l = i\n",
    "            \n",
    "        return b\n",
    "\n",
    "\n",
    "    \n",
    "    def load_species(self, verbose=False):\n",
    "        if not self.main.sp_doc or not self.main.index_map:\n",
    "            raise Exception(\"DNE\")\n",
    "\n",
    "        # Load Search Strings from Species Spans\n",
    "        search_strings = self.load_search_strings(verbose=verbose)\n",
    "\n",
    "        # Search for Species\n",
    "        # The results are stored in different \n",
    "        # forms below.\n",
    "        spans = []\n",
    "        tokens = []\n",
    "        token_to_span = {}\n",
    "\n",
    "        # Where we're searching for species.\n",
    "        text = self.main.sp_doc.text.lower()\n",
    "\n",
    "        for string in search_strings:\n",
    "            matches = re.finditer(re.escape(string), text, re.IGNORECASE)\n",
    "\n",
    "            for l_char_index, r_char_index, matched_text in [(match.start(), match.end(), match.group()) for match in matches]:    \n",
    "                # The full word must match, not just a substring inside of it.\n",
    "                # So, if the species we're looking for is \"ant\", only \"ant\"\n",
    "                # will match -- not \"pants\" or \"antebellum\". Therefore, the\n",
    "                # characters to the left and right of the matched string cannot\n",
    "                # be letters.\n",
    "                l_char_is_letter = l_char_index > 0 and text[l_char_index-1].isalpha()\n",
    "                r_char_is_letter = r_char_index < len(text) and text[r_char_index].isalpha()\n",
    "                \n",
    "                if l_char_is_letter or r_char_is_letter or not matched_text:\n",
    "                    continue\n",
    "\n",
    "                try:\n",
    "                    l_token_i = self.main.token_at_char(l_char_index).i\n",
    "                    r_token_i = self.main.token_at_char(r_char_index-1).i\n",
    "                except Exception as e:\n",
    "                    print(f\"Error: Unable to find token at index of {l_char_index}.\")\n",
    "                    print(f\"Error: Unable to find token at index of {r_char_index}.\")\n",
    "                    print(f\"\\tMatched: '{matched_text}'\")\n",
    "                    print(e)\n",
    "                    continue\n",
    "\n",
    "                # This is the matched substring (which would be\n",
    "                # a species) as a span in the parent document.\n",
    "                span = self.main.sp_doc[l_token_i:r_token_i+1]\n",
    "                \n",
    "                # Expand Species\n",
    "                # Let's say there's a word like \"squirrel\". That's a bit ambiguous. \n",
    "                # Is it a brown squirrel, a bonobo? If the species is possibly missing\n",
    "                # information (like an adjective to the left of it), we should expand\n",
    "                # in order to get a full picture of the species.\n",
    "                unclear_1 = len(span) == 1 and span[0].pos_ == \"NOUN\"\n",
    "                unclear_2 = span.start > 0 and self.main.sp_doc[span.start-1].pos_ in [\"ADJ\"]\n",
    "                \n",
    "                if unclear_1 or unclear_2:\n",
    "                    span = self.main.expand_unit(\n",
    "                        il_unit=span.start, \n",
    "                        ir_unit=span.end-1,\n",
    "                        il_boundary=0,\n",
    "                        ir_boundary=len(self.main.sp_doc),\n",
    "                        speech=[\"ADJ\", \"PROPN\"],\n",
    "                        literals=[\"-\"],\n",
    "                        include=True,\n",
    "                        direction=\"LEFT\",\n",
    "                        verbose=verbose\n",
    "                    )\n",
    "                \n",
    "                # Remove Outer Symbols\n",
    "                # There are times where a species is identified with a parenthesis\n",
    "                # nearby. Here, we remove that parenthesis (and any other symbols).\n",
    "                span = self.main.contract_unit(\n",
    "                    il_unit=span.start, \n",
    "                    ir_unit=span.end-1, \n",
    "                    speech=[\"PUNCT\", \"SYM\", \"DET\", \"PART\"],\n",
    "                    include=False,\n",
    "                    verbose=verbose\n",
    "                )\n",
    "\n",
    "                if not span:\n",
    "                    print(f\"Error: Span does not exist; left character index {l_char_index}.\")\n",
    "                    print(f\"\\tMatched: '{matched_text}'\")\n",
    "                    continue\n",
    "            \n",
    "                # A species must have a noun or a\n",
    "                # proper noun. This may help discard\n",
    "                # bad results.\n",
    "                letter_found = False\n",
    "                for token in span:\n",
    "                    if token.pos_ in [\"NOUN\", \"PROPN\"] or token.lower_ in self.dictionary:\n",
    "                        letter_found = True\n",
    "                        break\n",
    "\n",
    "                if not letter_found:\n",
    "                    continue\n",
    "\n",
    "                # Adding Species\n",
    "                spans.append(span)\n",
    "                for token in span:\n",
    "                    if token in tokens or token.pos_ in [\"PUNCT\", \"SYM\", \"DET\", \"PART\"]:\n",
    "                        continue\n",
    "                    tokens.append(token)\n",
    "                    token_to_span[token] = span\n",
    "        \n",
    "        spans = list({span.start: span for span in spans}.values())\n",
    "        spans.sort(key=lambda span: span.start)\n",
    "\n",
    "        # for span in spans:\n",
    "        #     print(span, span.start, span.end)\n",
    "        \n",
    "        # Remove Overlapping Spans\n",
    "        i = 0\n",
    "        while i < len(spans):\n",
    "            j = i + 1\n",
    "            while j < len(spans):\n",
    "                if spans[i].start <= spans[j].start <= spans[i].end and spans[i].start <= spans[j].end <= spans[i].end:\n",
    "                    # print(f\"Span at i: ({spans[i].start}, {spans[i].end})\")\n",
    "                    # print(f\"Span at j: ({spans[j].start}, {spans[j].end})\")\n",
    "                    # print(f'Overlap between \"{spans[i]}\" and \"{spans[j]}\"')\n",
    "                    spans.pop(j)\n",
    "                else:\n",
    "                    j += 1\n",
    "            i += 1\n",
    "        \n",
    "        \n",
    "        \n",
    "        span_starts = [span[0] for span in spans]\n",
    "        span_indices = self.bar([span[0].i for span in spans])\n",
    "        span_starts = [self.main.sp_doc[i] for i in span_indices]\n",
    "\n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(\"Output of load_species:\")\n",
    "            print(f\"Spans: {spans}\")\n",
    "            print(f\"Tokens: {tokens}\")\n",
    "            print(f\"Mapped Tokens: {token_to_span}\")\n",
    "            print(f\"Span Starts: {span_starts}\")\n",
    "        \n",
    "        return (spans, tokens, token_to_span, span_starts)\n",
    "\n",
    "\n",
    "\n",
    "    def is_alternate(self, sp_a, sp_b):\n",
    "        sp_b_text = sp_b.text.lower()\n",
    "        sp_a_text = sp_a.text.lower()\n",
    "            \n",
    "        # Species B is an alternate name for Species A\n",
    "        if sp_b_text in self.alternate_names.get(sp_a_text, []):\n",
    "            return True\n",
    "        \n",
    "        # Species A is an alternate name for Species B\n",
    "        if sp_a_text in self.alternate_names.get(sp_b_text, []):\n",
    "            return True\n",
    "\n",
    "        return False\n",
    "\n",
    "\n",
    "\n",
    "    def is_same_text(self, sp_a, sp_b):\n",
    "        sp_b_text = sp_b.text.lower()\n",
    "        sp_a_text = sp_a.text.lower()\n",
    "\n",
    "        if sp_a_text == sp_b_text:\n",
    "            return True\n",
    "            \n",
    "        sp_a_singular_texts = [sp_a_text] if sp_a[-1].tag_ in [\"NN\", \"NNP\"] else self.main.singularize(sp_a_text)\n",
    "        sp_b_singular_texts = [sp_b_text] if sp_b[-1].tag_ in [\"NN\", \"NNP\"] else self.main.singularize(sp_b_text)\n",
    "\n",
    "        if set(sp_a_singular_texts).intersection(sp_b_singular_texts):\n",
    "            return True\n",
    "        return False\n",
    "\n",
    "\n",
    "\n",
    "    def has_same_base_nouns(self, sp_a, sp_b):\n",
    "        sp_b_text = sp_b.text.lower()\n",
    "        sp_b_0_text = sp_b[0].lower_\n",
    "        sp_b_0_is_noun = sp_b[0].pos_ in [\"NOUN\", \"PROPN\"]\n",
    "        \n",
    "        sp_b_nouns = []\n",
    "        sp_b_num_adjectives = 0\n",
    "        \n",
    "        for token in sp_b:\n",
    "            if not sp_b_nouns and token.pos_ == \"ADJ\":\n",
    "                sp_b_num_adjectives += 1\n",
    "            elif token.pos_ in [\"PROPN\", \"NOUN\"]:\n",
    "                sp_b_nouns.append(token)\n",
    "\n",
    "        if not sp_b_nouns:\n",
    "            return False\n",
    "\n",
    "        sp_b_nouns_text = [noun.lower_ for noun in sp_b_nouns]\n",
    "        sp_b_singular_texts = [\" \".join(sp_b_nouns_text)] if sp_b_nouns[-1].tag_ in [\"NN\", \"NNP\"] else self.main.singularize(\" \".join(sp_b_nouns_text))\n",
    "\n",
    "        sp_a_text = sp_a.text.lower()\n",
    "        sp_a_0_text = sp_a[0].lower_\n",
    "        sp_a_0_is_noun = sp_a[0].pos_ in [\"NOUN\", \"PROPN\"]\n",
    "\n",
    "        # Case Example: 'Hyla' v. 'Hyla tadpoles'\n",
    "        if sp_a_0_text == sp_b_0_text and (sp_a_0_is_noun or sp_b_0_is_noun):\n",
    "            if sp_a_text in sp_b_text or sp_b_text in sp_a_text:\n",
    "                return True\n",
    "        \n",
    "        # Case Example: 'dogs' v. 'red dogs'\n",
    "        sp_a_nouns = []\n",
    "        sp_a_num_adjectives = 0\n",
    "        for token in sp_a:\n",
    "            if not sp_a_nouns and token.pos_ == \"ADJ\":\n",
    "                sp_a_num_adjectives += 1\n",
    "            elif token.pos_ in [\"PROPN\", \"NOUN\"]:\n",
    "                sp_a_nouns.append(token)\n",
    "        \n",
    "        if not sp_a_nouns:\n",
    "            return False\n",
    "        \n",
    "        sp_a_nouns_text = [noun.lower_ for noun in sp_a_nouns]\n",
    "        \n",
    "        if sp_a_nouns and sp_b_nouns and (\n",
    "            (sp_a_num_adjectives == 1 and sp_b_num_adjectives == 0) or \n",
    "            (sp_b_num_adjectives == 1 and sp_a_num_adjectives == 0)\n",
    "        ):\n",
    "            sp_a_singular_texts = [\" \".join(sp_a_nouns_text)] if sp_a_nouns[-1].tag_ in [\"NN\", \"NNP\"] else self.main.singularize(\" \".join(sp_a_nouns_text))\n",
    "            if set(sp_a_singular_texts).intersection(sp_b_singular_texts):\n",
    "                return True\n",
    "\n",
    "        return False\n",
    "\n",
    "\n",
    "\n",
    "    def find_same_species(self, sp_A, sp_b, verbose=False):\n",
    "        verbose=False\n",
    "        # METHOD 1: Check for Literal Matches\n",
    "        for sp_a in sp_A:\n",
    "            if self.is_same_text(sp_a, sp_b):\n",
    "                if verbose and VERBOSE_LEVEL >= 1:\n",
    "                    print(f\"Method 1: Match Between '{sp_a}' and '{sp_b}'\")\n",
    "                return sp_a\n",
    "\n",
    "        # METHOD 2: Check Alternate Names\n",
    "        for sp_a in sp_A:\n",
    "            if self.is_alternate(sp_a, sp_b):\n",
    "                if verbose and VERBOSE_LEVEL >= 1:\n",
    "                    print(f\"Method 2: Match Between '{sp_a}' and '{sp_b}'\")\n",
    "                return sp_a\n",
    "        \n",
    "        # METHOD 3: Check Nouns\n",
    "        # This is used if one or none of the species being compared\n",
    "        # has 1 adjective.\n",
    "        for sp_a in sp_A:\n",
    "            if self.has_same_base_nouns(sp_a, sp_b):\n",
    "                if verbose and VERBOSE_LEVEL >= 1:\n",
    "                    print(f\"Method 3: Match Between '{sp_a}' and '{sp_b}'\")\n",
    "                return sp_a\n",
    "\n",
    "        # METHOD 4: Last Ditch Effort\n",
    "        # If there's been no matches, we just look for one string inside of\n",
    "        # another.\n",
    "        for sp_a in sp_A:\n",
    "            sp_a_text = sp_a.text.lower()\n",
    "            sp_b_text = sp_b.text.lower()\n",
    "            \n",
    "            r_sp_a_text = re.compile(f\"(\\s|^){re.escape(sp_a_text)}(\\s|$)\", re.IGNORECASE)\n",
    "            r_sp_b_text = re.compile(f\"(\\s|^){re.escape(sp_b_text)}(\\s|$)\", re.IGNORECASE)\n",
    "            \n",
    "            if re.match(r_sp_a_text, sp_b_text) or re.match(r_sp_b_text, sp_a_text):\n",
    "                if verbose and VERBOSE_LEVEL >= 1:\n",
    "                    print(f\"Method 4: Match Between '{sp_a}' and '{sp_b}'\")\n",
    "                return sp_a\n",
    "\n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(f\"No Matches Between {sp_A} and {sp_b}\")\n",
    "        \n",
    "        return None\n",
    "\n",
    "\n",
    "\n",
    "    def span_at_token(self, token):\n",
    "        if token in self.token_to_span:\n",
    "            return self.token_to_span[token]\n",
    "        return None\n",
    "\n",
    "\n",
    "\n",
    "    def is_species(self, token):\n",
    "        return token in self.tokens\n",
    "\n",
    "\n",
    "\n",
    "    def has_species(self, tokens, verbose=False):\n",
    "        for token in tokens:\n",
    "            if token in self.tokens:\n",
    "                if verbose and VERBOSE_LEVEL >= 2:\n",
    "                    print(f\"\\tToken '{token}' is Species\")\n",
    "                return True\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d34749d9-17a5-4e17-80b5-d2fa095a4513",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Keywords:\n",
    "    REGEX = \"regex\"\n",
    "    VOCAB = \"vocab\"\n",
    "    RULES = \"rules\"\n",
    "\n",
    "\n",
    "\n",
    "    def __init__(self, main, *, regexes=[], vocab=[], patterns=[], def_pos=[], def_tag=[], def_threshold=0.7, def_weight=1.0):\n",
    "        self.main = main\n",
    "\n",
    "        # Constraints\n",
    "        self.def_threshold = def_threshold\n",
    "        self.def_tag = def_tag\n",
    "        self.def_pos = def_pos\n",
    "        self.def_weight = def_weight\n",
    "        \n",
    "        # Three Types of Matching\n",
    "        self.vocab, self.vocab_data = self.load_vocab(vocab)\n",
    "        self.regex, self.regex_data = self.load_regex(regexes)\n",
    "        self.rules, self.rules_data = self.load_rules(patterns)\n",
    "\n",
    "        # Quick Lookup\n",
    "        self.match_type_to_data = {\n",
    "            Keywords.REGEX: self.regex_data,\n",
    "            Keywords.VOCAB: self.vocab_data,\n",
    "            Keywords.RULES: self.rules_data\n",
    "        }\n",
    "\n",
    "    \n",
    "\n",
    "    def update(self, verbose=False):\n",
    "        # SpaCy Doc DNE or Indexing Map DNE\n",
    "        if not self.main.sp_doc or not self.main.index_map:\n",
    "            raise Exception(\"DNE\")\n",
    "        # Matched Tokens in Different Forms\n",
    "        self.token_data, self.mapped_token_data, self.tokens = self.match_tokens(verbose=verbose)\n",
    "\n",
    "\n",
    "\n",
    "    def load_regex(self, regexes):\n",
    "        r = []\n",
    "        r_data = {}\n",
    "\n",
    "        for unit in regexes:\n",
    "            if isinstance(unit, str):\n",
    "                r.append(unit)\n",
    "            else:\n",
    "                regex = unit[\"regex\"]\n",
    "                r.append(regex)\n",
    "                r_data[regex] = {\n",
    "                    \"types\": unit.get(\"types\", []),\n",
    "                    \"weight\": unit.get(\"weight\", self.def_weight)\n",
    "                }\n",
    "\n",
    "        return r, r_data\n",
    "\n",
    "\n",
    "\n",
    "    def load_vocab(self, vocab):\n",
    "        v = []\n",
    "        v_data = {}\n",
    "        \n",
    "        for unit in vocab:\n",
    "            if isinstance(unit, str):\n",
    "                doc = self.main.sp_nlp(unit)\n",
    "                v.append({\n",
    "                    \"doc\": doc,\n",
    "                    \"lemma\": \" \".join([t.lemma_ for t in doc])\n",
    "                })\n",
    "            else:\n",
    "                doc = self.main.sp_nlp(unit[\"word\"])\n",
    "                v.append({\n",
    "                    \"doc\": doc,\n",
    "                    \"tag\": unit.get(\"tag\", self.def_tag),\n",
    "                    \"pos\": unit.get(\"pos\", self.def_pos),\n",
    "                    \"threshold\": unit.get(\"threshold\", self.def_threshold),\n",
    "                    \"lemma\": \" \".join([t.lemma_ for t in doc])\n",
    "                })\n",
    "                v_data[unit[\"word\"]] = {\n",
    "                    \"types\": unit.get(\"types\") or [],\n",
    "                    \"weight\": unit.get(\"weight\", self.def_weight),\n",
    "                }\n",
    "        \n",
    "        return v, v_data\n",
    "\n",
    "\n",
    "\n",
    "    def load_rules(self, patterns):\n",
    "        r = Matcher(self.main.sp_nlp.vocab)\n",
    "        r_data = {}\n",
    "        \n",
    "        for i, unit in enumerate(patterns):\n",
    "            if isinstance(unit, list):\n",
    "                r.add(f\"{i}\", unit)\n",
    "            else:\n",
    "                r.add(unit[\"name\"], unit[\"pattern\"])\n",
    "                r_data[unit[\"name\"]] = {\n",
    "                    \"types\": unit.get(\"types\") or [],\n",
    "                    \"weight\": unit.get(\"weight\", self.def_weight),\n",
    "                }\n",
    "\n",
    "        return r, r_data\n",
    "\n",
    "\n",
    "\n",
    "    def get_match_data(self, token, match_id, match_type):\n",
    "        match_type_data = self.match_type_to_data[match_type]\n",
    "        \n",
    "        if match_id in match_type_data:\n",
    "            return {\n",
    "                \"token\": token,\n",
    "                \"types\": match_type_data[match_id].get(\"types\", []),\n",
    "                \"weight\": match_type_data[match_id].get(\"weight\", self.def_weight)\n",
    "            }\n",
    "        else:\n",
    "            return {\n",
    "                \"token\": token,\n",
    "                \"types\": [],\n",
    "                \"weight\": self.def_weight\n",
    "            }\n",
    "\n",
    "\n",
    "\n",
    "    def bad_pos(self, pos):\n",
    "        return self.def_pos and pos not in self.def_pos\n",
    "\n",
    "\n",
    "\n",
    "    def bad_tag(self, tag):\n",
    "        return self.def_tag and tag not in self.def_tag\n",
    "\n",
    "\n",
    "\n",
    "    def bad_token(self, token):\n",
    "        return self.bad_pos(token.pos_) or self.bad_tag(token.tag_)\n",
    "\n",
    "\n",
    "\n",
    "    def match_tokens(self, verbose=False):\n",
    "        # SpaCy Doc DNE or Indexing Map DNE\n",
    "        if not self.main.sp_doc or not self.main.index_map:\n",
    "            raise Exception(\"DNE\")\n",
    "        \n",
    "        matched_data = []\n",
    "        matched_tokens = []\n",
    "\n",
    "        # Match by Regex\n",
    "        text = self.main.sp_doc.text.lower()\n",
    "        \n",
    "        for regex in self.regex:\n",
    "            matches = [(match.start(), match.end()) for match in re.finditer(regex, text, re.IGNORECASE)]\n",
    "            \n",
    "            if verbose and VERBOSE_LEVEL >= 2:\n",
    "                print(f\"\\t'{regex}' Regex Matches: {matches}\")\n",
    "            \n",
    "            for l_char_index, r_char_index in matches:\n",
    "                span = self.main.get_span_at_indices(l_char_index, r_char_index - 1)\n",
    "\n",
    "                if verbose and VERBOSE_LEVEL >= 3:\n",
    "                    print(f\"\\t\\tSpan Matched: {span}\")\n",
    "\n",
    "                for token in span:\n",
    "                    if verbose and VERBOSE_LEVEL >= 3:\n",
    "                        print(f\"\\t\\tPossible Regex Match for Token '{token}' (Position: {token.pos_} and Tag: {token.tag_})\")\n",
    "                        \n",
    "                    if self.bad_token(token):\n",
    "                        continue\n",
    "                    \n",
    "                    if verbose and VERBOSE_LEVEL >= 3:\n",
    "                        print(f\"\\t\\tRegex Matched Token '{token}'\")\n",
    "                        \n",
    "                    matched_tokens.append(token)\n",
    "                    matched_data.append(self.get_match_data(token, regex, Keywords.REGEX))\n",
    "\n",
    "        # Match by Rules\n",
    "        matches = self.rules(self.main.sp_doc)\n",
    "\n",
    "        if verbose and VERBOSE_LEVEL >= 2:\n",
    "            print(f\"\\tRule Matches: {matches}\")\n",
    "        \n",
    "        for match_id, start, end in matches:\n",
    "            span = self.main.sp_doc[start:end]\n",
    "            name = self.main.sp_nlp.vocab.strings[match_id]\n",
    "\n",
    "            if verbose and VERBOSE_LEVEL >= 2:\n",
    "                print(f\"\\tPattern '{name}' Matched Span: {span}\")\n",
    "            \n",
    "            for token in span:\n",
    "                if verbose and VERBOSE_LEVEL >= 3:\n",
    "                    print(f\"\\t\\tPossible Rule Match for Token '{token}' (Position: {token.pos_} and Tag: {token.tag_})\")\n",
    "                    \n",
    "                if self.bad_token(token):\n",
    "                    continue\n",
    "                \n",
    "                if verbose and VERBOSE_LEVEL >= 3:\n",
    "                    print(f\"\\t\\tRule Matched Token '{token}'\")\n",
    "\n",
    "                matched_tokens.append(token)\n",
    "                matched_data.append(self.get_match_data(token, name, Keywords.RULES))\n",
    "\n",
    "        # Match by Vocab\n",
    "        for token in self.main.sp_doc:\n",
    "            if verbose and VERBOSE_LEVEL >= 3:\n",
    "                    print(f\"\\t\\tPossible Vocab Match for Token '{token}' (Position: {token.pos_} and Tag: {token.tag_})\")\n",
    "                    \n",
    "            if self.bad_token(token) or token in matched_tokens:\n",
    "                continue\n",
    "\n",
    "            token_doc = self.main.sp_nlp(token.lower_)\n",
    "            token_lemma = \" \".join([t.lemma_ for t in token_doc])\n",
    "            \n",
    "            for vocab_word in self.vocab:\n",
    "                # Ensure Correct Tag\n",
    "                if vocab_word.get(\"tag\"):\n",
    "                    if not [t for t in token_doc if t.tag_ in vocab_word.get(\"tag\")]:\n",
    "                        if verbose and VERBOSE_LEVEL >= 4:\n",
    "                            print(f\"\\t\\t\\tToken '{token_doc}' not in Vocab Word '{vocab_word['doc']}' Tags ({vocab_word.get('tag')})\")\n",
    "                        continue\n",
    "                \n",
    "                # Ensure Correct PoS\n",
    "                if vocab_word.get(\"pos\"):\n",
    "                    if not [t for t in token_doc if t.pos_ in vocab_word.get(\"pos\")]:\n",
    "                        if verbose and VERBOSE_LEVEL >= 4:\n",
    "                            print(f\"\\t\\t\\tToken '{token_doc}' not in Vocab Word '{vocab_word['doc']}' Speech ({vocab_word.get('pos')})\")\n",
    "                        continue\n",
    "\n",
    "                # Check Lemma\n",
    "                if verbose and VERBOSE_LEVEL >= 4:\n",
    "                    print(f\"\\t\\t\\t{token_doc} Lemma ({token_lemma}) and {vocab_word['doc']} Lemma ({vocab_word['lemma']})\")\n",
    "                    \n",
    "                if token_lemma == vocab_word[\"lemma\"]:\n",
    "                    matched_tokens.append(token)\n",
    "                    matched_data.append(self.get_match_data(token, vocab_word[\"doc\"].text, Keywords.VOCAB))\n",
    "\n",
    "                    if verbose and VERBOSE_LEVEL >= 3:\n",
    "                        print(f\"\\t\\tVocab (Lemma) Matched Token '{token}'\")\n",
    "                    \n",
    "                    break\n",
    "\n",
    "                # Check Similarity\n",
    "                similarity = vocab_word[\"doc\"].similarity(token_doc)\n",
    "\n",
    "                if verbose and VERBOSE_LEVEL >= 4:\n",
    "                    print(f\"\\t\\t\\t{token_doc} and {vocab_word['doc']} Similarity: {similarity}\")\n",
    "                    \n",
    "                if similarity >= vocab_word.get(\"threshold\", self.def_threshold):\n",
    "                    matched_tokens.append(token)\n",
    "                    matched_data.append(self.get_match_data(token, vocab_word[\"doc\"].text, Keywords.VOCAB))\n",
    "\n",
    "                    if verbose and VERBOSE_LEVEL >= 3:\n",
    "                        print(f\"\\t\\tVocab Matched Token '{token}'\")\n",
    "                        \n",
    "                    break\n",
    "\n",
    "        # Mapping Match(ed Token) Data\n",
    "        mapped_matched_data = {}\n",
    "        for matched_token_data in matched_data:\n",
    "            mapped_matched_data[matched_token_data[\"token\"]] = matched_token_data\n",
    "\n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(\"Output of match_tokens\")\n",
    "            print(f\"Token Data: {matched_data}\")\n",
    "            print(f\"Mapped Token Data: {mapped_matched_data}\")\n",
    "            print(f\"Token: {matched_tokens}\")\n",
    "        \n",
    "        return matched_data, mapped_matched_data, matched_tokens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4ea399b8-867b-4c28-8653-3d868b13c9e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "class ExperimentKeywords(Keywords):\n",
    "    def __init__(self, main):\n",
    "        super().__init__(\n",
    "            main, \n",
    "            vocab=[\n",
    "                \"study\", \n",
    "                \"hypothesis\", \n",
    "                \"experiment\", \n",
    "                \"found\", \n",
    "                \"discover\", \n",
    "                \"compare\", \n",
    "                \"finding\", \n",
    "                \"result\", \n",
    "                \"test\", \n",
    "                \"examine\", \n",
    "                \"model\",\n",
    "                \"measure\",\n",
    "                \"manipulate\",\n",
    "                \"assess\",\n",
    "                \"conduct\",\n",
    "                \"data\",\n",
    "                \"analyze\",\n",
    "                \"sample\",\n",
    "                \"observe\",\n",
    "                \"observation\",\n",
    "                \"predict\",\n",
    "                \"suggest\",\n",
    "                \"method\",\n",
    "                \"investigation\",\n",
    "                \"trial\",\n",
    "                \"experimental\",\n",
    "                \"evidence\",\n",
    "                \"demonstrate\",\n",
    "                \"analysis\",\n",
    "                \"show\",\n",
    "                \"compare\",\n",
    "                \"comparable\",\n",
    "                \"control group\", \n",
    "                \"independent\",\n",
    "                \"dependent\",\n",
    "                \"applied\",\n",
    "                \"treatment\",\n",
    "                \"survery\",\n",
    "                \"evaluate\",\n",
    "                \"ran\"\n",
    "            ],\n",
    "            def_pos=[\"VERB\", \"NOUN\", \"ADJ\"], \n",
    "            def_threshold=0.8\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "54ac5d3c-5a5f-471f-8f2a-8369acd32f8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "class NegativeExperimentKeywords(Keywords):\n",
    "    def __init__(self, main):\n",
    "        super().__init__(\n",
    "            main, \n",
    "            # regex=[\n",
    "            #     \"theory\",\n",
    "            #     \"review\",\n",
    "            #     \"meta-analysis\"\n",
    "            # ],\n",
    "            patterns=[\n",
    "                [[{\"LOWER\": {\"IN\": [\"theory\", \"theorized\", \"theories\", \"review\", \"reviews\", \"meta-analysis\"]}}]]\n",
    "            ],\n",
    "            def_pos=[\"VERB\", \"NOUN\", \"ADJ\"], \n",
    "            def_threshold=0.8\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d62adc31-f9b9-4c42-8ae2-7189b69c8643",
   "metadata": {},
   "outputs": [],
   "source": [
    "class NegativeTopicKeywords(Keywords):\n",
    "    def __init__(self, main):\n",
    "        super().__init__(\n",
    "            main, \n",
    "            regexes=[\n",
    "                r\"co-?evolution\",\n",
    "                r\"evolution\",\n",
    "            ],\n",
    "            def_pos=[\"VERB\", \"NOUN\", \"ADJ\"], \n",
    "            def_threshold=0.8\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a16e7ce6-d3b1-4996-9892-e1f7516356e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CauseKeywords(Keywords):\n",
    "    def __init__(self, main):\n",
    "        super().__init__(\n",
    "            main, \n",
    "            vocab=[\n",
    "                \"increase\", \n",
    "                \"decrease\", \n",
    "                \"change\", \n",
    "                \"shift\", \n",
    "                \"cause\", \n",
    "                \"produce\", \n",
    "                \"trigger\", \n",
    "                \"suppress\", \n",
    "                \"inhibit\",\n",
    "                \"encourage\",\n",
    "                \"allow\",\n",
    "                \"influence\",\n",
    "                \"affect\",\n",
    "                \"alter\",\n",
    "                \"induce\",\n",
    "                \"produce\",\n",
    "                \"result in\",\n",
    "                # \"associated with\",\n",
    "                # \"correlated with\",\n",
    "                \"contribute\",\n",
    "                \"impact\",\n",
    "                \"deter\",\n",
    "                \"depressed\",\n",
    "                \"when\",\n",
    "                \"because\",\n",
    "                \"reduce\",\n",
    "                \"killed\",\n",
    "                # \"supported\"\n",
    "            ],\n",
    "            def_pos=[\"VERB\", \"SCONJ\", \"NOUN\"],\n",
    "            # def_tag=[\"VB\", \"VBD\", \"WRB\", \"IN\", \"VBG\"],\n",
    "            # def_threshold=0.75\n",
    "            def_threshold=0.8\n",
    "        )\n",
    "\n",
    "\n",
    "    \n",
    "    def update(self, verbose=False):\n",
    "        Keywords.update(self, verbose)\n",
    "        self.tokens = self.filter_tokens(self.tokens, verbose)\n",
    "\n",
    "\n",
    "    \n",
    "    def filter_tokens(self, tokens, verbose=False):\n",
    "        if not self.main.sp_doc or not self.main.index_map:\n",
    "            raise Exception(\"DNE\")\n",
    "        \n",
    "        return tokens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "76739bda-d97c-4c24-b3ee-bbdf2b4036a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "class ChangeKeywords(Keywords):\n",
    "    def __init__(self, main):\n",
    "        super().__init__(\n",
    "            main, \n",
    "            vocab=[\n",
    "                \"few\", \n",
    "                \"more\", \n",
    "                \"increase\", \n",
    "                \"decrease\", \n",
    "                \"less\", \n",
    "                \"short\", \n",
    "                \"long\", \n",
    "                \"greater\"\n",
    "                \"shift\",\n",
    "                \"fluctuate\",\n",
    "                \"adapt\",\n",
    "                \"grow\",\n",
    "                \"rise\"\n",
    "                \"surge\",\n",
    "                \"intensify\",\n",
    "                \"amplify\",\n",
    "                \"multiply\",\n",
    "                \"decline\",\n",
    "                \"reduce\",\n",
    "                \"drop\",\n",
    "                \"diminish\",\n",
    "                \"fall\",\n",
    "                \"lessen\",\n",
    "                \"doubled\",\n",
    "                \"tripled\",\n",
    "                \"lower\",\n",
    "                \"adjust\",\n",
    "                \"reject\",\n",
    "            ],\n",
    "            regexes=[\n",
    "                # Match Examples:\n",
    "                # 1. \"one... as...\"\n",
    "                # 2. \"2x than...\"\n",
    "                r\"(one|two|three|four|five|six|seven|eight|nine|ten|twice|thrice|([0-9]+|[0-9]+.[0-9]+)(x|%))[\\s-]+[^\\s]*[\\s-]+(as|more|than|likely)([\\s-]+|$)\"\n",
    "            ],\n",
    "            def_pos=[\"NOUN\", \"ADJ\", \"ADV\", \"VERB\", \"NUM\"],\n",
    "            def_threshold=0.75\n",
    "        )\n",
    "\n",
    "\n",
    "    \n",
    "    def update(self, verbose=False):\n",
    "        Keywords.update(self, verbose=verbose)\n",
    "        self.tokens = self.filter_tokens(self.tokens, verbose)\n",
    "\n",
    "\n",
    "    \n",
    "    def filter_tokens(self, tokens, verbose=False):\n",
    "        if not self.main.sp_doc or not self.main.index_map:\n",
    "            raise Exception(\"DNE\")\n",
    "\n",
    "        filtered = []\n",
    "        for token in self.main.sp_doc:\n",
    "            # Already Matched\n",
    "            if token in tokens:\n",
    "                filtered.append(token)\n",
    "            \n",
    "            # Comparative Adjective\n",
    "            # Looking for words like \"bigger\" and \"better\".\n",
    "            elif token.pos_ == \"ADJ\" and token.tag_ == \"JJR\":\n",
    "                filtered.append(token)\n",
    "                continue\n",
    "            \n",
    "        return filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d878feea-bcfa-410b-a383-26f12f1a69bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "class TraitKeywords(Keywords):\n",
    "    FOOD = \"Food\"\n",
    "    PRESENT = \"Present\"\n",
    "    NOT_APPLICABLE = \"N/A\"\n",
    "    \n",
    "    def __init__(self, main):\n",
    "        super().__init__(\n",
    "            main, \n",
    "            regexes=[\n",
    "                r\"behaviou?r\", \n",
    "                r\"[^A-Za-z]+rate\", \n",
    "                \"colou?r\",\n",
    "                \"biomass\",\n",
    "                r\"[^A-Za-z]+mass\", \n",
    "                r\"[^A-Za-z]+size\",\n",
    "                \"number\",\n",
    "                \"length\", \n",
    "                \"pattern\", \n",
    "                \"weight\",\n",
    "                \"shape\", \n",
    "                \"efficiency\", \n",
    "                \"trait\",\n",
    "                \"phenotype\",\n",
    "                \"demography\",\n",
    "                \"scent\",\n",
    "                \"population (structure|mechanic)s?\",\n",
    "                \"ability\", \n",
    "                \"capacity\", \n",
    "                \"height\", \n",
    "                \"width\", \n",
    "                \"[A-Za-z]+span\",\n",
    "                {\"regex\": \"diet\", \"types\": [TraitKeywords.FOOD]},\n",
    "                {\"regex\": \"food\", \"types\": [TraitKeywords.FOOD, TraitKeywords.NOT_APPLICABLE]},\n",
    "                {\"regex\": \"feeding\", \"types\": [TraitKeywords.FOOD]},\n",
    "                \"nest\",\n",
    "                \"substrate\",\n",
    "                \"breeding\",\n",
    "                r\"[^A-Za-z]+age[^A-Za-z]+\",\n",
    "                \"lifespan\",\n",
    "                \"development\",\n",
    "                \"output\",\n",
    "                \"time\",\n",
    "                \"period\",\n",
    "                \"level\",\n",
    "                \"configuration\",\n",
    "                \"dimorphism\",\n",
    "                \"capability\",\n",
    "                \"regulation\",\n",
    "                \"excretion\",\n",
    "                \"luminescence\",\n",
    "                r\"[^A-Za-z]+role\",\n",
    "                \"sensitivity\",\n",
    "                \"resistance\",\n",
    "                r\"(un|(^|\\s)[A-Za-z]*-)infected\",\n",
    "                \"temperature\",\n",
    "                \"density\",\n",
    "                {\"regex\": \"present\", \"types\": [TraitKeywords.PRESENT]},\n",
    "                {\"regex\": \"presence\", \"types\": [TraitKeywords.PRESENT]},\n",
    "            ],\n",
    "            # Ideally, I would only include nouns, but sometimes\n",
    "            # they're recognized as adjectives.\n",
    "            def_pos=[\"NOUN\", \"ADJ\"]\n",
    "        )\n",
    "\n",
    "\n",
    "    \n",
    "    def update(self, verbose=False):\n",
    "        Keywords.update(self, verbose)\n",
    "        self.tokens = self.filter_tokens(self.tokens, verbose)\n",
    "\n",
    "\n",
    "    \n",
    "    def filter_tokens(self, tokens, verbose=False):\n",
    "        verbose=True\n",
    "        if not self.main.sp_doc or not self.main.index_map:\n",
    "            raise Exception(\"DNE\")\n",
    "\n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(f\"Unfiltered Trait Tokens: {tokens}\")\n",
    "        \n",
    "        filtered = []\n",
    "        for token in tokens:\n",
    "            expanded_token = self.main.expand_unit(\n",
    "                il_unit=token.i, \n",
    "                ir_unit=token.i, \n",
    "                il_boundary=token.sent.start, \n",
    "                ir_boundary=token.sent.end-1, \n",
    "                speech=[\"PUNCT\"],\n",
    "                include=False,\n",
    "                verbose=verbose\n",
    "            )\n",
    "\n",
    "            # print(\"Filtering Token\")\n",
    "            # print(token)\n",
    "            # print(expanded_token)\n",
    "            \n",
    "            valid_token = True\n",
    "            \n",
    "            if not self.main.species.has_species(expanded_token, verbose=verbose):\n",
    "                valid_token = False\n",
    "\n",
    "            for chunk in self.main.sp_doc.noun_chunks:\n",
    "                # print(chunk)\n",
    "                if token in chunk and chunk[-1] != token:\n",
    "                    valid_token = False\n",
    "\n",
    "            if valid_token:\n",
    "                filtered.append(token)\n",
    "            \n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(f\"Filtered Trait Tokens: {filtered}\")\n",
    "        \n",
    "        return filtered\n",
    "\n",
    "\n",
    "    def merge_traits(self, traits):\n",
    "        merged = {}\n",
    "        \n",
    "        for trait in traits:\n",
    "            found_trait = False\n",
    "            for m_trait in merged:\n",
    "                if self.main.has_same_base_nouns(trait, m_trait) or self.main.is_same_text(trait, m_trait):\n",
    "                    merged[m_trait].append(trait)\n",
    "                    found_trait = True\n",
    "                    break\n",
    "                \n",
    "                trait_types = []\n",
    "                for thing in trait:\n",
    "                    things = self.main.trait.mapped_token_data.get(thing)\n",
    "                    if things:\n",
    "                        trait_types.extend(things[\"types\"])\n",
    "                    \n",
    "                m_trait_types = []\n",
    "                for thing in m_trait:\n",
    "                    things = self.main.trait.mapped_token_data.get(thing)\n",
    "                    if things:\n",
    "                        m_trait_types.extend(things[\"types\"])\n",
    "    \n",
    "                # print(trait_types, m_trait_types)\n",
    "    \n",
    "                if not trait_types or not m_trait_types:\n",
    "                    continue\n",
    "    \n",
    "                type_intersection = set(trait_types).intersection(m_trait_types)\n",
    "                if type_intersection.intersection(['Food']):\n",
    "                    merged[m_trait].append(trait)\n",
    "                    found_trait = True\n",
    "                    break\n",
    "                \n",
    "            if not found_trait:\n",
    "                merged[trait] = [trait]\n",
    "    \n",
    "        return merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6d7c2f31-78e9-4254-939c-e257b59df743",
   "metadata": {},
   "outputs": [],
   "source": [
    "class TestKeywords(Keywords):\n",
    "    def __init__(self, main):\n",
    "        super().__init__(\n",
    "            main,\n",
    "            vocab=[\n",
    "                \"compare\",\n",
    "                \"examine\",\n",
    "                \"evaluate\",\n",
    "                \"assess\",\n",
    "            ],\n",
    "            def_pos=[\"VERB\", \"NOUN\"], \n",
    "            def_threshold=0.8\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ebb3e62e-b50e-4810-ab70-441ed56c673b",
   "metadata": {},
   "outputs": [],
   "source": [
    "class VariabilityKeywords(Keywords):\n",
    "    def __init__(self, main):\n",
    "        super().__init__(\n",
    "            main,\n",
    "            vocab=[\n",
    "                {\"word\": \"different\", \"pos\": [\"ADJ\", \"NOUN\"]},\n",
    "                {\"word\": \"vary\", \"pos\": [\"VERB\", \"NOUN\"]},\n",
    "                {\"word\": \"varied\", \"pos\": [\"VERB\", \"NOUN\"]},\n",
    "                {\"word\": \"compare\", \"pos\": [\"VERB\"]}\n",
    "            ],\n",
    "            regexes=[\n",
    "                r\"between\",\n",
    "                r\"against\",\n",
    "                r\"independen(t|ts|tly|cy)\",\n",
    "                r\"dependen(t|ts|tly|cy)\",\n",
    "                r\"treatments?\",\n",
    "                r\"effect\",\n",
    "                r\"control\",\n",
    "                r\"(with|without)[A-Za-z]*(with|without)\",\n",
    "                r\"(^| )(un|not)[-| ]?([A-Za-z]+) [^!;?.\\n]* \\3\",\n",
    "                r\"([A-Za-z]+) [^!;?.\\n]* (un|not)[-| ]?\\1( |$)\",\n",
    "                # I've added these two words because they can sometimes\n",
    "                # hint at a variable.\n",
    "                r\"when\",\n",
    "                r\"where\",\n",
    "                \n",
    "            ],\n",
    "            patterns=[\n",
    "                [[{\"LOWER\": {\"IN\": [\"neither\", \"either\", \"both\"]}}, {\"OP\": \"*\", \"TAG\": {\"NOT_IN\": [\".\"]}}, {\"LOWER\": {\"IN\": [\"or\", \"and\"]}}]],\n",
    "                [[{\"LOWER\": {\"IN\": [\"with\", \"without\"]}}, {\"OP\": \"*\", \"TAG\": {\"NOT_IN\": [\".\"]}}, {\"LOWER\": {\"IN\": [\"with\", \"without\"]}}]],\n",
    "                [[{\"LOWER\": {\"IN\": [\"at\"]}}, {\"POS\": \"NUM\"}]],\n",
    "                [[{\"LOWER\": {\"IN\": [\"at\"]}}, {\"LOWER\": {\"IN\": [\"several\", \"unique\", \"multiple\", \"different\"]}}]],\n",
    "            ],\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "29c257a4-08e4-4c57-b780-9f6a9ddac683",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Main(Base):\n",
    "    def __init__(self):\n",
    "        self.sp_nlp = spacy.load(\"en_core_web_lg\")\n",
    "        self.sp_doc = None\n",
    "        super().__init__(self)\n",
    "\n",
    "        # Maps Character Position to Token in Document\n",
    "        # Used to handle differences between different\n",
    "        # pipelines and tools.\n",
    "        self.index_map = None\n",
    "\n",
    "        # Parsers\n",
    "        # This is currently not being used, I may\n",
    "        # add it back later.\n",
    "        # self.parts = Parts(self)\n",
    "        self.species = Species(self)\n",
    "        self.trait = TraitKeywords(self)\n",
    "        self.cause = CauseKeywords(self)\n",
    "        self.change = ChangeKeywords(self)\n",
    "        self.experiment = ExperimentKeywords(self)\n",
    "        self.variability = VariabilityKeywords(self)\n",
    "        self.test = TestKeywords(self)\n",
    "        self.not_experiment = NegativeExperimentKeywords(self)\n",
    "        self.not_topic = NegativeTopicKeywords(self)\n",
    "\n",
    "\n",
    "    \n",
    "    def update_doc(self, doc, verbose=False):\n",
    "        self.sp_doc = doc\n",
    "        self.index_map = self.load_index_map()\n",
    "        # self.parts.update()\n",
    "        self.species.update(doc.text, verbose=False)\n",
    "        self.trait.update(verbose=False)\n",
    "        self.cause.update(verbose=False)\n",
    "        self.change.update(verbose=False)\n",
    "        self.experiment.update(verbose=False)\n",
    "        self.not_experiment.update(verbose=False)\n",
    "        self.not_topic.update(verbose=False)\n",
    "        self.variability.update(verbose=False)\n",
    "        self.test.update(verbose=False)\n",
    "\n",
    "\n",
    "    \n",
    "    def update_text(self, text, verbose=False):\n",
    "        self.sp_doc = self.sp_nlp(text)\n",
    "        self.update_doc(self.sp_doc, verbose=verbose)\n",
    "\n",
    "\n",
    "\n",
    "    def load_index_map(self):\n",
    "        # SpaCy Doc Not Found\n",
    "        if self.sp_doc is None:\n",
    "            raise Exception(\"DNE\")\n",
    "\n",
    "        # Map Character Index to Token\n",
    "        index_map = {}\n",
    "        for token in self.sp_doc:\n",
    "            l_char_index = token.idx\n",
    "            r_char_index = token.idx + len(token)\n",
    "\n",
    "            for i in range(l_char_index, r_char_index):\n",
    "                index_map[i] = token\n",
    "\n",
    "        return index_map\n",
    "\n",
    "\n",
    "\n",
    "    def token_at_char(self, char_index):\n",
    "        # SpaCy Doc or Indexing Map Not Found\n",
    "        if not self.sp_doc or not self.index_map:\n",
    "            raise Exception(\"DNE\")\n",
    "\n",
    "        if char_index in self.index_map:\n",
    "            return self.index_map[char_index]\n",
    "\n",
    "        raise Exception(f\"Token at Index {char_index} Not Found\")\n",
    "    \n",
    "\n",
    "    \n",
    "    def valid_trait_token(self, data, verbose=False):\n",
    "        verbose = True\n",
    "        token = data[\"token\"]\n",
    "        \n",
    "        # print(f\"Validate Trait Token '{token}'\")\n",
    "        if token not in self.trait.tokens:\n",
    "            return 0\n",
    "\n",
    "        # print(f\"Token '{token}' in Trait Tokens\")\n",
    "        \n",
    "        # Check if Applicable\n",
    "        token_data = self.trait.mapped_token_data[token]\n",
    "        \n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(f\"Token '{token}'\")\n",
    "            print(f\"Types: {token_data['types']}\")\n",
    "            \n",
    "        if TraitKeywords.NOT_APPLICABLE in token_data[\"types\"]:\n",
    "            # print(f\"\\tToken Not Applicable\")\n",
    "            return 0\n",
    "\n",
    "        token_context = set(self.find_unit_context(\n",
    "            il_unit=token.i, \n",
    "            ir_unit=token.i, \n",
    "            il_boundary=token.sent.start, \n",
    "            ir_boundary=token.sent.end-1, \n",
    "            verbose=verbose)\n",
    "        )\n",
    "\n",
    "        # print(f\"Token Context: {token_context}\")\n",
    "        \n",
    "        causes = set(data[\"sent_cause_tokens\"]).intersection(token_context)\n",
    "        changes = set(data[\"sent_change_tokens\"]).intersection(token_context)\n",
    "\n",
    "        # print(f\"Causes and Changes: {[*causes, *changes]}\")\n",
    "        \n",
    "        return 1.0 if causes or changes else 0.25\n",
    "\n",
    "\n",
    "\n",
    "    def valid_species_token(self, data, verbose=False):\n",
    "        token = data[\"token\"]\n",
    "        if token not in self.species.tokens:\n",
    "            return 0\n",
    "        \n",
    "        token_context = set(self.find_unit_context(\n",
    "            il_unit=token.i, \n",
    "            ir_unit=token.i, \n",
    "            il_boundary=token.sent.start, \n",
    "            ir_boundary=token.sent.end-1, \n",
    "            verbose=verbose)\n",
    "        )\n",
    "        \n",
    "        causes = set(data[\"sent_cause_tokens\"]).intersection(token_context)\n",
    "        changes = set(data[\"sent_change_tokens\"]).intersection(token_context)\n",
    "\n",
    "        return 1 if causes or changes else 0\n",
    "    \n",
    "    \n",
    "\n",
    "    def valid_trait(self, verbose=False):\n",
    "        traits = self.trait.merge_traits([self.sp_doc[token.i:token.i+1] for token in self.trait.tokens])\n",
    "\n",
    "        # For full points, you should have a trait that's mentioned\n",
    "        # at least two times in different areas.\n",
    "        two_mentions = False\n",
    "\n",
    "        for trait, instances in traits.items():\n",
    "            if len(instances) < 2:\n",
    "                continue\n",
    "\n",
    "            locations = set([instance.sent.start for instance in instances])\n",
    "            if len(locations) >= 2:\n",
    "                return 1\n",
    "\n",
    "        if traits:\n",
    "            return 0.33\n",
    "        return 0\n",
    "\n",
    "\n",
    "    \n",
    "    def valid_trait_variation(self, verbose=False):\n",
    "        verbose=True\n",
    "        max_trait_variation_points = 0\n",
    "        \n",
    "        sentences = list(self.sp_doc.sents)\n",
    "        num_sentences = len(sentences)\n",
    "\n",
    "        for i in range(num_sentences):\n",
    "            sent_i = sentences[i]\n",
    "            sent_i_tokens = set([token for token in sent_i])\n",
    "\n",
    "            if verbose and VERBOSE_LEVEL >= 2:\n",
    "                print(f\"\\tSentence I: {sent_i}\")\n",
    "            \n",
    "            sent_i_test_tokens = sent_i_tokens.intersection(self.test.tokens)\n",
    "            sent_i_experiment_tokens = sent_i_tokens.intersection(self.experiment.tokens)\n",
    "\n",
    "            if verbose and VERBOSE_LEVEL >= 2:\n",
    "                print(f\"\\tSentence I Test Tokens: {sent_i_test_tokens}\")\n",
    "                print(f\"\\tSentence I Experiment Tokens: {sent_i_experiment_tokens}\")\n",
    "\n",
    "            if not sent_i_test_tokens and not sent_i_experiment_tokens:\n",
    "                if verbose and VERBOSE_LEVEL >= 2:\n",
    "                    print(f\"\\tNo Experiment or Test Tokens in Sentence I\")\n",
    "                continue\n",
    "\n",
    "            trait_variation_points_i = 0\n",
    "\n",
    "            if sent_i_experiment_tokens:\n",
    "                trait_variation_points_i = 0.10\n",
    "            \n",
    "            if sent_i_test_tokens:\n",
    "                trait_variation_points_i = 0.25\n",
    "\n",
    "            if verbose and VERBOSE_LEVEL >= 2:\n",
    "                print(f\"\\tTrait Variation Points for I: {trait_variation_points_i}\")\n",
    "\n",
    "            sent_i_trait_tokens = sent_i_tokens.intersection(self.trait.tokens)\n",
    "            sent_i_species_tokens = sent_i_tokens.intersection(self.species.tokens)\n",
    "\n",
    "            if verbose and VERBOSE_LEVEL >= 2:\n",
    "                print(f\"\\tSentence I Trait Tokens: {sent_i_trait_tokens}\")\n",
    "\n",
    "            if not sent_i_trait_tokens and not sent_i_species_tokens:\n",
    "                if verbose and VERBOSE_LEVEL >= 2:\n",
    "                    print(f\"\\tNo Trait or Specie Tokens in Sentence I\")\n",
    "                continue\n",
    "\n",
    "            t_variables = []\n",
    "            s_variables = []\n",
    "            sent_i_variability_tokens = sent_i_tokens.intersection(self.variability.tokens)\n",
    "\n",
    "            if verbose and VERBOSE_LEVEL >= 2:\n",
    "                print(f\"\\tSentence I Variability Tokens: {sent_i_variability_tokens}\")\n",
    "\n",
    "            deduct_points = not sent_i_variability_tokens\n",
    "            \n",
    "            if sent_i_variability_tokens:\n",
    "                for token in sent_i_variability_tokens:\n",
    "                    trait_in_context = set(self.find_unit_context(\n",
    "                        il_unit=token.i, \n",
    "                        ir_unit=token.i, \n",
    "                        il_boundary=token.sent.start,\n",
    "                        ir_boundary=token.sent.end-1, \n",
    "                        speech=[\"PUNCT\"],\n",
    "                        include=False,\n",
    "                        comma_encloses=True,\n",
    "                        verbose=verbose\n",
    "                    )).intersection(self.trait.tokens)\n",
    "\n",
    "                    specie_in_context = set(self.find_unit_context(\n",
    "                        il_unit=token.i, \n",
    "                        ir_unit=token.i, \n",
    "                        il_boundary=token.sent.start,\n",
    "                        ir_boundary=token.sent.end-1, \n",
    "                        speech=[\"PUNCT\"],\n",
    "                        include=False,\n",
    "                        comma_encloses=True,\n",
    "                        verbose=verbose\n",
    "                    )).intersection(self.species.tokens)\n",
    "                    \n",
    "                    if verbose and VERBOSE_LEVEL >= 3:\n",
    "                        print(f\"\\t\\tVariability Token '{token}' Traits in Context: {trait_in_context}\") \n",
    "                        print(f\"\\t\\tVariability Token '{token}' Specie in Context: {specie_in_context}\") \n",
    "\n",
    "                    if not specie_in_context and not trait_in_context:\n",
    "                        if verbose and VERBOSE_LEVEL >= 3:\n",
    "                            print(f\"\\t\\tNo Traits in Variability Token '{token}' Context\")\n",
    "                        continue\n",
    "\n",
    "                    deduct_points = False\n",
    "                    t_variables.extend(trait_in_context)\n",
    "                    s_variables.extend(specie_in_context)\n",
    "\n",
    "            t_variables = list(set(t_variables))\n",
    "            s_variables = list(set(s_variables))\n",
    "\n",
    "            if verbose and VERBOSE_LEVEL >= 2:\n",
    "                print(f\"\\tt_variables: {t_variables}\")\n",
    "                print(f\"\\ts_variables: {s_variables}\")\n",
    "            \n",
    "            if t_variables or s_variables:\n",
    "                trait_variation_points_i += 0.25\n",
    "            else:\n",
    "                trait_variation_points_i += 0.15\n",
    "\n",
    "            if verbose and VERBOSE_LEVEL >= 2:\n",
    "                print(f\"\\tTrait Variation Points for I: {trait_variation_points_i}\")\n",
    "\n",
    "            assert trait_variation_points_i <= 0.5\n",
    "\n",
    "            # Apparently, Lysandra in the past believed all the\n",
    "            # points should be earned. I didn't end up removing this,\n",
    "            # so there's a chance I had somewhat of a reason.\n",
    "            # Perhaps, I did not in fact have a reason?\n",
    "            # if trait_variation_points_i <= 0.49:\n",
    "            #     continue\n",
    "\n",
    "            for j in range(i, num_sentences):\n",
    "                sent_j = sentences[j]\n",
    "                sent_j_tokens = set([token for token in sent_j])\n",
    "\n",
    "                if verbose and VERBOSE_LEVEL >= 2:\n",
    "                    print(f\"\\tSentence J: {sent_j}\")\n",
    "\n",
    "                sent_j_cause_tokens = sent_j_tokens.intersection(self.cause.tokens)\n",
    "                sent_j_change_tokens = sent_j_tokens.intersection(self.change.tokens)\n",
    "                sent_j_species_tokens = sent_j_tokens.intersection(self.species.span_starts)\n",
    "                sent_j_trait_tokens = sent_j_tokens.intersection(self.trait.tokens)\n",
    "\n",
    "                if verbose and VERBOSE_LEVEL >= 2:\n",
    "                    print(f\"\\tSentence J Cause Tokens: {sent_j_cause_tokens}\")\n",
    "                    print(f\"\\tSentence J Change Tokens: {sent_j_change_tokens}\")\n",
    "                    print(f\"\\tSentence J Species Tokens: {sent_j_species_tokens}\")\n",
    "                    print(f\"\\tSentence J Trait Tokens: {sent_j_trait_tokens}\")\n",
    "                \n",
    "                if not sent_j_species_tokens or (not sent_j_cause_tokens and not sent_j_change_tokens):\n",
    "                    if verbose and VERBOSE_LEVEL >= 2:\n",
    "                        print(f\"\\tUnsatisfied Conditions for Sentence J\")\n",
    "                    continue\n",
    "\n",
    "                trait_variation_points_j = 0\n",
    "                \n",
    "                if (not sent_j_species_tokens and not sent_j_trait_tokens) or (not t_variables and not s_variables):\n",
    "                    trait_variation_points_j += 0.25\n",
    "                elif i == j:\n",
    "                    trait_variation_points_j += 0.25\n",
    "                elif i != j:\n",
    "                    # Check if Variable Referenced Again via Types\n",
    "                    variable_types = set(self.flatten([self.trait.mapped_token_data[token][\"types\"] for token in t_variables]))\n",
    "                    sent_j_trait_types = set(self.flatten([self.trait.mapped_token_data[token][\"types\"] for token in sent_j_trait_tokens]))\n",
    "\n",
    "                    if verbose and VERBOSE_LEVEL >= 2:\n",
    "                        print(f\"\\tVariable Types: {variable_types}\")\n",
    "                        print(f\"\\tTrait Types in Sentence J: {sent_j_trait_types}\")\n",
    "                        \n",
    "                    # Check if Variable Referenced Again via Literals\n",
    "                    variable_strings = set([token.lower_ for token in t_variables])\n",
    "                    sent_j_trait_strings = set([token.lower_ for token in sent_j_trait_tokens])\n",
    "\n",
    "                    if verbose and VERBOSE_LEVEL >= 2:\n",
    "                        print(f\"\\tVariable Trait (as Strings): {variable_strings}\")\n",
    "                        print(f\"\\tTrait (as Strings) in Sentence J: {sent_j_trait_strings}\")\n",
    "\n",
    "                    # Check if Trait Referenced\n",
    "                    t_variable_referenced = bool(variable_types & sent_j_trait_types) or bool(variable_strings & sent_j_trait_strings)\n",
    "\n",
    "                    # Check if Species Rerferenced\n",
    "                    sent_j_species_spans = [self.sp_doc[token.i:token.i+1] for token in sent_j_species_tokens]\n",
    "                    s_variables_spans = [self.sp_doc[token.i:token.i+1] for token in s_variables]\n",
    "                    overlap = [self.species.find_same_species(sent_j_species_spans, species) for species in s_variables_spans]\n",
    "                    s_variable_referenced = any(overlap)\n",
    "\n",
    "                    if verbose and VERBOSE_LEVEL >= 2:\n",
    "                        print(f\"\\tT Variable Referenced? {t_variable_referenced}\")\n",
    "                        print(f\"\\tS Variable Referenced? {s_variable_referenced}\")\n",
    "                    \n",
    "                    if t_variable_referenced or s_variable_referenced:\n",
    "                        trait_variation_points_j += 0.50\n",
    "                    else:\n",
    "                        trait_variation_points_j += 0.25\n",
    "\n",
    "                if verbose and VERBOSE_LEVEL >= 2:\n",
    "                    print(f\"\\tTrait Variation Points for J: {trait_variation_points_j}\")\n",
    "\n",
    "                assert trait_variation_points_j <= 0.5\n",
    "\n",
    "                if verbose and VERBOSE_LEVEL >= 2:\n",
    "                    print(f\"\\ti: {i}\")\n",
    "                    print(f\"\\tj: {j}\")\n",
    "\n",
    "                trait_variation_points = trait_variation_points_i + trait_variation_points_j\n",
    "\n",
    "                if deduct_points:\n",
    "                    trait_variation_points *= 0.6375\n",
    "                \n",
    "                if verbose and VERBOSE_LEVEL >= 2:\n",
    "                    print(f\"\\tTrait Variation Points: {trait_variation_points}\")\n",
    "                \n",
    "                max_trait_variation_points = max(max_trait_variation_points, trait_variation_points)\n",
    "                \n",
    "                if max_trait_variation_points >= 1:\n",
    "                    return 1\n",
    "\n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(f\"Max Trait Variation Points: {max_trait_variation_points}\")\n",
    "            \n",
    "        return max_trait_variation_points\n",
    "\n",
    "\n",
    "    \n",
    "    def update_seen_species(self, data, verbose=False):\n",
    "        # Unpacking Data\n",
    "        token = data[\"token\"]\n",
    "        seen_species = data[\"seen_species\"]\n",
    "        sent_seen_species = data[\"sent_seen_species\"]\n",
    "        sent_num_unique_species = data[\"sent_num_unique_species\"]\n",
    "        \n",
    "        # Update Seen Species in Entire Text\n",
    "        span = self.species.span_at_token(token)\n",
    "        prev_ref = self.species.find_same_species(seen_species.keys(), span, verbose=verbose)\n",
    "        \n",
    "        if prev_ref:\n",
    "            seen_species[prev_ref] += 1\n",
    "        else:\n",
    "            seen_species[span] = 1\n",
    "\n",
    "        # Check Seen Species in Sentence\n",
    "        # We only add points if it's a species that has not been seen in the sentence. \n",
    "        # This is to avoid redundant points. The points are used so that a species that\n",
    "        # has been previously seen can still be used as long as it hasn't been awarded\n",
    "        # points already.\n",
    "        prev_sent_ref = self.species.find_same_species(sent_seen_species.keys(), span, verbose=verbose)\n",
    "        if prev_sent_ref:\n",
    "            sent_seen_species[prev_sent_ref][\"visits\"] += 1\n",
    "            ref = prev_sent_ref\n",
    "        else:\n",
    "            sent_seen_species[span] = {\n",
    "                \"visits\": 1,\n",
    "                \"points\": 0\n",
    "            }\n",
    "            ref = span\n",
    "        \n",
    "        # Update Number of Unique Species in Sentence\n",
    "        if not prev_sent_ref:\n",
    "            sent_num_unique_species += 1\n",
    "\n",
    "        return {\n",
    "            \"seen_species\": seen_species, \n",
    "            \"sent_seen_species\": sent_seen_species, \n",
    "            # Reference to the species, either its previous\n",
    "            # reference, or the given species.\n",
    "            \"ref\": ref,\n",
    "            # Reference to the species that was last seen\n",
    "            # in the sentence.\n",
    "            \"seen_in_sent\": prev_sent_ref, \n",
    "            \"sent_num_unique_species\": sent_num_unique_species, \n",
    "        }\n",
    "\n",
    "    \n",
    "    \n",
    "    def score(self, verbose=False):\n",
    "        NUM_CATEGORIES = 6\n",
    "\n",
    "        TRAIT = 0\n",
    "        SPECIES = 1\n",
    "        EXPERIMENT = 2\n",
    "        INTERACTION = 3\n",
    "        NOT_TOPIC = 4\n",
    "        TRAIT_VARIATION = 5\n",
    "\n",
    "        # Max # of Points of Category per Sentence (MPC)\n",
    "        # A category can collect points from each sentence. However,\n",
    "        # there's a maximum number of points it can collect. This is\n",
    "        # determined by the MPC.\n",
    "        MPC = [1] * NUM_CATEGORIES\n",
    "    \n",
    "        # Points per Instance of Category (PIC)\n",
    "        # Each token is evaluated to check whether a category\n",
    "        # can be given points. The number of points given, if\n",
    "        # the token is determined to be satisfactory, is the PIC.\n",
    "        # The PIC is less than or equal to the MPC for the corresponding\n",
    "        # category. The idea behind the PIC and MPC is similar to how\n",
    "        # sets work in tennis: you're not immediately awarded the full points\n",
    "        # for the set (MPC) if your opponent fails to return the ball,\n",
    "        # instead you're given a smaller # of points (PIC) that allow you to\n",
    "        # incrementally win the set (category).\n",
    "        PIC = [0] * NUM_CATEGORIES\n",
    "        PIC[TRAIT] = MPC[TRAIT]\n",
    "        PIC[SPECIES] = MPC[SPECIES]/2.0\n",
    "        PIC[EXPERIMENT] = MPC[EXPERIMENT]*1\n",
    "        PIC[INTERACTION] = MPC[INTERACTION]/2.0\n",
    "        PIC[NOT_TOPIC] = MPC[NOT_TOPIC]\n",
    "\n",
    "        for i in range(NUM_CATEGORIES):\n",
    "            assert 0 <= PIC[i] <= MPC[i]\n",
    "\n",
    "        # Category Weights (CW)\n",
    "        # It may be helpful to weigh a certain category's fraction of total points\n",
    "        # more or less than another's. Thus, at the end, we'll take a\n",
    "        # weighted average of the category's FTP. The weights must add up to 1.\n",
    "        CW = [0] * NUM_CATEGORIES\n",
    "        CW[TRAIT] = 0.3\n",
    "        CW[SPECIES] = 0.1\n",
    "        CW[EXPERIMENT] = 0.1\n",
    "        CW[INTERACTION] = 0.1\n",
    "        CW[NOT_TOPIC] = 0.1\n",
    "        CW[TRAIT_VARIATION] = 0.3\n",
    "\n",
    "        assert round(np.sum(CW)) == 1\n",
    "\n",
    "        # Leniency\n",
    "        # There are certain categories that aren't going to be as frequent as others.\n",
    "        # For example, the trait category. You could try and decrease the influence\n",
    "        # of said category by lowering its MPC and/or increasing the PIC (so that it's\n",
    "        # easier to achieve the FTP). However, this could make it harder to meaningfully\n",
    "        # represent the category. The idea of leniency is to remove (some) sentences that had 0\n",
    "        # points from the scoring. This increases the FTP as, for example, instead of comparing\n",
    "        # 0.5 points to a total of 2.5 points, you can compare 0.5 to 2.0 points, and so on.\n",
    "        # A leniency of 1 means that all sentences that received 0 points will be removed from\n",
    "        # the scoring. A leniency of 0 means that all the sentences are included in the scoring.\n",
    "        LEN = [0] * NUM_CATEGORIES\n",
    "        LEN[TRAIT] = 0\n",
    "        LEN[SPECIES] = 0.5\n",
    "        LEN[EXPERIMENT] = 0.5\n",
    "        LEN[INTERACTION] = 0.2\n",
    "        LEN[NOT_TOPIC] = 0\n",
    "\n",
    "        for i in range(NUM_CATEGORIES):\n",
    "            assert 0 <= LEN[i] <= 1\n",
    "\n",
    "        # Banned Sentences\n",
    "        # Not allowed to benefit from leniency.\n",
    "        banned = [[0] * len(list(self.sp_doc.sents)) for _ in range(NUM_CATEGORIES)]\n",
    "\n",
    "        # Points\n",
    "        points = [0] * NUM_CATEGORIES\n",
    "        binned_points = [0] * NUM_CATEGORIES\n",
    "        \n",
    "        num_zero_pt_sents = [0] * NUM_CATEGORIES\n",
    "        seen_species = {}\n",
    "\n",
    "        # For Testing\n",
    "        species_instances = []\n",
    "        interaction_instances = []\n",
    "        \n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(\"Extracted Information\")\n",
    "            print(f\"Cause Tokens: {self.cause.tokens}\")\n",
    "            print(f\"Change Tokens: {self.change.tokens}\")\n",
    "            print(f\"Trait Tokens: {self.trait.tokens}\")\n",
    "            print(f\"Species Tokens: {self.species.tokens}\")\n",
    "            print(f\"Experiment Tokens: {self.experiment.tokens}\")\n",
    "            print(f\"Not-Experiment Tokens: {self.not_experiment.tokens}\")\n",
    "            print(f\"Not-Topic Tokens: {self.not_topic.tokens}\")\n",
    "            print(f\"Variability Tokens: {self.variability.tokens}\")\n",
    "            print(f\"Test Tokens: {self.test.tokens}\")\n",
    "\n",
    "        # print(f\"Species Tokens: {self.species.tokens}\")\n",
    "        \n",
    "        for sent_i, sent in enumerate(self.sp_doc.sents):\n",
    "            # print()\n",
    "            # print()\n",
    "            # print(f\"Sentence I: {sent_i}\")\n",
    "            \n",
    "            # Current Points in Sentence\n",
    "            curr_points = [0] * NUM_CATEGORIES\n",
    "\n",
    "            # Sentence Local Info\n",
    "            sent_tokens = [token for token in sent]\n",
    "            # print(f\"Sentence Tokens: {sent_tokens}\")\n",
    "\n",
    "            data = {\n",
    "                \"sent_cause_tokens\": set(sent_tokens).intersection(self.cause.tokens),\n",
    "                \"sent_change_tokens\": set(sent_tokens).intersection(self.change.tokens),\n",
    "                \"sent_seen_species\": {},\n",
    "                \"sent_num_unique_species\": 0    \n",
    "            }\n",
    "            \n",
    "            # sent_cause_tokens = set(sent_tokens).intersection(self.cause.tokens)\n",
    "            # sent_change_tokens = set(sent_tokens).intersection(self.change.tokens)\n",
    "            # sent_seen_species = {}\n",
    "            # sent_num_unique_species = 0\n",
    "\n",
    "            species_instances.append([])\n",
    "            interaction_instances.append([])\n",
    "\n",
    "            if verbose and VERBOSE_LEVEL >= 2:\n",
    "                print(f\"\\tSentence: {sent}\")\n",
    "                print(f\"\\tSentence Cause Tokens: {data['sent_cause_tokens']}\")\n",
    "                print(f\"\\tSentence Change Tokens: {data['sent_change_tokens']}\")\n",
    "            \n",
    "            for token in sent_tokens:\n",
    "                # If each category has reached their maximum number of points,\n",
    "                # we can end the loop early.\n",
    "                all_maxed = True\n",
    "                for i in range(NUM_CATEGORIES):\n",
    "                    if i == TRAIT_VARIATION:\n",
    "                        continue\n",
    "                    if curr_points[i] < MPC[i]:\n",
    "                        all_maxed = False\n",
    "\n",
    "                if all_maxed:\n",
    "                    break\n",
    "\n",
    "                # if verbose and VERBOSE_LEVEL >= 3:\n",
    "                #     print(f\"\\t\\tToken in Sentence: {token}\")\n",
    "\n",
    "                # Update Token in Data\n",
    "                data[\"token\"] = token\n",
    "                \n",
    "                # Not Topic Points\n",
    "                if curr_points[NOT_TOPIC] < MPC[NOT_TOPIC]:\n",
    "                    if token in self.not_topic.tokens:\n",
    "                        curr_points[NOT_TOPIC] += PIC[NOT_TOPIC]\n",
    "\n",
    "                        if verbose and VERBOSE_LEVEL >= 3:\n",
    "                            print(f\"\\t\\t+ Points for Not-Topic\")\n",
    "\n",
    "                        \n",
    "                # Trait Points\n",
    "                if curr_points[TRAIT] < MPC[TRAIT]:\n",
    "                    if token in self.trait.tokens:\n",
    "                        scale = self.valid_trait_token(data, verbose=verbose)\n",
    "                        curr_points[TRAIT] += scale * PIC[TRAIT]\n",
    "\n",
    "                        if verbose and VERBOSE_LEVEL >= 3 and scale:\n",
    "                            print(f\"\\t\\t+ Points for Trait\")\n",
    "                \n",
    "                \n",
    "                # Not Experiment Points\n",
    "                if token in self.not_experiment.tokens:\n",
    "                    curr_points[EXPERIMENT] -= 2 * PIC[EXPERIMENT]\n",
    "                    banned[EXPERIMENT][sent_i] = 1\n",
    "                    \n",
    "                    if verbose and VERBOSE_LEVEL >= 3:\n",
    "                        print(f\"\\t\\t- Points for Experiment\")\n",
    "                \n",
    "                \n",
    "                # Experiment Points\n",
    "                elif curr_points[EXPERIMENT] < MPC[EXPERIMENT]:\n",
    "                    if token in self.experiment.tokens:\n",
    "                        scale = 1 if token.pos_ == \"VERB\" else 0.5\n",
    "                        curr_points[EXPERIMENT] += scale * PIC[EXPERIMENT]\n",
    "\n",
    "                        if verbose and VERBOSE_LEVEL >= 3:\n",
    "                            print(f\"\\t\\t+ Points for Experiment\")\n",
    "                \n",
    "                        \n",
    "                # Species and/or Interaction Points\n",
    "                if token in self.species.span_starts:\n",
    "                    # print()\n",
    "                    # print(f\"Species Token: {token}\")\n",
    "                    \n",
    "                    # Update Species\n",
    "                    update_data = data\n",
    "                    update_data[\"seen_species\"] = seen_species\n",
    "                    \n",
    "                    updated_data = self.update_seen_species(update_data, verbose=verbose)\n",
    "\n",
    "                    # Unpacking Updated Data\n",
    "                    # It's tiresome to repeatedly type updated_data[...].\n",
    "                    # Why did you use an object, then? Well!\n",
    "                    seen_species = updated_data[\"seen_species\"]\n",
    "                    seen_in_sent = updated_data[\"seen_in_sent\"]\n",
    "                    sent_seen_species = updated_data[\"sent_seen_species\"]\n",
    "                    sent_num_unique_species = updated_data[\"sent_num_unique_species\"]\n",
    "                    ref = updated_data[\"ref\"]\n",
    "                    \n",
    "                    data[\"sent_seen_species\"] = sent_seen_species\n",
    "                    data[\"sent_num_unique_species\"] = sent_num_unique_species\n",
    "                    \n",
    "                    # print(f\"Seen Species: {seen_species}\")\n",
    "                    # print(f\"Sent Seen Species: {sent_seen_species}\")\n",
    "                    # print(f\"Seen in Sent: {seen_in_sent}\")\n",
    "                    # print(f\"Sent Num Unique Species: {sent_num_unique_species}\")\n",
    "                    # print(f\"Ref: {ref}\")\n",
    "\n",
    "                    # print(updated_data)\n",
    "                    \n",
    "                    if not seen_in_sent and sent_num_unique_species == 1 and sent_seen_species[ref][\"visits\"] == 1:\n",
    "                        # print(ref.text)\n",
    "                        interaction_instances[-1].append(ref.text)\n",
    "\n",
    "                    if seen_in_sent:\n",
    "                        pass\n",
    "                        # print(f\"\\t\\tAlready Seen Species '{token}' in Sentence (No Interaction)\")\n",
    "                    else:\n",
    "                        # Interaction Points\n",
    "                        # It would make sense to add this fix here as well.\n",
    "                        # if sent_seen_species[ref][\"points\"] <= 0:\n",
    "                        # No it wouldn't\n",
    "                        if curr_points[INTERACTION] < MPC[INTERACTION]:\n",
    "                            if sent_num_unique_species == 2:\n",
    "                                \n",
    "                                curr_points[INTERACTION] = 2.0 * PIC[INTERACTION]\n",
    "                                # print(ref.text)\n",
    "                                # print(f\"\\tAdded\")\n",
    "\n",
    "                                interaction_instances[-1].append(ref.text)\n",
    "                                \n",
    "                                if verbose and VERBOSE_LEVEL >= 3:\n",
    "                                    print(f\"\\t\\t+ Points for Interaction\")\n",
    "                            \n",
    "                            elif sent_num_unique_species > 2:\n",
    "                                curr_points[INTERACTION] += PIC[INTERACTION]\n",
    "                                # print(ref.text)\n",
    "                                # print(f\"\\tAdded\")\n",
    "\n",
    "                                interaction_instances[-1].append(ref.text)\n",
    "    \n",
    "                                if verbose and VERBOSE_LEVEL >= 3:\n",
    "                                    print(f\"\\t\\t+ Points for Interaction\")\n",
    "\n",
    "                    # Species Points\n",
    "                    # I added this to fix something, but I can't remember\n",
    "                    # what it was supposed to fix. Shouldn't seen_in_sent handle this?\n",
    "                    # Did I forget to indent this chunk of code, and reinvent seen_in_sent?\n",
    "                    # Can't remember.\n",
    "                    # I remember now, the species may have been seen in the sentence, but if\n",
    "                    # it hasn't been awarded any points, it can still be used.\n",
    "                    if sent_seen_species[ref][\"points\"] <= 0:\n",
    "                        if curr_points[SPECIES] < MPC[SPECIES]:\n",
    "                            scale = self.valid_species_token(data)\n",
    "                            \n",
    "                            curr_points[SPECIES] += scale * PIC[SPECIES]\n",
    "                            sent_seen_species[ref][\"points\"] += scale * PIC[SPECIES]\n",
    "\n",
    "                            if scale:\n",
    "                                species_instances[-1].append(ref.text)\n",
    "                            \n",
    "                            if verbose and scale and VERBOSE_LEVEL >= 3:\n",
    "                                print(f\"\\t\\t+ Points for Species\")\n",
    "\n",
    "            \n",
    "            # Add Sentence Points to Total Points\n",
    "            for i in range(NUM_CATEGORIES):\n",
    "                # if i == INTERACTION and not data[\"sent_cause_tokens\"] and not data[\"sent_change_tokens\"]:\n",
    "                #     curr_points[INTERACTION] = 0\n",
    "                #     interaction_instances[-1] = []\n",
    "                \n",
    "                # if i == EXPERIMENT:\n",
    "                #     print(f\"Experiment Points for Sentence: {curr_points[i]}\")\n",
    "                \n",
    "                is_banned = banned[i][sent_i]\n",
    "                # if is_banned and i == EXPERIMENT:\n",
    "                #     print(f\"Sentence is Banned for Experiment\")\n",
    "                \n",
    "                if curr_points[i] <= 0 and not is_banned:\n",
    "                    num_zero_pt_sents[i] += 1\n",
    "                \n",
    "                if not is_banned:\n",
    "                    points[i] += max(0, min(curr_points[i], MPC[i]))\n",
    "\n",
    "        \n",
    "        # Trait Variation Points\n",
    "        points[TRAIT_VARIATION] = self.valid_trait_variation(verbose=False)\n",
    "        # print(f\"TRAIT VARIATION: {points[TRAIT_VARIATION]}\")\n",
    "        \n",
    "        points[TRAIT] = self.valid_trait(verbose=False)\n",
    "        \n",
    "        \n",
    "\n",
    "        # Bins\n",
    "        bins = []\n",
    "        for i in range(NUM_CATEGORIES):\n",
    "            bins.append([-math.inf, 0.33, 0.66, math.inf])\n",
    "        # bins[SPECIES] = [-math.inf, 0.33, 0.66, math.inf]\n",
    "        # bins[EXPERIMENT] = [-math.inf, 0.7, math.inf]\n",
    "        bins[TRAIT_VARIATION] = [-math.inf, 0.5, 1, math.inf]\n",
    "        # print(f\"Bins: {bins}\")\n",
    "        \n",
    "        # Calculating Score            \n",
    "        NUM_SENTENCES = len(list(self.sp_doc.sents))\n",
    "        score = 0\n",
    "        binned_score = 0\n",
    "\n",
    "        for i in range(NUM_CATEGORIES):\n",
    "            if i not in [TRAIT, TRAIT_VARIATION]:\n",
    "                num_non_zero_pt_sents = NUM_SENTENCES - num_zero_pt_sents[i]\n",
    "                \n",
    "                banned_tax = 0\n",
    "                for b in banned[i]:\n",
    "                    if b:\n",
    "                        banned_tax += 1\n",
    "                \n",
    "                # print(f\"Banned Tax: {banned_tax}\")\n",
    "                lenient_num_sentences = max(num_non_zero_pt_sents, (1 - LEN[i]) * (NUM_SENTENCES) + banned_tax)\n",
    "    \n",
    "                # Calculating FTP\n",
    "                points[i] = points[i] / (MPC[i] * lenient_num_sentences)\n",
    "    \n",
    "                # Take the Inverse for Not-Topic\n",
    "                if i == NOT_TOPIC:\n",
    "                    points[i] = 1 - points[i]\n",
    "\n",
    "            # Bin Points\n",
    "            for b in range(len(bins[i]) - 2, -1, -1):\n",
    "                # print(bins[i][b], \"<=\", points[i], \"<=\", bins[i][b+1])\n",
    "\n",
    "                if bins[i][b] <= points[i] <= bins[i][b+1]:\n",
    "                    binned_points[i] = b + 1\n",
    "                    # binned_points[i] = bins[i][b]\n",
    "                    break\n",
    "\n",
    "            points[i] = max(0, min(points[i], 1))\n",
    "            binned_points[i] = max(0, min(binned_points[i], math.inf))\n",
    "            \n",
    "            score += max(0, min(points[i], 1)) * CW[i]\n",
    "            # Redundant, but we move\n",
    "            binned_score += max(0, min(binned_points[i], math.inf)) * CW[i]\n",
    "\n",
    "        # Enforcing 3 or More Species            \n",
    "        # if len(seen_species) < 3:\n",
    "        #     return 0, 0\n",
    "\n",
    "        # Removing, it's now on a scale from 0.0 to 3.0.\n",
    "        # I could just change the values, but eeeh.\n",
    "        # assert 0.0 <= score <= 1.0\n",
    "        # assert 0.0 <= binned_score <= 1.0\n",
    "\n",
    "        if verbose and VERBOSE_LEVEL >= 1:\n",
    "            print(f\"Score, Points: {score}, {points}\")\n",
    "            print(f\"Binned Score, Binned Points: {binned_score}, {binned_points}\")\n",
    "    \n",
    "        return {\n",
    "            \"s_score\": 0 if len(seen_species) < 3 else score,\n",
    "            \"s_bin_score\": 0 if len(seen_species) < 3 else binned_score,\n",
    "            \"score\": score,\n",
    "            \"bin_score\": binned_score,\n",
    "            \"points\": {\n",
    "                \"trait\": points[TRAIT],\n",
    "                \"species\": points[SPECIES],\n",
    "                \"experiment\": points[EXPERIMENT],\n",
    "                \"interaction\": points[INTERACTION],\n",
    "                \"not_topic\": points[NOT_TOPIC],\n",
    "                \"trait_variation\": points[TRAIT_VARIATION],\n",
    "            },\n",
    "            \"binned_points\": {\n",
    "                \"trait\": binned_points[TRAIT],\n",
    "                \"species\": binned_points[SPECIES],\n",
    "                \"experiment\": binned_points[EXPERIMENT],\n",
    "                \"interaction\": binned_points[INTERACTION],\n",
    "                \"not_topic\": binned_points[NOT_TOPIC],\n",
    "                \"trait_variation\": binned_points[TRAIT_VARIATION],\n",
    "            },\n",
    "            \"interaction_instances\": interaction_instances,\n",
    "            \"species_instances\": species_instances,\n",
    "            \"causes\": self.cause.tokens,\n",
    "            \"changes\": self.change.tokens,\n",
    "            \"traits\": self.trait.tokens,\n",
    "            \"species\": self.species.tokens,\n",
    "            \"experiments\": self.experiment.tokens,\n",
    "            \"not_experiments\": self.not_experiment.tokens,\n",
    "            \"not_topics\": self.not_topic.tokens,\n",
    "            \"variability\": self.variability.tokens,\n",
    "            \"tests\": self.test.tokens,\n",
    "            \"seen_species\": seen_species\n",
    "        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a7a9f282-6fea-4463-a9c9-023403a8978e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def store_scored_dataset(dataset, name, version=''):\n",
    "    filename = f\"{name}{'' if not version else f'-{version}'}\"\n",
    "    dataset.to_csv(f\"{filename}.csv\", index=False)\n",
    "    dataset.to_excel(f\"{filename}.xlsx\", index=None, header=True)\n",
    "\n",
    "def score_dataset(name=None, save_output=False, version=\"\", df=None, prefix=\"\"):\n",
    "    # Redirect Print Statements\n",
    "    # https://stackoverflow.com/questions/7152762/how-to-redirect-print-output-to-a-file\n",
    "    if save_output:\n",
    "        initial_stdout = sys.stdout\n",
    "        f = open(f'./Print{name}{\"\" if not version else f\"-{version}\"}.txt', 'w')\n",
    "        sys.stdout = f\n",
    "        sys.stdout.reconfigure(encoding='utf-8')\n",
    "\n",
    "    # Load Dataset\n",
    "    if type(df) == type(None):\n",
    "        data = load_preprocessed_dataset(name)\n",
    "    else:\n",
    "        data = df\n",
    "\n",
    "    # We'll be running the points algorithm\n",
    "    # on the abstracts of these papers.\n",
    "    texts = list(data['Abstract'].to_numpy())\n",
    "    \n",
    "    # The scores for each paper will be stored here,\n",
    "    # we'll set this as a column of the dataframe.\n",
    "    # For testing, I'm storing a lot of information.\n",
    "    # It looks untidy, for now.\n",
    "\n",
    "    # REGULAR POINTS\n",
    "    s_scores = []\n",
    "    scores = []\n",
    "    points_trait = []\n",
    "    points_species = []\n",
    "    points_experiment = []\n",
    "    points_interaction = []\n",
    "    points_not_topic = []\n",
    "    points_trait_variation = []\n",
    "\n",
    "    # BINNED POINTS\n",
    "    s_binned_scores = []\n",
    "    binned_scores = []\n",
    "    binned_points_trait = []\n",
    "    binned_points_species = []\n",
    "    binned_points_experiment = []\n",
    "    binned_points_interaction = []\n",
    "    binned_points_not_topic = []\n",
    "    binned_points_trait_variation = []\n",
    "\n",
    "    # OTHER INFORMATION\n",
    "    causes = []\n",
    "    changes = []\n",
    "    traits = []\n",
    "    species = []\n",
    "    seen_species = []\n",
    "    tests = []\n",
    "    experiments = []\n",
    "    not_experiments = []\n",
    "    not_topics = []\n",
    "    variability = []\n",
    "    interaction_instances = []\n",
    "    species_instances = []\n",
    "    \n",
    "    # Scan and Evaluate Documents\n",
    "    main = Main()\n",
    "    for i, doc in enumerate(main.sp_nlp.pipe(texts)):\n",
    "        print(f\"{i+1}/{data.shape[0]} - {data.iloc[i]['Title']}\\n\")\n",
    "        main.update_doc(doc, verbose=save_output)\n",
    "\n",
    "        # Empty string literals cause errors, so it's\n",
    "        # being handled here.\n",
    "        if not main.sp_doc or not main.species.tn_doc:\n",
    "            points.append(0)\n",
    "            binned_points.append(0)\n",
    "        else:\n",
    "            output = main.score(verbose=save_output)\n",
    "\n",
    "            s_scores.append(output[\"s_score\"])\n",
    "            scores.append(output[\"score\"])\n",
    "            points_trait.append(output[\"points\"][\"trait\"])\n",
    "            points_species.append(output[\"points\"][\"species\"])\n",
    "            points_experiment.append(output[\"points\"][\"experiment\"])\n",
    "            points_interaction.append(output[\"points\"][\"interaction\"])\n",
    "            points_not_topic.append(output[\"points\"][\"not_topic\"])\n",
    "            points_trait_variation.append(output[\"points\"][\"trait_variation\"])\n",
    "\n",
    "            s_binned_scores.append(output[\"s_bin_score\"])\n",
    "            binned_scores.append(output[\"bin_score\"])\n",
    "            binned_points_trait.append(output[\"binned_points\"][\"trait\"])\n",
    "            binned_points_species.append(output[\"binned_points\"][\"species\"])\n",
    "            binned_points_experiment.append(output[\"binned_points\"][\"experiment\"])\n",
    "            binned_points_interaction.append(output[\"binned_points\"][\"interaction\"])\n",
    "            binned_points_not_topic.append(output[\"binned_points\"][\"not_topic\"])\n",
    "            binned_points_trait_variation.append(output[\"binned_points\"][\"trait_variation\"])\n",
    "            \n",
    "            causes.append(output[\"causes\"])\n",
    "            changes.append(output[\"changes\"])\n",
    "            traits.append(output[\"traits\"])\n",
    "            species.append(output[\"species\"])\n",
    "            seen_species.append(output[\"seen_species\"])\n",
    "            tests.append(output[\"tests\"])\n",
    "            experiments.append(output[\"experiments\"])\n",
    "            not_experiments.append(output[\"not_experiments\"])\n",
    "            not_topics.append(output[\"not_topics\"])\n",
    "            variability.append(output[\"variability\"])\n",
    "            species_instances.append(output[\"species_instances\"])\n",
    "            interaction_instances.append(output[\"interaction_instances\"])\n",
    "\n",
    "        # if not save_output:\n",
    "        #     clear_output(wait=True)\n",
    "\n",
    "    # Reset Standard Output\n",
    "    if save_output:\n",
    "        sys.stdout = initial_stdout\n",
    "        f.close()\n",
    "\n",
    "    # Trait DF\n",
    "    # trait_df = data.copy()\n",
    "    # trait_df[\"Points\"] = points_trait\n",
    "    # trait_df[\"Binned Points\"] = binned_points_trait\n",
    "    # trait_df[\"Traits\"] = [json.dumps([trait.lower_ for trait in _traits]) for _traits in traits]\n",
    "    # trait_df[\"Causes\"] = [json.dumps([_.lower_ for _ in __]) for __ in causes]\n",
    "    # trait_df[\"Changes\"] = [json.dumps([_.lower_ for _ in __]) for __ in changes]\n",
    "    # trait_df.sort_values(by=\"Points\", ascending=False, inplace=True)\n",
    "    # store_scored_dataset(trait_df, f\"{prefix}Trait\", version=version)\n",
    "    \n",
    "    # # Species DF\n",
    "    # species_df = data.copy()\n",
    "    # species_df[\"Points\"] = points_species\n",
    "    # species_df[\"Binned Points\"] = binned_points_species\n",
    "    # species_df[\"Species\"] = [json.dumps([_.lower_ for _ in _species]) for _species in species]\n",
    "    # species_df[\"Seen Species\"] = [json.dumps([(k.text, v) for k, v in _species.items()]) for _species in seen_species]\n",
    "    # species_df.sort_values(by=\"Points\", ascending=False, inplace=True)\n",
    "    # store_scored_dataset(species_df, \"Examples Species\", version=version)\n",
    "    \n",
    "    # # Interaction DF\n",
    "    # interaction_df = data.copy()\n",
    "    # interaction_df[\"Points\"] = points_interaction\n",
    "    # interaction_df[\"Binned Points\"] = binned_points_interaction\n",
    "    # interaction_df[\"Instances\"] = [json.dumps(instance) for instance in interaction_instances]\n",
    "    # interaction_df[\"Causes\"] = [json.dumps([_.lower_ for _ in __]) for __ in causes]\n",
    "    # interaction_df[\"Changes\"] = [json.dumps([_.lower_ for _ in __]) for __ in changes]\n",
    "    # interaction_df.sort_values(by=\"Points\", ascending=False, inplace=True)\n",
    "    # store_scored_dataset(interaction_df, \"Examples Interaction\", version=version)\n",
    "\n",
    "    # # Experiment DF\n",
    "    # experiment_df = data.copy()\n",
    "    # experiment_df[\"Points\"] = points_experiment\n",
    "    # experiment_df[\"Binned Points\"] = binned_points_experiment\n",
    "    # experiment_df[\"Experiments\"] = [json.dumps([_.lower_ for _ in __]) for __ in experiments]\n",
    "    # experiment_df[\"Not Experiments\"] = [json.dumps([_.lower_ for _ in __]) for __ in not_experiments]\n",
    "    # experiment_df.sort_values(by=\"Points\", ascending=False, inplace=True)\n",
    "    # store_scored_dataset(experiment_df, \"Examples Experiment\", version=version)\n",
    "\n",
    "    # # Trait Variation DF\n",
    "    # trait_variation_df = data.copy()\n",
    "    # trait_variation_df[\"Points\"] = points_trait_variation\n",
    "    # trait_variation_df[\"Binned Points\"] = binned_points_trait_variation\n",
    "    # trait_variation_df.sort_values(by=\"Points\", ascending=False, inplace=True)\n",
    "    # store_scored_dataset(trait_variation_df, f\"{prefix}Trait Variation\", version=version)\n",
    "\n",
    "    # Regular DF\n",
    "    reg_data = data.copy()\n",
    "    reg_data[\"Real Score\"] = s_scores\n",
    "    reg_data[\"Score\"] = scores\n",
    "    reg_data[\"Trait Points\"] = points_trait\n",
    "    reg_data[\"Species Points\"] = points_species\n",
    "    reg_data[\"Experiment Points\"] = points_experiment\n",
    "    reg_data[\"Interaction Points\"] = points_interaction\n",
    "    reg_data[\"Topic Points\"] = points_not_topic\n",
    "    reg_data[\"Trait Variation Points\"] = points_trait_variation\n",
    "    reg_data.sort_values(by=\"Score\", ascending=False, inplace=True)\n",
    "    store_scored_dataset(reg_data, f\"{prefix}Aggregate\", version=version)\n",
    "\n",
    "    # Binned DF\n",
    "    bin_data = data.copy()\n",
    "    bin_data[\"Real Score\"] = s_binned_scores\n",
    "    bin_data[\"Score\"] = binned_scores\n",
    "    bin_data[\"Trait Points\"] = binned_points_trait\n",
    "    bin_data[\"Species Points\"] = binned_points_species\n",
    "    bin_data[\"Experiment Points\"] = binned_points_experiment\n",
    "    bin_data[\"Interaction Points\"] = binned_points_interaction\n",
    "    bin_data[\"Topic Points\"] = binned_points_not_topic\n",
    "    bin_data[\"Trait Variation Points\"] = binned_points_trait_variation\n",
    "    bin_data.sort_values(by=\"Score\", ascending=False, inplace=True)\n",
    "    store_scored_dataset(bin_data, f\"{prefix}Aggregate Binned\", version=version)\n",
    "    \n",
    "    return data\n",
    "\n",
    "def score_dataset_cont(fp=None, version=\"\", df=None, name=\"\"):\n",
    "    # Load Dataset\n",
    "    if df is None:\n",
    "        data = load_preprocessed_dataset(fp)\n",
    "    else:\n",
    "        data = df\n",
    "\n",
    "    columns = [\n",
    "        (\"Adjusted Score\", float),\n",
    "        (\"Adjusted Binned Score\", float),\n",
    "        (\"Score\", float),\n",
    "        (\"Binned Score\", float),\n",
    "        (\"Trait Points\", float),\n",
    "        (\"Binned Trait Points\", float),\n",
    "        (\"Species Points\", float),\n",
    "        (\"Binned Species Points\", float),\n",
    "        (\"Experiment Points\", float),\n",
    "        (\"Binned Experiment Points\", float),\n",
    "        (\"Interaction Points\", float),\n",
    "        (\"Binned Interaction Points\", float),\n",
    "        (\"Not Topic Points\", float),\n",
    "        (\"Binned Not Topic Points\", float), \n",
    "        (\"Trait Variation Points\", float),\n",
    "        (\"Binned Trait Variation Points\", float),\n",
    "        (\"Cause Keywords\", str),\n",
    "        (\"Change Keywords\", str),\n",
    "        (\"Trait Keywords\", str),\n",
    "        (\"Species\", str),\n",
    "        (\"Seen Species\", str),\n",
    "        (\"Test Keywords\", str),\n",
    "        (\"Experiment Keywords\", str),\n",
    "        (\"Not Experiment Keywords\", str),\n",
    "        (\"Not Topic Keywords\", str),\n",
    "        (\"Variability Keywords\", str),\n",
    "        (\"Interaction Instances\", str),\n",
    "        (\"Species Instances\", str),\n",
    "    ]\n",
    "\n",
    "    for col in columns:\n",
    "        if col[0] not in data.columns:\n",
    "            def_val = -1 if col[1] == float else \"\"\n",
    "            data[col[0]] = def_val\n",
    "            data[col[0]].astype(col[1])\n",
    "\n",
    "    # We'll be running the points algorithm\n",
    "    # on the abstracts of these papers.\n",
    "    texts = list(data['Abstract'].to_numpy())\n",
    "    \n",
    "    # Scan and Evaluate Documents\n",
    "    main = Main()\n",
    "    print(f\"Main Initialized\")\n",
    "    \n",
    "    for i, doc in enumerate(main.sp_nlp.pipe(texts)):\n",
    "        # I'm assuming that the lists don't change order,\n",
    "        # if it does, then this will be wrong.\n",
    "        assert data.loc[i, \"Abstract\"] == texts[i]\n",
    "        print(f\"{i+1}/{data.shape[0]} - {data.iloc[i]['Title']}\")\n",
    "\n",
    "        # If it's already scored, we will not waste time scoring\n",
    "        # it again.\n",
    "        scored = data.loc[i, \"Score\"] != -1\n",
    "        if scored:\n",
    "            print(f\"\\tAlready Scored, Skipped\")\n",
    "            continue\n",
    "        \n",
    "        main.update_doc(doc, verbose=False)\n",
    "        if main.sp_doc and main.species.tn_doc:\n",
    "            output = main.score(verbose=False)\n",
    "            data.loc[i, \"Adjusted Score\"] = output[\"s_score\"]\n",
    "            data.loc[i, \"Adjusted Binned Score\"] = output[\"s_bin_score\"]\n",
    "            data.loc[i, \"Score\"] = output[\"score\"]\n",
    "            data.loc[i, \"Binned Score\"] = output[\"bin_score\"]\n",
    "            data.loc[i, \"Trait Points\"] = output[\"points\"][\"trait\"]\n",
    "            data.loc[i, \"Binned Trait Points\"] = output[\"binned_points\"][\"trait\"]\n",
    "            data.loc[i, \"Species Points\"] = output[\"points\"][\"species\"]\n",
    "            data.loc[i, \"Binned Species Points\"] = output[\"binned_points\"][\"species\"]\n",
    "            data.loc[i, \"Experiment Points\"] = output[\"points\"][\"experiment\"]\n",
    "            data.loc[i, \"Binned Experiment Points\"] = output[\"binned_points\"][\"experiment\"]\n",
    "            data.loc[i, \"Interaction Points\"] = output[\"points\"][\"interaction\"]\n",
    "            data.loc[i, \"Binned Interaction Points\"] = output[\"binned_points\"][\"interaction\"]\n",
    "            data.loc[i, \"Not Topic Points\"] = output[\"points\"][\"not_topic\"]\n",
    "            data.loc[i, \"Binned Not Topic Points\"] = output[\"binned_points\"][\"not_topic\"]\n",
    "            data.loc[i, \"Trait Variation Points\"] = output[\"points\"][\"trait_variation\"]\n",
    "            data.loc[i, \"Binned Trait Variation Points\"] = output[\"binned_points\"][\"trait_variation\"]\n",
    "            data.loc[i, \"Cause Keywords\"] = json.dumps([_.lower_ for _ in output[\"causes\"]])\n",
    "            data.loc[i, \"Change Keywords\"] = json.dumps([_.lower_ for _ in output[\"changes\"]])\n",
    "            data.loc[i, \"Trait Keywords\"] = json.dumps([_.lower_ for _ in output[\"traits\"]])\n",
    "            data.loc[i, \"Species\"] = json.dumps([_.text for _ in output[\"species\"]])\n",
    "            data.loc[i, \"Seen Species\"] = json.dumps([_.text for _ in output[\"seen_species\"]])\n",
    "            data.loc[i, \"Test Keywords\"] = json.dumps([_.lower_ for _ in output[\"tests\"]])\n",
    "            data.loc[i, \"Experiment Keywords\"] = json.dumps([_.lower_ for _ in output[\"experiments\"]])\n",
    "            data.loc[i, \"Not Experiment Keywords\"] = json.dumps([_.lower_ for _ in output[\"not_experiments\"]])\n",
    "            data.loc[i, \"Not Topic Keywords\"] = json.dumps([_.lower_ for _ in output[\"not_topics\"]])\n",
    "            data.loc[i, \"Variability Keywords\"] = json.dumps([_.text for _ in output[\"variability\"]])\n",
    "            data.loc[i, \"Interaction Instances\"] = json.dumps([json.dumps(instance) for instance in output[\"interaction_instances\"]])\n",
    "            data.loc[i, \"Species Instances\"] = json.dumps([json.dumps(instance) for instance in output[\"species_instances\"]])\n",
    "\n",
    "        # Auto Save\n",
    "        if i % 10 == 0:\n",
    "            store_scored_dataset(data, f\"{name}Aggregate\", version=version)\n",
    "\n",
    "        store_scored_dataset(data, f\"{name}Aggregate\", version=version)\n",
    "    \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "fe8b1c6e-8352-4490-a1b7-85a843922b38",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\lbeln\\anaconda3\\envs\\3.10\\lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Main Initialized\n",
      "1/1983 - Trophic cascades: the primacy of trait‐mediated indirect interactions\n",
      "\tAlready Scored, Skipped\n",
      "2/1983 - RESOURCE COMPETITION MODIFIES THE STRENGTH OF TRAIT-MEDIATED PREDATOR–PREY INTERACTIONS: A META-ANALYSIS\n",
      "\tAlready Scored, Skipped\n",
      "3/1983 - TRAIT-MEDIATED INDIRECT EFFECTS IN LARVAL ANURANS: REVERSING COMPETITION WITH THE THREAT OF PREDATION\n",
      "\tAlready Scored, Skipped\n",
      "4/1983 - Cascading trait-mediation: disruption of a trait-mediated mutualism by parasite-induced behavioral modification\n",
      "\tAlready Scored, Skipped\n",
      "5/1983 - Trophic promiscuity, intraguild predation and the problem of omnivores\n",
      "\tAlready Scored, Skipped\n",
      "6/1983 - Leaf pubescence mediates intraguild predation between predatory mites\n",
      "\tAlready Scored, Skipped\n",
      "7/1983 - Trait‐mediated effects between predators: ant chemical cues induce spider dispersal\n",
      "\tAlready Scored, Skipped\n",
      "8/1983 - Herbivore density mediates the indirect effect of herbivores on plants via induced resistance and apparent competition\n",
      "\tAlready Scored, Skipped\n",
      "9/1983 - Density‐ and trait‐mediated effects of a parasite and a predator in a tri‐trophic food web\n",
      "\tAlready Scored, Skipped\n",
      "10/1983 - Trait and density mediated indirect interactions in simple food webs\n",
      "\tAlready Scored, Skipped\n",
      "11/1983 - Trait-Mediated Indirect Effects of Phorid Flies on Ants\n",
      "\tAlready Scored, Skipped\n",
      "12/1983 - Moving beyond linear food chains: trait-mediated indirect interactions in a rocky intertidal food web\n",
      "\tAlready Scored, Skipped\n",
      "13/1983 - Trait-Mediated Effects of Parasitic Phorid Flies (Diptera: Phoridae) on Ant (Hymenoptera: Formicidae) Competition and Resource Access in Coffee Agro-ecosystems\n",
      "\tAlready Scored, Skipped\n",
      "14/1983 - Using plant functional traits to restore Hawaiian rainforest\n",
      "\tAlready Scored, Skipped\n",
      "15/1983 - Controlling species densities in structurally perturbed intransitive cycles with higher-order interactions\n",
      "\tAlready Scored, Skipped\n",
      "16/1983 - Trait-Mediated Effects of Parasitic Phorid Flies (Diptera: Phoridae) on Ant (Hymenoptera: Formicidae) Competition and Resource Access in Coffee Agro-ecosystems\n",
      "\tAlready Scored, Skipped\n",
      "17/1983 - Herkogamy, a Principal Functional Trait of Plant Reproductive Biology\n",
      "\tAlready Scored, Skipped\n",
      "18/1983 - Trait-Mediated Indirect Effects in Larval Anurans: Reversing Competition with the Threat of Predation\n",
      "\tAlready Scored, Skipped\n",
      "19/1983 - Contrasting impacts of competition on ecological and social trait evolution in songbirds\n",
      "\tAlready Scored, Skipped\n",
      "20/1983 - Mistletoe generates non‐trophic and trait‐mediated indirect interactions through a shared host of herbivore consumers\n",
      "\tAlready Scored, Skipped\n",
      "21/1983 - From apparent competition to facilitation: Impacts of consumer niche construction on the coexistence and stability of consumer‐resource communities\n",
      "\tAlready Scored, Skipped\n",
      "22/1983 - Trait-mediated competition drives an ant invasion and alters functional diversity\n",
      "\tAlready Scored, Skipped\n",
      "23/1983 - Leaf pubescence mediates intraguild predation between predatory mites\n",
      "\tAlready Scored, Skipped\n",
      "24/1983 - ECOLOGICAL MECHANISMS OF A RESOURCE POLYMORPHISM IN ZENAIDA DOVES OF BARBADOS\n",
      "\tAlready Scored, Skipped\n",
      "25/1983 - SCARED TO DEATH? THE EFFECTS OF INTIMIDATION AND CONSUMPTION IN PREDATOR–PREY INTERACTIONS\n",
      "\tAlready Scored, Skipped\n",
      "26/1983 - Species-Specific Traits Rather Than Resource Partitioning Mediate Diversity Effects on Resource Use\n",
      "\tAlready Scored, Skipped\n",
      "27/1983 - EVOLUTIONARILY INDUCED ALTERNATIVE STATES AND COEXISTENCE IN SYSTEMS WITH APPARENT COMPETITION\n",
      "\tAlready Scored, Skipped\n",
      "28/1983 - A multifaceted trophic cascade in a detritus‐based system: density‐, trait‐, or processing‐chain‐mediated effects?\n",
      "\tAlready Scored, Skipped\n",
      "29/1983 - Symbiont-conferred immunity interacts with effects of parasitoid genotype and intraguild predation to affect aphid immunity in a clone-specific fashion\n",
      "\tAlready Scored, Skipped\n",
      "30/1983 - A combination of developmental plasticity, parental effects, and genetic differentiation mediates divergences in life history traits between dung beetle populations\n",
      "\tAlready Scored, Skipped\n",
      "31/1983 - Trait variability differs between leaf and wood tissues across ecological scales in subtropical forests\n",
      "\tAlready Scored, Skipped\n",
      "32/1983 - Architectural and growth traits differ in effects on performance of clonal plants: an analysis using a field‐parameterized simulation model\n",
      "\tAlready Scored, Skipped\n",
      "33/1983 - Architectural and growth traits differ in effects on performance of clonal plants: an analysis using a field-parameterized simulation model\n",
      "\tAlready Scored, Skipped\n",
      "34/1983 - Ventral colour polymorphism correlates with alternative behavioural patterns in female common lizards (<i>Lacerta vivipara</i>)\n",
      "\tAlready Scored, Skipped\n",
      "35/1983 - Fine‐root and mycorrhizal traits help explain ecosystem processes and responses to global change\n",
      "\tAlready Scored, Skipped\n",
      "36/1983 - Competition drives trait evolution and character displacement between <i>Mimulus</i> species along an environmental gradient\n",
      "\tAlready Scored, Skipped\n",
      "37/1983 - The sign of cascading predator effects varies with prey traits in a detrital system\n",
      "\tAlready Scored, Skipped\n",
      "38/1983 - Can Functional Traits Explain Plant Coexistence? A Case Study with Tropical Lianas and Trees\n",
      "\tAlready Scored, Skipped\n",
      "39/1983 - Parasites that change predator or prey behaviour can have keystone effects on community composition\n",
      "\tAlready Scored, Skipped\n",
      "40/1983 - Direct litter interference and indirect soil competitive effects of two contrasting phenotypes of a spiny legume shrub drive the forb composition of an oromediterranean community\n",
      "\tAlready Scored, Skipped\n",
      "41/1983 - FORAGING TRADE-OFFS ALONG A PREDATOR–PERMANENCE GRADIENT IN SUBALPINE WETLANDS\n",
      "\tAlready Scored, Skipped\n",
      "42/1983 - Testing functional trait-based mechanisms underpinning plant responses to grazing and linkages to ecosystem functioning in grasslands\n",
      "\tAlready Scored, Skipped\n",
      "43/1983 - Diet quality and conspecific larval density predict functional trait variation and performance in a polyphagous frugivorous fly\n",
      "\tAlready Scored, Skipped\n",
      "44/1983 - Plant trait - environmental linkages among contrasting landscapes and climate regimes in temperate eucalypt woodlands\n",
      "\tAlready Scored, Skipped\n",
      "45/1983 - Trait values and not invasive status determine competitive outcomes between native and invasive species under varying soil nutrient availability\n",
      "\tAlready Scored, Skipped\n",
      "46/1983 - The community effects of phenotypic and genetic variation within a predator population\n",
      "\tAlready Scored, Skipped\n",
      "47/1983 - Trophic consequences of terrestrial eutrophication for a threatened ungulate\n",
      "\tAlready Scored, Skipped\n",
      "48/1983 - Biotic interactions shape trait assembly of marine communities across time and latitude\n",
      "\tAlready Scored, Skipped\n",
      "49/1983 - REVISITING THE CLASSICS: CONSIDERING NONCONSUMPTIVE EFFECTS IN TEXTBOOK EXAMPLES OF PREDATOR–PREY INTERACTIONS\n",
      "\tAlready Scored, Skipped\n",
      "50/1983 - Pelagic–benthic resource polymorphism in <i>Schizopygopsis thermalis</i> Herzenstein 1891 (Pisces, Cyprinidae) in a headwater lake in the Salween River system on the Tibetan Plateau\n",
      "\tAlready Scored, Skipped\n",
      "51/1983 - Character Shifts of Prey Species That Share Predators\n",
      "\tAlready Scored, Skipped\n",
      "52/1983 - Decline in top predator body size and changing climate alter trophic structure in an oceanic ecosystem\n",
      "\tAlready Scored, Skipped\n",
      "53/1983 - HOW FLUCTUATING COMPETITION AND PHENOTYPIC PLASTICITY MEDIATE SPECIES DIVERGENCE\n",
      "\tAlready Scored, Skipped\n",
      "54/1983 - Functional traits shape plant–plant interactions and recruitment in a hotspot of woody plant diversity\n",
      "\tAlready Scored, Skipped\n",
      "55/1983 - RESOURCE IDENTITY MODIFIES THE INFLUENCE OF PREDATION RISK ON ECOSYSTEM FUNCTION\n",
      "\tAlready Scored, Skipped\n",
      "56/1983 - Maternal antibodies in a wild altricial bird: effects on offspring immunity, growth and survival\n",
      "\tAlready Scored, Skipped\n",
      "57/1983 - Evidence for an ecological cost of enhanced herbicide metabolism in<i>Lolium rigidum</i>\n",
      "\tAlready Scored, Skipped\n",
      "58/1983 - Food resources and competition rather than eco-geographic rules explain trait variations in two contrasting rat species: Implications for future climate change\n",
      "\tAlready Scored, Skipped\n",
      "59/1983 - Response of the residential piscivorous fish community to introduction of a new predator type in a mesotrophic lake\n",
      "\tAlready Scored, Skipped\n",
      "60/1983 - Prey state shapes the effects of temporal variation in predation risk\n",
      "\tAlready Scored, Skipped\n",
      "61/1983 - Neutral theory in community ecology and the hypothesis of functional equivalence\n",
      "\tAlready Scored, Skipped\n",
      "62/1983 - Associational decomposition: After‐life traits and interactions among decomposing litters control during‐life aggregation of plant species\n",
      "\tAlready Scored, Skipped\n",
      "63/1983 - Non‐consumptive effects of predatory mites on thrips and its host plant\n",
      "\tAlready Scored, Skipped\n",
      "64/1983 - Intraspecific competition between co‐infecting parasite strains enhances host survival in African trypanosomes\n",
      "\tAlready Scored, Skipped\n",
      "65/1983 - Extending trait dispersion across trophic levels: Predator assemblages act as top‐down filters on prey communities\n",
      "\tAlready Scored, Skipped\n",
      "66/1983 - Diet and hormonal manipulation reveal cryptic genetic variation: implications for the evolution of novel feeding strategies\n",
      "\tAlready Scored, Skipped\n",
      "67/1983 - The shape of the predator biomass distribution affects biological pest control services in agricultural landscapes\n",
      "\tAlready Scored, Skipped\n",
      "68/1983 - Modelling coevolution in ecological networks with adaptive dynamics\n",
      "\tAlready Scored, Skipped\n",
      "69/1983 - Allometric scaling of indirect effects: body size ratios predict non‐consumptive effects in multi‐predator systems\n",
      "\tAlready Scored, Skipped\n",
      "70/1983 - Bioenergetic theory predicts infection dynamics of human schistosomes in intermediate host snails across ecological gradients\n",
      "\tAlready Scored, Skipped\n",
      "71/1983 - Leaf trait variation in grassland plant species in response to soil phosphorus\n",
      "\tAlready Scored, Skipped\n",
      "72/1983 - Herbivore release through cascading risk effects\n",
      "\tAlready Scored, Skipped\n",
      "73/1983 - Herbivores as drivers of negative density dependence in tropical forest saplings\n",
      "\tAlready Scored, Skipped\n",
      "74/1983 - CONDITION-SPECIFIC COMPETITION: IMPLICATIONS FOR THE ALTITUDINAL DISTRIBUTION OF STREAM FISHES\n",
      "\tAlready Scored, Skipped\n",
      "75/1983 - Competition‐induced transgenerational plasticity influences competitive interactions and leaf decomposition of offspring\n",
      "\tAlready Scored, Skipped\n",
      "76/1983 - Coexistence of coinvading species with mutualism and competition\n",
      "\tAlready Scored, Skipped\n",
      "77/1983 - Patch Dynamics and Islands of Fertility in a High Mountain Mediterranean Community\n",
      "\tAlready Scored, Skipped\n",
      "78/1983 - Indirect interactions in the microbial world: specificities and similarities to plant–insect systems\n",
      "\tAlready Scored, Skipped\n",
      "79/1983 - Evolution as a Coexistence Mechanism: Does Genetic Architecture Matter?\n",
      "\tAlready Scored, Skipped\n",
      "80/1983 - Positive effect of shade on plant growth: amelioration of stress or active regulation of growth rate?\n",
      "\tAlready Scored, Skipped\n",
      "81/1983 - Adaptive foragers and community ecology: linking individuals to communities and ecosystems\n",
      "\tAlready Scored, Skipped\n",
      "82/1983 - Predicting coexistence in species with continuous ontogenetic niche shifts and competitive asymmetry\n",
      "\tAlready Scored, Skipped\n",
      "83/1983 - Chemical similarity and local community assembly in the species rich tropical genus <i>Piper</i>\n",
      "\tAlready Scored, Skipped\n",
      "84/1983 - TRITROPHIC EFFECTS OF BIRDS AND ANTS ON A CANOPY FOOD WEB, TREE GROWTH, AND PHYTOCHEMISTRY\n",
      "\tAlready Scored, Skipped\n",
      "85/1983 - Water‐use efficiency and relative growth rate mediate competitive interactions in Sonoran Desert winter annual plants\n",
      "\tAlready Scored, Skipped\n",
      "86/1983 - Evolution of virulence: triggering host inflammation allows invading pathogens to exclude competitors\n",
      "\tAlready Scored, Skipped\n",
      "87/1983 - GEOGRAPHIC VARIATION IN RESOURCE DOMINANCE–DISCOVERY IN BRAZILIAN ANT COMMUNITIES\n",
      "\tAlready Scored, Skipped\n",
      "88/1983 - The negative effects of pathogen‐infected prey on predators: a meta‐analysis\n",
      "\tAlready Scored, Skipped\n",
      "89/1983 - EVALUATING THE TARGETS OF SELECTION DURING CHARACTER DISPLACEMENT\n",
      "\tAlready Scored, Skipped\n",
      "90/1983 - Inter‐strain competition and dispersal in aphids: evidence from a greenhouse study\n",
      "\tAlready Scored, Skipped\n",
      "91/1983 - Cascading effects of a top predator on intraspecific competition at intermediate and basal trophic levels\n",
      "\tAlready Scored, Skipped\n",
      "92/1983 - Exploring multitrophic interactions in oilseed rape fields reveals the prevailing role of Carabidae\n",
      "\tAlready Scored, Skipped\n",
      "93/1983 - Non‐consumptive effects of a top‐predator decrease the strength of the trophic cascade in a four‐level terrestrial food web\n",
      "\tAlready Scored, Skipped\n",
      "94/1983 - ASYMMETRIC COMPETITION VIA INDUCED RESISTANCE: SPECIALIST HERBIVORES INDIRECTLY SUPPRESS GENERALIST PREFERENCE AND POPULATIONS\n",
      "\tAlready Scored, Skipped\n",
      "95/1983 - Trophic cascades in streams: effects of nutrient enrichment on autotrophic and consumer benthic communities under two different fish predation regimes\n",
      "\tAlready Scored, Skipped\n",
      "96/1983 - ASSORTATIVE MATE CHOICE AND DOMINANCE MODIFICATION: ALTERNATIVE WAYS OF REMOVING HETEROZYGOTE DISADVANTAGE\n",
      "\tAlready Scored, Skipped\n",
      "97/1983 - Genetic and morphological divergence along the littoral-pelagic axis in two common and sympatric fishes: perch,<i>Perca fluviatilis</i>(Percidae) and roach,<i>Rutilus rutilus</i>(Cyprinidae)\n",
      "\tAlready Scored, Skipped\n",
      "98/1983 - Disentangling multiple predator effects in biodiversity and ecosystem functioning research\n",
      "\tAlready Scored, Skipped\n",
      "99/1983 - Eco‐evolutionary feedbacks between prey densities and linkage disequilibrium in the predator maintain diversity\n",
      "\tAlready Scored, Skipped\n",
      "100/1983 - Spillover of an alien parasite reduces expression of costly behaviour in native host species\n",
      "\tAlready Scored, Skipped\n",
      "101/1983 - Environmental and plant genetic effects on tri‐trophic interactions\n",
      "\tAlready Scored, Skipped\n",
      "102/1983 - Cascading effects of predator–detritivore interactions depend on environmental context in a <scp>T</scp>ibetan alpine meadow\n",
      "\tAlready Scored, Skipped\n",
      "103/1983 - Palatability and profitability of co‐occurring seeds alter indirect interactions among rodent‐dispersed trees\n",
      "\tAlready Scored, Skipped\n",
      "104/1983 - HOW FLUCTUATING COMPETITION AND PHENOTYPIC PLASTICITY MEDIATE SPECIES DIVERGENCE\n",
      "\tAlready Scored, Skipped\n",
      "105/1983 - An exotic herbivore reinforces competition between exotic and native plants\n",
      "\tAlready Scored, Skipped\n",
      "106/1983 - Dietary flexibility aids Asian earthworm invasion in North American forests\n",
      "\tAlready Scored, Skipped\n",
      "107/1983 - Intraspecific adaptive radiation: Competition, ecological opportunity, and phenotypic diversification within species\n",
      "\tAlready Scored, Skipped\n",
      "108/1983 - The Influence of Competition Trait-Mediated Effects on Intra-Guild Predation System\n",
      "\tAlready Scored, Skipped\n",
      "109/1983 - Growth tradeoffs produce complex microbial communities on a single limiting resource\n",
      "\tAlready Scored, Skipped\n",
      "110/1983 - The roles of history: age and prior exploitation in aquatic container habitats have immediate and carry‐over effects on mosquito life history\n",
      "\tAlready Scored, Skipped\n",
      "111/1983 - Pollinator competition and the structure of floral resources\n",
      "\tAlready Scored, Skipped\n",
      "112/1983 - Community Ecology and the Evolution of Molecules of Keystone Significance\n",
      "\tAlready Scored, Skipped\n",
      "113/1983 - Aardvarck to <i>Zyzyxia</i>– functional groups across kingdoms\n",
      "\tAlready Scored, Skipped\n",
      "114/1983 - An Ecological and Evolutionary Framework for Commensalism in Anthropogenic Environments\n",
      "115/1983 - Differences in prey personality mediate trophic cascades\n",
      "116/1983 - Interactive effects of competition and social environment on the expression of sexual dimorphism\n",
      "117/1983 - Variegated desert vegetation: Covariation of edaphic and fire variables provides a framework for understanding mulga‐spinifex coexistence\n",
      "118/1983 - Marine reserve designation, trophic cascades and altered community dynamics\n",
      "119/1983 - An experimental test of how parasites of predators can influence trophic cascades and ecosystem functioning\n",
      "120/1983 - Long‐term nitrogen input reduces soil bacterial network complexity by shifts in life history strategy in temperate grassland\n",
      "121/1983 - Determinants of Northerly Range Limits along the Himalayan Bird Diversity Gradient\n",
      "122/1983 - Interspecific Competition between Two Stream Insect Grazers Mediated by Non-Feeding Predatory Fish\n",
      "123/1983 - The Last Trees Standing: Climate modulates tree survival factors during a prolonged bark beetle outbreak in Europe\n",
      "124/1983 - Interactions between the introduced European honey bee and native bees in urban areas varies by year, habitat type and native bee guild\n",
      "125/1983 - Preference for feeding at habitat edges declines among juvenile blue crabs as oyster reef patchiness increases and predation risk grows\n",
      "126/1983 - Heterospecific Aggression and Dominance in a Guild of Coral-Feeding Fishes: The Roles of Dietary Ecology and Phylogeny\n",
      "127/1983 - Body shape evolution among ploidy levels of theSqualius alburnoideshybrid complex (Teleostei, Cyprinidae)\n",
      "128/1983 - Predator modulation of plant pathogen spread through induced changes in vector development rates\n",
      "129/1983 - Interactions between Climate and Nutrient Cycles on Forest Response to Global Change: The Role of Mixed Forests\n",
      "130/1983 - Intransitivity increases plant functional diversity by limiting dominance in drylands worldwide\n",
      "131/1983 - REPRODUCTIVE ECOLOGY AND MATING SUCCESS OF MALE LIMNONECTES KUHLII, A FANGED FROG FROM TAIWAN\n",
      "132/1983 - Evolutionary change in agriculture: the past, present and future\n",
      "133/1983 - Multiple fish predators: effects of identity, density, and nutrients on lower trophic levels\n",
      "134/1983 - Resource levels and prey state influence antipredator behavior and the strength of nonconsumptive predator effects\n",
      "135/1983 - Swimming performance trade-offs across a gradient in community composition in Trinidadian killifish (Rivulus hartii)\n",
      "136/1983 - Diet composition does not affect ant colony tempo\n",
      "137/1983 - Selection against Accumulating Mutations in Niche-Preference Genes Can Drive Speciation\n",
      "138/1983 - Restricting Prey Dispersal Can Overestimate the Importance of Predation in Trophic Cascades\n",
      "139/1983 - Some neighbours are better than others: Variation in associational effects among plants in an old field community\n",
      "140/1983 - Functional importance of sacred forest patches in the altered landscape of Palakkad region, Kerala, India\n",
      "141/1983 - Turnover in Life-Strategies Recapitulates Marine Microbial Succession Colonizing Model Particles\n",
      "142/1983 - Associational resistance or susceptibility: the indirect interaction between chemically‐defended and non‐defended herbivore prey via a shared predator\n",
      "143/1983 - CD4 <sup>+</sup> T cells do not mediate within-host competition between genetically diverse malaria parasites\n",
      "144/1983 - Fine-scale behavioural differences distinguish resource use by ecomorphs in a closed ecosystem\n",
      "145/1983 - Distinct Assembly Processes and Microbial Communities Constrain Soil Organic Carbon Formation\n",
      "146/1983 - Determinants of emigration and their impact on survival during dispersal in fox and jackal populations\n",
      "147/1983 - Ru<scp>BPC</scp>ase activase (<scp>RCA</scp>) mediates growth–defense trade‐offs: silencing <scp>RCA</scp> redirects jasmonic acid (<scp>JA</scp>) flux from <scp>JA</scp>‐isoleucine to methyl jasmonate (Me<scp>JA</scp>) to attenuate induced defense responses in <i>Nicotiana attenuata</i>\n",
      "148/1983 - Herbivore Diet Breadth and Host Plant Defense Mediate the Tri-Trophic Effects of Plant Toxins on Multiple Coccinellid Predators\n",
      "149/1983 - Tree, sex and size: Ecological determinants of male vs. female fecundity in three<i>Fagus sylvatica</i>stands\n",
      "150/1983 - Dynamic character displacement among a pair of bacterial phyllosphere commensals in situ\n",
      "151/1983 - Ecological Interactions Involving Feral Horses and Predators: Review with Implications for Biodiversity Conservation\n",
      "152/1983 - SEX-SPECIFIC EFFECTS OF ALBUMEN REMOVAL AND NEST ENVIRONMENT MANIPULATION ON BARN SWALLOW NESTLINGS\n",
      "153/1983 - Color mimicry of empty seeds influences the probability of predation by birds\n",
      "154/1983 - Niche Packing and Local Coexistence in a Megadiverse Guild of Frugivorous Birds Are Mediated by Fruit Dependence and Shifts in Interaction Frequencies\n",
      "155/1983 - Sexual selection's impacts on ecological specialization: an experimental test\n",
      "156/1983 - Biological invasion modifies the co‐occurrence patterns of insects along a stress gradient\n",
      "157/1983 - Effects of biotic interactions on plant fecundity depend on spatial and functional structure of communities and time since disturbance\n",
      "158/1983 - Not attackable or not crackable—How pre‐ and post‐attack defenses with different competition costs affect prey coexistence and population dynamics\n",
      "159/1983 - Biophysical Interactions Control the Size and Abundance of Large Phytoplankton Chains at the Ushant Tidal Front\n",
      "160/1983 - Using Resistant Prey Demonstrates That Bt Plants Producing Cry1Ac, Cry2Ab, and Cry1F Have No Negative Effects on<i>Geocoris punctipes</i>and<i>Orius insidiosus</i>\n",
      "161/1983 - Exploring the competitive dynamic enzyme allocation scheme through enzyme cost minimization\n",
      "162/1983 - Trophic cascades alter eco-evolutionary dynamics and body size evolution\n",
      "163/1983 - Can Top-Down Controls Expand the Ecological Niche of Marine N2 Fixers?\n",
      "164/1983 - BIOLOGICAL CONTROL OF APHIDS BY THE PREDATORY MIDGE APHIDOLETES APHIDIMYZA IN THE PRESENCE OF INTRAGUILD PREDATORY BUGS AND THRIPS\n",
      "165/1983 - The Cinderella syndrome: why do malaria-infected cells burst at midnight?\n",
      "166/1983 - DO HABITAT SHIFTS ALTER FLOWERING PHENOLOGY OVERLAP IN CLOSE RELATIVES? IMPLICATIONS FOR LOCAL COEXISTENCE\n",
      "167/1983 - Effects of elevated CO<sub>2</sub> and N on tree–grass interactions: an experimental test using <i>Fraxinus excelsior</i> and <i>Dactylis glomerata</i>\n",
      "168/1983 - Can diet niche partitioning enhance sexual dimorphism?\n",
      "169/1983 - Ecological Traits and Intraguild Competition Mediate Spatial and Temporal Overlaps Among Sympatric Mesocarnivores\n",
      "170/1983 - Carotenoid-based plumage coloration and aggression during molt in male house finches\n",
      "171/1983 - Experimental Addition of Green Plants to the Nest Increases Testosterone Levels in Female Spotless Starlings\n",
      "172/1983 - Breaking down the components of the competition‐colonization trade‐off: New insights into its role in diverse systems\n",
      "173/1983 - Feeding evolution of a herbivore influences an arthropod community through plants: implications for plant-mediated eco-evolutionary feedback loop\n",
      "174/1983 - Ecological release leads to novel ontogenetic diet shift in kokanee (<i>Oncorhynchus nerka</i>)\n",
      "175/1983 - Destabilizing effects on a classic tri-trophic oyster-reef cascade\n",
      "176/1983 - DENSITY-DEPENDENT COOPERATION AS A MECHANISM FOR PERSISTENCE AND COEXISTENCE\n",
      "177/1983 - Ant body size mediates functional performance and species interactions in carrion decomposer communities\n",
      "178/1983 - Landscape and local effects on multiparasitoid coexistence\n",
      "179/1983 - Love Thy Neighbour? Tropical Tree Growth and Its Response to Climate Anomalies Is Mediated by Neighbourhood Hierarchy and Dissimilarity in Carbon‐ and Water‐Related Traits\n",
      "180/1983 - Prey handling performance facilitates competitive dominance of an invasive over native keystone ant\n",
      "181/1983 - Investigating the morphological and genetic divergence of arctic char (<i>Salvelinus alpinus)</i> populations in lakes of arctic Alaska\n",
      "182/1983 - Foraging behaviour of brown trout in wild populations: can population density cause behaviourally-mediated foraging specializations?\n",
      "183/1983 - Changes in ecological conditions may influence intraguild competition: inferring interaction patterns of snow leopard with co-predators\n",
      "184/1983 - Stream Salamanders Accurately Assess Size-Dependent Predation Threats\n",
      "185/1983 - N-Induced Species Loss Dampened by Clipping Mainly Through Suppressing Dominant Species in an Alpine Meadow\n",
      "186/1983 - Nocturnal Bees Feed on Diurnal Leftovers and Pay the Price of Day – Night Lifestyle Transition\n",
      "187/1983 - The <i>Ecology Underground</i> coalition: building a collaborative future of belowground ecology and ecologists\n",
      "188/1983 - Character Shifts of Prey Species That Share Predators\n",
      "189/1983 - Clonal plant research: proliferation, integration, but not much evolution\n",
      "190/1983 - Parasitoids use chemical footprints to track down caterpillars\n",
      "191/1983 - Here come the commensals\n",
      "192/1983 - Seed Traits and Curculio Weevil Infestation: A Study in Quercus mongolica\n",
      "193/1983 - Co‐evolution of foraging behaviour in herbivores and their natural enemies predicts multifunctionality of herbivore‐induced plant volatiles\n",
      "194/1983 - Ghosts of Competition and Predation Past: Why Ecologists Value Negative Over Positive Interactions\n",
      "195/1983 - Condition-Specific Competition: Implications for the Altitudinal Distribution of Stream Fishes\n",
      "196/1983 - Pay it forward: refuse dump from leaf‐cutting ants promotes caterpillar digestive performance by increasing plant nitrogen content\n",
      "197/1983 - Disturbance size and frequency mediate the coexistence of benthic spatial competitors\n",
      "198/1983 - When the window is a mirror: how do dominant theories limit our understanding of nature? (<scp>ESA</scp> 2023 <scp>INS23</scp>)\n",
      "199/1983 - Factor in Fear: Interference Competition in Polymorphic Spadefoot Toad Tadpoles and Its Potential Role in Disruptive Selection\n",
      "200/1983 - The evolution of invasiveness: a mechanistic view of trade‐offs involving defenses\n",
      "201/1983 - Towards closer integration between ecology and evolution\n",
      "202/1983 - Systemic acquired resistance: available information and open ecological questions\n",
      "203/1983 - Phenological change and ecological interactions: an introduction\n",
      "204/1983 - Towards a mechanistic understanding of global change ecology\n",
      "205/1983 - How Plant Functional Traits of Dominant Species Respond to Fencing and Water-Nitrogen Addition in Horqin Grassland, China\n",
      "206/1983 - Distinct phytoplankton size classes respond differently to biotic and abiotic factors\n",
      "207/1983 - Act locally, think globally: Ant studies in Argentina in the context of ecological theory\n",
      "208/1983 - Neotyphodium impacts on soil nematode communities: rhizosphere- and litter-mediated interactions\n",
      "209/1983 - Out of sight, out of mind: the impacts of root‐feeding insects in natural and managed systems\n",
      "210/1983 - Competition between mixo‐ and heterotrophic ciliates under dynamic resource supply\n",
      "211/1983 - Resource supply and organismal dominance are associated with high secondary production in temperate agricultural streams\n",
      "212/1983 - Systems biology or the biology of systems: routes to reducing hunger\n",
      "213/1983 - Selection for domestication favored taxa characterized by fast growth and tolerance of high intraspecific density\n",
      "214/1983 - Differences in Tri‐Trophic Community Responses to Temperature‐Dependent Vital Rates, Thermal Niche Mismatches and Temperature‐Size Rule\n",
      "215/1983 - Organized Oral Session 44: Impacts of Species Addition and Species Loss on Ecosystem Function in Freshwater Systems\n",
      "216/1983 - Is It What We Know or Who We Know? Choice of Organism and Robustness of Inference in Ecology and Evolutionary Biology\n",
      "217/1983 - Plant neighbors differentially alter a focal species' biotic interactions through changes to resource allocation\n",
      "218/1983 - Toward “Rules” for Studying Biological Invasions\n",
      "219/1983 - The Drosophila Modifier of Variegationmodulo Gene Product Binds Specific RNA Sequences at the Nucleolus and Interacts with DNA and Chromatin in a Phosphorylation-dependent Manner\n",
      "220/1983 - Co‐infections and the third trophic level\n",
      "221/1983 - Individual control over reproduction: an underestimated element in the maintenance of sex?\n",
      "222/1983 - Keystone predator and keystone intransitivity and the rescue of a competitively subdominant species\n",
      "223/1983 - Introduction: Special issue on species interactions, ecological networks and community dynamics – Untangling the entangled bank using molecular techniques\n",
      "224/1983 - Reconstitution in Vitro of the V1 Complex from the Yeast Vacuolar Proton-translocating ATPase\n",
      "225/1983 - The Evolution of Social Behaviour\n",
      "226/1983 - Plastisphere microbiome: Methodology, diversity, and functionality\n",
      "227/1983 - The many faces of plant carbon relations: forging an ecophysiological identity in the age of human influence\n",
      "228/1983 - Non-additive effects of multiple predators alter the biological control of the coffee leaf miner (Leucoptera coffeella)\n",
      "229/1983 - Farmers' delivery of floral resources: to “bee” or not to “bee”\n",
      "230/1983 - Is It What We Know or Who We Know? Choice of Organism and Robustness of Inference in Ecology and Evolutionary Biology. (American Society of Naturalists Presidential Address)\n",
      "231/1983 - Evolution of plasticity and character displacement in a fluctuating environment\n",
      "232/1983 - Intracellular pathogens: living in an extreme environment\n",
      "233/1983 - Patterns of Ecological Specialisation Among Multiple Piscivorous Morphs of a Diverse Assemblage of <i>Salvelinus</i> Fishes\n",
      "234/1983 - THE ROLE OF SNAKE VENOM IN THE BIOTIC INTERACTION\n",
      "235/1983 - Competitive interactions affect introgression and population viability amidst maladaptive hybridization\n",
      "236/1983 - The role of habitat in predator–prey dynamics with applications to restoration\n",
      "237/1983 - Microclimatic effects on Central European deciduous tree species and their interactions with arthropod herbivory\n",
      "238/1983 - Mentoring and Education: A Lifetime of Experience and Learning\n",
      "239/1983 - Evolution of foraging behaviour induces variable complexity-stability relationships in mutualist-exploiter-predator communities\n",
      "240/1983 - Fitness differences override variation‐dependent coexistence mechanisms in California grasslands\n",
      "241/1983 - Using individual-based modeling to understand grassland diversity and resilience in the Anthropocene\n",
      "242/1983 - In Memoriam: Barry Sinervo 1961–2021\n",
      "243/1983 - Local‐ and Regional‐Scale Climate Variability Drives Complex Patterns of Growth Synchrony and Asynchrony in Deep‐Sea Snappers Across the Indo‐Pacific\n",
      "244/1983 - Comment on ‘Kidron, G. J. (2018). Biocrust research: A critical view on eight common hydrological‐related paradigms and dubious theses. <i>Ecohydrology</i>, e2061’\n",
      "245/1983 - <i>EcoSampler:</i>A Learning Object for Community Sampling, Community Structure, and Succession\n",
      "246/1983 - Editorial and Retrospective 2007\n",
      "247/1983 - Predictive Role of Species Diversity on Aboveground Carbon via Functional Diversity and Dominance in the Moist Afromontane Forest: Implications for Conservation Policies\n",
      "248/1983 - When lettuce bolts: natural selection vs artificial selection and beyond\n",
      "249/1983 - Response of fish biota to dams in the Lower Colorado River Basin: empirical findings and utility for predicting responses to climate and water use change\n",
      "250/1983 - Effect of Resource Abundance on Woodland Rodents' Demography at Latitudinal Extremes in Europe\n",
      "251/1983 - Understanding feedback between ecological systems requires feedback among scientists\n",
      "252/1983 - Invaders on the Wing\n",
      "253/1983 - \"Gently Caress Me, I Love Chris Jericho\": Pro Wrestling Fans \"Marking Out\"\n",
      "254/1983 - Review\n",
      "255/1983 - OOS 38: Ecological Dimensions of Biofuel Production\n",
      "256/1983 - Introduction: Honoring the contribution of Charles H. Peterson (1946–2020) to the field of marine ecology\n",
      "257/1983 - Trophic cascades: the primacy of trait‐mediated indirect interactions\n",
      "258/1983 - RESOURCE COMPETITION MODIFIES THE STRENGTH OF TRAIT-MEDIATED PREDATOR–PREY INTERACTIONS: A META-ANALYSIS\n",
      "259/1983 - TRAIT-MEDIATED INDIRECT EFFECTS IN LARVAL ANURANS: REVERSING COMPETITION WITH THE THREAT OF PREDATION\n",
      "260/1983 - Cascading trait-mediation: disruption of a trait-mediated mutualism by parasite-induced behavioral modification\n",
      "261/1983 - Trophic promiscuity, intraguild predation and the problem of omnivores\n",
      "262/1983 - Leaf pubescence mediates intraguild predation between predatory mites\n",
      "263/1983 - Trait‐mediated effects between predators: ant chemical cues induce spider dispersal\n",
      "264/1983 - Herbivore density mediates the indirect effect of herbivores on plants via induced resistance and apparent competition\n",
      "265/1983 - Density‐ and trait‐mediated effects of a parasite and a predator in a tri‐trophic food web\n",
      "266/1983 - Trait and density mediated indirect interactions in simple food webs\n",
      "267/1983 - Trait-Mediated Indirect Effects of Phorid Flies on Ants\n",
      "268/1983 - Moving beyond linear food chains: trait-mediated indirect interactions in a rocky intertidal food web\n",
      "269/1983 - Trait-Mediated Effects of Parasitic Phorid Flies (Diptera: Phoridae) on Ant (Hymenoptera: Formicidae) Competition and Resource Access in Coffee Agro-ecosystems\n",
      "270/1983 - Using plant functional traits to restore Hawaiian rainforest\n",
      "271/1983 - Controlling species densities in structurally perturbed intransitive cycles with higher-order interactions\n",
      "272/1983 - Trait-Mediated Effects of Parasitic Phorid Flies (Diptera: Phoridae) on Ant (Hymenoptera: Formicidae) Competition and Resource Access in Coffee Agro-ecosystems\n",
      "273/1983 - Herkogamy, a Principal Functional Trait of Plant Reproductive Biology\n",
      "274/1983 - Trait-Mediated Indirect Effects in Larval Anurans: Reversing Competition with the Threat of Predation\n",
      "275/1983 - Contrasting impacts of competition on ecological and social trait evolution in songbirds\n",
      "276/1983 - Mistletoe generates non‐trophic and trait‐mediated indirect interactions through a shared host of herbivore consumers\n",
      "277/1983 - From apparent competition to facilitation: Impacts of consumer niche construction on the coexistence and stability of consumer‐resource communities\n",
      "278/1983 - Trait-mediated competition drives an ant invasion and alters functional diversity\n",
      "279/1983 - Leaf pubescence mediates intraguild predation between predatory mites\n",
      "280/1983 - ECOLOGICAL MECHANISMS OF A RESOURCE POLYMORPHISM IN ZENAIDA DOVES OF BARBADOS\n",
      "281/1983 - SCARED TO DEATH? THE EFFECTS OF INTIMIDATION AND CONSUMPTION IN PREDATOR–PREY INTERACTIONS\n",
      "282/1983 - Species-Specific Traits Rather Than Resource Partitioning Mediate Diversity Effects on Resource Use\n",
      "283/1983 - EVOLUTIONARILY INDUCED ALTERNATIVE STATES AND COEXISTENCE IN SYSTEMS WITH APPARENT COMPETITION\n",
      "284/1983 - A multifaceted trophic cascade in a detritus‐based system: density‐, trait‐, or processing‐chain‐mediated effects?\n",
      "285/1983 - Symbiont-conferred immunity interacts with effects of parasitoid genotype and intraguild predation to affect aphid immunity in a clone-specific fashion\n",
      "286/1983 - A combination of developmental plasticity, parental effects, and genetic differentiation mediates divergences in life history traits between dung beetle populations\n",
      "287/1983 - Trait variability differs between leaf and wood tissues across ecological scales in subtropical forests\n",
      "288/1983 - Architectural and growth traits differ in effects on performance of clonal plants: an analysis using a field‐parameterized simulation model\n",
      "289/1983 - Architectural and growth traits differ in effects on performance of clonal plants: an analysis using a field-parameterized simulation model\n",
      "290/1983 - Ventral colour polymorphism correlates with alternative behavioural patterns in female common lizards (<i>Lacerta vivipara</i>)\n",
      "291/1983 - Fine‐root and mycorrhizal traits help explain ecosystem processes and responses to global change\n",
      "292/1983 - Competition drives trait evolution and character displacement between <i>Mimulus</i> species along an environmental gradient\n",
      "293/1983 - The sign of cascading predator effects varies with prey traits in a detrital system\n",
      "294/1983 - Can Functional Traits Explain Plant Coexistence? A Case Study with Tropical Lianas and Trees\n",
      "295/1983 - Parasites that change predator or prey behaviour can have keystone effects on community composition\n",
      "296/1983 - Direct litter interference and indirect soil competitive effects of two contrasting phenotypes of a spiny legume shrub drive the forb composition of an oromediterranean community\n",
      "297/1983 - FORAGING TRADE-OFFS ALONG A PREDATOR–PERMANENCE GRADIENT IN SUBALPINE WETLANDS\n",
      "298/1983 - Testing functional trait-based mechanisms underpinning plant responses to grazing and linkages to ecosystem functioning in grasslands\n",
      "299/1983 - Diet quality and conspecific larval density predict functional trait variation and performance in a polyphagous frugivorous fly\n",
      "300/1983 - Plant trait - environmental linkages among contrasting landscapes and climate regimes in temperate eucalypt woodlands\n",
      "301/1983 - Trait values and not invasive status determine competitive outcomes between native and invasive species under varying soil nutrient availability\n",
      "302/1983 - The community effects of phenotypic and genetic variation within a predator population\n",
      "303/1983 - Trophic consequences of terrestrial eutrophication for a threatened ungulate\n",
      "304/1983 - Biotic interactions shape trait assembly of marine communities across time and latitude\n",
      "305/1983 - REVISITING THE CLASSICS: CONSIDERING NONCONSUMPTIVE EFFECTS IN TEXTBOOK EXAMPLES OF PREDATOR–PREY INTERACTIONS\n",
      "306/1983 - Pelagic–benthic resource polymorphism in <i>Schizopygopsis thermalis</i> Herzenstein 1891 (Pisces, Cyprinidae) in a headwater lake in the Salween River system on the Tibetan Plateau\n",
      "307/1983 - Character Shifts of Prey Species That Share Predators\n",
      "308/1983 - Decline in top predator body size and changing climate alter trophic structure in an oceanic ecosystem\n",
      "309/1983 - HOW FLUCTUATING COMPETITION AND PHENOTYPIC PLASTICITY MEDIATE SPECIES DIVERGENCE\n",
      "310/1983 - Functional traits shape plant–plant interactions and recruitment in a hotspot of woody plant diversity\n",
      "311/1983 - RESOURCE IDENTITY MODIFIES THE INFLUENCE OF PREDATION RISK ON ECOSYSTEM FUNCTION\n",
      "312/1983 - Maternal antibodies in a wild altricial bird: effects on offspring immunity, growth and survival\n",
      "313/1983 - Evidence for an ecological cost of enhanced herbicide metabolism in<i>Lolium rigidum</i>\n",
      "314/1983 - Food resources and competition rather than eco-geographic rules explain trait variations in two contrasting rat species: Implications for future climate change\n",
      "315/1983 - Response of the residential piscivorous fish community to introduction of a new predator type in a mesotrophic lake\n",
      "316/1983 - Prey state shapes the effects of temporal variation in predation risk\n",
      "317/1983 - Neutral theory in community ecology and the hypothesis of functional equivalence\n",
      "318/1983 - Associational decomposition: After‐life traits and interactions among decomposing litters control during‐life aggregation of plant species\n",
      "319/1983 - Non‐consumptive effects of predatory mites on thrips and its host plant\n",
      "320/1983 - Intraspecific competition between co‐infecting parasite strains enhances host survival in African trypanosomes\n",
      "321/1983 - Extending trait dispersion across trophic levels: Predator assemblages act as top‐down filters on prey communities\n",
      "322/1983 - Diet and hormonal manipulation reveal cryptic genetic variation: implications for the evolution of novel feeding strategies\n",
      "323/1983 - The shape of the predator biomass distribution affects biological pest control services in agricultural landscapes\n",
      "324/1983 - Modelling coevolution in ecological networks with adaptive dynamics\n",
      "325/1983 - Allometric scaling of indirect effects: body size ratios predict non‐consumptive effects in multi‐predator systems\n",
      "326/1983 - Bioenergetic theory predicts infection dynamics of human schistosomes in intermediate host snails across ecological gradients\n",
      "327/1983 - Leaf trait variation in grassland plant species in response to soil phosphorus\n",
      "328/1983 - Herbivore release through cascading risk effects\n",
      "329/1983 - Herbivores as drivers of negative density dependence in tropical forest saplings\n",
      "330/1983 - CONDITION-SPECIFIC COMPETITION: IMPLICATIONS FOR THE ALTITUDINAL DISTRIBUTION OF STREAM FISHES\n",
      "331/1983 - Competition‐induced transgenerational plasticity influences competitive interactions and leaf decomposition of offspring\n",
      "332/1983 - Coexistence of coinvading species with mutualism and competition\n",
      "333/1983 - Patch Dynamics and Islands of Fertility in a High Mountain Mediterranean Community\n",
      "334/1983 - Indirect interactions in the microbial world: specificities and similarities to plant–insect systems\n",
      "335/1983 - Evolution as a Coexistence Mechanism: Does Genetic Architecture Matter?\n",
      "336/1983 - Positive effect of shade on plant growth: amelioration of stress or active regulation of growth rate?\n",
      "337/1983 - Adaptive foragers and community ecology: linking individuals to communities and ecosystems\n",
      "338/1983 - Predicting coexistence in species with continuous ontogenetic niche shifts and competitive asymmetry\n",
      "339/1983 - Chemical similarity and local community assembly in the species rich tropical genus <i>Piper</i>\n",
      "340/1983 - TRITROPHIC EFFECTS OF BIRDS AND ANTS ON A CANOPY FOOD WEB, TREE GROWTH, AND PHYTOCHEMISTRY\n",
      "341/1983 - Water‐use efficiency and relative growth rate mediate competitive interactions in Sonoran Desert winter annual plants\n",
      "342/1983 - Evolution of virulence: triggering host inflammation allows invading pathogens to exclude competitors\n",
      "343/1983 - GEOGRAPHIC VARIATION IN RESOURCE DOMINANCE–DISCOVERY IN BRAZILIAN ANT COMMUNITIES\n",
      "344/1983 - The negative effects of pathogen‐infected prey on predators: a meta‐analysis\n",
      "345/1983 - EVALUATING THE TARGETS OF SELECTION DURING CHARACTER DISPLACEMENT\n",
      "346/1983 - Inter‐strain competition and dispersal in aphids: evidence from a greenhouse study\n",
      "347/1983 - Cascading effects of a top predator on intraspecific competition at intermediate and basal trophic levels\n",
      "348/1983 - Exploring multitrophic interactions in oilseed rape fields reveals the prevailing role of Carabidae\n",
      "349/1983 - Non‐consumptive effects of a top‐predator decrease the strength of the trophic cascade in a four‐level terrestrial food web\n",
      "350/1983 - ASYMMETRIC COMPETITION VIA INDUCED RESISTANCE: SPECIALIST HERBIVORES INDIRECTLY SUPPRESS GENERALIST PREFERENCE AND POPULATIONS\n",
      "351/1983 - Trophic cascades in streams: effects of nutrient enrichment on autotrophic and consumer benthic communities under two different fish predation regimes\n",
      "352/1983 - ASSORTATIVE MATE CHOICE AND DOMINANCE MODIFICATION: ALTERNATIVE WAYS OF REMOVING HETEROZYGOTE DISADVANTAGE\n",
      "353/1983 - Genetic and morphological divergence along the littoral-pelagic axis in two common and sympatric fishes: perch,<i>Perca fluviatilis</i>(Percidae) and roach,<i>Rutilus rutilus</i>(Cyprinidae)\n",
      "354/1983 - Disentangling multiple predator effects in biodiversity and ecosystem functioning research\n",
      "355/1983 - Eco‐evolutionary feedbacks between prey densities and linkage disequilibrium in the predator maintain diversity\n",
      "356/1983 - Spillover of an alien parasite reduces expression of costly behaviour in native host species\n",
      "357/1983 - Environmental and plant genetic effects on tri‐trophic interactions\n",
      "358/1983 - Cascading effects of predator–detritivore interactions depend on environmental context in a <scp>T</scp>ibetan alpine meadow\n",
      "359/1983 - Palatability and profitability of co‐occurring seeds alter indirect interactions among rodent‐dispersed trees\n",
      "360/1983 - HOW FLUCTUATING COMPETITION AND PHENOTYPIC PLASTICITY MEDIATE SPECIES DIVERGENCE\n",
      "361/1983 - An exotic herbivore reinforces competition between exotic and native plants\n",
      "362/1983 - Dietary flexibility aids Asian earthworm invasion in North American forests\n",
      "363/1983 - Intraspecific adaptive radiation: Competition, ecological opportunity, and phenotypic diversification within species\n",
      "364/1983 - The Influence of Competition Trait-Mediated Effects on Intra-Guild Predation System\n",
      "365/1983 - Growth tradeoffs produce complex microbial communities on a single limiting resource\n",
      "366/1983 - The roles of history: age and prior exploitation in aquatic container habitats have immediate and carry‐over effects on mosquito life history\n",
      "367/1983 - Pollinator competition and the structure of floral resources\n",
      "368/1983 - Community Ecology and the Evolution of Molecules of Keystone Significance\n",
      "369/1983 - Aardvarck to <i>Zyzyxia</i>– functional groups across kingdoms\n",
      "370/1983 - An Ecological and Evolutionary Framework for Commensalism in Anthropogenic Environments\n",
      "371/1983 - Differences in prey personality mediate trophic cascades\n",
      "372/1983 - Interactive effects of competition and social environment on the expression of sexual dimorphism\n",
      "373/1983 - Variegated desert vegetation: Covariation of edaphic and fire variables provides a framework for understanding mulga‐spinifex coexistence\n",
      "374/1983 - Marine reserve designation, trophic cascades and altered community dynamics\n",
      "375/1983 - An experimental test of how parasites of predators can influence trophic cascades and ecosystem functioning\n",
      "376/1983 - Long‐term nitrogen input reduces soil bacterial network complexity by shifts in life history strategy in temperate grassland\n",
      "377/1983 - Determinants of Northerly Range Limits along the Himalayan Bird Diversity Gradient\n",
      "378/1983 - Interspecific Competition between Two Stream Insect Grazers Mediated by Non-Feeding Predatory Fish\n",
      "379/1983 - The Last Trees Standing: Climate modulates tree survival factors during a prolonged bark beetle outbreak in Europe\n",
      "380/1983 - Interactions between the introduced European honey bee and native bees in urban areas varies by year, habitat type and native bee guild\n",
      "381/1983 - Preference for feeding at habitat edges declines among juvenile blue crabs as oyster reef patchiness increases and predation risk grows\n",
      "382/1983 - Heterospecific Aggression and Dominance in a Guild of Coral-Feeding Fishes: The Roles of Dietary Ecology and Phylogeny\n",
      "383/1983 - Body shape evolution among ploidy levels of theSqualius alburnoideshybrid complex (Teleostei, Cyprinidae)\n",
      "384/1983 - Predator modulation of plant pathogen spread through induced changes in vector development rates\n",
      "385/1983 - Interactions between Climate and Nutrient Cycles on Forest Response to Global Change: The Role of Mixed Forests\n",
      "386/1983 - Intransitivity increases plant functional diversity by limiting dominance in drylands worldwide\n",
      "387/1983 - REPRODUCTIVE ECOLOGY AND MATING SUCCESS OF MALE LIMNONECTES KUHLII, A FANGED FROG FROM TAIWAN\n",
      "388/1983 - Evolutionary change in agriculture: the past, present and future\n",
      "389/1983 - Multiple fish predators: effects of identity, density, and nutrients on lower trophic levels\n",
      "390/1983 - Resource levels and prey state influence antipredator behavior and the strength of nonconsumptive predator effects\n",
      "391/1983 - Swimming performance trade-offs across a gradient in community composition in Trinidadian killifish (Rivulus hartii)\n",
      "392/1983 - Diet composition does not affect ant colony tempo\n",
      "393/1983 - Selection against Accumulating Mutations in Niche-Preference Genes Can Drive Speciation\n",
      "394/1983 - Restricting Prey Dispersal Can Overestimate the Importance of Predation in Trophic Cascades\n",
      "395/1983 - Some neighbours are better than others: Variation in associational effects among plants in an old field community\n",
      "396/1983 - Functional importance of sacred forest patches in the altered landscape of Palakkad region, Kerala, India\n",
      "397/1983 - Turnover in Life-Strategies Recapitulates Marine Microbial Succession Colonizing Model Particles\n",
      "398/1983 - Associational resistance or susceptibility: the indirect interaction between chemically‐defended and non‐defended herbivore prey via a shared predator\n",
      "399/1983 - CD4 <sup>+</sup> T cells do not mediate within-host competition between genetically diverse malaria parasites\n",
      "400/1983 - Fine-scale behavioural differences distinguish resource use by ecomorphs in a closed ecosystem\n",
      "401/1983 - Distinct Assembly Processes and Microbial Communities Constrain Soil Organic Carbon Formation\n",
      "402/1983 - Determinants of emigration and their impact on survival during dispersal in fox and jackal populations\n",
      "403/1983 - Ru<scp>BPC</scp>ase activase (<scp>RCA</scp>) mediates growth–defense trade‐offs: silencing <scp>RCA</scp> redirects jasmonic acid (<scp>JA</scp>) flux from <scp>JA</scp>‐isoleucine to methyl jasmonate (Me<scp>JA</scp>) to attenuate induced defense responses in <i>Nicotiana attenuata</i>\n",
      "404/1983 - Herbivore Diet Breadth and Host Plant Defense Mediate the Tri-Trophic Effects of Plant Toxins on Multiple Coccinellid Predators\n",
      "405/1983 - Tree, sex and size: Ecological determinants of male vs. female fecundity in three<i>Fagus sylvatica</i>stands\n",
      "406/1983 - Dynamic character displacement among a pair of bacterial phyllosphere commensals in situ\n",
      "407/1983 - Ecological Interactions Involving Feral Horses and Predators: Review with Implications for Biodiversity Conservation\n",
      "408/1983 - SEX-SPECIFIC EFFECTS OF ALBUMEN REMOVAL AND NEST ENVIRONMENT MANIPULATION ON BARN SWALLOW NESTLINGS\n",
      "409/1983 - Color mimicry of empty seeds influences the probability of predation by birds\n",
      "410/1983 - Niche Packing and Local Coexistence in a Megadiverse Guild of Frugivorous Birds Are Mediated by Fruit Dependence and Shifts in Interaction Frequencies\n",
      "411/1983 - Sexual selection's impacts on ecological specialization: an experimental test\n",
      "412/1983 - Biological invasion modifies the co‐occurrence patterns of insects along a stress gradient\n",
      "413/1983 - Effects of biotic interactions on plant fecundity depend on spatial and functional structure of communities and time since disturbance\n",
      "414/1983 - Not attackable or not crackable—How pre‐ and post‐attack defenses with different competition costs affect prey coexistence and population dynamics\n",
      "415/1983 - Biophysical Interactions Control the Size and Abundance of Large Phytoplankton Chains at the Ushant Tidal Front\n",
      "416/1983 - Using Resistant Prey Demonstrates That Bt Plants Producing Cry1Ac, Cry2Ab, and Cry1F Have No Negative Effects on<i>Geocoris punctipes</i>and<i>Orius insidiosus</i>\n",
      "417/1983 - Exploring the competitive dynamic enzyme allocation scheme through enzyme cost minimization\n",
      "418/1983 - Trophic cascades alter eco-evolutionary dynamics and body size evolution\n",
      "419/1983 - Can Top-Down Controls Expand the Ecological Niche of Marine N2 Fixers?\n",
      "420/1983 - BIOLOGICAL CONTROL OF APHIDS BY THE PREDATORY MIDGE APHIDOLETES APHIDIMYZA IN THE PRESENCE OF INTRAGUILD PREDATORY BUGS AND THRIPS\n",
      "421/1983 - The Cinderella syndrome: why do malaria-infected cells burst at midnight?\n",
      "422/1983 - DO HABITAT SHIFTS ALTER FLOWERING PHENOLOGY OVERLAP IN CLOSE RELATIVES? IMPLICATIONS FOR LOCAL COEXISTENCE\n",
      "423/1983 - Effects of elevated CO<sub>2</sub> and N on tree–grass interactions: an experimental test using <i>Fraxinus excelsior</i> and <i>Dactylis glomerata</i>\n",
      "424/1983 - Can diet niche partitioning enhance sexual dimorphism?\n",
      "425/1983 - Ecological Traits and Intraguild Competition Mediate Spatial and Temporal Overlaps Among Sympatric Mesocarnivores\n",
      "426/1983 - Carotenoid-based plumage coloration and aggression during molt in male house finches\n",
      "427/1983 - Experimental Addition of Green Plants to the Nest Increases Testosterone Levels in Female Spotless Starlings\n",
      "428/1983 - Breaking down the components of the competition‐colonization trade‐off: New insights into its role in diverse systems\n",
      "429/1983 - Feeding evolution of a herbivore influences an arthropod community through plants: implications for plant-mediated eco-evolutionary feedback loop\n",
      "430/1983 - Ecological release leads to novel ontogenetic diet shift in kokanee (<i>Oncorhynchus nerka</i>)\n",
      "431/1983 - Destabilizing effects on a classic tri-trophic oyster-reef cascade\n",
      "432/1983 - DENSITY-DEPENDENT COOPERATION AS A MECHANISM FOR PERSISTENCE AND COEXISTENCE\n",
      "433/1983 - Ant body size mediates functional performance and species interactions in carrion decomposer communities\n",
      "434/1983 - Landscape and local effects on multiparasitoid coexistence\n",
      "435/1983 - Love Thy Neighbour? Tropical Tree Growth and Its Response to Climate Anomalies Is Mediated by Neighbourhood Hierarchy and Dissimilarity in Carbon‐ and Water‐Related Traits\n",
      "436/1983 - Prey handling performance facilitates competitive dominance of an invasive over native keystone ant\n",
      "437/1983 - Investigating the morphological and genetic divergence of arctic char (<i>Salvelinus alpinus)</i> populations in lakes of arctic Alaska\n",
      "438/1983 - Foraging behaviour of brown trout in wild populations: can population density cause behaviourally-mediated foraging specializations?\n",
      "439/1983 - Changes in ecological conditions may influence intraguild competition: inferring interaction patterns of snow leopard with co-predators\n",
      "440/1983 - Stream Salamanders Accurately Assess Size-Dependent Predation Threats\n",
      "441/1983 - N-Induced Species Loss Dampened by Clipping Mainly Through Suppressing Dominant Species in an Alpine Meadow\n",
      "442/1983 - Nocturnal Bees Feed on Diurnal Leftovers and Pay the Price of Day – Night Lifestyle Transition\n",
      "443/1983 - The <i>Ecology Underground</i> coalition: building a collaborative future of belowground ecology and ecologists\n",
      "444/1983 - Character Shifts of Prey Species That Share Predators\n",
      "445/1983 - Clonal plant research: proliferation, integration, but not much evolution\n",
      "446/1983 - Parasitoids use chemical footprints to track down caterpillars\n",
      "447/1983 - Here come the commensals\n",
      "448/1983 - Seed Traits and Curculio Weevil Infestation: A Study in Quercus mongolica\n",
      "449/1983 - Co‐evolution of foraging behaviour in herbivores and their natural enemies predicts multifunctionality of herbivore‐induced plant volatiles\n",
      "450/1983 - Ghosts of Competition and Predation Past: Why Ecologists Value Negative Over Positive Interactions\n",
      "451/1983 - Condition-Specific Competition: Implications for the Altitudinal Distribution of Stream Fishes\n",
      "452/1983 - Pay it forward: refuse dump from leaf‐cutting ants promotes caterpillar digestive performance by increasing plant nitrogen content\n",
      "453/1983 - Disturbance size and frequency mediate the coexistence of benthic spatial competitors\n",
      "454/1983 - When the window is a mirror: how do dominant theories limit our understanding of nature? (<scp>ESA</scp> 2023 <scp>INS23</scp>)\n",
      "455/1983 - Factor in Fear: Interference Competition in Polymorphic Spadefoot Toad Tadpoles and Its Potential Role in Disruptive Selection\n",
      "456/1983 - The evolution of invasiveness: a mechanistic view of trade‐offs involving defenses\n",
      "457/1983 - Towards closer integration between ecology and evolution\n",
      "458/1983 - Systemic acquired resistance: available information and open ecological questions\n",
      "459/1983 - Phenological change and ecological interactions: an introduction\n",
      "460/1983 - Towards a mechanistic understanding of global change ecology\n",
      "461/1983 - How Plant Functional Traits of Dominant Species Respond to Fencing and Water-Nitrogen Addition in Horqin Grassland, China\n",
      "462/1983 - Distinct phytoplankton size classes respond differently to biotic and abiotic factors\n",
      "463/1983 - Act locally, think globally: Ant studies in Argentina in the context of ecological theory\n",
      "464/1983 - Neotyphodium impacts on soil nematode communities: rhizosphere- and litter-mediated interactions\n",
      "465/1983 - Out of sight, out of mind: the impacts of root‐feeding insects in natural and managed systems\n",
      "466/1983 - Competition between mixo‐ and heterotrophic ciliates under dynamic resource supply\n",
      "467/1983 - Resource supply and organismal dominance are associated with high secondary production in temperate agricultural streams\n",
      "468/1983 - Systems biology or the biology of systems: routes to reducing hunger\n",
      "469/1983 - Selection for domestication favored taxa characterized by fast growth and tolerance of high intraspecific density\n",
      "470/1983 - Differences in Tri‐Trophic Community Responses to Temperature‐Dependent Vital Rates, Thermal Niche Mismatches and Temperature‐Size Rule\n",
      "471/1983 - Organized Oral Session 44: Impacts of Species Addition and Species Loss on Ecosystem Function in Freshwater Systems\n",
      "472/1983 - Is It What We Know or Who We Know? Choice of Organism and Robustness of Inference in Ecology and Evolutionary Biology\n",
      "473/1983 - Plant neighbors differentially alter a focal species' biotic interactions through changes to resource allocation\n",
      "474/1983 - Toward “Rules” for Studying Biological Invasions\n",
      "475/1983 - The Drosophila Modifier of Variegationmodulo Gene Product Binds Specific RNA Sequences at the Nucleolus and Interacts with DNA and Chromatin in a Phosphorylation-dependent Manner\n",
      "476/1983 - Co‐infections and the third trophic level\n",
      "477/1983 - Individual control over reproduction: an underestimated element in the maintenance of sex?\n",
      "478/1983 - Keystone predator and keystone intransitivity and the rescue of a competitively subdominant species\n",
      "479/1983 - Introduction: Special issue on species interactions, ecological networks and community dynamics – Untangling the entangled bank using molecular techniques\n",
      "480/1983 - Reconstitution in Vitro of the V1 Complex from the Yeast Vacuolar Proton-translocating ATPase\n",
      "481/1983 - The Evolution of Social Behaviour\n",
      "482/1983 - Plastisphere microbiome: Methodology, diversity, and functionality\n",
      "483/1983 - The many faces of plant carbon relations: forging an ecophysiological identity in the age of human influence\n",
      "484/1983 - Non-additive effects of multiple predators alter the biological control of the coffee leaf miner (Leucoptera coffeella)\n",
      "485/1983 - Farmers' delivery of floral resources: to “bee” or not to “bee”\n",
      "486/1983 - Is It What We Know or Who We Know? Choice of Organism and Robustness of Inference in Ecology and Evolutionary Biology. (American Society of Naturalists Presidential Address)\n",
      "487/1983 - Evolution of plasticity and character displacement in a fluctuating environment\n",
      "488/1983 - Intracellular pathogens: living in an extreme environment\n",
      "489/1983 - Patterns of Ecological Specialisation Among Multiple Piscivorous Morphs of a Diverse Assemblage of <i>Salvelinus</i> Fishes\n",
      "490/1983 - THE ROLE OF SNAKE VENOM IN THE BIOTIC INTERACTION\n",
      "491/1983 - Competitive interactions affect introgression and population viability amidst maladaptive hybridization\n",
      "492/1983 - The role of habitat in predator–prey dynamics with applications to restoration\n",
      "493/1983 - Microclimatic effects on Central European deciduous tree species and their interactions with arthropod herbivory\n",
      "494/1983 - Mentoring and Education: A Lifetime of Experience and Learning\n",
      "495/1983 - Evolution of foraging behaviour induces variable complexity-stability relationships in mutualist-exploiter-predator communities\n",
      "496/1983 - Fitness differences override variation‐dependent coexistence mechanisms in California grasslands\n",
      "497/1983 - Using individual-based modeling to understand grassland diversity and resilience in the Anthropocene\n",
      "498/1983 - In Memoriam: Barry Sinervo 1961–2021\n",
      "499/1983 - Local‐ and Regional‐Scale Climate Variability Drives Complex Patterns of Growth Synchrony and Asynchrony in Deep‐Sea Snappers Across the Indo‐Pacific\n",
      "500/1983 - Comment on ‘Kidron, G. J. (2018). Biocrust research: A critical view on eight common hydrological‐related paradigms and dubious theses. <i>Ecohydrology</i>, e2061’\n",
      "501/1983 - <i>EcoSampler:</i>A Learning Object for Community Sampling, Community Structure, and Succession\n",
      "502/1983 - Editorial and Retrospective 2007\n",
      "503/1983 - Predictive Role of Species Diversity on Aboveground Carbon via Functional Diversity and Dominance in the Moist Afromontane Forest: Implications for Conservation Policies\n",
      "504/1983 - When lettuce bolts: natural selection vs artificial selection and beyond\n",
      "505/1983 - Response of fish biota to dams in the Lower Colorado River Basin: empirical findings and utility for predicting responses to climate and water use change\n",
      "506/1983 - Effect of Resource Abundance on Woodland Rodents' Demography at Latitudinal Extremes in Europe\n",
      "507/1983 - Understanding feedback between ecological systems requires feedback among scientists\n",
      "508/1983 - Invaders on the Wing\n",
      "509/1983 - \"Gently Caress Me, I Love Chris Jericho\": Pro Wrestling Fans \"Marking Out\"\n",
      "510/1983 - Review\n",
      "511/1983 - OOS 38: Ecological Dimensions of Biofuel Production\n",
      "512/1983 - Introduction: Honoring the contribution of Charles H. Peterson (1946–2020) to the field of marine ecology\n",
      "513/1983 - Recruitment to a seabird population depends on environmental factors and on population size\n",
      "514/1983 - The North Atlantic Oscillation\n",
      "515/1983 - The role of industrial fisheries and oceanographic change in the decline of North Sea black‐legged kittiwakes\n",
      "516/1983 - Adult survival rates of Shag (Phalacrocorax aristotelis), Common Guillemot (Uria aalge), Razorbill (Alca torda), Puffin (Fratercula arctica) and Kittiwake (Rissa tridactyla) on the Isle of May 1986-96\n",
      "517/1983 - Biomass Yields And Geography Of Large Marine Ecosystems\n",
      "518/1983 - Variability And Management Of Large Marine Ecosystems\n",
      "519/1983 - Invasiveness potential of <i>Miscanthus sinensis</i>: implications for bioenergy production in the United States\n",
      "520/1983 - Biomass for energy: supply prospects.\n",
      "521/1983 - Biofuels: Effects on Land and Fire\n",
      "522/1983 - What drives tropical deforestation?: a meta-analysis of proximate and underlying causes of deforestation based on subnational case study evidence\n",
      "523/1983 - Biomass from crop residues: cost and supply estimates.\n",
      "524/1983 - Does proximity to conspecific adults influence the establishment of ectomycorrhizal trees in rain forest?\n",
      "525/1983 - The Influence of Functional Diversity and Composition on Ecosystem Processes\n",
      "526/1983 - Plant diversity and ecosystem productivity: Theoretical considerations\n",
      "527/1983 - Tropical Soil Fertility Changes Under Monocultures and Successional Communities of Different Structure\n",
      "528/1983 - Compositional Analysis of Selected Second-Growth Species from Lowland Guatemala and Panama\n",
      "529/1983 - Biodiversity: Population Versus Ecosystem Stability\n",
      "530/1983 - Phenology and dynamics of an African rainforest at Korup, Cameroon\n",
      "531/1983 - Vesicular–arbuscular mycorrhizae in the Chenopodiaceae and Cruciferae: do they occur?\n",
      "532/1983 - THE INFLUENCE OF VESICULAR‐ARBUSCULAR MYCORRHIZAS ON PHOSPHORUS TRANSFER BETWEEN PLANTS\n",
      "533/1983 - Incorporating the Soil Community into Plant Population Dynamics: The Utility of the Feedback Approach\n",
      "534/1983 - Deer farming in New Zealand\n",
      "535/1983 - Direct and indirect cues of predatory risk and patch use by fox squirrels and thirteen-lined ground squirrels\n",
      "536/1983 - A REVIEW OF TRAIT-MEDIATED INDIRECT INTERACTIONS IN ECOLOGICAL COMMUNITIES\n",
      "537/1983 - Predator facilitation: the combined effect of snakes and owls on the foraging behavior of gerbils\n",
      "538/1983 - Direct Observations of Owls and Heteromyid Rodents: Can Predation Risk Explain Microhabitat Use?\n",
      "539/1983 - Effects of Habitat Loss and Fragmentation on the Behavior and Demography of Gray‐Tailed Voles\n",
      "540/1983 - Predatory Fish Removal and Native Fish Recovery in the Colorado River Mainstem\n",
      "541/1983 - Nonnative Fish Control and Endangered Fish Recovery: Lessons from the Colorado River\n",
      "542/1983 - Origin of Brood Stock and Allozyme Variation in Hatchery-Reared Bonytail, an Endangered North American Cyprinid Fish\n",
      "543/1983 - Extinctions of North American Fishes During the past Century\n",
      "544/1983 - The relative importance of climate, stand variables and liana abundance for carbon storage in tropical forests\n",
      "545/1983 - Diversity enhances carbon storage in tropical forests\n",
      "546/1983 - Tropical tree seedling growth responses to nitrogen, phosphorus and potassium addition\n",
      "547/1983 - Disentangling stand and environmental correlates of aboveground biomass in Amazonian forests\n",
      "548/1983 - Patterns in potassium dynamics in forest ecosystems\n",
      "549/1983 - Tropical Andean forest derives calcium and magnesium from Saharan dust\n",
      "550/1983 - Liana Abundance Patterns: The Role of Ecological Filters during Development\n",
      "551/1983 - The partitioning of diversity: showing Theseus a way out of the labyrinth\n",
      "552/1983 - Additive partitioning of diversity including species differences: a comment on Hardy &amp; Senterre (2007)\n",
      "553/1983 - Interpreting and estimating measures of community phylogenetic structuring\n",
      "554/1983 - Characterizing the phylogenetic structure of communities by an additive partitioning of phylogenetic diversity\n",
      "555/1983 - Spatial population genetic structure and limited dispersal in a Rocky Mountain alpine stream insect\n",
      "556/1983 - Genetic variation and population differentiation in the lichen‐forming ascomycete<i>Xanthoria parietina</i>on the island Storfosna, central Norway\n",
      "557/1983 - Preface to the special issue: ecological and evolutionary genomics of populations in nature\n",
      "558/1983 - Parallelism in gene transcription among sympatric lake whitefish (<i>Coregonus clupeaformis</i>Mitchill) ecotypes\n",
      "559/1983 - Genetic differentiation of a European caddisfly: past and present gene flow among fragmented larval habitats\n",
      "560/1983 - The complete trees of North America : field guide and natural history\n",
      "561/1983 - Factors Influencing Development of Cryptogamic Soil Crusts in Utah Deserts\n",
      "562/1983 - Soil Properties in Relation to Cryptogamic Groundcover in Canyonlands National Park\n",
      "563/1983 - Geo‐ecological soil features and the vegetation pattern in an arid dune area in the Northern Negev, Israel\n",
      "564/1983 - Desert Ecosystems: Their Resources in Space and Time\n",
      "565/1983 - Variation in Population Synchrony in a Multi-Species Seabird Community: Response to Changes in Predator Abundance\n",
      "566/1983 - Visual tracking from a rigid-hulled inflatable boat to determine foraging movements of breeding terns\n",
      "567/1983 - Severity of the Effects of Invasive Rats on Seabirds: A Global Review\n",
      "568/1983 - Variations in the diet of introduced Norway rats (<i>Rattus norvegicus</i>) inferred using stable isotope analysis\n",
      "569/1983 - Reflections of a specialist : Patterns in food provisioning and foraging conditions in sandwich terns Sterna sandvicensis\n",
      "570/1983 - GPS tracking of the foraging movements of Manx Shearwaters <i>Puffinus puffinus</i> breeding on Skomer Island, Wales\n",
      "571/1983 - Radio telemetry as a tool for impact assessment of wind farms: the case of Little Terns <i>Sterna albifrons</i> at Scroby Sands, Norfolk, UK\n",
      "572/1983 - Physiology and Climate Change\n",
      "573/1983 - Optimal growth temperature hypothesis: Why do anchovy flourish and sardine collapse or vice versa under the same ocean regime?\n",
      "574/1983 - Energetic Responses of Salmon to Temperature. A Study of Some Thermal Relations in the Physiology and Freshwater Ecology of Sockeye Salmon (<i>Oncorhynchus nerkd</i>)\n",
      "575/1983 - Controls over native perennial grass exclusion and persistence in California grasslands invaded by annuals\n",
      "576/1983 - Virus infection in remnant native bunchgrasses from invaded California grasslands\n",
      "577/1983 - Dynamics of competition between wild oats (<i>Avena fatua</i> L.) and winter cereals\n",
      "578/1983 - Temporal Dynamics and Ecological Process\n",
      "579/1983 - Consequences of adaptive behaviour for the structure and dynamics of food webs\n",
      "580/1983 - Diversity, Complexity, and Persistence in Large Model Ecosystems\n",
      "581/1983 - Experimental Ecology: Issues and Perspectives\n",
      "582/1983 - Allee effects driven by predation\n",
      "583/1983 - The future role of large carnivores in terrestrial trophic interactions: the northern temperate view\n",
      "584/1983 - The stability of model ecosystems\n",
      "585/1983 - The Alaska Salmon Enhancement Program: A Cost/Benefit Analysis\n",
      "586/1983 - Using an International Econometric Model to Forecast Alaska Salmon Revenues\n",
      "587/1983 - Can interbreeding of wild and artificially propagated animals be prevented by using broodstock selected for a divergent life history?\n",
      "588/1983 - Run Timing, Spawn Timing, and Spawning Distribution of Hatchery‐ and Natural‐Origin Spring Chinook Salmon in the Imnaha River, Oregon\n",
      "589/1983 - The Role of Hybrid Vigor in the Replacement of Pecos Pupfish by Its Hybrids with Sheepshead Minnow\n",
      "590/1983 - Genetic Introgression of Endemic Taxa by Non‐natives: A Case Study with Leon Springs Pupfish and Sheepshead Minnow\n",
      "591/1983 - Molecular evidence for multiple origins of hybridogenetic fish clones (Poeciliidae:Poeciliopsis).\n",
      "592/1983 - Predator–prey interactions mediated by prey personality and predator hunting mode\n",
      "593/1983 - Predator and prey functional traits: understanding the adaptive machinery driving predator–prey interactions\n",
      "594/1983 - Outrun or Outmaneuver: Predator–Prey Interactions as a Model System for Integrating Biomechanical Studies in a Broader Ecological and Evolutionary Context\n",
      "595/1983 - Predators determine how weather affects the spatial niche of lizard prey: exploring niche dynamics at a fine scale\n",
      "596/1983 - HABITAT COMPLEXITY DISRUPTS PREDATOR–PREY INTERACTIONS BUT NOT THE TROPHIC CASCADE ON OYSTER REEFS\n",
      "597/1983 - Associational Plant Defenses and the Maintenance of Species Diversity: Turning Competitors Into Accomplices\n",
      "598/1983 - Intraguild predation between an introduced lady beetle, Harmonia axyridis (Coleoptera: Coccinellidae), and a native lady beetle, Coleomegilla maculata (Coleoptera: Coccinellidae).\n",
      "599/1983 - Chronobiology: biological timekeeping\n",
      "600/1983 - Observations on the pollination of red clover by bees\n",
      "601/1983 - An entangled bank: the origins of ecosystem ecology\n",
      "602/1983 - Predicting species occurrences : issues of accuracy and scale\n",
      "603/1983 - Phylogeny, Lifeform, And Habitat Dependence Of Ant-Defended Plants In A Panamanian Forest\n",
      "604/1983 - Multitrophic/multispecies mutualistic interactions: the role of non-mutualists in shaping and mediating mutualisms\n",
      "605/1983 - <i>Bacillus thuringiensis</i>serovar<i>israelensis</i>is highly toxic to the coffee berry borer,<i>Hypothenemus hampei</i>Ferr. (Coleoptera: Scolytidae)\n",
      "606/1983 - Trait‐mediated interactions: influence of prey size, density and experience\n",
      "607/1983 - Functional Redundancy, Non‐Additive Interactions, and Supply‐Side Dynamics in Experimental Pond Communities\n",
      "608/1983 - Declining Amphibian Populations: The Problem of Separating Human Impacts from Natural Fluctuations\n",
      "609/1983 - Interaction Between Food Availability and Predation Mortality Mediated by Adaptive Behavior\n",
      "610/1983 - Physiological Plant Ecology: Ecophysiology and Stress Physiology of Functional Groups\n",
      "611/1983 - The Rise of Plants and Their Effect on Weathering and Atmospheric CO <sub>2</sub>\n",
      "612/1983 - Early Forest Soils and Their Role in Devonian Global Change\n",
      "613/1983 - Seasonal and inter-annual variations of PCB and DDT contents in the oyster Crassostrea angulata from the Sado Estuary (Portugal)\n",
      "614/1983 - The Introduction of Polychaetes Hydroides elegans (Haswell), Polydora limicola Annenkova, and Pseudopotamilla occelata Moore to the Northwestern Part of the East Sea\n",
      "615/1983 - Female ambrosia beetles adjust their offspring sex ratio according to outbreeding opportunities for their sons\n",
      "616/1983 - First-Cohort Advantage Hypothesis: A New Twist on Facultative Sex Ratio Adjustment\n",
      "617/1983 - Ecology: The Experimental Analysis of Distribution and Abundance\n",
      "618/1983 - Exposing the structure of an Arctic food web\n",
      "619/1983 - Improved detection of an alien invasive species through environmental DNA barcoding: the example of the American bullfrog<i>Lithobates catesbeianus</i>\n",
      "620/1983 - Browsed twig environmental DNA: diagnostic PCR to identify ungulate species\n",
      "621/1983 - “Sight-unseen” detection of rare aquatic species using environmental DNA\n",
      "622/1983 - Can filter‐feeding Asian carp invade the Laurentian Great Lakes? A bioenergetic modelling exercise\n",
      "623/1983 - The roles of large herbivores in ecosystem nutrient cycles\n",
      "624/1983 - Occupancy Estimation and Modeling: Inferring Patterns and Dynamics of Species Occurrence\n",
      "625/1983 - Physical Conditions, Carbon Transport, and Climate Change Impacts in a Northeast Greenland Fjord\n",
      "626/1983 - Global Change and Arctic Ecosystems: Conclusions and Predictions from Experiments with Terrestrial Invertebrates on Spitsbergen\n",
      "627/1983 - Planning for success: Serengeti lions seek prey accessibility rather than abundance\n",
      "628/1983 - The Natural History and Behavior of the California Sea Lion\n",
      "629/1983 - Bacteriophages Encode Factors Required for Protection in a Symbiotic Mutualism\n",
      "630/1983 - An Elicitor of Plant Volatiles from Beet Armyworm Oral Secretion\n",
      "631/1983 - Exploitation of Herbivore-Induced Plant Odors by Host-Seeking Parasitic Wasps\n",
      "632/1983 - Environmental variation has stronger effects than plant genotype on competition among plant species\n",
      "633/1983 - Mutual Feedbacks Maintain Both Genetic and Species Diversity in a Plant Community\n",
      "634/1983 - Invasive Plant Suppresses the Growth of Native Tree Seedlings by Disrupting Belowground Mutualisms\n",
      "635/1983 - Multitrophic interactions below and above ground: <i>en route</i> to the next level\n",
      "636/1983 - Tadpoles enhance microbial activity and leaf decomposition in a neotropical headwater stream\n",
      "637/1983 - Tangled webs: reciprocal flows of invertebrate prey link streams and riparian zones\n",
      "638/1983 - FISH DISTRIBUTIONS AND NUTRIENT CYCLING IN STREAMS: CAN FISH CREATE BIOGEOCHEMICAL HOTSPOTS\n",
      "639/1983 - Ecological Studies in Tropical Fish Communities\n",
      "640/1983 - Effects of Biomass, Light, and Grazing on Phosphorus Cycling in Stream Periphyton Communities\n",
      "641/1983 - Role of Nutrient Cycling and Herbivory in Regulating Periphyton Communities in Laboratory Streams\n",
      "642/1983 - Do tadpoles affect leaf decomposition in neotropical streams?\n",
      "643/1983 - Bacterial Biovolume and Biomass Estimations\n",
      "644/1983 - Warming-induced reductions in body size are greater in aquatic than terrestrial species\n",
      "645/1983 - Recent increase in body size of the American marten Martes americana in Alaska\n",
      "646/1983 - Climatic change and body size in two species of Japanese rodents\n",
      "647/1983 - The temperature‐size rule emerges from ontogenetic differences between growth and development rates\n",
      "648/1983 - Is Physical Space a Soil Resource?\n",
      "649/1983 - Restricted Rooting Decreases Tillering and Growth of Winter Wheat<sup>1</sup>\n",
      "650/1983 - Physiological Ecology of Three Codominant Successional Annuals\n",
      "651/1983 - Understory Response to Canopy Gaps of Varying Size in a Mature Oak Forest\n",
      "652/1983 - Ecological Relationships of Co-Occurring Goldenrods (Solidago: Compositae)\n",
      "653/1983 - Berseem clover is getting a second chance\n",
      "654/1983 - Beyond taxonomy: a review of macroinvertebrate trait‐based community descriptors as tools for freshwater biomonitoring\n",
      "655/1983 - Macroinvertebrate community structure, secondary production and trophic‐level dynamics in urban streams affected by non‐point‐source pollution\n",
      "656/1983 - Resource quality and stoichiometric constraints on stream ecosystem functioning\n",
      "657/1983 - PRODUCTION AND BIOENERGETIC ROLE OF THE MIDGE GLYPTOTENDIPES BARBIPES (STAEGER) IN A WASTE STABILIZATION LAGOON1\n",
      "658/1983 - Synthesis of nutrient and sediment data for watersheds within the Chesapeake Bay drainage basin\n",
      "659/1983 - The effects of acidic deposition on streams in the Appalachian Mountain and Piedmont Region of the Mid‐Atlantic United States\n",
      "660/1983 - TESTING THE SPECIES TRAITS–ENVIRONMENT RELATIONSHIPS: THE FOURTH‐CORNER PROBLEM REVISITED\n",
      "661/1983 - Effects of exotic amphipod invasions on fish diet in the Lower Rhine\n",
      "662/1983 - CONSISTENCY BETWEEN ORDINATION TECHNIQUES AND DIVERSITY MEASUREMENTS: TWO STRATEGIES FOR SPECIES OCCURRENCE DATA\n",
      "663/1983 - Photoadaptation Alters the Ingestion Rate of <i>Paramecium bursaria,</i> a Mixotrophic Ciliate\n",
      "664/1983 - A model of annual plankton cycles\n",
      "665/1983 - Phytoplankton modeling involving random rate constants\n",
      "666/1983 - Use of Monodispersed, Fluorescently Labeled Bacteria to Estimate In Situ Protozoan Bacterivory\n",
      "667/1983 - Determination of bacterial number and biomass in the marine environment\n",
      "668/1983 - Use of Genetically Marked Minicells as a Probe in Measurement of Predation on Bacteria in Aquatic Environments\n",
      "669/1983 - Use of Metabolic Inhibitors to Estimate Protozooplankton Grazing and Bacterial Production in a Monomictic Eutrophic Lake with an Anaerobic Hypolimnion\n",
      "670/1983 - Below‐ground herbivory promotes soil nutrient transfer and root growth in grassland\n",
      "671/1983 - Grasslands and montane communities\n",
      "672/1983 - Endophytic fungi and interactions among host plants, herbivores, and natural enemies\n",
      "673/1983 - Bottom–up cascades induced by fungal endophytes in multitrophic systems\n",
      "674/1983 - Barley yellow dwarf viruses in Japanese pasture grasses and lack of correlation with the presence of fungal endophytes\n",
      "675/1983 - Ant Abundance along a Productivity Gradient: Addressing Two Conflicting Hypotheses\n",
      "676/1983 - Effect of harvester ants of the genus <i>Pogonomyrmex</i> on the soil seed bank around their nests in the central Monte desert, Argentina\n",
      "677/1983 - Influence of temporal fluctuations in seed abundance on the foraging behaviour of harvester ants (<i>Pogonomyrmex</i> spp.) in the central Monte desert, Argentina\n",
      "678/1983 - Consequences of climate and body size on the foraging performance of seed‐eating ants\n",
      "679/1983 - The Nature of the Taxon Cycle in the Melanesian Ant Fauna\n",
      "680/1983 - Dynamic matching of forager size to resources in the continuously polymorphic leaf‐cutter ant,<i>Atta colombica</i>(Hymenoptera, Formicidae)\n",
      "681/1983 - The Depression of Reptile Biomass by Large Herbivores\n",
      "682/1983 - Biology and ecology of the leaf cutting ant Atta vollenweideri Forel, 1893\n",
      "683/1983 - INDIRECT FACILITATION: EVIDENCE AND PREDICTIONS FROM A RIPARIAN COMMUNITY\n",
      "684/1983 - Induced Resistance and Interspecific Competition Between Spider Mites and a Vascular Wilt Fungus\n",
      "685/1983 - COMPETITIVE RELATIONSHIPS BETWEEN ATRAZINE RESISTANT AND SUSCEPTIBLE POPULATIONS OF <i>AMARANTHUS RETROFLEXUS</i> AND <i>A. POWELLII</i> FROM SOUTHERN ONTARIO\n",
      "686/1983 - Community assembly and shifts in plant trait distributions across an environmental gradient in coastal California\n",
      "687/1983 - Ecological Assembly Rules\n",
      "688/1983 - THE SUMMER WATER BUDGET AND ITS IMPORTANCE IN THE ALPINE TUNDRA OF COLORADO\n",
      "689/1983 - Niche overlap reveals the effects of competition, disturbance and contrasting assembly processes in experimental grassland communities\n",
      "690/1983 - Opposing effects of competitive exclusion on the phylogenetic structure of communities\n",
      "691/1983 - Photosynthesis, respiration, and carbon turnover in sinking marine snow from surface waters of Southern California Bight:implications for the carbon cycle in the ocean\n",
      "692/1983 - Acquisition and allocation of resources in two waterlogging‐tolerant grasses\n",
      "693/1983 - Ecotypic differentiation in relation to soil nitrogen in northern Scandinavian Cochlearia officinalis\n",
      "694/1983 - Root‐Shoot Ratios on Native Forest Herbs and Zea Mays at Different Soil‐Moisture Levels\n",
      "695/1983 - Growth response of grasses to elevated CO<sub>2</sub>: a physiological plurispecific analysis\n",
      "696/1983 - Rapid Range Shifts of Species Associated with High Levels of Climate Warming\n",
      "697/1983 - Invasive Plants Have Scale-Dependent Effects on Diversity by Altering Species-Area Relationships\n",
      "698/1983 - Expansion of forest stands into tundra in the Noatak National Preserve, northwest Alaska\n",
      "699/1983 - Black Spruce Growth Forms as a Record of a Changing Winter Environment at Treeline, Quebec, Canada\n",
      "700/1983 - Change and stability in the altitude of the birch tree-limit in the southern Swedish Scandes 1915-1975\n",
      "701/1983 - Regional impacts of climate change in the Arctic and Antarctic\n",
      "702/1983 - Diminished Plant Richness and Abundance Below Lonicera maackii, an Invasive Shrub\n",
      "703/1983 - DISTRIBUTION OF THE NON-NATIVE SHRUB LONICERA MAACKII IN KRAMER WOODS, OHIO\n",
      "704/1983 - Biological invasions : a global perspective\n",
      "705/1983 - Threats to Imperiled Freshwater Fauna\n",
      "706/1983 - Changing Perspectives on Pearly Mussels, North America's Most Imperiled Animals\n",
      "707/1983 - Extinction vulnerability in marine populations\n",
      "708/1983 - Effects of Soil Resources on Plant Invasion and Community Structure in Californian Serpentine Grassland\n",
      "709/1983 - Effects of Different Resource Additions of Species Diversity in an Annual Plant Community\n",
      "710/1983 - A weak upward elevational shift in the distributions of breeding birds in the Italian Alps\n",
      "711/1983 - Do species’ traits predict recent shifts at expanding range edges?\n",
      "712/1983 - Climate, climate change and range boundaries\n",
      "713/1983 - The Structure and Dynamics of Geographic Ranges\n",
      "714/1983 - Climate effects on population fluctuations of the white-throated dipper Cinclus cinclus\n",
      "715/1983 - Reconstructing community relationships: the impact of sampling error, ordination approach, and gradient length\n",
      "716/1983 - Robustness of mutualistic networks under phenological change and habitat destruction\n",
      "717/1983 - Unifying sources and sinks in ecology and <scp>E</scp>arth sciences\n",
      "718/1983 - Tolerance of pollination networks to species extinctions\n",
      "719/1983 - The Distribution and Decline of British Bumble Bees (<i>Bombus</i>Latr.)\n",
      "720/1983 - D: Fragmented plant populations and their lost interactions\n",
      "721/1983 - Sources, Sinks and Sustainability\n",
      "722/1983 - Which species will succesfully track climate change? The influence of intraspecific competition and density dependent dispersal on range shifting dynamics\n",
      "723/1983 - Limits to the niche and range margins of alien species\n",
      "724/1983 - A dispersal-induced paradox: synchrony and stability in stochastic metapopulations\n",
      "725/1983 - Conspecific attraction and nest site selection in a nomadic species, the zebra finch\n",
      "726/1983 - Comparing the seasonal survival of resident and migratory oystercatchers: carry‐over effects of habitat quality and weather conditions\n",
      "727/1983 - Dispersal evolution during invasions\n",
      "728/1983 - The Role of Local Species Abundance in the Evolution of Pollinator Attraction in Flowering Plants\n",
      "729/1983 - Generalization in Pollination Systems, and Why it Matters\n",
      "730/1983 - Functional Diversity of Plant–Pollinator Interaction Webs Enhances the Persistence of Plant Communities\n",
      "731/1983 - Resilience and Stability of Ecological Systems\n",
      "732/1983 - Intraspecific variability and trait‐based community assembly\n",
      "733/1983 - Fine‐scale Microhabitat Heterogeneity in a French Guianan Forest\n",
      "734/1983 - The ecology of tropical forest tree seedlings\n",
      "735/1983 - Seasonal and long‐term soil water regime in West African tropical forest\n",
      "736/1983 - Patterns in specific leaf area and the structure of a temperate heath community\n",
      "737/1983 - Individuals and the Variation Needed for High Species Diversity in Forest Trees\n",
      "738/1983 - AIR-MASS CLIMATOLOGY OF ONTARIO NORTH OF LAKE HURON AND LAKE SUPERIOR BEFORE OUTBREAKS OF THE SPRUCE BUDWORM, <i>CHORISTONEURA FUMIFERANA</i> (CLEM.), AND THE FOREST TENT CATERPILLAR, <i>MALACOSOMA DISSTRIA</i> HBN. (LEPIDOPTERA: TORTRICIDAE; LASIOCAMPIDAE)\n",
      "739/1983 - PHYSICAL AND BIOLOGICAL INDICATORS OF THE DEVELOPMENT OF OUTBREAKS OF THE SPRUCE BUDWORM, <i>CHORISTONEURA FUMIFERANA</i> (CLEM.) (LEPIDOPTERA: TORTRICIDAE)\n",
      "740/1983 - Tolerance to herbivory, and not resistance, may explain differential success of invasive, naturalized, and native North American temperate vines\n",
      "741/1983 - A meta‐analysis of tradeoffs between plant tolerance and resistance to herbivores: combining the evidence from ecological and agricultural studies\n",
      "742/1983 - The Continuum of Plant Responses to Herbivory: The Influence of Plant Association, Nutrient Availability, and Timing\n",
      "743/1983 - Recovery of a <i>Betula pubescens</i> forest in northern Sweden after severe defoliation by <i>Epirrita autumnata</i>\n",
      "744/1983 - Interrelated Causes of Plant Invasion\n",
      "745/1983 - Interacting effects of herbivory and fertility on a synthesized plant community\n",
      "746/1983 - Toward a Causal Explanation of Plant Invasiveness: Seedling Growth and Life‐History Strategies of 29 Pine (<i>Pinus</i>) Species\n",
      "747/1983 - Evolution of antipredator behavior in an island lizard species,<i>Podarcis erhardii</i>(Reptilia: Lacertidae): The sum of all fears?\n",
      "748/1983 - Frequency of tail loss reflects variation in predation levels, predator efficiency, and the behaviour of three populations of brown anoles\n",
      "749/1983 - Blue Tails and Autotomy: Enhancement of Predation Avoidance in Juvenile Skinks\n",
      "750/1983 - Ecological studies of an epikarst community in Snežna jama na planini Arto – an ice cave in north central Slovenia\n",
      "751/1983 - The ecological causes of individual specialisation\n",
      "752/1983 - Field and experimental evidence that competition and ecological opportunity promote resource polymorphism\n",
      "753/1983 - Frequency-Dependent Natural Selection in the Handedness of Scale-Eating Cichlid Fish\n",
      "754/1983 - Population Density Fluctuations Change the Selection Gradient in Eurasian Perch\n",
      "755/1983 - Stabilization of Population Fluctuations due to Cannibalism Promotes Resource Polymorphism in Fish\n",
      "756/1983 - Individual-based Modeling and Ecology\n",
      "757/1983 - Generalist foraging of pollinators: diet expansion at high density\n",
      "758/1983 - Gut length plasticity in perch: into the bowels of resource polymorphisms\n",
      "759/1983 - Soil history as a primary control on plant invasion in abandoned agricultural fields\n",
      "760/1983 - Standard Soil Methods for Long-Term Ecological Research\n",
      "761/1983 - Finding Endemic Soil-Based Controls for Weed Growth<sup>1</sup>\n",
      "762/1983 - Seed Predation and the Coexistence of Tree Species in Tropical Forests\n",
      "763/1983 - Spacing Dynamics of a Tropical Rain Forest Tree: Evaluation of the Janzen-Connell Model\n",
      "764/1983 - Sexual recruitment in <i>Z</i> <i>ostera marina</i> : A patch to landscape-scale investigation\n",
      "765/1983 - Dispersal and population structure of Neotrypaea californiensis\n",
      "766/1983 - Seed dispersal and seedling establishment in grass populations\n",
      "767/1983 - Effect of “brown tide” shading on eelgrass (Zostera marina L.) distributions\n",
      "768/1983 - Fitness-consequences of geitonogamous selfing in a clonal marine angiosperm (<i>Zostera marina</i>)\n",
      "769/1983 - EELGRASS WASTING DISEASE: CAUSE AND RECURRENCE OF A MARINE EPIDEMIC\n",
      "770/1983 - Silica in grasses as a defence against insect herbivores: contrasting effects on folivores and a phloem feeder\n",
      "771/1983 - Effects of increased soil nitrogen on the dominance of alien annual plants in the Mojave Desert\n",
      "772/1983 - Silicon Uptake and Distribution in Agropyron smithii as Related to Grazing History and Defoliation\n",
      "773/1983 - Resource deprival as an anti-herbivore strategy in plants, with particular reference to aphids\n",
      "774/1983 - Variation in pea aphid population development in three different habitats\n",
      "775/1983 - CHEMORECEPTION OF HOST PLANT CHEMICALS BY OVIPOSITING FEMALES OF <i>DELIA (HYLEMYA) BRASSICAE</i>\n",
      "776/1983 - Yield, Stand, And Volume Tables For Even-Aged Upland Oak Forests\n",
      "777/1983 - Reproduction of upland hardwood forests in the central states\n",
      "778/1983 - Seed Biology and Technology of &lt;i&gt;Quercus&lt;/i&gt;\n",
      "779/1983 - THE JACK OF ALL TRADES IS MASTER OF NONE: A PATHOGEN'S ABILITY TO INFECT A GREATER NUMBER OF HOST GENOTYPES COMES AT A COST OF DELAYED REPRODUCTION\n",
      "780/1983 - Changes in scab susceptibility of apple leaves as influenced by age.\n",
      "781/1983 - Sexual reproduction as an adaptation to resist parasites (a review).\n",
      "782/1983 - Ecological drift and local exposures drive enteric bacterial community differences within species of Galápagos iguanas\n",
      "783/1983 - Neutral assembly of bacterial communities\n",
      "784/1983 - Context dependent effects of plant species and functional group loss on vegetation invasibility across an island area gradient\n",
      "785/1983 - Gnotobiotic zebrafish reveal evolutionarily conserved responses to the gut microbiota\n",
      "786/1983 - Biodiversity as both a cause and consequence of resource availability: a study of reciprocal causality in a predator–prey system\n",
      "787/1983 - Polyphagy complicates conservation biological control that targets generalist predators\n",
      "788/1983 - Prey selection by linyphiid spiders: molecular tracking of the effects of alternative prey on rates of aphid consumption in the field\n",
      "789/1983 - GENERALIST PREDATORS DISRUPT BIOLOGICAL CONTROL BY A SPECIALIST PARASITOID\n",
      "790/1983 - INTERACTIONS BETWEEN SPECIALIST AND GENERALIST NATURAL ENEMIES: PARASITOIDS, PREDATORS, AND PEA APHID BIOCONTROL\n",
      "791/1983 - Testing the Holy Grail framework: using functional traits to predict ecosystem change\n",
      "792/1983 - Evolutionary trade‐offs among decomposers determine responses to nitrogen enrichment\n",
      "793/1983 - An evaluation of three quick methods commonly used to assess sward height in ecology\n",
      "794/1983 - Floral adaptations to nocturnal moth pollination in <i>Diplomorpha</i> (Thymelaeaceae)\n",
      "795/1983 - Plant Reproductive Ecology\n",
      "796/1983 - REPRODUCTIVE BIOLOGY OF TROPICAL LOWLAND RAIN FOREST TREES. II. POLLINATION SYSTEMS\n",
      "797/1983 - Bats as pollinators: foraging energetics and floral adaptations\n",
      "798/1983 - Relative importance of burning, mowing and species translocation in the restoration of a former boreal hayfield: responses of plant diversity and the microbial community\n",
      "799/1983 - Plant diversity partitioning in grazed Mediterranean grassland at multiple spatial and temporal scales\n",
      "800/1983 - Assessing grassland restoration success: relative roles of seed additions and native ungulate activities\n",
      "801/1983 - Additive partitioning of diversity across hierarchical spatial scales in a forested landscape\n",
      "802/1983 - Local and landscape effects of organic farming on butterfly species richness and abundance\n",
      "803/1983 - Facilitating grassland diversification using the hemiparasitic plant <i>Rhinanthus minor</i>\n",
      "804/1983 - Soil seed banks and the effects of meadow management on vegetation change in a 10‐year meadow field trial\n",
      "805/1983 - Restoration of species‐rich grassland on arable land: assessing the limiting processes using a multi‐site experiment\n",
      "806/1983 - Impact of Clipping on Root Systems of 3 Grasses Species in Tunisia\n",
      "807/1983 - Testing the Threat‐Sensitive Hypothesis with Predator Familiarity and Dietary Specificity\n",
      "808/1983 - Behavioral Interactions Among Four Species of the Salamander Genus Desmognathus\n",
      "809/1983 - Dear Enemy Recognition and the Costs of Aggression between Salamanders\n",
      "810/1983 - Rock‐Face Populations of the Mountain Salamander, Desmognathus Ochrophaeus, In North Carolina\n",
      "811/1983 - Empirical Evaluation of the Hypothesis of Food Competition in Salamanders of the Genus Plethodon\n",
      "812/1983 - Predator-Induced Phenotypical Change in Body Morphology in Crucian Carp\n",
      "813/1983 - Empirical Relationships between Bacterial Abundance and Chlorophyll Concentration in Fresh and Marine Waters\n",
      "814/1983 - Wildlife of the Tibetan steppe\n",
      "815/1983 - Designing occupancy studies: general advice and allocating survey effort\n",
      "816/1983 - Assessing the accuracy of species distribution models: prevalence, kappa and the true skill statistic (TSS)\n",
      "817/1983 - Vegetation patterns and abundances of amphibians and small mammals along small streams in a Northwestern California watershed\n",
      "818/1983 - Range Shifts and Adaptive Responses to Quaternary Climate Change\n",
      "819/1983 - Global patterns of environmental synchrony and the Moran effect\n",
      "820/1983 - Effects of stream size on bird community structure in coastal temperate forests of the Pacific Northwest, U.S.A.\n",
      "821/1983 - The continuing challenges of testing species distribution models\n",
      "822/1983 - REGIONAL DYNAMICS OF WETLAND‐BREEDING FROGS AND TOADS: TURNOVER AND SYNCHRONY\n",
      "823/1983 - The statistical analysis of the Canadian Lynx cycle.\n",
      "824/1983 - Ontogenetic Dietary Shifts in a Predatory Freshwater Fish Species: The Brown Trout as an Example of a Dynamic Fish Species\n",
      "825/1983 - Density‐dependent diet composition of juvenile Atlantic salmon (<i>Salmo salar</i>)\n",
      "826/1983 - Nest distribution shaping within‐stream variation in Atlantic salmon juvenile abundance and competition over small spatial scales\n",
      "827/1983 - Ecological Effects of Winter Conditions on Trout and Trout Foods in Convict Creek, California, 1951\n",
      "828/1983 - Ontogenetic and spatial variations in brown trout habitat selection\n",
      "829/1983 - Numerical changes in stream-resident brown trout (<i>Salmo trutta</i>): uncovering the roles of density-dependent and density-independent factors across space and time\n",
      "830/1983 - Evaluation of the suitability of samples collected in vivo for investigations of juvenile sturgeon stomach contents\n",
      "831/1983 - Rapid food web recovery in response to removal of an introduced apex predator\n",
      "832/1983 - Profitable stream positions for salmonids: relating specific growth rate to net energy gain\n",
      "833/1983 - Territory Size as a Predictor of the Upper Limit to Population Density of Juvenile Salmonids in Streams\n",
      "834/1983 - Genome‐wide association genetics of an adaptive trait in lodgepole pine\n",
      "835/1983 - Recent changes in burbot growth in Lake Erie\n",
      "836/1983 - Forest Fire History in the Northern Rockies\n",
      "837/1983 - The discovery–dominance trade‐off is the exception, rather than the rule\n",
      "838/1983 - Artificial Diet for Rearing Various Species of Ants\n",
      "839/1983 - Condition‐specific competition between invasive Argentine ants and Australian <i>Iridomyrmex</i>\n",
      "840/1983 - Australian Ants\n",
      "841/1983 - Diversity by Default\n",
      "842/1983 - Herbivores and the Number of Tree Species in Tropical Forests\n",
      "843/1983 - Plant strategies and the dynamics and structure of plant communities\n",
      "844/1983 - Tree species richness promotes productivity in temperate forests through strong complementarity between species\n",
      "845/1983 - The determinants of leaf turgor loss point and prediction of drought tolerance of species and biomes: a global meta‐analysis\n",
      "846/1983 - Calculating competition in thinned northern hardwoods.\n",
      "847/1983 - Biodiversity effects increase linearly with biotope space\n",
      "848/1983 - Silvical requirements for natural birch regeneration\n",
      "849/1983 - Microsatellite variation and structure of 28 populations of the common wetland plant, <i>Lychnis flos‐cuculi</i> L., in a fragmented landscape\n",
      "850/1983 - Forest dynamics: an ecological model\n",
      "851/1983 - Optimal clutch size of the gall wasp &lt;i&gt;Diplolepis rosae&lt;/i&gt; (Hymenoptera: Cynipidae)\n",
      "852/1983 - Lack's Clutch Size Hypothesis: An Examination of the Evidence Using Meta‐Analysis\n",
      "853/1983 - Forest and forest succession in Hong Kong, China\n",
      "854/1983 - Foundations of ecological resilience\n",
      "855/1983 - Using plant functional traits to understand the landscape distribution of multiple ecosystem services\n",
      "856/1983 - Remote sensing of plant functional types\n",
      "857/1983 - Functional identity is more important than diversity in influencing ecosystem processes in a temperate native grassland\n",
      "858/1983 - Body mass constraints on feeding rates determine the consequences of predator loss\n",
      "859/1983 - The role of body mass in diet contiguity and food-web structure\n",
      "860/1983 - Warming effects on consumption and intraspecific interference competition depend on predator metabolism\n",
      "861/1983 - Serengeti: Dynamics of an Ecosystem.\n",
      "862/1983 - Serengeti II: Dynamics, Management and Conservation of an Ecosystem.\n",
      "863/1983 - The effect of temperature on the activity of Carabidae (Coleoptera) in a fallow field\n",
      "864/1983 - Effects of intra‐ and interspecific interactions on species responses to environmental change\n",
      "865/1983 - Multiple stable reference points in oyster populations: biological relationships for the eastern oyster (Crassostrea virginica)in Delaware Bay\n",
      "866/1983 - DISAPPEARANCE OF THE NATURAL EMERGENT 3-DIMENSIONAL OYSTER REEF SYSTEM OF THE JAMES RIVER, VIRGINIA, 1871-1948\n",
      "867/1983 - Interaction Between Circulation of the Estuary of the James River and Transport of Oyster Larvae\n",
      "868/1983 - Decline of the Chesapeake Bay oyster population: a century of habitat destruction and overfishing\n",
      "869/1983 - Multiple Stable Reference Points in Oyster Populations: Implications for Reference Point-Based Management\n",
      "870/1983 - Interaction strength combinations and the overfishing of a marine food web\n",
      "871/1983 - Morphological divergence and origin of sympatric populations of European whitefish (<i>Coregonus lavaretus</i>L.) in Lake Femund, Norway\n",
      "872/1983 - The behavior and ecology of Pacific salmon and trout\n",
      "873/1983 - Eco-evolutionary interactions between predators and prey: can predator-induced changes to prey communities feed back to shape predator foraging traits?\n",
      "874/1983 - Survival of Fall Armyworm, Spodoptera frugiperda, (Lepidoptera: Noctuidae) Exposed to Cold Temperatures\n",
      "875/1983 - Herbivore‐induced plant volatiles induce an indirect defence in neighbouring plants\n",
      "876/1983 - Seasonal Populations of Armyworms and Loopers at Hastings, Florida\n",
      "877/1983 - Optimal Foraging by Parasitoid Wasps\n",
      "878/1983 - Niche Construction: The Neglected Process in Evolution\n",
      "879/1983 - Effect of Turbidity on the Temporal and Spatial Utilization of the Inner Bay of Fundy by American Shad (<i>Alosa sapidissima</i>) (Pisces: Clupeidae) and its Relationship to Local Fisheries\n",
      "880/1983 - Filling a gap:dynamics of space occupancy on a mussel-dominated subtropical rocky shore\n",
      "881/1983 - Competition, Disturbance, and Community Organization: The Provision and Subsequent Utilization of Space in a Rocky Intertidal Community\n",
      "882/1983 - Islands of Mytilus as a habitat for small intertidal animals: effect of island size on community structure\n",
      "883/1983 - Comparative Species Richness in Fluctuating and Constant Environments: Coral-Associated Decapod Crustaceans\n",
      "884/1983 - THE BREEDING BIOLOGY OF THE SWIFT APUS APUS\n",
      "885/1983 - Dispersal as a limiting factor in the colonization of restored mountain streams by plants and macroinvertebrates\n",
      "886/1983 - Incorporating asymmetric connectivity into spatial decision making for conservation\n",
      "887/1983 - Habitat fragmentation reduces grassland connectivity for both short‐distance and long‐distance wind‐dispersed forbs\n",
      "888/1983 - Propagule deposition along river margins: linking hydrology and ecology\n",
      "889/1983 - Effect of brood size manipulations on parents and offspring in the Barn owl, Tyto alba\n",
      "890/1983 - The effects of castration and androgen replacement on song, courtship, and aggression in zebra finches (<i>Poephila guttata</i>)\n",
      "891/1983 - Lethal interactions among vertebrate top predators: a review of concepts, assumptions and terminology\n",
      "892/1983 - The Structure of Lizard Communities\n",
      "893/1983 - Nonsynchronous Spatial Overlap of Lizards in Patchy Habitats\n",
      "894/1983 - Niche Width and Niche Overlap: A Method for Measuring Them\n",
      "895/1983 - Bird Censuses from Desert Localities in Western Australia\n",
      "896/1983 - An Analysis of the Avifauna of the Southern Kalahari Gemsbok National Park\n",
      "897/1983 - Intraguild predation and interspecific co-existence between predatory endotherms\n",
      "898/1983 - Raptors of the Hanford Site and nearby areas of southcentral Washington\n",
      "899/1983 - Intraguild predation in raptor assemblages: a review\n",
      "900/1983 - Movements and associated observations of bait-attracted White Sharks, Carcharodon carcharias: A preliminary report\n",
      "901/1983 - Preliminary observations on the biology of the White Shark, Carcharodon carcharias, in South Australian waters\n",
      "902/1983 - Nest site selection by Holarctic waterfowl: a multi-level review\n",
      "903/1983 - A walk-in trap for nesting ducks\n",
      "904/1983 - The Effect of Variable Spring Water Conditions on Mallard Reproduction\n",
      "905/1983 - Generalized hydrology of prairie potholes on the Coteau du Missouri, North Dakota\n",
      "906/1983 - Evolution of old and new world migration systems\n",
      "907/1983 - Greater Scaup, &lt;em&gt;Aythya marila&lt;/em&gt;, Nest Site Characteristics on Grassy Island, New Brunswick\n",
      "908/1983 - Spatial and Temporal Scales in Habitat Selection\n",
      "909/1983 - The Comparison of Usage and Availability Measurements for Evaluating Resource Preference\n",
      "910/1983 - Estimating Nest Success: The Mayfield Method and an Alternative\n",
      "911/1983 - Ecology of seedling regeneration.\n",
      "912/1983 - The shared influence of phylogeny and ecology on the reproductive patterns of Myrteae (Myrtaceae)\n",
      "913/1983 - The seasons in a tropical rain-forest (New Hebrides).-Part 3. Fruit-bats (Pteropidae).\n",
      "914/1983 - Phenology of a tropical rain forest in Malaya\n",
      "915/1983 - ECOLOGICAL OPPORTUNITY AND THE RATE OF MORPHOLOGICAL EVOLUTION IN THE DIVERSIFICATION OF GREATER ANTILLEAN ANOLES\n",
      "916/1983 - Evolution on islands\n",
      "917/1983 - Ecological opportunity and the origin of adaptive radiations\n",
      "918/1983 - Patterns of Leaf Development and Herbivory in a Tropical Understory Community\n",
      "919/1983 - Forest environments in tropical life zones: a pilot study.\n",
      "920/1983 - Movements of grey-headed flying foxes (Pteropus poliocephalus) to and from colony site on the central coast of New South Wales\n",
      "921/1983 - Gregarious flowering in a lowland tropical rainforest: A possible response to disturbance by Cyclone Winifred\n",
      "922/1983 - A Theoretical Framework for Intraguild Predation\n",
      "923/1983 - Positive Indirect Effects Between Prey Species that Share Predators\n",
      "924/1983 - POPULATION DYNAMICS OF <i>LEPTINOTARSA DECEMLINEATA</i> (SAY) IN EASTERN ONTARIO: III. MAJOR POPULATION PROCESSES\n",
      "925/1983 - Role of Trichodesmium spp. in the productivity of the subtropical North Pacific Ocean\n",
      "926/1983 - Amino Acid Cycling in Colonies of the Planktonic Marine Cyanobacterium <i>Trichodesmium thiebautii</i>\n",
      "927/1983 - Population Dynamics of Chesapeake Bay Virioplankton: Total-Community Analysis by Pulsed-Field Gel Electrophoresis\n",
      "928/1983 - Evidence by electron micrographs for a high incidence of bacteriophage particles in the waters of Yaquina Bay, oregon: ecological and taxonomical implications\n",
      "929/1983 - Distribution of Viruses and Dissolved DNA along a Coastal Trophic Gradient in the Northern Adriatic Sea\n",
      "930/1983 - Viruses as Partners in Spring Bloom Microbial Trophodynamics\n",
      "931/1983 - Effect of Food Quantity on Fecundity of Rainbow Trout, <i>Salmo gairdneri</i>\n",
      "932/1983 - An Experimental Analysis of Reproductive Effort and Cost in the Japanese Medaka, Oryzias Latipes\n",
      "933/1983 - Effect of rising atmospheric carbon dioxide on the marine nitrogen fixer <i>Trichodesmium</i>\n",
      "934/1983 - Modelling regional responses by marine pelagic ecosystems to global climate change\n",
      "935/1983 - Utilization of Dissolved Nitrogen by Heterotrophic Bacterioplankton: a Comparison of Three Ecosystems\n",
      "936/1983 - Bacterioplankton Growth Yield: Seasonal Variations and Coupling to Substrate Lability and β-Glucosidase Activity\n",
      "937/1983 - Natural enemies of cabbage looper on cotton in the San Joaquin Valley\n",
      "938/1983 - Does Long-Term Elevation of CO2 Concentration Increase Photosynthesis in Forest Floor Vegetation? (Indiana Strawberry in a Maryland Forest)\n",
      "939/1983 - Adult Susceptibility of Nebraska Western Corn Rootworm (Coleoptera: Chrysomelidae) Populations to Selected Insecticides\n",
      "940/1983 - The Ecological Risks and Benefits of Genetically Engineered Plants\n",
      "941/1983 - Thrips (Thysanoptera) Species Associated with Developing Citrus Flowers in Florida and A Key to Adult Terebrantian Females\n",
      "942/1983 - In situ video and fluorescence analysis (VFA) of marine particles: applications to phytoplankton ecological studies.\n",
      "943/1983 - A submersible imaging‐in‐flow instrument to analyze nano‐and microplankton: Imaging FlowCytobot\n",
      "944/1983 - A winch and cable for the Autonomous Vertically Profiling Plankton Observatory\n",
      "945/1983 - An imaging-in-flow system for automated analysis of marine microplankton\n",
      "946/1983 - Snails can survive passage through a bird’s digestive system\n",
      "947/1983 - How the ladybird got its spots: effects of resource limitation on the honesty of aposematic signals\n",
      "948/1983 - Interspecific interactions between the native and introduced White‐eyes in the Bonin Islands\n",
      "949/1983 - Interaction and Coevolution\n",
      "950/1983 - Ecology of Protozoa: The Biology of Free-living Phagotrophic Protists\n",
      "951/1983 - Interactions between Planktonic Microalgae and Protozoan Grazers<sup>1</sup>\n",
      "952/1983 - Plant ecophysiological diversity\n",
      "953/1983 - Vegetation studies on Secretary Island, Fiordland\n",
      "954/1983 - Vegetation studies on Secretary Island, Fiordland\n",
      "955/1983 - Spatial Variation in Nitrogen Availability in Three Successional Plant Communities\n",
      "956/1983 - How fundamental plant functional trait relationships scale‐up to trade‐offs and synergies in ecosystem services\n",
      "957/1983 - Species- and community-level patterns in fine root traits along a 120 000-year soil chronosequence in temperate rain forest\n",
      "958/1983 - The ecological significance of plasticity.\n",
      "959/1983 - A probabilistic model for analysing species co‐occurrence\n",
      "960/1983 - Ecological factors determining the distribution and assemblages of the aquatic Hemiptera (Gerromorpha &amp; Nepomorpha) in the Segura River basin (Spain)\n",
      "961/1983 - Benthic invertebrate distributions in the San Joaquin River, California, in relation to physical and chemical factors\n",
      "962/1983 - Estudio faunístico y ecológico de los coleópteros y heterópteros acuáticos de las lagunas y humedales de Albacete (Lagunas de Ruidera, Salinas de Pinilla, Laguna del Saladar, Laguna del Salobralejo, Lagunas de Corral Rubio, Fuente de Isso y Fuente de Agua Ramos)\n",
      "963/1983 - Coleópteros acuáticos del sur de Córdoba (España) (Haliplidae, Gyrinidae, Noteridae, Dytiscidae, Hydraenidae, Hydrochidae, Helophoridae, Hydrophilidae, Dryopidae y Elmidae)\n",
      "964/1983 - Pairwise co-existence of Bismarck and Solomon landbird species\n",
      "965/1983 - Assembly rules as general constraints on community composition\n",
      "966/1983 - Binary matrices and checkerboard distributions of birds in the Bismarck Archipelago\n",
      "967/1983 - Neotropical Land-Bridge Avifaunas: New Approaches to Null Hypotheses in Biogeography\n",
      "968/1983 - Colonization of Exploded Volcanic Islands by Birds: The Supertramp Strategy\n",
      "969/1983 - The multidimensionality of the niche reveals functional diversity changes in benthic marine biotas across geological time\n",
      "970/1983 - Predator control of ecosystem nutrient dynamics\n",
      "971/1983 - Taxonomic and functional homogenization of an endemic desert fish fauna\n",
      "972/1983 - Partitioning global patterns of freshwater fish beta diversity reveals contrasting signatures of past climate changes\n",
      "973/1983 - Determinants of fine‐scale homogenization and differentiation of native freshwater fish faunas in a Mediterranean Basin: implications for conservation\n",
      "974/1983 - Sex-Biased Hatching Order and Adaptive Population Divergence in a Passerine Bird\n",
      "975/1983 - Yolk is a source of maternal testosterone for developing birds.\n",
      "976/1983 - Constraints on the Coevolution of Bacteria and Virulent Phage: A Model, Some Experiments, and Predictions for Natural Communities\n",
      "977/1983 - Functional and phylogenetic diversity and assemblage structure of frugivorous birds along an elevational gradient in the tropical Andes\n",
      "978/1983 - Beyond taxonomic diversity patterns: how do α, β and γ components of bird functional and phylogenetic diversity respond to environmental gradients across France?\n",
      "979/1983 - Are communities saturated? On the relationship between α, β and γ diversity\n",
      "980/1983 - Spatial mismatch and congruence between taxonomic, phylogenetic and functional diversity: the need for integrative conservation strategies in a changing world\n",
      "981/1983 - Phylogenetic beta diversity: linking ecological and evolutionary processes across space in time\n",
      "982/1983 - The Short-Tailed Fruit Bat: A Study in Plant-Animal Interactions\n",
      "983/1983 - On Intraspecific Competition for Avian Dispersers in Tropical Trees\n",
      "984/1983 - Vergleich von Lebenszyklus und Populationsstruktur bei Höhenvikarianten der Gattung Epilobium\n",
      "985/1983 - Bimodal colour pattern of individual<i>Pinus halepensis</i> Mill. seeds: a new type of crypsis\n",
      "986/1983 - Caterpillars: ecological and evolutionary constraints on foraging\n",
      "987/1983 - Supervivencia de las plántulas de abies pinsapo boiss. En su hábitat natural\n",
      "988/1983 - Effects of seed provenance and mycorrhizal fungi on early seedling growth in <i>Picea</i><i>mariana</i>\n",
      "989/1983 - Early maternal effects and antibacterial immune factors in the eggs, nestlings and adults of the barn swallow\n",
      "990/1983 - Size matters: competition between male and female great tit offspring\n",
      "991/1983 - Immunoglobulin plasma concentration in relation to egg laying and mate ornamentation of female barn swallows (<i>Hirundo rustica</i>)\n",
      "992/1983 - Male Attractiveness and Differential Testosterone Investment in Zebra Finch Eggs\n",
      "993/1983 - Chapter 19 Elk Group Size and Wolf Predation\n",
      "994/1983 - Conservation of caribou (<i>Rangifer tarandus</i>) in Canada: an uncertain future<sup>1</sup>This review is part of the virtual symposium “Flagship Species – Flagship Problems” that deals with ecology, biodiversity and management issues, and climate impacts on species at risk and of Canadian importance, including the polar bear (<i>Ursus maritimus</i>), Atlantic cod (<i>Gadus morhua</i>), Piping Plover (<i>Charadrius melodus</i>), and caribou (<i>Rangifer tarandus</i>).\n",
      "995/1983 - Biased adult sex ratio can affect fecundity in primiparous moose <i>Alces alces</i>\n",
      "996/1983 - Minimum viable populations: coping with uncertainty\n",
      "997/1983 - Ecology of an African rain forest: logging in Kibale and the conflict between conservation and exploitation\n",
      "998/1983 - The Red Colobus Monkeys\n",
      "999/1983 - Evolutionary consequences of gender plasticity in genetically dimorphic breeding systems\n",
      "1000/1983 - Selection on floral characters in natural Spanish populations of <i>Silene latifolia</i>\n",
      "1001/1983 - Mixing of propagules from discrete sources at long distance: comparing a dispersal tail to an exponential\n",
      "1002/1983 - Propagule dispersal and the scales of marine community process\n",
      "1003/1983 - Ecology And Evolution Of Plant Reproduction\n",
      "1004/1983 - The multiple faces of indirect defences and their agents of natural selection\n",
      "1005/1983 - VOLATILE PLANT CHEMICALS AND THEIR EFFECT ON HOST PLANT FINDING BY THE CABBAGE ROOT FLY <i>(DELIA BRASSICAE)</i>\n",
      "1006/1983 - Constitutive and induced subterranean plant volatiles attract both entomopathogenic and plant parasitic nematodes\n",
      "1007/1983 - Invertebrate biodiversity (ants, bees, and others) associated with stem domatia of the Indian myrmecophyte Humboldtia brunonis Wallich (Magnoliophyta: Fabaceae).\n",
      "1008/1983 - Co(i)‐ordinating defenses: NaCOI1 mediates herbivore‐ induced resistance in <i>Nicotiana attenuata</i> and reveals the role of herbivore movement in avoiding defenses\n",
      "1009/1983 - Inhibition and Acclimation of Photosynthesis to Heat Stress Is Closely Correlated with Activation of Ribulose-1,5-Bisphosphate Carboxylase/Oxygenase\n",
      "1010/1983 - Condition‐dependent, phenotype‐dependent and genetic‐dependent factors in the natal dispersal of a solitary rodent\n",
      "1011/1983 - Abrupt spatial and numerical responses of overabundant foxes to a reduction in anthropogenic resources\n",
      "1012/1983 - Integrating across Life‐History Stages: Consequences of Natal Habitat Effects on Dispersal\n",
      "1013/1983 - Scaling of Natal Dispersal Distances in Terrestrial Birds and Mammals\n",
      "1014/1983 - Conserving Biodiversity in Managed Forests\n",
      "1015/1983 - MIGRATION AND DISPERSAL OF LAUGHING GULLS IN THE UNITED STATES\n",
      "1016/1983 - Natal Dispersal in the Nuthatch\n",
      "1017/1983 - Impact of exotic generalist predators on the native fauna of Australia\n",
      "1018/1983 - Innate predator recognition in newly-hatched Atlantic salmon\n",
      "1019/1983 - Biology of gliding mammals\n",
      "1020/1983 - Home ranges and habitat use in the declining flying squirrel <i>Pteromys volans</i> in managed forests\n",
      "1021/1983 - Wood : decay, pests, and protection\n",
      "1022/1983 - Lipid Analysis in Microbial Ecology\n",
      "1023/1983 - Fine-scale habitat use and behaviour of Arctic charr (Salvelinus alpinus) morphs in a remote Arctic lake\n",
      "1024/1983 - Epigenetic regulation of trophic morphology through feeding behaviour in Arctic charr, Salvelinus alpinus\n",
      "1025/1983 - Parallel divergence of sympatric genetic and body size forms of Arctic charr, Salvelinus alpinus, from two Scottish lakes\n",
      "1026/1983 - Morphometric analysis of two ecologically distinct forms of Arctic charr, <i>Salvelinus alpinus</i> (L.), in Loch Rannoch, Scotland\n",
      "1027/1983 - Evolution: Essays in Honour of John Maynard Smith\n",
      "1028/1983 - Ecological polymorphism in Arctic charr\n",
      "1029/1983 - Ecological speciation in postglacial fishes\n",
      "1030/1983 - Habitat and Food Segregation of Dwarf and Normal Arctic Charr (<i>Salvelinus alpinus</i>) from Vangsvatnet Lake, Western Norway\n",
      "1031/1983 - Evidence for environmental modulation of gametocytogenesis in Plasmodium falciparum in continuous culture.\n",
      "1032/1983 - Effects of diversity and identity of the neighbouring plant community on the abundance of arthropods on individual ragwort (<i>Jacobaea vulgaris</i>) plants\n",
      "1033/1983 - Macronutrient status of tall and short forms of Spartina alterniflora in a South Carolina salt marsh\n",
      "1034/1983 - Glucosinolates and Herbivory by Specialists (Coleoptera: Chrysomelidae, Lepidoptera: Plutellidae): Consequences of Concentration and Induced Resistance\n",
      "1035/1983 - ANTAGONISTIC INTERACTIONS BETWEEN PLANT COMPETITION AND INSECT HERBIVORY\n",
      "1036/1983 - Patterns in Population Change and the Organization of the Insect Community Associated with Goldenrod\n",
      "1037/1983 - Joint Effects of Competitors and Herbivores on Growth and Reproduction in Aristolochia reticulata\n",
      "1038/1983 - Interactions between Invertebrate Herbivores and Plant Competition\n",
      "1039/1983 - Linking Statistical and Ecological Theory: Hubbell's Unified Neutral Theory of Biodiversity as a Hierarchical Dirichlet Process\n",
      "1040/1983 - Niche and neutral models predict asymptotically equivalent species abundance distributions in high-diversity ecological communities\n",
      "1041/1983 - mvabund– an R package for model‐based analysis of multivariate abundance data\n",
      "1042/1983 - Do not log‐transform count data\n",
      "1043/1983 - Functional diversity indices reveal the impacts of land use intensification on plant community assembly\n",
      "1044/1983 - The diversity and conservation of plant reproductive and dispersal functional traits in human‐dominated tropical landscapes\n",
      "1045/1983 - The Forgotten Pollinators\n",
      "1046/1983 - Community-wide distribution of predator–prey interaction strength in kelp forests\n",
      "1047/1983 - Nontrophic Interactions, Biodiversity, and Ecosystem Functioning: An Interaction Web Model\n",
      "1048/1983 - Constant Connectance in Community Food Webs\n",
      "1049/1983 - Stability in Real Food Webs: Weak Links in Long Loops\n",
      "1050/1983 - Predicting Direct and Indirect Effects: An Integrated Approach Using Experiments and Path Analysis\n",
      "1051/1983 - Comments on Fluctuations of Animal Populations and Measures of Community Stability\n",
      "1052/1983 - Approaches for testing herbivore effects on plant population dynamics\n",
      "1053/1983 - Growth, Reproduction and Chemical Composition of Horsenettle (<i>Solanum carolinense</i>)\n",
      "1054/1983 - Movement of 2,4-Dichlorophenoxyacetic Acid Stimulus and Its Relation to the Translocation of Organic Food Materials in Plants\n",
      "1055/1983 - Predicting the impacts of harvesting using structured population models: the importance of density‐dependence and timing of harvest for a tropical palm tree\n",
      "1056/1983 - Experiments in Ecology: Their Logical Design and Interpretation Using Analysis of Variance\n",
      "1057/1983 - The phosphorus‐chlorophyll relationship in lakes1,2\n",
      "1058/1983 - Significance of predator influence on composition of Bosmina spp. populations1\n",
      "1059/1983 - Predation and Risk in Foraging Minnows: Balancing Conflicting Demands\n",
      "1060/1983 - Establishment of Invertebrate Communities on Log Substrates in the Kaskaskia River, Illinois\n",
      "1061/1983 - Aquatic Insect Communities of a Small Stream on Mont St. Hilaire, Quebec\n",
      "1062/1983 - The influence of behavioral differences among tent-caterpillar larvae on predation by a pentatomid bug\n",
      "1063/1983 - Is Predaceous Stonefly Behavior Affected by Competition?\n",
      "1064/1983 - Biotic and Abiotic Controls in River and Stream Communities\n",
      "1065/1983 - Colony dynamics of the green tree ant (Oecophylla smaragdina Fab.) in a seasonal tropical climate\n",
      "1066/1983 - Low intraspecific aggression level in the polydomous and facultative polygynous ant<i>Ectatomma tuberculatum</i>\n",
      "1067/1983 - Size-selective grazing on bacteria by natural assemblages of estuarine flagellates and ciliates\n",
      "1068/1983 - Differential Rates of Digestion of Bacteria by Freshwater and Marine Phagotrophic Protozoa\n",
      "1069/1983 - Misprescription and misuse of one‐tailed tests\n",
      "1070/1983 - EXPERIMENTS WITH FRESHWATER INVERTEBRATE ZOOPLANKTIVORES: QUALITY OF STATISTICAL ANALYSES\n",
      "1071/1983 - Positive effect of predators on prey growth rate through induced modifications of prey behaviour\n",
      "1072/1983 - Landscape of fear influences the relative importance of consumptive and nonconsumptive predator effects\n",
      "1073/1983 - On Defining Assembly Space: A Reply to Grover and Lawton\n",
      "1074/1983 - The ecology of rocky shores\n",
      "1075/1983 - COMPLEMENTARY FEEDING NICHES SUSTAINED BY SIZE‐SELECTIVE PREDATION1\n",
      "1076/1983 - Adaptive change in plankton morphology in response to size‐selective predation: A new hypothesis of cyclomorphosis\n",
      "1077/1983 - Effects of Predators on Prey Growth Rate: Relative Contributions of Thinning and Reduced Activity\n",
      "1078/1983 - INTERACTIONS BETWEEN EPIPHYTES, MACROPHYTES AND FRESHWATER SNAILS: A REVIEW\n",
      "1079/1983 - Predation, Competition, and the Composition of Larval Anuran Guilds\n",
      "1080/1983 - A Predator‐Prey System in the Marine Intertidal Region. I. Balanus glandula and Several Predatory Species of Thais\n",
      "1081/1983 - The Biology of<i>Purpura Lapillus</i>. I. Shell Variation in Relation to Environment\n",
      "1082/1983 - Herbivore and predator diversity interactively affect ecosystem properties in an experimental marine community\n",
      "1083/1983 - Cascading effects from predator removal depend on resource availability in a benthic food web\n",
      "1084/1983 - Experimental Aquatic Food Webs: Interactions between Two Predators and Two Prey\n",
      "1085/1983 - Food habits and dietary overlap of perlid stoneflies (Plecoptera) in a tributary of Lake Ontario\n",
      "1086/1983 - On the Prevalence and Relative Importance of Interspecific Competition: Evidence from Field Experiments\n",
      "1087/1983 - Ideal Free Habitat Selection and Consumer-Resource Dynamics\n",
      "1088/1983 - Marine community ecology\n",
      "1089/1983 - INGESTION, ASSIMILATION, SURVIVAL, AND REPRODUCTION BY DAPHNIA PULEX FED SEVEN SPECIES OF BLUE‐GREEN ALGAE1,2\n",
      "1090/1983 - The ecology of marine microbenthos IV. Structure and function of the benthic ecosystem, its chemical and physical factors and the microfauna commuities with special reference to the ciliated protozoa\n",
      "1091/1983 - A farewell to Bonferroni: the problems of low statistical power and publication bias\n",
      "1092/1983 - Terrestrial Trophic Cascades: How Much Do They Trickle?\n",
      "1093/1983 - Fishing Down Marine Food Webs\n",
      "1094/1983 - STUDIES ON THE PHYSIOLOGIC SPECIALIZATION OF YELLOW RUST (PUCCINIA GLUMARUM (SCHM.) ERIKSS. &amp; HENN.) IN GREAT BRITAIN\n",
      "1095/1983 - The Tungara Frog: A Study in Sexual Selection and Communication\n",
      "1096/1983 - Territoriality and Male Mating Success in the Green Frog (Rana Clamitans)\n",
      "1097/1983 - Territorial Behavior among Puku in Zambia\n",
      "1098/1983 - Intransitive competition is widespread in plant communities and maintains their species richness\n",
      "1099/1983 - Mechanisms contributing to stability in ecosystem function depend on the environmental context\n",
      "1100/1983 - Environmental Variability Promotes Coexistence in Lottery Competitive Systems\n",
      "1101/1983 - Ration, Growth, and Measures of Somatic and Organ Condition in Relation to Meal Frequency in Winter Flounder, <i>Pseudopleuronectes americanus</i>, with Hypotheses Regarding Population Homeostasis\n",
      "1102/1983 - A Consumer's Guide to Evenness Indices\n",
      "1103/1983 - Stability and Variability in Competitive Communities\n",
      "1104/1983 - From protégé to nurse plant: establishment of thorny shrubs in grazed temperate woodlands\n",
      "1105/1983 - A field test of the stress-gradient hypothesis along an aridity gradient\n",
      "1106/1983 - Strong facilitation in mild environments: the stress gradient hypothesis revisited\n",
      "1107/1983 - Herbivores as mediators of their environment: the impact of large and small species on vegetation dynamics\n",
      "1108/1983 - Facilitation of <i>Quercus ilex</i> recruitment by shrubs in Mediterranean open woodlands\n",
      "1109/1983 - Direct and indirect interactions among plants explain counterintuitive positive drought effects on an eastern Mediterranean shrub species\n",
      "1110/1983 - Resistance of European tree species to drought stress in mixed <i>versus</i> pure forests: evidence of stress release by inter‐specific facilitation\n",
      "1111/1983 - Are Mixed Tropical Tree Plantations More Resistant to Drought than Monocultures?\n",
      "1112/1983 - The functional role of biodiversity in the context of global change\n",
      "1113/1983 - Global warming and biological diversity\n",
      "1114/1983 - Forest Resources in Europe 1950–1990\n",
      "1115/1983 - Will a 385 million year-struggle for light become a struggle for water and for carbon? - How trees may cope with more frequent climate change-type drought events\n",
      "1116/1983 - Growth performance, windthrow, and insects: meta-analyses of parameters influencing performance of mixed-species stands in boreal and northern temperate biomes\n",
      "1117/1983 - Dynamics of global forest area: Results from the FAO Global Forest Resources Assessment 2015\n",
      "1118/1983 - Global Forest Resources Assessment 2015: What, why and how?\n",
      "1119/1983 - Terrestrial and Inland Water Systems\n",
      "1120/1983 - Baseline Map of Carbon Emissions from Deforestation in Tropical Regions\n",
      "1121/1983 - Tropical Forests and the Global Carbon Cycle\n",
      "1122/1983 - Status and trends in global primary forest, protected areas, and areas designated for conservation of biodiversity from the Global Forest Resources Assessment 2015\n",
      "1123/1983 - Size‐dependent predation risk in tree‐feeding insects with different colouration strategies: a field experiment\n",
      "1124/1983 - Reactions of passerine birds to aposematic and non-aposematic firebugs (Pyrrhocoris apterus ; Heteroptera)\n",
      "1125/1983 - Evolutionary convergence of body shape and trophic morphology in cichlids from Lake Tanganyika\n",
      "1126/1983 - Coexistence of Clones in a Heterogeneous Environment\n",
      "1127/1983 - Stability of Small Unisexual‐Bisexual Populations of Poecilipsis (Pisces: Poeciliidae)\n",
      "1128/1983 - Variation and Clonal Structure in a Unisexual Fish\n",
      "1129/1983 - Effects Of Climate-Induced Coral Bleaching On Coral-Reef Fishes ‚Äî Ecological And Economic Consequences\n",
      "1130/1983 - A comparative study of ecological specialization estimators\n",
      "1131/1983 - Coral reefs as drivers of cladogenesis: expanding coral reefs, cryptic extinction events, and the development of biodiversity hotspots\n",
      "1132/1983 - Effect of macroalgal reduction on coral-reef fish in the Watamu Marine National Park, Kenya\n",
      "1133/1983 - Contrasting patterns of individual specialization and trophic coupling in two marine apex predators\n",
      "1134/1983 - Diversity and abundance of insect herbivores foraging on seedlings in a rainforest in Guyana\n",
      "1135/1983 - AN EXPERIMENTAL APPROACH TO THE PRODUCTION DYNAMICS AND STRUCTURE OF FRESHWATER ANIMAL COMMUNITIES1\n",
      "1136/1983 - Density‐Dependent Settler‐Recruit‐Juvenile Relationships in Blue Crabs\n",
      "1137/1983 - Recruitment of Aurelia aurita (Cnidaria: Scyphozoa) larvae is position-dependent, and independent of conspecific density, within a settling surface\n",
      "1138/1983 - Refuges, Disturbance, and Community Structure: A Marine Soft‐Bottom Example\n",
      "1139/1983 - Population dynamics and habitat partitioning by size, sex, and molt stage of blue crabs Callinectes sapidus in a sub-estuary of central Chesapeake Bay\n",
      "1140/1983 - An experimentalist's challenge: when artifacts of intervention interact with treatments\n",
      "1141/1983 - Effects of Artificial Fertilization on Plankton and Benthos Abundance in Four Experimental Ponds\n",
      "1142/1983 - Seagrass Patch Size and Survivorship of an Infaunal Bivalve\n",
      "1143/1983 - The ecology of eelgrass meadows of the Atlantic coast: a community profile\n",
      "1144/1983 - Towards scientific contributions in applying the precautionary principle: an example from southwestern Australia\n",
      "1145/1983 - The conservation of bees.\n",
      "1146/1983 - Sex and the Single Mustard: Population Density and Pollinator Behavior Effects on Seed‐Set\n",
      "1147/1983 - Introduction to WinBUGS for Ecologists: Bayesian approach to regression, ANOVA, mixed models and related analyses\n",
      "1148/1983 - Predator‐‐Induced Life History Changes: Antipredator Behavior Costs or Facultative Life History Shifts?\n",
      "1149/1983 - Relationship between Diurnal Activity Patterns, Cryptic Coloration, and Subsequent Avoidance of Predaceous Fish by Perlid Stoneflies\n",
      "1150/1983 - Phonotaxis in Mole Crickets: Its Reproductive Significance\n",
      "1151/1983 - Size Variation and the Distribution of Hemimetabolous Aquatic Insects: Two Thermal Equilibrium Hypotheses\n",
      "1152/1983 - Predator Avoidance and Community Structure: Interactions among Piscivores, Planktivores, and Plankton\n",
      "1153/1983 - Seasonal Variation in Feeding Habits of Darwin's Ground Finches\n",
      "1154/1983 - Within‐ and Between‐Habitat Avian Diversity in African and Neotropical Lowland Habitats\n",
      "1155/1983 - Darwin's Finches\n",
      "1156/1983 - Switching in General Predators: Experiments on Predator Specificity and Stability of Prey Populations\n",
      "1157/1983 - Drought in the Sahara: A Biogeophysical Feedback Mechanism\n",
      "1158/1983 - Nature's Services: Societal Dependence on Natural Ecosystems\n",
      "1159/1983 - Biodiversity and ecosystem productivity in a fluctuating environment: The insurance hypothesis\n",
      "1160/1983 - Generalized Models Reveal Stabilizing Factors in Food Webs\n",
      "1161/1983 - Generalized models as a universal approach to the analysis of nonlinear dynamical systems\n",
      "1162/1983 - Scale-Dependent Constraints on Food-Web Structure\n",
      "1163/1983 - Divergent Macroparasite Infections in Parapatric Swiss Lake-Stream Pairs of Threespine Stickleback (Gasterosteus aculeatus)\n",
      "1164/1983 - Hybridization between distant lineages increases adaptive variation during a biological invasion: stickleback in Switzerland\n",
      "1165/1983 - A world without parasites: exploring the hidden ecology of infection\n",
      "1166/1983 - Genetic relationships among marine and freshwater populations of the European three‐spined stickleback (<i>Gasterosteus aculeatus</i>) revealed by microsatellites\n",
      "1167/1983 - Some aspects of adaptation in a polymorphism for breeding colours in the Threespine stickleback <i>(Gasterosteus aculeatus)</i>)\n",
      "1168/1983 - Lateral plate asymmetry, diet and parasitism in threespine stickleback\n",
      "1169/1983 - Multiple parasites are driving major histocompatibility complex polymorphism in the wild\n",
      "1170/1983 - Role of epibenthic resource opportunities in the parallel evolution of lake whitefish species pairs (<i>Coregonus</i> sp.)\n",
      "1171/1983 - Quantitative genetic inheritance of morphological divergence in a lake-stream stickleback ecotype pair: implications for reproductive isolation\n",
      "1172/1983 - Designing nature reserves in the face of uncertainty\n",
      "1173/1983 - PLUGGING A HOLE IN THE OCEAN: THE EMERGING SCIENCE OF MARINE RESERVES<sup>1</sup>\n",
      "1174/1983 - The ecological impact of the sea urchin <i>Paracentrotus lividus</i> (Lamarck) in Lough Ine, Ireland\n",
      "1175/1983 - Some aspects of the ecology of strongylocentrotus droebachiensis in eastern Newfoundland\n",
      "1176/1983 - SUBLITTORAL ECOLOGY OF KELP BEDS OF THE OPEN COAST AREA NEAR CARMEL, CALIFORNIA\n",
      "1177/1983 - The ecology of<i>Echinus esculentus</i>L. Quantitative distribution and rate of feeding\n",
      "1178/1983 - Aspects Of the biology of<i>Laminaria hyperborea</i>I. vertical distribution\n",
      "1179/1983 - THE BIOLOGY OF THE TROPICAL SEA URCHIN TRIPNEUSTES ESCULENTUS LESKE IN BARBADOS, BRITISH WEST INDIES\n",
      "1180/1983 - A Comparison of the Biology of<i>Echinus esculentus</i>in Different Habitats\n",
      "1181/1983 - UPTAKE OF AMINO ACIDS BY MARINE INVERTEBRATES1\n",
      "1182/1983 - Marine ecosystem services\n",
      "1183/1983 - ISLAND BIOGEOGRAPHY OF ANTS\n",
      "1184/1983 - EUTROPHICATION AND PREDATION RISK INTERACT TO AFFECT SEXUAL TRAIT EXPRESSION AND MATING SUCCESS\n",
      "1185/1983 - Sexual size dimorphism and the integration of phenotypically plastic traits\n",
      "1186/1983 - Limnology: Lake and River Ecosystems\n",
      "1187/1983 - Body size and fitness in Odonata, stabilising selection and a meta‐analysis too far?\n",
      "1188/1983 - Dimorphic growth patterns and sex-specific reaction norms in the butterfly <i>Lycaena hippothoe sumadiensis</i>\n",
      "1189/1983 - SYNTHESIS: Life history change in commercially exploited fish stocks: an analysis of trends across studies\n",
      "1190/1983 - How do consumers deal with stoichiometric constraints? Lessons from functional genomics using Daphnia pulex\n",
      "1191/1983 - DIRECT AND INDIRECT FITNESS CONSEQUENCES OF FEMALE CHOICE IN A CRUSTACEAN\n",
      "1192/1983 - Sex in a material world: why the study of sexual reproduction and sex-specific traits should become more nutritionally-explicit\n",
      "1193/1983 - wFleaBase: the Daphnia genome database\n",
      "1194/1983 - Guidelines for estimating repeatability\n",
      "1195/1983 - Baboon Mothers and Infants.\n",
      "1196/1983 - Behavioural trade‐offs between growth and mortality explain evolution of submaximal growth rates\n",
      "1197/1983 - RANGE AND DIET OF STOATS (MUSTELA ERMINEA )I N A NEW ZEALAND BEECH FOREST\n",
      "1198/1983 - Weak or strong invaders? A comparison of impact between the native and invaded ranges of mammals and birds alien to Europe\n",
      "1199/1983 - Methods for population control within a silver gull colony.\n",
      "1200/1983 - Ammotragus lervia: a review on systematics, biology, ecology and distribution\n",
      "1201/1983 - Invasion success of vertebrates in Europe and North America\n",
      "1202/1983 - Invasive exotic aoudad (<i>Ammotragus lervia</i>) as a major threat to native Iberian ibex (<i>Capra pyrenaica</i>): a habitat suitability model approach\n",
      "1203/1983 - Handbook of New Zealand Mammals\n",
      "1204/1983 - Relating Species Diversity to Ecosystem Functioning: Mechanistic Backgrounds and Experimental Approach with a Decomposer Food Web\n",
      "1205/1983 - The soil fauna of beech forests: comparison between a mull and a moder soil\n",
      "1206/1983 - Beyond the pollination syndrome: nectar ecology and the role of diurnal and nocturnal pollinators in the reproductive success of <i>Inga sessilis</i> (Fabaceae)\n",
      "1207/1983 - Comparing the efficacy of agri‐environment schemes to enhance bumble bee abundance and diversity on arable field margins\n",
      "1208/1983 - Long foraging distances impose high costs on offspring production in solitary bees\n",
      "1209/1983 - Developing and establishing bee species as crop pollinators: the example of <i>Osmia</i> spp. (Hymenoptera: Megachilidae) and fruit trees\n",
      "1210/1983 - RESOURCE DISTRIBUTIONS AMONG HABITATS DETERMINE SOLITARY BEE OFFSPRING PRODUCTION IN A MOSAIC LANDSCAPE\n",
      "1211/1983 - A landscape‐scale study of bumble bee foraging range and constancy, using harmonic radar\n",
      "1212/1983 - A meta‐analysis of crop pest and natural enemy response to landscape complexity\n",
      "1213/1983 - Delivery of crop pollination services is an insufficient argument for wild pollinator conservation\n",
      "1214/1983 - Insect pollination of crops\n",
      "1215/1983 - Effects of habitat amount and isolation on biodiversity in fragmented traditional orchards\n",
      "1216/1983 - Reproductive biology of <i>Echinopsis terscheckii</i> (Cactaceae): the role of nocturnal and diurnal pollinators\n",
      "1217/1983 - THE PREDICTIVE VALUE OF NECTAR CHEMISTRY TO THE RECOGNITION OF POLLINATOR TYPES\n",
      "1218/1983 - The evolution of bat pollination: a phylogenetic perspective\n",
      "1219/1983 - Larval feeding behavior of three co-occurring species of container mosquitoes.\n",
      "1220/1983 - Consequences of detritus type in an aquatic microsystem: effects on water quality, micro‐organisms and performance of the dominant consumer\n",
      "1221/1983 - Food Webs and Container Habitats\n",
      "1222/1983 - Annual Experimental Evaluation of the Effect of Invasion History on Community Structure\n",
      "1223/1983 - Productivity, Disturbance and Food Web Structure at a Local Spatial Scale in Experimental Container Habitats\n",
      "1224/1983 - Does temperature affect the outcome of larval competition between Aedes aegypti and Aedes albopictus?\n",
      "1225/1983 - Vertical Transmission of West Nile Virus by Three California &lt;I&gt;Culex&lt;/I&gt; (Diptera: Culicidae) Species\n",
      "1226/1983 - Larval Competition Between the Introduced Vector of Dengue Fever in Australia, Aedes-Aegypti (L), and a Native Container-Breeding Mosquito, Aedes-Notoscriptus (Skuse) (Diptera, Culicidae)\n",
      "1227/1983 - Parasitism of Aedes aegypti and Ae. albopictus (Diptera: Culicidae) by Ascogregarina spp. (Apicomplexa: Lecudinidae) in Florida\n",
      "1228/1983 - Host manipulation by parasites in the world of dead-end predators: adaptation to enhance transmission?\n",
      "1229/1983 - Host manipulation by parasites and risk of non-host predation: is manipulation costly in an eye fluke–fish interaction?\n",
      "1230/1983 - Lattice Models and Pair Approximation in Ecology\n",
      "1231/1983 - EFFECT OF HABITAT FRAGMENTATION ON THE EXTINCTION THRESHOLD: A SYNTHESIS<sup>*</sup>\n",
      "1232/1983 - IMPORTANCE OF SPATIAL SCALE TO HABITAT USE BY BREEDING BIRDS IN RIPARIAN FORESTS: A HIERARCHICAL ANALYSIS\n",
      "1233/1983 - One hundred years of eruptions of house mice in Australia - a natural biological curio\n",
      "1234/1983 - Concerns regarding a call for pluralism of information theory and hypothesis testing\n",
      "1235/1983 - Alternatives To Statistical Hypothesis Testing In Ecology: A Guide To Self Teaching\n",
      "1236/1983 - Earthworm Ecology and Biogeography in North America\n",
      "1237/1983 - Studies of earthworm populations in orchards\n",
      "1238/1983 - Test of the enemy release hypothesis: The native magpie moth prefers a native fireweed (<i>Senecio pinnatifolius</i>) to its introduced congener (<i>S. madagascariensis</i>)\n",
      "1239/1983 - The effects of herbivory and competition on the invasive alien plant <i>Senecio inaequidens</i> (Asteraceae)\n",
      "1240/1983 - Biological control of ragwort ( Senecio jacobaea L.) in Australia\n",
      "1241/1983 - Minimal Community Structure: An Explanation of Species Abundance Patterns\n",
      "1242/1983 - Bird Species Diversity: Components of Shannon's Formula\n",
      "1243/1983 - The Structure of Diatom Communities in Similar Ecological Conditions\n",
      "1244/1983 - Stocks are Chameleons: Plasticity in Gill Rakers of Coregonid Fishes\n",
      "1245/1983 - Comparison of Meristic and Morphometric Characters among Spring- and Fall-Spawning Ecotypes of Cisco (<i>Coregonus artedii</i>) in Southern Quebec, Canada\n",
      "1246/1983 - Linking biogeography to physiology: Evolutionary and acclimatory adjustments of thermal limits.\n",
      "1247/1983 - Maintenance of tree diversity in tropical forests.\n",
      "1248/1983 - Determination of total phenolics in acorns from different species of oak trees in conjunction with acorn poisoning in cattle.\n",
      "1249/1983 - Global change: effects on coniferous forests and grasslands.\n",
      "1250/1983 - Global patterns of root turnover for terrestrial ecosystems\n",
      "1251/1983 - Responses of tree fine roots to temperature\n",
      "1252/1983 - Effects of endomycorrhizas on the competitive ability of white clover\n",
      "1253/1983 - Spatial Variation in the Strength of a Trophic Cascade Involving<i>Ruellia nudiflora</i>(Acanthaceae), an Insect Seed Predator and Associated Parasitoid Fauna in Mexico\n",
      "1254/1983 - Specialization, Speciation, and RadiationThe Evolutionary Biology of Herbivorous Insects\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "[E040] Attempt to access token at 281, max length 281.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[27], line 2\u001b[0m\n\u001b[0;32m      1\u001b[0m df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mV3-PapersAggregate.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m----> 2\u001b[0m \u001b[43mscore_dataset_cont\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdf\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mV3-Papers\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mversion\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[1;32mIn[26], line 255\u001b[0m, in \u001b[0;36mscore_dataset_cont\u001b[1;34m(fp, version, df, name)\u001b[0m\n\u001b[0;32m    252\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124mAlready Scored, Skipped\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    253\u001b[0m     \u001b[38;5;28;01mcontinue\u001b[39;00m\n\u001b[1;32m--> 255\u001b[0m \u001b[43mmain\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mupdate_doc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdoc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[0;32m    256\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m main\u001b[38;5;241m.\u001b[39msp_doc \u001b[38;5;129;01mand\u001b[39;00m main\u001b[38;5;241m.\u001b[39mspecies\u001b[38;5;241m.\u001b[39mtn_doc:\n\u001b[0;32m    257\u001b[0m     output \u001b[38;5;241m=\u001b[39m main\u001b[38;5;241m.\u001b[39mscore(verbose\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "Cell \u001b[1;32mIn[25], line 32\u001b[0m, in \u001b[0;36mMain.update_doc\u001b[1;34m(self, doc, verbose)\u001b[0m\n\u001b[0;32m     30\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindex_map \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mload_index_map()\n\u001b[0;32m     31\u001b[0m \u001b[38;5;66;03m# self.parts.update()\u001b[39;00m\n\u001b[1;32m---> 32\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mspecies\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mupdate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdoc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtext\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[0;32m     33\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtrait\u001b[38;5;241m.\u001b[39mupdate(verbose\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[0;32m     34\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcause\u001b[38;5;241m.\u001b[39mupdate(verbose\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "Cell \u001b[1;32mIn[15], line 37\u001b[0m, in \u001b[0;36mSpecies.update\u001b[1;34m(self, text, verbose)\u001b[0m\n\u001b[0;32m     35\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtn_doc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtn_nlp(text)\n\u001b[0;32m     36\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mspans, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtokens, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtoken_to_span, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mspan_starts \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mload_species(verbose\u001b[38;5;241m=\u001b[39mverbose)\n\u001b[1;32m---> 37\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39malternate_names \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mload_alternate_names\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mspans\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[1;32mIn[15], line 144\u001b[0m, in \u001b[0;36mSpecies.load_alternate_names\u001b[1;34m(self, spans, verbose)\u001b[0m\n\u001b[0;32m    141\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m next_species_span\u001b[38;5;241m.\u001b[39mstart \u001b[38;5;241m-\u001b[39m species_span\u001b[38;5;241m.\u001b[39mend \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[0;32m    142\u001b[0m     \u001b[38;5;66;03m# Token Before and After the Next Species\u001b[39;00m\n\u001b[0;32m    143\u001b[0m     before_next \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmain\u001b[38;5;241m.\u001b[39msp_doc[next_species_span\u001b[38;5;241m.\u001b[39mstart\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m--> 144\u001b[0m     after_next \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmain\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msp_doc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mnext_species_span\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mend\u001b[49m\u001b[43m]\u001b[49m\n\u001b[0;32m    146\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m before_next\u001b[38;5;241m.\u001b[39mlower_ \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m(\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;129;01mand\u001b[39;00m after_next\u001b[38;5;241m.\u001b[39mlower_ \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m)\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[0;32m    147\u001b[0m         sp_1_text \u001b[38;5;241m=\u001b[39m species_span\u001b[38;5;241m.\u001b[39mtext\u001b[38;5;241m.\u001b[39mlower()\n",
      "File \u001b[1;32m~\\anaconda3\\envs\\3.10\\lib\\site-packages\\spacy\\tokens\\doc.pyx:505\u001b[0m, in \u001b[0;36mspacy.tokens.doc.Doc.__getitem__\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32m~\\anaconda3\\envs\\3.10\\lib\\site-packages\\spacy\\tokens\\token.pxd:25\u001b[0m, in \u001b[0;36mspacy.tokens.token.Token.cinit\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mIndexError\u001b[0m: [E040] Attempt to access token at 281, max length 281."
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(\"V3-PapersAggregate.csv\")\n",
    "score_dataset_cont(df=df, name='V3-Papers', version=2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
